<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="2.6 Power of using a single structured ‘Project’ directory for storing and tracking research project files | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing.">

<title>2.6 Power of using a single structured ‘Project’ directory for storing and tracking research project files | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="toc.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#overview"><span class="toc-section-number">1</span> Overview</a></li>
<li><a href="2-experimental-data-recording.html#experimental-data-recording"><span class="toc-section-number">2</span> Experimental Data Recording</a></li>
<li><a href="3-experimental-data-preprocessing.html#experimental-data-preprocessing"><span class="toc-section-number">3</span> Experimental Data Preprocessing</a></li>
<li><a href="4-references.html#references"><span class="toc-section-number">4</span> References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="module6" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Power of using a single structured ‘Project’ directory for storing and tracking research project files</h2>
<p>To improve the computational reproducibility of a research project, researchers
can use a single ‘Project’ directory to collectively store all research data,
meta-data, pre-processing code, and research products (e.g., paper drafts,
figures). We will explain how this practice improves the reproducibility and
list some of the common components and subdirectories to include in the
structure of a ‘Project’ directory, including subdirectories for raw and
pre-processed experimental data.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Describe a ‘Project’ directory, including common components and subdirectories</li>
<li>List how a single ‘Project’ directory improves reproducibility</li>
</ul>
<p>In previous modules, we’ve discussed how you can separate the steps of data
collection from steps of data cleaning, management, pre-processing, and
analysis. In the case of data recorded in a laboratory, this separation
will often result in moving from a single file (for example, a spreadsheet),
that combines all the steps to using separate files for the different
steps. As you move to separate data recording and analysis, then, it
is very important to store the set of files for a project in a way that
is clear and easy to manage. If the file directory for the project is
well-designed, it can even allow you to create software tools and report templates
that can be reused over many experiments. In the next few modules, we’ll
discuss how you can organize the files for an experiment using R’s “Project”
system. The modules will discuss the advantages of well-designed project
directories, tips for arranging files within a project directory,
and how to create templates for R Projects that allow you to use consistent
file organization across many experiments.</p>
<p>As a motivating example, for the next few modules, we’ll use an example based on
a set of real immunology experiments. This example highlights how a research
laboratory will often conduct a similar type of experiment many times, so it
lets us demonstrate how the design of the project’s files within a project
directory can be reused across similar experiments. It will allow us to show you
how you can move from designing a file directory for a single experiment to
designing one that can be used repeatedly, and then how you can take advantage
of consistency in the directory structure across projects to make tools and
templates that can be reused.</p>
<div id="example-project" class="section level3" number="2.6.1">
<h3><span class="header-section-number">2.6.1</span> Example project</h3>
<p>The examples for this and the next few modules are based on a collection of
studies that were conducted with similar designs and similar goals—all aimed
to test candidate treatments for tuberculosis. Most studies in this set tested
one or more treatments as well as one or more controls. The controls could
include negative controls, like saline solution, or positive controls, like a
drug already in use to treat the disease, isoniazid. A few of the studies tested
only controls, to help in developing baseline expectations for things like the
bacterial load in different mouse strains used in studies in the set. The
set of studies tested some treatments that were monotherapies (only one drug
given to the animal) as well as some that were combinations of two or three
different drugs. For many of the drugs that were tested, they were tested at
different doses and, in some cases, different methods of delivery or different
mouse models.</p>
<p>Each of the treatments were given to several mice that had been infected with
<em>Mycobacterium tuberculosis</em>. During the treatment, the mice were weighed
regularly. This weight measurement helps to determine if a particular treatment
is well-tolerated by the animals—if not, it may show through the treated mice
losing weight during treatment. For convenience, the mice were not weighed
individually. Instead, mice with the same treatment were kept in a single cage,
and the entire cage was weighed, the weight of the cage itself factored out, and
the average weight of mice for that treatment determined by dividing the weight
of all mice in the cage by the number of mice in the cage. After a period of
time, the mice were sacrificed and one lobe from their lungs was used to
determine each mouse’s bacterial load, through plating the material from the
lobe and counting the colony forming units (CFUs). One aim of the data analysis
is to compare the bacterial load of mice under various treatments to the
bacterial load of mice in the control group.</p>
<p>The full set of studies included 19 different studies. These were conducted at
different times, but the data for all of the studies can be collected using a
common format. In this module, as well as the following two, we’ll be exploring
how you can use RStudio’s Project functionality to organize data from one or
more studies. We’ll particularly focus on how, by using a common format for data
collection, you can create tools that can be used repeatedly for different
experiments to ensure that methods are the same across all studies of a similar
type, as well as to improve the reproducibility of the studies. Let’s walk
through the types of data that were collected for each study.</p>
<p>First, there was some metadata recorded for each study. Figure
<a href="2.6-module6.html#fig:metadata">2.23</a> gives an example. This includes information about the strain
of mouse that was used in the study, treatment details (including the method of
giving the drug or drugs, how often they were given each week, and for how many
weeks), how much bacteria the animals were exposed to (measured both in terms of
the inoculum they were given and their bacterial load one day after they were
given that inoculum, which was based on sacrificing one animal the day after
challenging all the animals with the bacteria), and, if the study included a
novel drug as part of the tested treatment, the batch number of that drug.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:metadata"></span>
<img src="figures/project_metadata.png" alt="Example of recording metadata for a study in the set of example studies for this module." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.23: Example of recording metadata for a study in the set of example studies for this module.
</p>
</div>
<p>Next, the researchers recorded some information about each treatment group
within the experiment. This typically included at least one negative control. In
some cases, there was also a positive control, in which the animals were treated
with a drug that’s in standard use against tuberculosis already (e.g.,
isoniazid). Most studies would also test one or more treatments, which could
include monotherapies or combined therapies. Figure <a href="2.6-module6.html#fig:treatmentdetails">2.24</a>
shows an example of the data that were recorded on each treatment in the study.
These data include the names and doses of up to three drugs in each treatment,
as well as a column where the researcher can provide detailed specifications of
the treatment.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:treatmentdetails"></span>
<img src="figures/project_treatment_details.png" alt="Example of recording treatment details for a study in the set of example studies for this module." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.24: Example of recording treatment details for a study in the set of example studies for this module.
</p>
</div>
<p>Once the animals were challenged with the bacteria, treatment began, and two
main types of data were measured and recorded. First, the mice were weighed once
a week. These weights were converted to a measure of the percent change in
weight since the start of treatment. If the animals’ weights decrease during the
treatment, it is a marker that the treatment is not well-tolerated by the
animals. Figure <a href="2.6-module6.html#fig:mouseweight">2.25</a> shows an example of how these data were
recorded. All animals within a treatment group were kept in the same cage, and
this cage was measured once a week. By dividing the weight of all animals in the
cage by the number of animals, the researchers could estimate the average weight
of animals in that treatment group, which is recorded as shown in Figure
<a href="#fig:mouseweights"><strong>??</strong></a>.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:mouseweight"></span>
<img src="figures/project_metadata.png" alt="Example of recording weekly weights of mice in each treatment group for the example set of studies." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.25: Example of recording weekly weights of mice in each treatment group for the example set of studies.
</p>
</div>
<p>Finally, after the treatment period, the mice were sacrificed and a portion of
each mouse’s lung was used to estimate the bacterial load in that mouse. Figure
<a href="2.6-module6.html#fig:bacterialload">2.26</a> shows an example of how the data on the bacterial load
in each mouse was recorded.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:bacterialload"></span>
<img src="figures/project_bacterial_load.png" alt="Example of recording the bacterial load in the lungs of each mouse at the end of treatment for the example set of studies." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.26: Example of recording the bacterial load in the lungs of each mouse at the end of treatment for the example set of studies.
</p>
</div>
<p>As you can see, these data were all recorded using templates that were designed
for the tidy collection of laboratory data (see modules 2.4 and 2.5). These
spreadsheets were used only to record the data, and then processing, analysis,
and visualization were done in a separate file. Specifically, for this set of
studies a preliminary report was designed, with an example shown in Figure [x].
This report uses the first page to provide a nicely format version of the
metadata for the study, including a table with overall details and a table with
details for each specific treatment that was tested. The second page provides a
graph that shows the percent weight change for mice in each treatment group
compared to the weight of that group at the start of treatment. The third page
provides a graph that shows the bacterial loads in each mouse, grouped by
treatment, as well as the results of running a statistical test to test, for
each treatment group, the hypothesis that the mean of a transformed version of
the measure of bacterial load (log-10) for the group was the same as for the
untreated control group.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:prelimreport"></span>
<img src="figures/project_prelim_report.png" alt="Example of the preliminary report generated for each study in the set of example studies for this module. The first page includes metadata on the study, as well as details on each treatment that was tested. The second page shows how mouse weights in each treatment group changed over the course of treatment, to help identify if a treatment was well-tolerated. The third page graphs the bacterial load in each mouse, grouped by treatment, and gives the result of a statistical analysis to test which treatment groups had outcomes that were significantly different from the untreated control group." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.27: Example of the preliminary report generated for each study in the set of example studies for this module. The first page includes metadata on the study, as well as details on each treatment that was tested. The second page shows how mouse weights in each treatment group changed over the course of treatment, to help identify if a treatment was well-tolerated. The third page graphs the bacterial load in each mouse, grouped by treatment, and gives the result of a statistical analysis to test which treatment groups had outcomes that were significantly different from the untreated control group.
</p>
</div>
<p>Let’s take a closer look at a few of these elements. For example, Figure
<a href="2.6-module6.html#fig:studytable">2.28</a> shows the tables from the first page of the report shown
in Figure <a href="2.6-module6.html#fig:prelimreport">2.27</a>. If you look back to the data collection for
this study (e.g., Figures <a href="2.6-module6.html#fig:metadata">2.23</a> and <a href="2.6-module6.html#fig:treatmentdetails">2.24</a>),
you can see that all of the information in these tables was pulled from data
recorded at the start of the study.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:studytable"></span>
<img src="figures/project_study_info_table.png" alt="Example of one element of the preliminary report generated for each study in the set of example studies for this module. The first page provides tables with metadata about the study and details about each treatment that was tested." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.28: Example of one element of the preliminary report generated for each study in the set of example studies for this module. The first page provides tables with metadata about the study and details about each treatment that was tested.
</p>
</div>
<p>Figure <a href="2.6-module6.html#fig:mouseweightsplot">2.29</a> shows the second page of the report. This
figure has taken the mouse weights—which were recorded in one of the data
collection templates for the project (Figure <a href="#fig:mouseweights"><strong>??</strong></a>)—and used
them to generate a plot of how average mouse weight in each treatment group
changed over the course of the treatment.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:mouseweightsplot"></span>
<img src="figures/project_mouse_weights_graph.png" alt="Example of one element of the preliminary report generated for each study in the set of example studies for this module. The second page provides a plot of how the weights of mice in each treatment changed over the course of treatment." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.29: Example of one element of the preliminary report generated for each study in the set of example studies for this module. The second page provides a plot of how the weights of mice in each treatment changed over the course of treatment.
</p>
</div>
<p>Figure <a href="2.6-module6.html#fig:bactcompare">2.30</a> shows the last page of the report. This page
starts with a figure that shows the bacterial load in the lungs of each mouse in
the study at the end of the treatment period. In this figure, the measurement
for each mouse is shown with a point, and these points are grouped by the
treatment group of the mouse. Boxplots are added to show the distribution across
the mice in each group. The color is used to show whether the treatment was a
negative control, a positive control, a monotherapy, or a combined therapy. The
second part of the page gives a table with the results from running a
statistical analysis to compare the bacterial load for mice in each treatment
group to the bacterial load in the mice in the untreated control group. Color is
added to the table to highlight treatments that had a large difference in
bacterial load from the untreated control, as well as treatments for which the
difference from the untreated control was estimated to be statistically
significant. All the data for these results, including the labels for the plot,
are from the data collected in the data collection templates shown earlier.</p>
<div class="figure fullwidth"><span style="display:block;" id="fig:bactcompare"></span>
<img src="figures/project_bact_compare_plot.png" alt="Example of one element of the preliminary report generated for each study in the set of example studies for this module. The third page provides results on how bacterial load in the lungs compares among treatments at the end of the treatment period." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 2.30: Example of one element of the preliminary report generated for each study in the set of example studies for this module. The third page provides results on how bacterial load in the lungs compares among treatments at the end of the treatment period.
</p>
</div>
</div>
<div id="creating-an-organized-directory-for-project-files" class="section level3" number="2.6.2">
<h3><span class="header-section-number">2.6.2</span> Creating an organized directory for project files</h3>
<p>In the example project that we just described, you can see that we ended up
with a lot of different files. We used separate files for collecting data and
for creating a preliminary report on that data. For collecting the data, we could
either use separate sheets in one spreadsheet file, or entirely separate files,
or some combination. The end result is that we have several project files,
in comparison to if we had used a single sheet of a spreadsheet file to
both record and work with the data.</p>
<p>This is a step in the right direction—by separating data collection and
analysis into separate files, we can make the file for each step simpler.
Further, by separating, we can save the files in plain text, which makes it
easier to track them using version controlled (discussed in later modules),
which helps create a record of changes made to the data or analysis code during
the research process. As we continue work on a project, we’ll likely generate
more types of files. These can include other data files (both “raw”
data—directly output from measurement equipment or directly recorded from
observations, as well as any “cleaned” version of this data, after steps have
been taken to preprocess the data to prepare it for visualization and analysis
in papers and reports). These files also include the files with writing and
presentations (posters and slides) associated with the project, as well as code
scripts for preprocessing data, for conducting data analysis, and for creating
and sharing final figures and tables.</p>
<p>However, now that we have multiple files, we should think about how we store and
organize them. It is very easy otherwise for separate files for a research
project to become disorganized. One study, for example, surveyed over 250
biomedical researchers at the University of Washington. They noted that, “a
common theme surrounding data management and analysis was that may researchers
preferred to utilize their own individual methods to organize data. The varied
ways of managing data were accepted as functional for most present needs. Some
researchers admitted to having no organizational methodology at all, while
others used whatever method best suited their individual needs.”
<span class="citation">(Anderson et al. 2007)</span> One respondent answered, “They’re not organized in any
way—they’re just thrown into files under different projects,” while another
said “I grab them when I need them, they’re not organized in any decent way,”
and another, “It’s not even organized—a file on a central computer of
protocols that we use, common lab protocols but those are just individual Word
files within a folder so it’s not searchable per se.” <span class="citation">(Anderson et al. 2007)</span></p>
<p><label for="tufte-mn-21" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-21" class="margin-toggle"><span class="marginnote"><span style="display: block;">“Just as a well-organized laboratory makes a scientist’s life easier, a well-organized and well-documented project makes a bioinformatician’s life easier. Regardless of the particular project you’re working on, your project directory should be laid out in a consistent and understandable fashion. Clear project organization makes it easier for both you and collaborators to figure out exactly where and what everything is. Additionally, it’s much easier to automate tasks when files are organized and clearly named. For example, processing 300 gene sequences stored in separate FASTA files with a script is trivial if these files are organized in a single directory and are consistently named.”</span>
<span style="display: block;"><span class="citation"><span class="citation">(Buffalo 2015)</span></span></span></span></p>
<p>First, and at a minimum, your research team
should get in the habit of storing all the files for a project (for example,
for an experiment or a study) in the same file directory.
There are a number of advantages to keeping all files related to a single
project inside a dedicated file directory on your computer. First, this provides
a clear and obvious place to search for all project files throughout your work
on the project, including after lulls in activity (for example, while waiting
for reviews from a paper submission). By keeping all project files within a
single directory, you also make it easier to share the collection of files for
the project. There are several reasons you might want to share these files. An
obvious one is that you likely will want to share the project files across
members in your research team, so they can collaborate together on the project.
However, there are also other reasons you’d need to share files, and one that is
growing in popularity is that you may be asked to share files (data, code
scripts, etc.) when you publish a paper describing your results.</p>
<p><label for="tufte-mn-22" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-22" class="margin-toggle"><span class="marginnote"><span style="display: block;">“All files and directories used in your project should live in a single project directory with a clear name. During the course of a project, you’ll have amassed data files, notes, scripts, and so on—if these were scattered all over your hard drive (or worse, across many computers’ hard drives), it would be a nightmare to keep track of everything. Even worse, such a disordered project would later make your research nearly impossible to reproduce.”</span>
<span style="display: block;"><span class="citation"><span class="citation">(Buffalo 2015)</span></span></span></span></p>
<p>When files are all stored in one directory, the directory can be compressed and
shared as an email attachment or through a file sharing platform like Google Drive.
As you learn more tools for reproducibility, you can also share the directory through
some more dynamic platforms, that let all those sharing access continue to change
and contribute to the files in the directory in a way that is tracked and
reversible. In later modules in this book, we will introduce <code>git</code> version control
software and the GitHub platform for sharing files under this type of version
control—this is one example of this more dynamic way of sharing files within
a directory. To gain the advantages of directory-based project file organization, all the
files need to be within a single directory, but they don’t all have to be within
the same “level” in that directory. Instead, you can use subdirectories to
structure and organize these files, while still retaining all the advantages of
directory-based file organization. This will help limit the number of files in
each “level” of the directory, so none becomes an overwhelming slew of files of
different types. It can help you navigate the files in the directory, and also
help someone you share the directory with figure out what’s in it and where
everything is.</p>
<p>Subdirectory organizations can also, it turns out, be used in clever ways within
code scripts applied to files in the directory. For example, there are functions
in all scripting languages that will list all the files in a specified subdirectory.
If you keep all your raw data files of a certain type (for example, all output from
running flow cytometry for the project) within a single subdirectory, you can
use this type of function with code scripts to list all the files in that directory
and then apply code that you’ve developed to preprocess or visualize the data
across all those files. This code would continue to work as you added files to that
directory, since it starts by looking in that subdirectory each time it runs and
working with all files there as of that moment.</p>
<p><label for="tufte-mn-23" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-23" class="margin-toggle"><span class="marginnote"><span style="display: block;">“Organizing data files into a single directory with consistent filenames prepares us to iterate over <em>all</em> of our data, whether it’s the four example files used in this example, or 40,000 files in a real project. Think of it this way: remember when you discovered you could select many files with your mouse cursor? With this trick, you could move 60 files as easily as six files. You could also select certain file types (e.g., photos) and attach them all to an email with one movement. By using consistent file naming and directory organization, you can do the same programatically using the Unix shell and other programming languages.”</span>
<span style="display: block;"><span class="citation"><span class="citation">(Buffalo 2015)</span></span></span></span></p>
<p>As you decide how to organize project files within a directory, it is worthwhile
to take some time to think about the types of files that are often generated by
your research projects, because there are also big advantages to creating a
standard structure of subdirectories that you can use consistently across the
directories for all the projects in your research program. Of course, some
projects may not include certain files, and some might have a new or unusual
type of file, so you can customize the directory structure to some degree for
these types of cases, but it is still a big advantage to include as many common
elements as possible across all your projects.</p>
<p>For example, you may want to always include a subdirectory called “data,” and
consistently call it “data,” to store data directly from observations or
directly output from laboratory equipment. You may want to include
subdirectories in that “data” subdirectory for each type of data—maybe a
“recorded_data” subdirectory for data that were recorded by a researcher in the
laboratory, and another subdirectory called “flow_data” for output from a flow
cytometer. By using the same structure and the same subdirectory names for
different experiments with similar aims or designs, you will find that code
scripts are easier to reuse from one project to another. Again, most scripting
languages allow you to leverage this order in how you’ve arranged your files in
the file system, and so using the same order across different projects lets you
repeat and reuse code scripts more easily from one project to another. This
strategy is used often in handling complex bioinformatics data
<span class="citation">(Buffalo 2015)</span>, but it can also be leveraged to improve the
reproducibility and reliability when only using less complex data recorded in
the laboratory, as with the data shown in the example for this module.</p>
<p><label for="tufte-mn-24" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-24" class="margin-toggle"><span class="marginnote"><span style="display: block;">“Because automating file processing tasks is an integral part of bioinformatics, organizing our projects to facilitate this is essential. Organizing data into subdirectories and using clear and consistent file naming schemes is imperative—both of these practices allow us to <em>programmatically</em> refer to files, the first step to automating a task. Doing something programatically means doing it through code rather than manually, using a method that can effortlessly scale to multiple objects (e.g., files). Programatically referring to multiple files is easier and safer than typing them all out (because it’s less error prone.)”</span>
<span style="display: block;"><span class="citation"><span class="citation">(Buffalo 2015)</span></span></span></span></p>
<p>Even within a subdirectory, it is worthwhile to give a lot of thought to how you
name each file. First, special tools in languages like R can use
thoughtfully-designed files name to extract and use information in the file
names themselves, based on a technique called <em>regular expressions</em>. Second, the
file name is a valuable way to be very clear about exactly what is stored in the
file. The less murky the organization and naming of subdirectories and files within
the project directory is, the more likely that all research team members will be
able to correctly navigate and use the data stored for the project.</p>
<p><label for="tufte-mn-25" class="margin-toggle">⊕</label><input type="checkbox" id="tufte-mn-25" class="margin-toggle"><span class="marginnote"><span style="display: block;">“In addition to simplifying working with files, consistent naming is an often overlooked component of robust bioinformatics. Bad naming schemes can easily lead to switched samples. Poorly chosen filenames can also cause serious errors when you or collaborators think you’re working with the correct data, but it’s actually outdate or the wrong file. I guarantee that out of all the papers published in the past decade, at least a few and likely many more contain erroneous results because of a file naming issue.”</span>
<span style="display: block;"><span class="citation"><span class="citation">(Buffalo 2015)</span></span></span></span></p>
<p>Finally, if you create a clear and clean organization structure for your project
directories, you will find it is much easier to navigate your files in all
directories, and also that new lab members and others you share the directories
with will be able to quickly learn to navigate them. In other areas of science
and engineering, this idea of standardized directory structures has allowed the
development of powerful techniques for open-source software developers to work
together. For example, anyone may create their own extensions to the R
programming language and share these with others through GitHub or several large
repositories. This is coordinated by enforcing a common directory structure on
these extension “packages”—to create a new package, you must put certain types
of files in certain subdirectories within a project directory. With these
standardized rules of directory structure and content, each of these packages
can interact with the base version of R, since there are functions that can tap
into any of these new packages by assuming where each type of file will be
within the package’s directory of files. In a similar way, if you impose a
common directory structure across all the project directories in your research
lab, your collaborators will quickly be able to learn where to find each
element, even in projects they are new to, and you will all be able to write
code that can be easily applied across all project directories, allowing you to
improve reproducibility and comparability across all projects by assuring that
you are conducting the same preprocessing and analysis across all projects (or,
if you are conducting things differently for different projects, that you are
deliberate and aware that you are doing so).</p>
<p>Figure [x] gives an example of a project directory organization that might make
sense for the example set of studies described at the beginning of this module. …</p>
<p>Once you have decided on a structure for your directory, you can create a
template of it—a file directory with all the subdirectories included, but
without any files (or only template files you’d want to use as a starting
point in each project). When you start a new project, you can then just
copy this template and rename it. If you are using R and begin to use
R Project (described in the next section), you can also create an R Studio
Project template to serve as this kind of starting point each time you
start a new project. In module 2.8, we’ll walk through an example of
creating and using this kind of project template for the example set of
studies described earlier in the module.</p>
</div>
<div id="using-rstudio-projects-with-project-file-directories" class="section level3" number="2.6.3">
<h3><span class="header-section-number">2.6.3</span> Using RStudio Projects with project file directories</h3>
<p>If you are using the R programming language for data preprocessing, analysis,
and visualization—as well as RMarkdown for writing reports and
presentations—then you can use RStudio’s “Project” functionality to make it
even more convenient to work with files within a research project’s directory.
You can make any file directory a “Project” in RStudio by chosing “File” -&gt;
“New Project” in RStudio’s menu. This gives you the option to create a
project from scratch or to make an existing directory and RStudio Project.</p>
<p>When you make a file directory an RStudio Project, it doesn’t change much in
the directory itself except adding a “.RProj” file. This file keeps track of
some things about the file directory for RStudio, including preferred settings
for RStudio to use when working in that project. Also, when you
open one of these Projects in RStudio, it will move your working directory
into that projects top-level directory. This makes it very easy and practical
to write code using relative pathnames that start from this top-level of the
project directory. This is very good practice, because these relative pathnames
will work equally well on someone else’s computer, whereas if you use file
pathnames that are absolute (i.e., giving directions to the file from the root
directory on your computer), then when someone else tries on run the code on their
own computer, it won’t work and they’ll need to change the filepaths in the code,
since everyone’s computer has its files organized differently. For example, if you,
on your personal computer, have the project directory stored in your “Documents”
folder, while a colleague has stored the project directory in his or her “Desktop”
directory, then the absolute filepaths for each file in the directory will be
different for each of you. The relative pathnames, starting from the top level of
the project directory, will be the same for both of you, though, regardless of
where you each stored the project directory on your computer.</p>
<p>There are some other advantages, as well, to turning each of your research
project directories into RStudio Projects. One is that it is very easy to
connect each of these Projects with GitHub, which facilitates collaborative work
on the project across multiple team members while tracking all changes under
version control. This functionality is described in a later module in this book.</p>
<p>As you continue to use R and RStudio’s Project functionality, you may want to
take the template directory for your project and create an RStudio Project
template based on its structure. Once you do, when you start a new research
project, you can create the full directory for your project’s files from within
RStudio by going to “File” -&gt; “New Project” and then choosing to create a new
project based on that template. The new project will already be set up with the
“.RProj” file that allows you to easily navigate into and out of that project,
to connect it to GitHub, and all the other advantages of setting a file
directory as an RStudio Project. This takes a bit of time to set-up, but can
be a powerful tool in ensuring that researchers in your laboratory use a
standardized format for project directories across many experiments.</p>
<p>The next module gives step-by-step directions
for making a directory an RStudio Project, and also how to create you own
RStudio Project template to quickly create a new directory for project files
each time you start a new research project.</p>
<!-- ### Practice quiz -->

</div>
</div>
<p style="text-align: center;">
<a href="2.5-module5.html"><button class="btn btn-default">Previous</button></a>
<a href="2.7-module7.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>



</body>
</html>
