<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="3.10 RMarkdown for creating reproducible data pre-processing protocols | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing.">

<title>3.10 RMarkdown for creating reproducible data pre-processing protocols | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>

<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="toc.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#overview"><span class="toc-section-number">1</span> Overview</a>
<ul>
<li><a href="1-1-license.html#license"><span class="toc-section-number">1.1</span> License</a></li>
</ul></li>
<li><a href="2-experimental-data-recording.html#experimental-data-recording"><span class="toc-section-number">2</span> Experimental Data Recording</a>
<ul>
<li><a href="2-1-separating-data-recording-and-analysis.html#separating-data-recording-and-analysis"><span class="toc-section-number">2.1</span> Separating data recording and analysis</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#principles-and-power-of-structured-data-formats"><span class="toc-section-number">2.2</span> Principles and power of structured data formats</a></li>
<li><a href="2-3-the-tidy-data-format.html#the-tidy-data-format"><span class="toc-section-number">2.3</span> The ‘tidy’ data format</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#designing-templates-for-tidy-data-collection"><span class="toc-section-number">2.4</span> Designing templates for “tidy” data collection</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#example-creating-a-template-for-tidy-data-collection"><span class="toc-section-number">2.5</span> Example: Creating a template for “tidy” data collection</a></li>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><span class="toc-section-number">2.6</span> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a></li>
<li><a href="2-7-creating-project-templates.html#creating-project-templates"><span class="toc-section-number">2.7</span> Creating ‘Project’ templates</a></li>
<li><a href="2-8-example-creating-a-project-template.html#example-creating-a-project-template"><span class="toc-section-number">2.8</span> Example: Creating a ‘Project’ template</a></li>
<li><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#harnessing-version-control-for-transparent-data-recording"><span class="toc-section-number">2.9</span> Harnessing version control for transparent data recording</a></li>
<li><a href="2-10-enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><span class="toc-section-number">2.10</span> Enhance the reproducibility of collaborative research with version control platforms</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#using-git-and-gitlab-to-implement-version-control"><span class="toc-section-number">2.11</span> Using git and GitLab to implement version control</a></li>
</ul></li>
<li><a href="3-experimental-data-preprocessing.html#experimental-data-preprocessing"><span class="toc-section-number">3</span> Experimental Data Preprocessing</a>
<ul>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><span class="toc-section-number">3.1</span> Principles and benefits of scripted pre-processing of experimental data</a></li>
<li><a href="3-2-introduction-to-scripted-data-pre-processing-in-r.html#introduction-to-scripted-data-pre-processing-in-r"><span class="toc-section-number">3.2</span> Introduction to scripted data pre-processing in R</a></li>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><span class="toc-section-number">3.3</span> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a></li>
<li><a href="3-4-complex-data-types-in-experimental-data-pre-processing.html#complex-data-types-in-experimental-data-pre-processing"><span class="toc-section-number">3.4</span> Complex data types in experimental data pre-processing</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#complex-data-types-in-r-and-bioconductor"><span class="toc-section-number">3.5</span> Complex data types in R and Bioconductor</a></li>
<li><a href="3-6-example-converting-from-complex-to-tidy-data-formats.html#example-converting-from-complex-to-tidy-data-formats"><span class="toc-section-number">3.6</span> Example: Converting from complex to ‘tidy’ data formats</a></li>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#introduction-to-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.7</span> Introduction to reproducible data pre-processing protocols</a></li>
<li><a href="3-8-introducing-reproducible-data-pre-processing-protocols.html#introducing-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.8</span> Introducing reproducible data pre-processing protocols</a></li>
<li><a href="3-9-from-pre-processing-scripts-to-pre-processing-protocols.html#from-pre-processing-scripts-to-pre-processing-protocols"><span class="toc-section-number">3.9</span> From pre-processing scripts to pre-processing protocols</a></li>
<li><a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.10</span> RMarkdown for creating reproducible data pre-processing protocols</a></li>
<li><a href="3-11-example-creating-a-reproducible-data-pre-processing-protocol.html#example-creating-a-reproducible-data-pre-processing-protocol"><span class="toc-section-number">3.11</span> Example: Creating a reproducible data pre-processing protocol</a></li>
</ul></li>
<li><a href="4-references.html#references"><span class="toc-section-number">4</span> References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="rmarkdown-for-creating-reproducible-data-pre-processing-protocols" class="section level2" number="3.10">
<h2><span class="header-section-number">3.10</span> RMarkdown for creating reproducible data pre-processing protocols</h2>
<p>The R extension package RMarkdown can be used to create documents that combine
code and text in a ‘knitted’ document, and it has become a popular tool for
improving the computational reproducibility and efficiency of the data analysis
stage of research. This tool can also be used earlier in the research process,
however, to improve reproducibility of pre-processing steps. In this module, we
will provide detailed instructions on how to use RMarkdown in RStudio to create
documents that combine code and text. We will show how an RMarkdown document
describing a data pre-processing protocol can be used to efficiently apply the
same data pre-processing steps to different sets of raw data.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Define RMarkdown and the documents it can create</li>
<li>Explain how RMarkdown can be used to improve the reproducibility of research
projects at the data pre-processing phase</li>
<li>Create a document in RStudio using RMarkdown</li>
<li>Apply it to several different datasets with the same format</li>
</ul>
<div id="creating-knitted-documents-in-r" class="section level3" number="3.10.1">
<h3><span class="header-section-number">3.10.1</span> Creating knitted documents in R</h3>
<p>In the last module, we described what knitted documents are, as well as the
advantages of using knitted documents to create data pre-processing protocols
for common pre-processing tasks in your research group. In this module, we will
go into more detail about how you can create these documents using R and
RStudio, and in the next module we will walk through an example data
pre-processing protocol created using this method.</p>
<p>R has a special format for creating knitted documents called <strong>Rmarkdown</strong>.
Later in this module, we will provide details on creating these documents, but
first, we will provide some details of some of the specific conventions used in
this type of knitted documents. As with other knitted documents, Rmarkdown
documents are originally written in plain text. In the previous module, we
discussed how knitted documents will use a markup language within the plain text
for formatting; for RMarkdown files, the markup language used is Markdown. Since
the laguage is Markdown, the preamble for each document uses YAML. We described
in the previous module how knitted documents include executable code, along with
the formatted text. By default, executable code for RMarkdown files will be in
R, but there are also options to include executable code in the document in a
number of other programming languages.</p>
</div>
<div id="creating-and-exploring-an-rmarkdown-file" class="section level3" number="3.10.2">
<h3><span class="header-section-number">3.10.2</span> Creating and exploring an Rmarkdown file</h3>
<p><strong>Creating a new Rmarkdown file and exploring the template.</strong></p>
<p>First, we will explain how you can create a new Rmarkdown document using
RStudio. Like other plain text documents, an Rmarkdown file should be edited
using a text editor, rather than a word processor like Word or Google Docs. It
is easiest to use the Rstudio IDE as the text editor when creating and editing
an R markdown document, as this IDE has incorporated some helpful functionality
for working with plain text documents for Rmarkdown. The RStudio IDE can be
downloaded and installed as a free software, as long as you use the personal
version (RStudio creates higher-powered versions for corporate use). Since
RStudio is free and has many helpful tools for working with Rmarkdown, we will
focus on using this interface in our advice in this module.</p>
<p>In RStudio, you can create a number of types of new files through the
“File” menu. To create a new R markdown file, choose “New File” and then choose
“Rmarkdown” from the choices in that menu. Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownnewfile">3.5</a>
shows an example of what this menu option looks like.</p>
<div class="figure"><span id="fig:rmarkdownnewfile"></span>
<p class="caption marginnote shownote">
Figure 3.5: RStudio pull-down menus to help you navigate to open a new Rmarkdown file.
</p>
<img src="figures/rmarkdown_newfile.png" alt="RStudio pull-down menus to help you navigate to open a new Rmarkdown file." width="\textwidth"  />
</div>
<p>This will open a window with some options you can specify some of the overall
information about the document (Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownchoices">3.6</a>), including
the title and the author, and you can specify the output format that you would
like. Possible output formats include HTML, Word, and PDF. You should be able to
use the HTML and Word output formats without any additional software. If you
would like to use the PDF output, you will need to install one other piece of
software: Miktex for Windows, MacTex for Mac, or TeX Live for Linux. These are
all pieces of software with an underlying TeX engine and all are open-source and
free.</p>
<div class="figure"><span id="fig:rmarkdownchoices"></span>
<p class="caption marginnote shownote">
Figure 3.6: Options available when you create a new Rmarkdown file in RStudio. You can specify information that will go into the document’s preamble, including the title and authors and the format that the document will be output to (HTML, Word, or PDF).
</p>
<img src="figures/rmarkdown_choices.png" alt="Options available when you create a new Rmarkdown file in RStudio. You can specify information that will go into the document's preamble, including the title and authors and the format that the document will be output to (HTML, Word, or PDF)." width="\textwidth"  />
</div>
<p>Once you have selected the options in this menu you can choose the “Okay” button
(Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownchoices">3.6</a>). This will open a new document. This
document, however, won’t be blank. Instead it will include an example document
written in Rmarkdown (Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdowntemplate">3.7</a>). This example
document helps you navigate how the Rmarkdown process works, by letting you test
out a sample document. It also gives you a starting point—once you understand
how the example document works, you can edit it and change it to convert it
into the document you would like to create.</p>
<div class="figure"><span id="fig:rmarkdowntemplate"></span>
<p class="caption marginnote shownote">
Figure 3.7: Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document.
</p>
<img src="figures/rmarkdown_template.png" alt="Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document." width="\textwidth"  />
</div>
<p>If you have not used Rmarkdown before, it is very helpful to try knitting this
example document before making changes, to explore how pieces in the document
align with elements in the rendered output document. Once you are
familiar with the line-up between elements in this file in the output document,
you delete parts of the example file and insert your own text and code.</p>
<p>We will walk you through exploring this example document, as well as aligning it
with the formatted output document (Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>).
First, to render this or any Rmarkdown document, if you are in RStudio you can
use the “Knit” button at the top of the file, as shown in Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdowntemplate2">3.9</a>. When you click on this button, it will render the
entire document to the output format you’ve selected (HTML, PDF, or Word). This
rendering process will both run the executable code and apply all formatting.
The final output (Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>, right) will pop up
in a new window. As you start with Rmarkdown, it is useful to look at this
output to see how it compares with the plain text Rmarkdown file (Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>, left).</p>
<div class="figure fullwidth"><span id="fig:rmarkdownoriginalfinal"></span>
<img src="figures/rmarkdownoriginalfinal.png" alt="Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 3.8: Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document.
</p>
</div>
<div class="figure"><span id="fig:rmarkdowntemplate2"></span>
<p class="caption marginnote shownote">
Figure 3.9: Example of the template Rmarkdown document, highlighting buttons in RStudio that you can use to facilitate working with the document. The ‘knit’ button, highlighted at the top of the figure, will render the entire document. The green arrow, highlighted lower in the figure within a code chunk, can be used to run the code in that specific code chunk.
</p>
<img src="figures/rmarkdown_template2.png" alt="Example of the template Rmarkdown document, highlighting buttons in RStudio that you can use to facilitate working with the document. The 'knit' button, highlighted at the top of the figure, will render the entire document. The green arrow, highlighted lower in the figure within a code chunk, can be used to run the code in that specific code chunk." width="\textwidth"  />
</div>
<p>You will also notice, after you first render the document, that your working
directory has a new file with this output document. For example, if you are
working to create an HTML document using an Rmarkdown file called
“my_report.Rmd,” once you knit your Rmarkdown file, you will notice a new file
in your working directory called “my_report.html.” This new file is your output
file, the one that you would share with colleagues as a report. You should
consider this output document to be read only—in other words, you can read and
share this document, but you should not make any changes directly to this
document, since they will be overwritten anytime you re-render the original
Rmarkdown document.</p>
<p>Next, let’s compare the example Rmarkdown document (the one that is given when
you first open an Rmarkdown file in RStudio) with the output file that is
created when you render this example document (Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>). If you look at the output document (Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>, right), you can notice how different elements
align with pieces in the original Rmarkdown file (Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdownoriginalfinal">3.8</a>). For example, the output document includes a
header with the text “R Markdown.” This second-level header is created by the
Markdown notation in the original file of:</p>
<pre><code>## R Markdown</code></pre>
<p>This header is formatted in a larger font than other text, and on a separate
line—the exact formatting is specified within the style file for the Rmarkdown
document, and will be applied to all second-level headers in the document. You
can also see formatting specified through things like bold font for the word
“Knit,” through the Markdown syntax <code>**Knit**</code>, and a clickable link specified
through the syntax <code>&lt;http://rmarkdown.rstudio.com&gt;</code>. At the beginning of the
original document, you can see how elements like the title, author, date, and
output format are specified in the YAML. Finally, you can see that special
character combinations demarcate sections of executable code.</p>
<p><strong>Formatting text with Markdown.</strong></p>
<p>For the main text in an Rmarkdown document, all formatting is done using
Markdown as the markup language. Markdown is a popular markup language, in part
because it is a good bit simpler than other markup languages like HTML or LaTeX.
This simplicity means that it is not quite as expressive as other markup
languages. However, Markdown provides adequate formatting for 90% of the
formatting you will typically want to do for a research report or
pre-processeing protocol, and by staying simpler, it is much easier to learn the
Markdown syntax quickly compared to other markup languages.</p>
<p>As with other markup languages, Markdown uses special characters or combinations of characters to indicate formatting within the plain text of the original document. When the document is rendered, these markings are used by the software to create the formatting that you have specified in the final output document. Some example formatting symbols and conventions for Markdown include:</p>
<ul>
<li>to format a word or phrase in bold, surround it with two asterisks (<code>**</code>)</li>
<li>to format a word or phrase in italics, surround it with one asterisk (<code>*</code>)</li>
<li>to create a first-level header, put the header text on its own line, starting the line with <code>#</code></li>
<li>to create a second-level header, put the header text on its own line, starting the line with <code>##</code></li>
<li>separate paragraphs with empty lines</li>
<li>use hyphens to create bulleted lists</li>
</ul>
<p>One thing to keep in mind when using Markdown, in terms of formatting, is that
white space can be very important in specifying the formatting. For example when
you specify a new paragraph, you must leave a blank line from your previous
text. Similarly when you use a hash (<code>#</code>) to indicate a header, you must leave a
blank space after the hash before the word or phrase that you want to be used in
that header. To create a section header, you would write:</p>
<pre><code># Initial Data Inspection</code></pre>
<p>Meanwhile this:</p>
<pre><code>#Initial Data Inspection</code></pre>
<p>Would render to:</p>
<p>#Initial Data Inspection</p>
<p>Similarly, white space is needed to separate paragraphs. For example, this would create two paragraphs:</p>
<pre><code>This is a first paragraph. 

This is a second.</code></pre>
<p>Meanwhile this would create one:</p>
<pre><code>This is a first paragraph.
This is still part of the first paragraph.</code></pre>
<p>The syntax of Markdown is fairly simple and can be learned quickly. For more
details on this syntax, you can refer to the Rmarkdown reference guide at
<a href="https://rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf" class="uri">https://rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf</a>. The
basic formatting rules for Markdown are also covered any number of more
extensive resources for Rmarkdown that we will point you to later in this
module.</p>
<p><strong>Creating a preamble with YAML.</strong></p>
<p>In the previous module, we explained how knitted documents include a
preamble to specify some metadata about the document, including elements
like the title, authors, and output format. In R, this preamble is
created using YAML. In this subsection, we provide some more details
on using this YAML section in Rmarkdown documents.</p>
<p>When you initially create an Rmarkdown document, you select the output format
for it (e.g., pdf, Word, HTML). You can change this output selection once you’ve
created the document, however, by making a change in a special section at the
top of the RMarkdown document called the “YAML.” The YAML is a special section
at the top of an RMarkdown document (the original, plain text file, not the
rendered version). It is set off from the rest of the document using a special
combination of characters, using a process very similar to how executable code
is set off from other text with a special set of characters so it can be easily
identified by the software program that renders the document. For the YAML, this
combination of characters is three hyphens (<code>---</code>) on a line by themselves to
start the YAML section and then another three on a line by themselves to end it.
Here is an example of what the YAML might look like at the top of an RMarkdown
document:</p>
<pre><code>---
title: &quot;Laboratory report for example project&quot;
author: &quot;Brooke Anderson&quot;
date: &quot;1/12/2020&quot;
output: word_document
---</code></pre>
<p>Within the YAML itself, you can specify different options for your document.
You can change simple things like the title, author, and date, but you can
also change more complex things, including how the output document is rendered.
For each thing that you want to specify, you specify it with a special
keyword for that option and then a valid choice for that keyword. The idea
is very similar to setting parameter values in a function call in R. For
example, the <code>title:</code> keyword is a valid one in RMarkdown YAML. It allows you
to set the words that will be printed in the title space, using title formatting,
in your output document. It can take any string of characters, so you can put in
any text for the title that you’d like, as long as you surround it with quotation
marks. The <code>author:</code> and <code>date:</code> keywords work in similar ways. The <code>output:</code>
keyword allows you to specify the output that the document should be rendered to.
In this case, the keyword can only take one of a few set values, including
<code>word_document</code> to output a Word document, <code>pdf_document</code> to output a pdf
document (see later in this section for some more set-up required to make that
work), and <code>html_document</code> to output an HTML document.</p>
<p>As you start using RMarkdown, you will be able to do a lot without messing with
the YAML much. In fact, you can get a long way without ever changing the values
in the YAML from the default values they are given when you first create an
RMarkdown document. As you become more familiar with R, you may want to learn
more about how the YAML works and how you can use it to customize your
document—it turns out that quite a lot can be set in the YAML to do very
interesting customizations in your final rendered document. The book <em>R
Markdown: The Definitive Guide</em> [ref], which is available free online, has
sections discussing YAML choices for both HTML and pdf output, at
<a href="https://bookdown.org/yihui/rmarkdown/html-document.html" class="uri">https://bookdown.org/yihui/rmarkdown/html-document.html</a> and
<a href="https://bookdown.org/yihui/rmarkdown/pdf-document.html" class="uri">https://bookdown.org/yihui/rmarkdown/pdf-document.html</a>, respectively. There is
also a talk that Yihui Xie, the creator of RMarkdown, gave on this topic at a
past RStudio conference, available at
<a href="https://rstudio.com/resources/rstudioconf-2017/customizing-extending-r-markdown/" class="uri">https://rstudio.com/resources/rstudioconf-2017/customizing-extending-r-markdown/</a>.</p>
<p><strong>Executable code in Rmarkdown files.</strong></p>
<p>In the previous module, we described how knitted documents use special markers
to indicate where sections of executable code start and stop. In RMarkdown,
the markers you will use to indicate executable code look like this:</p>
<pre><code>```r{}
my_object &lt;- c(1, 2, 3)
```</code></pre>
<p>The first part of this marker (<code>```{r}</code>) indicates that a section of
code is starting. Then end part (<code>```</code>) indicates that the file is
moving back to regular Markdown. In the first marker, you can use the simplest
form (<code>```{r}</code>) to indicate the start of the code. However, you can also
include options to customize actions when the code in that section is executed.
There are many options that you can set for each code chunk. These options will
specify how the code in that section of executable code will be run and how the
output from running the code will be presented. These specifications are called
<strong>chunk options</strong>, and you specify them in the special character combination
where you mark the start of executable code. For example, you can specify that
the code should be printed in the document, but not executed, by setting the
<code>eval</code> parameter to <code>FALSE</code> with <code>```{r eval = FALSE}</code> as the marker to
start the code section.</p>
<p>The chunk options also include <code>echo</code>, which can be used to specify whether to
print the code in that code chunk when the document is rendered. For some
documents, it is useful to print out the code that is executed, where for other
documents you may not want that printed. For example, for a pre-processing
protocol, you are aiming to show yourself and others how the pre-processing was
done. In this case, it is very helpful to print out all of the code, so that
future researchers who read that protocol can clearly see each step. By
contrast, if you are using Rmarkdown to create a report or an article that is
focused on the results of your analysis, it may make more sense to instead hide
the code in the final document.</p>
<p>As part of the code options, you can also specify whether messages and warnings
created when running the code should be included in the document output, and
there are number of code chunk options that specify how tables and figures
rendered by the code should be shown. For more details on the possible options
that can be specified for how code is evaluated within an executable chunk of
code, you refer to the Rmarkdown cheat sheet available at
<a href="https://rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf" class="uri">https://rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf</a></p>
<p>RStudio has some functionality that is useful when you are working with code in
Rmarkdown documents. Within each code chuck are some buttons that can be used to
test out the code in that chunk of executable code. One is the green right arrow
key to the right at the top of the code chunk, highlighted in Figure
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rmarkdowntemplate2">3.9</a>. This button will run all of the code in that
chunk and show you the output in an output field that will open directly below
the code chunk. This functionality allows you to explore the code in your
document as you build it, rather than waiting until you are ready to render the
entire document. The button directly to the left of that button, which looks
like an upward-pointing arrow over a rectangle, will execute all code that comes
before this chunk in the document. This can be very helpful in making sure that
you have set up your environment to run this particular chunk of code.</p>
</div>
<div id="more-advanced-rmarkdown-functionality" class="section level3" number="3.10.3">
<h3><span class="header-section-number">3.10.3</span> More advanced Rmarkdown functionality</h3>
<p>The details and resources that we have covered so far focus on the basics of
Rmarkdown. You can get a lot done just with these basics. However, the Rmarkdown
system is very rich and allows complex functionality beyond these basics. In
this subsection, we will highlight just a few of the ways Rmarkdown can be used
in a more advanced way. Since this topic is so broad, we will focus on elements
that we have found to be particularly useful for biomedical researchers as they
become more advanced Rmarkdown users. For the most part, we will not go into
extensive detail about how to use these more advanced features in this module,
but instead point to resources where you can learn more as you are ready. If you
are just learning Rmarkdown, at this point it will be helpful to just know that
some of these advanced features are available, so you can come back and explore
them when you become familiar with the basics. However, we will provide more
details for two elements that we find particularly useful in creating data
pre-processing protocols: including bibliographical references and including
mathematical equations. In the next module, we walk through an example of using
RMarkdown, and in this example we show how you can use these two more advanced
features when creating a pre-processing protocol.</p>
<p><strong>Including bibliographical references.</strong></p>
<p>You can use Rmarkdown, in combination with something called BibTeX, as a
referencing system. This allows you to include bibliographical references in the
documents that you create, and Rmarkdown will handle the creation of the
references section and the numbering of the documents within your text.</p>
<p>To include references in RMarkdown documents, you can use something called
<strong>BibTeX</strong>. This has three components:</p>
<ol style="list-style-type: decimal">
<li>Create a plain text file with listings for each of
your references (<strong>BibTeX file</strong>). Save it with the
extension <code>.bib</code>.</li>
<li>In your RMarkdown document, include the filepath
to this BibTeX file.</li>
<li>In the text of the RMarkdown file, include a key and special character
combination anytime you want to reference a paper.</li>
</ol>
<p>Once you have a BibTeX file, you will include its
path in the YAML:</p>
<pre><code>  ---
  title: &quot;Reproducible Research with R&quot;
  author: &quot;Brooke Anderson&quot;
  date: &quot;1/25/2021&quot;
  output: beamer_presentation
  bibliography: mybibliography.bib
  ---</code></pre>
<p>It is easiest if you save the BibTeX file in the same directory
as the RMarkdown document.</p>
<p>You can add references in the text using a key for each paper
within special characters (<code>[@paper1, @paper2]</code>).</p>
<p>For example, you would write:</p>
<pre><code>This technique follows earlier work [@fox2020, 
@anderson2019].</code></pre>
<p>To create:</p>
<blockquote>
<p>``This technique follows earlier work (Fox et al. 2020, Anderson et al. 2019)."</p>
</blockquote>
<p>With full paper details included at the end of the document.</p>
<p>Put all the bibliographical details in the BibTeX file:</p>
<p>You will have a file named something like <code>mybibliography.bib</code>
with entries for each paper like this:</p>
<pre><code>@article{fox2020,
  title={Cyto-feature engineering: A pipeline for flow cytometry 
    analysis to uncover immune populations and associations with 
    disease},
  author={Fox, Amy and Dutt, Taru S and Karger, Burton and Rojas, 
    Mauricio and Obreg{\&#39;o}n-Henao, Andr{\&#39;e}s and 
    Anderson, G Brooke and Henao-Tamayo, Marcela},
  journal={Scientific Reports},
  volume={10},
  number={1},
  pages={1--12},
  year={2020}
}</code></pre>
<p>You can get these details from Google Scholar:</p>
<div class="figure fullwidth"><span id="fig:unnamed-chunk-23"></span>
<img src="figures/google_scholar.png" alt="Example of using Google Scholar to get bibliographical information for a BibTeX file. When you look up an article on Google Scholar, there is an option (the quotation mark icon under the article listing) to open a pop-up window with bibliographical information. At the bottom of this pop-up box, you can click on 'BibTeX' to get a plain text version of the BibTeX entry for the article. You can copy and paste this into you BibTeX file." width="\textwidth"  />
<p class="caption marginnote shownote">
Figure 3.10: Example of using Google Scholar to get bibliographical information for a BibTeX file. When you look up an article on Google Scholar, there is an option (the quotation mark icon under the article listing) to open a pop-up window with bibliographical information. At the bottom of this pop-up box, you can click on ‘BibTeX’ to get a plain text version of the BibTeX entry for the article. You can copy and paste this into you BibTeX file.
</p>
</div>
<p><strong>Including mathematical equations.</strong></p>
<p>Next, you can include mathematical equations within an
Rmarkdown document. This allows you to use nice formatting whenever you need
to include an equation within your document.</p>
<p>You can include math using a system developed for specifying mathematical
equations in LaTeX. [Add more on math]</p>
<p><strong>Including executable code in other languages.</strong></p>
<p>In your RMarkdown documents, you include executable code in special sections
(“chunks”) that are separated from the regular text using a special combination
of characters, as described earlier in this module and in the previous module.
By default, in Rmarkdown files the code in these chunks are executed using the R
programming language. However, you can also include executable code in a number
of other programming languages. For example, you could set some code chunks to
run Python, others to run Julia, and still others (e.g., bash) to run a shell script.</p>
<p>This can be very helpful if you have steps in you Python that use code in
different languages. For example, there may be a module in Python that
works well for an early step in your data preprocessing, and then later
steps that are easier with general R functions. This presents no problem in
creating an RMarkdown data pre-processing protocol, as you can include
different steps using different languages.</p>
<p>The program that is used to run the code in a specific chunk is called the
“engine” for that chunk [ref—R Markdown def guide]. You can change the engine
by changing the combination of characters you use to demarcate the start of
executable code. When you are including a chunk of R code, you mark it off
starting with the character combination <code>```{r}</code>. You change this to
give the engine you would like to use—for example, you would include a chunk
of Python code using <code>```{python}</code> [ref—R Markdown def guide]. When
your RMarkdown document is rendered, your computer will use the specified
software to run each code chunk. Of course, to run that piece of code, your
computer must have that type of software installed and available. For example,
if you include a chunk of code that you’d like to run with a Python engine, you
must have Python on your computer.</p>
<p>While you can use many different software programs as the engine for each code
chunk, there are a few limitations with some programs. For many open-source
software programs, the results from running a chunk of code with that engine
will be available for later code chunks that also use that engine to use as an
input [ref—R Markdown def guide]. This is not the case, however, for most of
the available engines. For example, if you use the SAS software program as the
engine for one of your code chunks, the output from running that code will not
be available to input to later code in the document.</p>
<p><strong>Caching code results.</strong></p>
<p>Some code can take a while to run, particularly if it is processing very large
datasets. By default, RMarkdown will re-run all code in the document every time
you render it. This is usually the best set-up, since it allows you to confirm
that the code is all executing as desired each time the code is rendered.
However, if you have steps that take a long time, this can make it so the
RMarkdown document takes a long time to render each time you render it.</p>
<p>To help with this problem, RMarkdown has a system that allows you to <em>cache</em>
results from some or all code chunks in the document. This is a really nice
system—it will check the inputs to that part of the code each time the
document is run. If those inputs have changed, it will take the time to re-run
that piece of code, to use the updated inputs. However, if the inputs have not
changed since the last time the document was rendered, then the last results for
that chunk of code will be pulled from memory and used, without re-running the
code in that chunk. This saves time most of the times that you render the
document, while taking the time to re-run the code when necessary, because the
inputs have changed and so the outputs may be different.</p>
<p>There are some downsides to caching. For example, caching can increase the
storage space it takes to save Rmarkdown work, as intermediate results are
saved. However, if some of your code is very time-intensive to run, it may make
sense to look into caching options with Rmarkdown. For more on caching with
Rmarkdown documents, see <a href="https://bookdown.org/yihui/rmarkdown-cookbook/cache.html">this
section</a> of the <em>R
Markdown Cookbook</em> [ref].</p>
<p><strong>Outputting to other formats.</strong></p>
<p>You can use RMarkdown to create documents other than traditional reports.
Scientists might find the outputs of presentations and posters particularly
useful.</p>
<p>RMarkdown has allowed a pdf slide output for a long time. This output leverages
the “beamer” format from LaTeX. You can create a series of presentation slides
in RMarkdown, using Markdown to specify formatting, and then the document will
be rendered to pdf slides. These slides can be shown using pdf viewer software,
like Adobe Acrobat, set either to full screen or to the presentation option.
More recently, capability has been added to RMarkdown that allows you to create
PowerPoint slides. Again, you will start from an RMarkdown document, using
Markdown syntax to do things like divide content into separate slides.
Regardless of the output format you choose (pdf slides or PowerPoint), the code
to generate figures and tables in the presentation can be included directly in
the RMarkdown file, so it is re-run with the latest data each time you render
the presentation.</p>
<p>It is also possible to use RMarkdown to create scientific posters, although this
is a bit less common and there are fewer tutorial resources with instructions on
doing this. To find out more about creating scientific posters with Rmarkdown,
you can start by looking at the documentation for some R packages that have been
created for this process. Two include the <code>posterdown</code> package [ref], with
documentation available at
<a href="https://reposhub.com/python/miscellaneous/brentthorne-posterdown.html" class="uri">https://reposhub.com/python/miscellaneous/brentthorne-posterdown.html</a>, and the
<code>pagedown</code> package [ref], with documentation available at
<a href="https://github.com/rstudio/pagedown" class="uri">https://github.com/rstudio/pagedown</a>. There are also some blog posts available
where researchers describe how they created a poster with Rmarkdown; one
thorough one is “How to make a poster in R” by Wei Yang Tham, available at
<a href="https://wytham.rbind.io/post/making-a-poster-in-r/" class="uri">https://wytham.rbind.io/post/making-a-poster-in-r/</a>.</p>
<p>This idea of customizing Rmarkdown documents has evolved in another useful way
through the idea of Rmarkdown templates. These are templates that are
customized—often very highly customized—while allowing you to write the
content using Rmarkdown. One area where these templates can be very useful to
scientists is with article templates that are customized for specific
scientific journals. A number of scientific journals have created LaTeX
templates that can be used when writing drafts to submit to the journal.
These templates produce a draft that is nicely formatted, following all
the journal’s guidelines for submission, and in some cases formatted as the
final article would be for the journal. These templates have existed for
a long time, particularly for journals in fields in which LaTeX is commonly
used for document formatting, including physics and statistics. However,
the templates traditionally required you to use LaTeX, which is a complex
markup language with a high threshold for learning to use it.</p>
<p>Now, many of these article templates have been wrapped within an Rmarkdown
template, allowing you to leverage them while writing all the content in
Rmarkdown syntax, and allowing you to include executable code directly in the
draft. An example of the first page of an article created in Rmarkdown using
one of these article templates is shown in Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rticleexample">3.11</a>.</p>
<div class="figure"><span id="fig:rticleexample"></span>
<p class="caption marginnote shownote">
Figure 3.11: Example of a manuscript written in Rmarkdown using a templat. This figure shows the first page of an article written for submission to <em>PLoS Computation Biology</em>, written in Rmarkdown while using the <em>PLoS</em> template from the <code>rticles</code> package. The full article, including the Rmarkdown input and final pdf, are available on GitHub at <a href="https://github.com/cjwendt/plos_ten" class="uri">https://github.com/cjwendt/plos_ten</a>.
</p>
<img src="figures/rticles_example.png" alt="Example of a manuscript written in Rmarkdown using a templat. This figure shows the first page of an article written for submission to *PLoS Computation Biology*, written in Rmarkdown while using the *PLoS* template from the `rticles` package. The full article, including the Rmarkdown input and final pdf, are available on GitHub at https://github.com/cjwendt/plos_ten." width="\textwidth"  />
</div>
<p>These Rmarkdown templates are typically available through R packages, which you
can install on your computer in the same way you would install any R package
(i.e., with the <code>install.packages</code> function). Many journal article templates are
available through the <code>rticles</code> package [ref], including the template used to
create the manuscript shown in Figure <a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#fig:rticleexample">3.11</a>. You can find
more information about the <code>rticles</code> package on its GitHub page, at
<a href="https://github.com/rstudio/rticles" class="uri">https://github.com/rstudio/rticles</a>. There is also a section in the book <em>R
Markdown: A Definitive Guide</em> [ref] on writing manuscripts for scientific
journals using Rmarkdown, available online at
<a href="https://bookdown.org/yihui/rmarkdown/rticles-templates.html" class="uri">https://bookdown.org/yihui/rmarkdown/rticles-templates.html</a>.</p>
<p>You can use RMarkdown to create much larger outputs, compared to simpler reports
and protocols. RMarkdown can now be used to create very large and dynamic
documents, including online books (which can also be rendered to pdf versions
suitable for printing), dashboard-style websites, and blogs. Once members of
your research group are familiar with the basics of RMarkdown, you may want to
explore using it to create these more complex outputs. The book format divides
content into chapters and special sections like appendices and references. It
includes a table of contents based on weblinks, so readers can easily navigate
the content. It uses a book format as its base that allows readers to do things
like change the font size and search the book text for keywords. The book
containing these modules is one example of using bookdown. If you would like to
explore using bookdown to create online books based on Rmarkdown files, there
are a number of resources available. There is an online book available with
extensive instructions on using this package, available at
<a href="https://bookdown.org/yihui/bookdown/" class="uri">https://bookdown.org/yihui/bookdown/</a>. There is also a helpful website with more
details on this package, <a href="https://bookdown.org/" class="uri">https://bookdown.org/</a>. The website include a gallery of
example books created with bookdown <a href="https://bookdown.org/home/archive/" class="uri">https://bookdown.org/home/archive/</a>, which
you can use to explore the types of books that can be created.</p>
<p>You can also use Rmarkdown documents to create webpages, with pages included for
blogs. This format allows you to create a very attractive website that includes
a blog section, where you can write and regularly post new blogs, keeping the
site dynamic. It is a nice entry point to developing and maintaining a website
for people who are learning to code in R but otherwise haven’t done much coding,
as you can do all the steps within RStudio. There are templates for these blogs
that are appropriate for creating personal or research group websites for
academics. These websites can be created to highlight the research and people in
your research lab. You can encourage students and postdocs to create personal
sites, to raise the profile of their research. In the past, we have even used
one as a central, unifying spot for a group study, with students contributing
blog posts as their graded assignment
(<a href="https://kind-neumann-789611.netlify.app/" class="uri">https://kind-neumann-789611.netlify.app/</a>). To learn how to create websites with
blogs, you can check the book <em>blogdown: Creating Websites with R Markdown</em>
[ref], which is available both in print and free online at
<a href="https://bookdown.org/yihui/blogdown/" class="uri">https://bookdown.org/yihui/blogdown/</a>. This process takes a bit of work to
initially get the website set up, but then allows for easy and straightforward
maintenance.</p>
<p>Finally, a simpler way to make basic web content with RMarkdown is through their
<em>flexdashboard</em> format. This format creates a smaller website that is focused on
sharing data results—you can see a gallery of examples at
<a href="https://rmarkdown.rstudio.com/flexdashboard/examples.html" class="uri">https://rmarkdown.rstudio.com/flexdashboard/examples.html</a>. This format is
excellent for creating a webpage that allows users to view complex, and
potentially interactive, results from data you’ve collected. It can be
particularly helpful for groups that need to quickly communicate regularly
updated data to viewers. During 2020, for example, many public health
departments maintained dashboard-style websites to share evolving data on
COVID-19 in the community. Using RMarkdown in this case has the key advantage of
allowing you to easily update the dashboard webpage as you get new or updated
data, since it is easy to re-run any data processing, analysis, and
visualization code in the document. To learn how to use RMarkdown to create
dashboard websites, you can check out RStudio’s flexdashboard site at
<a href="https://rmarkdown.rstudio.com/flexdashboard/index.html" class="uri">https://rmarkdown.rstudio.com/flexdashboard/index.html</a>. There is also guidance
available in one of the chapters of <em>R Markdown: The Definitive Guide</em> [ref]:
<a href="https://bookdown.org/yihui/rmarkdown/dashboards.html" class="uri">https://bookdown.org/yihui/rmarkdown/dashboards.html</a>.</p>
<p><strong>More complex formatting.</strong></p>
<p>As mentioned earlier, Markdown is a fairly simple markup language. Occasionally,
this simplicity means that you might not be able to create fancier formatting
that you might desire. There is a method that allows you to work around this
constraint in RMarkdown.</p>
<p>In Rmarkdown documents, when you need more complex formatting, you can shift
into a more complex markup language for part of the document. Markup languages
like LaTeX and HTML are much more expressive than Markdown, with many more
formatting choices possible. For example, there is functionality within
LaTeX and HTML to create much more complex tables than in Markdown. However,
there is a downside—when you include formatting specified in these
more complex markup languages, you will limit the output formats that you can
render the document to. For example, if you include LaTeX formatting within
an RMarkdown document, you must output the document to PDF, while if you
include HTML, you must output to an HTML file. Conversely, if you stick with
the simpler formatting available through the Markdown syntax, you can easily
switch the output format for your document among several choices.</p>
<p>The <em>R Markdown Cookbook</em> [ref] includes chapters on how to customize Rmarkdown
output through LaTeX
(<a href="https://bookdown.org/yihui/rmarkdown-cookbook/latex-output.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/latex-output.html</a>) and HTML
(<a href="https://bookdown.org/yihui/rmarkdown-cookbook/html-output.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/html-output.html</a>). These
customizations can include creating custom formats for the entire document (for
example, you can customize the appearance of a whole HTML document by
customizing the CSS style file for the document). They can also include
smaller-level customizations, like changing the citation style that is used in
conjunction with a BibTeX file by adding to the preamble for LaTeX output.</p>
<p>One area of customization that is particularly useful and simple to implement is
with customized tables. The Markdown syntax can create very simple tables, but
does not allow the creation of more complex tables. There is an R package called
<code>kableExtra</code> [ref] that allows you to create very attractive and complex tables
in RMarkdown documents.</p>
<p>This package leverages more of the power of underlying markup languages, rather
than the simpler Markdown language. If you remember, Markdown is pretty easy to
learn because it has a somewhat limited set of special characters and special
markings that you can use to specify formatting in your output document. This
basic set of functionality is often all you need, but for complex table
formatting, you will need more. There is much more available in the deeper
markup languages that you can use specifically to render pdf documents (software
derived from TeX) and the one that you can use specifically to render HTML (the
HTML markup language). As a result, you will need to create RMarkdown files that
are customized to a single output format (pdf or HTML) to take advantage of this
package.</p>
<p>You can install this package the same as any other R package from CRAN, using
<code>install.packages</code>. You will need to use then need to use <code>library("kableExtra)</code>
within your RMarkdown document before you use functions from the package. The
<code>kableExtra</code> package is extensively documented through two vignettes that come
with package, one if the output will be in pdf
(<a href="https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_pdf.pdf" class="uri">https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_pdf.pdf</a>)
and one if it will be in HTML
(<a href="https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html" class="uri">https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html</a>).
There is also information on using <code>kableExtra</code> available through <em>R Markdown
Cookbook</em> [ref]: <a href="https://bookdown.org/yihui/rmarkdown-cookbook/kableextra.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/kableextra.html</a>.</p>
</div>
<div id="learning-more-about-rmarkdown." class="section level3" number="3.10.4">
<h3><span class="header-section-number">3.10.4</span> Learning more about Rmarkdown.</h3>
<p>To learn more about RMarkdown, you can explore a number of excellent resources.
The most comprehensive are shared by RStudio, where RMarkdown’s developer
and maintainer, Yihui Xie, works. These resources are all freely available
online, and some are also available to buy as print books, if you prefer that
format.</p>
<p>First, you should check out the online tutorials that are provided by RStudio on
RMarkdown. These are available at RStudio’s RMarkdown page:
<a href="https://rmarkdown.rstudio.com/" class="uri">https://rmarkdown.rstudio.com/</a>. The page’s “Getting Started” section
(<a href="https://rmarkdown.rstudio.com/lesson-1.html" class="uri">https://rmarkdown.rstudio.com/lesson-1.html</a>) provides a nice introduction you
can work through to try out RMarkdown and practice the overview provided in the
last subsection of this module. The “Articles” section
(<a href="https://rmarkdown.rstudio.com/articles.html" class="uri">https://rmarkdown.rstudio.com/articles.html</a>) provides a number of other
documents to help you learn RMarkdown. RStudio’s RMarkdown page also includes a
“Gallery” (<a href="https://rmarkdown.rstudio.com/gallery.html" class="uri">https://rmarkdown.rstudio.com/gallery.html</a>). This resource allows you
to browse through example documents, so you can get a visual idea of what you
might want to create and then access the example code for a similar document.
This is a great resource for exploring the variety of documents that you can
create using RMarkdown.</p>
<p>To go more deeply into RMarkdown, there are two online books from some of the
same team that are available online. The first is <em>R Markdown: The Definitive
Guide</em> by Yihui Xie, J. J. Allaire, and Garrett Grolemund [ref]. This book is
available free online at <a href="https://bookdown.org/yihui/rmarkdown/" class="uri">https://bookdown.org/yihui/rmarkdown/</a>. It moves from
basics through very advanced functionality that you can implement with
RMarkdown, including several of the topics we highlight later in this
subsection.</p>
<p>The second online book to explore from this team is <em>R Markdown Cookbook</em>, by
Yihui Xie, Christophe Dervieux, and Emily Riederer [ref]. This book is available
free online at <a href="https://bookdown.org/yihui/rmarkdown-cookbook/" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/</a>. This book is a
helpful resource for dipping in to a specific section when you want to learn how
to achieve a specific task. Just like a regular cookbook has recipes that you
can explore and use one at a time, this book does not require a comprehensive
end-to-end read, but instead provides “recipes” with advice and instructions for
doing specific things. For example, if you want to figure out how to align a
figure that you create in the center of the page, rather than the left, you can
find a “recipe” in this book to do that.</p>
<!-- ### Potential quotes / refs  -->
<!-- > "WordPerfect was always the best word processor. Because it allowed for insight into -->
<!-- its very structure. You could hit a certain key combination and suddenly the screen -->
<!-- would split and you'd reveal the codes, the bolds and italics, and so forth,  -->
<!-- that would define your text when it was printed. It was beloved of legal secretaries -->
<!-- and journalists alike. Because when you work with words, at the practical, everyday -->
<!-- level, the ability to look under the hood is essential. Words are not simple. And -->
<!-- WordPerfect acknowledged that. Microsoft Word did not. Microsoft kept insisting that -->
<!-- what you saw on your screen was the way things *were*, and if your fonts just kept -->
<!-- sort of randonmly changing, well, you must have wanted it that way. Then along came  -->
<!-- HTML, and what I remember most was that sense of being back inside the file. Sure,  -->
<!-- HTML was a typographic nightmare, a bunch of unjustified Times New Roman in 12 pt on  -->
<!-- screens with chiclet-size pixels, but under the hood you could see all the pieces.  -->
<!-- Just like WordPerfect. That transparency was a wonderful thing, and it renewed -->
<!-- computing for me." [@ford2015on] -->
<!-- > "TeX was created by Donald E. Knuth, a professor at Stanford University who has -->
<!-- achieved international renown as a mathematician and computer scientist. -->
<!-- Knuth also has an aesthetic sense uncommon in his field, and his work output is -->
<!-- truly phenomenal. TeX is a happy byproduct of Knuth's mammoth enterprise,  -->
<!-- *The Art of Computer Programming*. This series of reference books, designed -->
<!-- to cover the whole gamut of programming concepts and techniques, is a  -->
<!-- *sine qua non* for all computer scientists." [@seroul2012beginner] -->
<!-- > "Roughly speaking, text processors fall into two categories: -->
<!-- (1) WYSIWYG systems: what you see is what you get. You see on the screen at all -->
<!-- times what the printed document will look like, and what you type has immediate -->
<!-- effect on the appearance of the document. (2)  markup systems, where you type your text  -->
<!-- interspersed with formatting instructions, but don't see their effect right away. You must run a program to examine the -->
<!-- resulting image, whether on paper or on the screen. In computer science jargon, -->
<!-- markup systems must compile the source file you type.  WYSIWYG systems have the obvious  -->
<!-- advantage of immediate feedback, but they -->
<!-- are not very precise: what is acceptable at a resolution of 300 dots per inch, for an -->
<!-- ephemeral publication such as a newsletter or flier, is no longer so for a book that -->
<!-- will be phototypeset at high resolution. The human eye is extraordinarily sensitive: -->
<!-- you can be bothered by the appearance of a text without being able to pinpoint why, -->
<!-- just as you can tell when someone plays the wrong note in an orchestra, without -->
<!-- being able to identify the CUlprit. One quickly leams in typesetting that the beauty, -->
<!-- legibility and comfortable reading of a text depend on minute details: each element -->
<!-- must be placed exactly right, within thousandths of an inch. For this type of work, -->
<!-- the advantage of immediate feedback vanishes: fine details of spacing, alignment, -->
<!-- and so on are much too small to be discernible at the screen's relatively low  -->
<!-- resolution, and even if it such were not the case, it would still be a monumental chore -->
<!-- to find the right place for everything by hand. For this reason it is not surprising that in the world of professional typesetting -->
<!-- markup systems are preferred. They automate the task of finding the right place -->
<!-- for each character with great precision. Naturally, this approach is less attractive for -->
<!-- beginners, since one can't see the results as one types, and must develop a feeling -->
<!-- for what the system will do. But nowadays, you can have the best of both worlds -->
<!-- by using a markup system with a WYSIWYG *front end*; we'll talk about such front -->
<!-- ends for TEX later on. TEX was developed in the late seventies and early eighties,  -->
<!-- before WYSIWYG systems were widespread. But were it to be redesigned now, it would  -->
<!-- still be a markup -->
<!-- language. To give you an idea of the precision with which TEX operates: the internal  -->
<!-- unit it uses for its calculations is about a hundred times smaller than the -->
<!-- wavelength of visible light! (That's right, a hundred times.) In other words, any -->
<!-- round-off error introduced in the calculations is invisible to the naked eye."  -->
<!-- [@seroul2012beginner] -->
<!-- > "You should be sure to understand the difference between a text editor and a text -->
<!-- processor. A text processor is a text editor together with formatting software that -->
<!-- allows you to switch fonts, do double columns, indent, and so on. A text editor -->
<!-- puts your text in a file on disk, and displays a portion of it on the screen. It doesn't -->
<!-- format your text at all. We insist on the difference because those accustomed to WYSIWYG systems are -->
<!-- often not aware of it: they only know text processors. Where can you find a text -->
<!-- editor? Just about everywhere. Every text processor includes a text editor which -->
<!-- you can use. But if you use your text processor as a text editor, be sure to save your -->
<!-- file using a 'save ASCII' or 'save text only' option, so that the text processor's own -->
<!-- formatting commands are stripped off. If you give TEX a file created without this -->
<!-- precaution, you'll get garbage, because TEX cannot digest your text processor's -->
<!-- commands." [@seroul2012beginner] -->
<!-- > "TeX enabled authors to encode their precise intent into their manuscripts: -->
<!-- This block of text is a computer program, while this word is a keyword in that -->
<!-- program. The language it used, called TeX markup, formalized the slow, -->
<!-- error-prone communication that is normally carried out with the printer over -->
<!-- repeated galley proofs." [@apte2019lingua] -->
<!-- > "The idea of writing markup inside text wasn’t especially novel; it has been -->
<!-- used from 1970’s runoff (the UNIX family of printer-preparation utilities) to -->
<!-- today’s HTML tags. TeX was new in that it captured key concepts necessary for -->
<!-- realistic typesetting and formalized them." [@apte2019lingua] -->
<!-- > "With these higher-level commands, the free TeX engine, and the LaTeX book, -->
<!-- the use of TeX exploded. The macro file has since evolved and changed names, but -->
<!-- authors still typically run the program called latex or its variants. Hence, -->
<!-- most people who write TeX manuscripts know the program as LaTeX and the commands -->
<!-- they use as LaTeX commands." [@apte2019lingua] -->
<!-- > "The effect of LaTeX on scientific and technical publishing has been profound. -->
<!-- Precise typesetting is critical, particularly for conveying concepts using -->
<!-- chemical and mathematical formulas, algorithms, and similar constructs. The -->
<!-- sheer volume of papers, journals, books, and other publications generated in the -->
<!-- modern world is far beyond the throughput possible via manual typesetting. And -->
<!-- TeX enables automation without losing precision. Thanks to LaTeX, book authors -->
<!-- can generate camera-ready copy on their own. Most academic and journal -->
<!-- publishers accept article manuscripts written in LaTeX, and there’s even an open -->
<!-- archive maintained by Cornell University where authors of papers in physics, -->
<!-- chemistry, and other disciplines can directly submit their LaTeX manuscripts for -->
<!-- open viewing. Over 10,000 manuscripts are submitted to this archive every month -->
<!-- from all over the world." [@apte2019lingua] -->
<!-- > "For many users, a practical difficulty with typesetting using TeX is -->
<!-- preparing the manuscripts. When TeX was first developed, technical authors were -->
<!-- accustomed to using plain-text editors like WordStar, vi, or Emacs with a -->
<!-- computer keyboard. The idea of marking up their text with commands and running -->
<!-- the manuscript through a typesetting engine felt natural to them. Today’s -->
<!-- typesetters, particularly desktop publishers, have a different mental model. -->
<!-- They expect to see the output in graphical form and then to visually make edits -->
<!-- with a mouse and keyboard, as they would in any WYSIWYG program. They might not -->
<!-- be too picky about the quality of the output, but they appreciate design -->
<!-- capabilities, such as the ability to flow text around curved outlines. Many -->
<!-- print products are now produced with tools like Microsoft Word for this very -->
<!-- reason. TeX authors cannot do the same work as easily." [@apte2019lingua] -->
<!-- > "Poor documentation can lead to irreproducibility and serious errors. There's -->
<!-- a vast amount of lurking complexity in bioinformatics work: complex workflows,  -->
<!-- multiple files, countless program parameters, and different software versions. -->
<!-- The best way to prevent this complexity from causing problems is to document -->
<!-- everything extensively. Documentation also makes your life easier when you need to  -->
<!-- go back and rerun an analysis, write detailed methods about your steps for a  -->
<!-- paper, or find the origin of some data in a directory." [@buffalo2015bioinformatics] -->
<!-- > "Scatterplots are useful for visualizing treatment-response comparisons ...,  -->
<!-- assocations between variables ..., or paired data (e.g., a disease biomarker in  -->
<!-- several patients before and after treatment)." [@holmes2018modern] -->
<!-- > "Sometimes we want to show the relationshiips between more than two variables. -->
<!-- Obvious choices for including additional dimensions are plot symbol shapes and -->
<!-- colors. ... Another way to show additional dimensions of the data is to show -->
<!-- multiple plots that result from repeatedly subsetting (or 'slicing') the data -->
<!-- based on one (or more) of the variables, so that we can visualize each part -->
<!-- separately. This is called faceting and it enables us to visualize data in up to -->
<!-- four or five dimensions. So we can, for instance, investigate whether the -->
<!-- observed patterns among the other variables are the same or different across the -->
<!-- range of the faceting variable." [@holmes2018modern] -->
<!-- > "You can add an enormous amount of information and expressivity by making your -->
<!-- plots interactive. ... The package `ggvis` is an attempt to extend the good -->
<!-- features of `ggplot2` into the realm of interactive graphics. In contrast to -->
<!-- `ggplot2`, which produces graphics into R's traditional graphics devices (PDF, -->
<!-- PNG, etc.), `ggvis` builds upon a JavaScript infrastructure called *Vega*, and -->
<!-- its plots are intended to be viewed in an HTML browser." [@holmes2018modern] -->
<!-- > "Heatmaps are a powerful way of visualizing large, matrix-like datasets and  -->
<!-- providing a quick overview of the patterns that might be in the data. There  -->
<!-- are a number of heatmap drawing functions in R; one that is convenient and  -->
<!-- produces good-looking output is the function `pheatmap` from the eponymous  -->
<!-- package." [@holmes2018modern] -->
<!-- > "Plots in which most points are huddled in one area, with much of the available -->
<!-- spaces sparesly populated, are difficult to read. If the histogram of the  -->
<!-- marginal distribution of a variable has a sharp peak and then long tails to  -->
<!-- one or both sides, transforming the data can be helpful. ... The plots in  -->
<!-- this chapter that involve microarray data use the logarithmic transformation -->
<!-- [footnote: 'We used it implicitly, since the data in the `ExpressionSet` object -->
<!-- `x` already came log-transformed']---not only in scatterplots... for the x- and -->
<!-- y-coordinates but also ... for the color scale that represents the  -->
<!-- expression fold change. The logarithm transformation is attractive because it has -->
<!-- a definite meaning---a move up or down by the same amount on a log-transformed -->
<!-- scale corresponds to the same multiplicative change on the original scale:  -->
<!-- log(ax) = log a + log x. Sometimes, however, the logarithm is not good enough,  -->
<!-- for instance when the data include zero or negative values, or when even on the  -->
<!-- logarithmic scale the data distribution is highly uneven." [@holmes2018modern] -->
<!-- > "To visualize genomic data, in addition to the general principles we have -->
<!-- discussed in this chapter, there are some specific considerations. The data are -->
<!-- usually associated with genomic coordinates. In fact, genomic coordinates offer -->
<!-- a great organizing principle for the integration of genomic data. ... The main  -->
<!-- challenge of genomic data visualization is the size of the genomes. We need -->
<!-- visualization at multiple scales, from whole genome down to the  -->
<!-- nucleotide level. It should be easy to zoom in and out, and we may need different -->
<!-- visualization strategies for the different size scales. It can be convenient to  -->
<!-- visualize biological molecules (genomes, genes, transcripts, proteins) in a linear -->
<!-- manner, although their embedding in the physical world can matter (a great deal)." -->
<!-- [@holmes2018modern] -->
<!-- > "Visualizing the data, either 'raw' or along the various steps of processing,  -->
<!-- summarization, and inference, is one of the most important activities in applied -->
<!-- statistics and, indeed, in science. It sometimes gets short shrift in textbooks -->
<!-- since there is not much deductive theory. However, there are many good (and bad) -->
<!-- practices, and once you pay attention to it, you will quickly see whether a  -->
<!-- certain graphic is effective in conveying its message, or what -->
<!-- choices you could make to create powerful and aesthetically attractive data -->
<!-- visualizations." [@holmes2018modern] -->
</div>
<div id="applied-exercise-7" class="section level3" number="3.10.5">
<h3><span class="header-section-number">3.10.5</span> Applied exercise</h3>

</div>
</div>
<p style="text-align: center;">
<a href="3-9-from-pre-processing-scripts-to-pre-processing-protocols.html"><button class="btn btn-default">Previous</button></a>
<a href="3-11-example-creating-a-reproducible-data-pre-processing-protocol.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>



</body>
</html>
