<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="3.11 Learning more about Rmarkdown. | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing.">

<title>3.11 Learning more about Rmarkdown. | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>

<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="toc.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#overview"><span class="toc-section-number">1</span> Overview</a>
<ul>
<li><a href="1-1-license.html#license"><span class="toc-section-number">1.1</span> License</a></li>
</ul></li>
<li><a href="2-experimental-data-recording.html#experimental-data-recording"><span class="toc-section-number">2</span> Experimental Data Recording</a>
<ul>
<li><a href="2-1-separating-data-recording-and-analysis.html#separating-data-recording-and-analysis"><span class="toc-section-number">2.1</span> Separating data recording and analysis</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#principles-and-power-of-structured-data-formats"><span class="toc-section-number">2.2</span> Principles and power of structured data formats</a></li>
<li><a href="2-3-the-tidy-data-format.html#the-tidy-data-format"><span class="toc-section-number">2.3</span> The ‘tidy’ data format</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#designing-templates-for-tidy-data-collection"><span class="toc-section-number">2.4</span> Designing templates for “tidy” data collection</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#example-creating-a-template-for-tidy-data-collection"><span class="toc-section-number">2.5</span> Example: Creating a template for “tidy” data collection</a></li>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><span class="toc-section-number">2.6</span> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a></li>
<li><a href="2-7-creating-project-templates.html#creating-project-templates"><span class="toc-section-number">2.7</span> Creating ‘Project’ templates</a></li>
<li><a href="2-8-example-creating-a-project-template.html#example-creating-a-project-template"><span class="toc-section-number">2.8</span> Example: Creating a ‘Project’ template</a></li>
<li><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#harnessing-version-control-for-transparent-data-recording"><span class="toc-section-number">2.9</span> Harnessing version control for transparent data recording</a></li>
<li><a href="2-10-enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><span class="toc-section-number">2.10</span> Enhance the reproducibility of collaborative research with version control platforms</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#using-git-and-gitlab-to-implement-version-control"><span class="toc-section-number">2.11</span> Using git and GitLab to implement version control</a></li>
</ul></li>
<li><a href="3-experimental-data-preprocessing.html#experimental-data-preprocessing"><span class="toc-section-number">3</span> Experimental Data Preprocessing</a>
<ul>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><span class="toc-section-number">3.1</span> Principles and benefits of scripted pre-processing of experimental data</a></li>
<li><a href="3-2-introduction-to-scripted-data-pre-processing-in-r.html#introduction-to-scripted-data-pre-processing-in-r"><span class="toc-section-number">3.2</span> Introduction to scripted data pre-processing in R</a></li>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><span class="toc-section-number">3.3</span> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a></li>
<li><a href="3-4-complex-data-types-in-experimental-data-pre-processing.html#complex-data-types-in-experimental-data-pre-processing"><span class="toc-section-number">3.4</span> Complex data types in experimental data pre-processing</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#complex-data-types-in-r-and-bioconductor"><span class="toc-section-number">3.5</span> Complex data types in R and Bioconductor</a></li>
<li><a href="3-6-example-converting-from-complex-to-tidy-data-formats.html#example-converting-from-complex-to-tidy-data-formats"><span class="toc-section-number">3.6</span> Example: Converting from complex to ‘tidy’ data formats</a></li>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#introduction-to-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.7</span> Introduction to reproducible data pre-processing protocols</a></li>
<li><a href="3-8-introducing-reproducible-data-pre-processing-protocols.html#introducing-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.8</span> Introducing reproducible data pre-processing protocols</a></li>
<li><a href="3-9-from-pre-processing-scripts-to-pre-processing-protocols.html#from-pre-processing-scripts-to-pre-processing-protocols"><span class="toc-section-number">3.9</span> From pre-processing scripts to pre-processing protocols</a></li>
<li><a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.10</span> RMarkdown for creating reproducible data pre-processing protocols</a></li>
<li><a href="3-11-learning-more-about-rmarkdown-.html#learning-more-about-rmarkdown."><span class="toc-section-number">3.11</span> Learning more about Rmarkdown.</a></li>
<li><a href="3-12-example-creating-a-reproducible-data-pre-processing-protocol.html#example-creating-a-reproducible-data-pre-processing-protocol"><span class="toc-section-number">3.12</span> Example: Creating a reproducible data pre-processing protocol</a></li>
</ul></li>
<li><a href="4-references.html#references"><span class="toc-section-number">4</span> References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="learning-more-about-rmarkdown." class="section level2" number="3.11">
<h2><span class="header-section-number">3.11</span> Learning more about Rmarkdown.</h2>
<p>To learn more about RMarkdown, you can explore a number of excellent resources.
The most comprehensive are shared by RStudio, where RMarkdown’s developer
and maintainer, Yihui Xie, works. These resources are all freely available
online, and some are also available to buy as print books, if you prefer that
format.</p>
<p>First, you should check out the online tutorials that are provided by RStudio on
RMarkdown. These are available at RStudio’s RMarkdown page:
<a href="https://rmarkdown.rstudio.com/" class="uri">https://rmarkdown.rstudio.com/</a>. The page’s “Getting Started” section
(<a href="https://rmarkdown.rstudio.com/lesson-1.html" class="uri">https://rmarkdown.rstudio.com/lesson-1.html</a>) provides a nice introduction you
can work through to try out RMarkdown and practice the overview provided in the
last subsection of this module. The “Articles” section
(<a href="https://rmarkdown.rstudio.com/articles.html" class="uri">https://rmarkdown.rstudio.com/articles.html</a>) provides a number of other
documents to help you learn RMarkdown. RStudio’s RMarkdown page also includes a
“Gallery” (<a href="https://rmarkdown.rstudio.com/gallery.html" class="uri">https://rmarkdown.rstudio.com/gallery.html</a>). This resource allows you
to browse through example documents, so you can get a visual idea of what you
might want to create and then access the example code for a similar document.
This is a great resource for exploring the variety of documents that you can
create using RMarkdown.</p>
<p>To go more deeply into RMarkdown, there are two online books from some of the
same team that are available online. The first is <em>R Markdown: The Definitive
Guide</em> by Yihui Xie, J. J. Allaire, and Garrett Grolemund [ref].
This book is available at <a href="https://bookdown.org/yihui/rmarkdown/" class="uri">https://bookdown.org/yihui/rmarkdown/</a>. It moves from
basics through very advanced functionality that you can implement with RMarkdown,
including several of the topics we highlight later in this subsection.</p>
<p>The second online book to explore from this team is <em>R Markdown Cookbook</em>,
by Yihui Xie, Christophe Dervieux, and Emily Riederer. This book is a helpful
resource for dipping in to a specific section when you want to learn how to
achieve a specific task. Just like a regular cookbook has recipes that you can
explore and use one at a time, this book does not require a comprehensive
end-to-end read, but instead provides “recipes” with advice and instructions
for doing specific things. For example, if you want to figure out how to
align a figure that you create in the center of the page, rather than the
left, you can find a “recipe” in this book to do that.</p>
<!-- ### Potential quotes / refs  -->
<!-- > "WordPerfect was always the best word processor. Because it allowed for insight into -->
<!-- its very structure. You could hit a certain key combination and suddenly the screen -->
<!-- would split and you'd reveal the codes, the bolds and italics, and so forth,  -->
<!-- that would define your text when it was printed. It was beloved of legal secretaries -->
<!-- and journalists alike. Because when you work with words, at the practical, everyday -->
<!-- level, the ability to look under the hood is essential. Words are not simple. And -->
<!-- WordPerfect acknowledged that. Microsoft Word did not. Microsoft kept insisting that -->
<!-- what you saw on your screen was the way things *were*, and if your fonts just kept -->
<!-- sort of randonmly changing, well, you must have wanted it that way. Then along came  -->
<!-- HTML, and what I remember most was that sense of being back inside the file. Sure,  -->
<!-- HTML was a typographic nightmare, a bunch of unjustified Times New Roman in 12 pt on  -->
<!-- screens with chiclet-size pixels, but under the hood you could see all the pieces.  -->
<!-- Just like WordPerfect. That transparency was a wonderful thing, and it renewed -->
<!-- computing for me." [@ford2015on] -->
<!-- > "TeX was created by Donald E. Knuth, a professor at Stanford University who has -->
<!-- achieved international renown as a mathematician and computer scientist. -->
<!-- Knuth also has an aesthetic sense uncommon in his field, and his work output is -->
<!-- truly phenomenal. TeX is a happy byproduct of Knuth's mammoth enterprise,  -->
<!-- *The Art of Computer Programming*. This series of reference books, designed -->
<!-- to cover the whole gamut of programming concepts and techniques, is a  -->
<!-- *sine qua non* for all computer scientists." [@seroul2012beginner] -->
<!-- > "Roughly speaking, text processors fall into two categories: -->
<!-- (1) WYSIWYG systems: what you see is what you get. You see on the screen at all -->
<!-- times what the printed document will look like, and what you type has immediate -->
<!-- effect on the appearance of the document. (2)  markup systems, where you type your text  -->
<!-- interspersed with formatting instructions, but don't see their effect right away. You must run a program to examine the -->
<!-- resulting image, whether on paper or on the screen. In computer science jargon, -->
<!-- markup systems must compile the source file you type.  WYSIWYG systems have the obvious  -->
<!-- advantage of immediate feedback, but they -->
<!-- are not very precise: what is acceptable at a resolution of 300 dots per inch, for an -->
<!-- ephemeral publication such as a newsletter or flier, is no longer so for a book that -->
<!-- will be phototypeset at high resolution. The human eye is extraordinarily sensitive: -->
<!-- you can be bothered by the appearance of a text without being able to pinpoint why, -->
<!-- just as you can tell when someone plays the wrong note in an orchestra, without -->
<!-- being able to identify the CUlprit. One quickly leams in typesetting that the beauty, -->
<!-- legibility and comfortable reading of a text depend on minute details: each element -->
<!-- must be placed exactly right, within thousandths of an inch. For this type of work, -->
<!-- the advantage of immediate feedback vanishes: fine details of spacing, alignment, -->
<!-- and so on are much too small to be discernible at the screen's relatively low  -->
<!-- resolution, and even if it such were not the case, it would still be a monumental chore -->
<!-- to find the right place for everything by hand. For this reason it is not surprising that in the world of professional typesetting -->
<!-- markup systems are preferred. They automate the task of finding the right place -->
<!-- for each character with great precision. Naturally, this approach is less attractive for -->
<!-- beginners, since one can't see the results as one types, and must develop a feeling -->
<!-- for what the system will do. But nowadays, you can have the best of both worlds -->
<!-- by using a markup system with a WYSIWYG *front end*; we'll talk about such front -->
<!-- ends for TEX later on. TEX was developed in the late seventies and early eighties,  -->
<!-- before WYSIWYG systems were widespread. But were it to be redesigned now, it would  -->
<!-- still be a markup -->
<!-- language. To give you an idea of the precision with which TEX operates: the internal  -->
<!-- unit it uses for its calculations is about a hundred times smaller than the -->
<!-- wavelength of visible light! (That's right, a hundred times.) In other words, any -->
<!-- round-off error introduced in the calculations is invisible to the naked eye."  -->
<!-- [@seroul2012beginner] -->
<!-- > "You should be sure to understand the difference between a text editor and a text -->
<!-- processor. A text processor is a text editor together with formatting software that -->
<!-- allows you to switch fonts, do double columns, indent, and so on. A text editor -->
<!-- puts your text in a file on disk, and displays a portion of it on the screen. It doesn't -->
<!-- format your text at all. We insist on the difference because those accustomed to WYSIWYG systems are -->
<!-- often not aware of it: they only know text processors. Where can you find a text -->
<!-- editor? Just about everywhere. Every text processor includes a text editor which -->
<!-- you can use. But if you use your text processor as a text editor, be sure to save your -->
<!-- file using a 'save ASCII' or 'save text only' option, so that the text processor's own -->
<!-- formatting commands are stripped off. If you give TEX a file created without this -->
<!-- precaution, you'll get garbage, because TEX cannot digest your text processor's -->
<!-- commands." [@seroul2012beginner] -->
<!-- > "TeX enabled authors to encode their precise intent into their manuscripts: -->
<!-- This block of text is a computer program, while this word is a keyword in that -->
<!-- program. The language it used, called TeX markup, formalized the slow, -->
<!-- error-prone communication that is normally carried out with the printer over -->
<!-- repeated galley proofs." [@apte2019lingua] -->
<!-- > "The idea of writing markup inside text wasn’t especially novel; it has been -->
<!-- used from 1970’s runoff (the UNIX family of printer-preparation utilities) to -->
<!-- today’s HTML tags. TeX was new in that it captured key concepts necessary for -->
<!-- realistic typesetting and formalized them." [@apte2019lingua] -->
<!-- > "With these higher-level commands, the free TeX engine, and the LaTeX book, -->
<!-- the use of TeX exploded. The macro file has since evolved and changed names, but -->
<!-- authors still typically run the program called latex or its variants. Hence, -->
<!-- most people who write TeX manuscripts know the program as LaTeX and the commands -->
<!-- they use as LaTeX commands." [@apte2019lingua] -->
<!-- > "The effect of LaTeX on scientific and technical publishing has been profound. -->
<!-- Precise typesetting is critical, particularly for conveying concepts using -->
<!-- chemical and mathematical formulas, algorithms, and similar constructs. The -->
<!-- sheer volume of papers, journals, books, and other publications generated in the -->
<!-- modern world is far beyond the throughput possible via manual typesetting. And -->
<!-- TeX enables automation without losing precision. Thanks to LaTeX, book authors -->
<!-- can generate camera-ready copy on their own. Most academic and journal -->
<!-- publishers accept article manuscripts written in LaTeX, and there’s even an open -->
<!-- archive maintained by Cornell University where authors of papers in physics, -->
<!-- chemistry, and other disciplines can directly submit their LaTeX manuscripts for -->
<!-- open viewing. Over 10,000 manuscripts are submitted to this archive every month -->
<!-- from all over the world." [@apte2019lingua] -->
<!-- > "For many users, a practical difficulty with typesetting using TeX is -->
<!-- preparing the manuscripts. When TeX was first developed, technical authors were -->
<!-- accustomed to using plain-text editors like WordStar, vi, or Emacs with a -->
<!-- computer keyboard. The idea of marking up their text with commands and running -->
<!-- the manuscript through a typesetting engine felt natural to them. Today’s -->
<!-- typesetters, particularly desktop publishers, have a different mental model. -->
<!-- They expect to see the output in graphical form and then to visually make edits -->
<!-- with a mouse and keyboard, as they would in any WYSIWYG program. They might not -->
<!-- be too picky about the quality of the output, but they appreciate design -->
<!-- capabilities, such as the ability to flow text around curved outlines. Many -->
<!-- print products are now produced with tools like Microsoft Word for this very -->
<!-- reason. TeX authors cannot do the same work as easily." [@apte2019lingua] -->
<!-- > "Poor documentation can lead to irreproducibility and serious errors. There's -->
<!-- a vast amount of lurking complexity in bioinformatics work: complex workflows,  -->
<!-- multiple files, countless program parameters, and different software versions. -->
<!-- The best way to prevent this complexity from causing problems is to document -->
<!-- everything extensively. Documentation also makes your life easier when you need to  -->
<!-- go back and rerun an analysis, write detailed methods about your steps for a  -->
<!-- paper, or find the origin of some data in a directory." [@buffalo2015bioinformatics] -->
<!-- > "Scatterplots are useful for visualizing treatment-response comparisons ...,  -->
<!-- assocations between variables ..., or paired data (e.g., a disease biomarker in  -->
<!-- several patients before and after treatment)." [@holmes2018modern] -->
<!-- > "Sometimes we want to show the relationshiips between more than two variables. -->
<!-- Obvious choices for including additional dimensions are plot symbol shapes and -->
<!-- colors. ... Another way to show additional dimensions of the data is to show -->
<!-- multiple plots that result from repeatedly subsetting (or 'slicing') the data -->
<!-- based on one (or more) of the variables, so that we can visualize each part -->
<!-- separately. This is called faceting and it enables us to visualize data in up to -->
<!-- four or five dimensions. So we can, for instance, investigate whether the -->
<!-- observed patterns among the other variables are the same or different across the -->
<!-- range of the faceting variable." [@holmes2018modern] -->
<!-- > "You can add an enormous amount of information and expressivity by making your -->
<!-- plots interactive. ... The package `ggvis` is an attempt to extend the good -->
<!-- features of `ggplot2` into the realm of interactive graphics. In contrast to -->
<!-- `ggplot2`, which produces graphics into R's traditional graphics devices (PDF, -->
<!-- PNG, etc.), `ggvis` builds upon a JavaScript infrastructure called *Vega*, and -->
<!-- its plots are intended to be viewed in an HTML browser." [@holmes2018modern] -->
<!-- > "Heatmaps are a powerful way of visualizing large, matrix-like datasets and  -->
<!-- providing a quick overview of the patterns that might be in the data. There  -->
<!-- are a number of heatmap drawing functions in R; one that is convenient and  -->
<!-- produces good-looking output is the function `pheatmap` from the eponymous  -->
<!-- package." [@holmes2018modern] -->
<!-- > "Plots in which most points are huddled in one area, with much of the available -->
<!-- spaces sparesly populated, are difficult to read. If the histogram of the  -->
<!-- marginal distribution of a variable has a sharp peak and then long tails to  -->
<!-- one or both sides, transforming the data can be helpful. ... The plots in  -->
<!-- this chapter that involve microarray data use the logarithmic transformation -->
<!-- [footnote: 'We used it implicitly, since the data in the `ExpressionSet` object -->
<!-- `x` already came log-transformed']---not only in scatterplots... for the x- and -->
<!-- y-coordinates but also ... for the color scale that represents the  -->
<!-- expression fold change. The logarithm transformation is attractive because it has -->
<!-- a definite meaning---a move up or down by the same amount on a log-transformed -->
<!-- scale corresponds to the same multiplicative change on the original scale:  -->
<!-- log(ax) = log a + log x. Sometimes, however, the logarithm is not good enough,  -->
<!-- for instance when the data include zero or negative values, or when even on the  -->
<!-- logarithmic scale the data distribution is highly uneven." [@holmes2018modern] -->
<!-- > "To visualize genomic data, in addition to the general principles we have -->
<!-- discussed in this chapter, there are some specific considerations. The data are -->
<!-- usually associated with genomic coordinates. In fact, genomic coordinates offer -->
<!-- a great organizing principle for the integration of genomic data. ... The main  -->
<!-- challenge of genomic data visualization is the size of the genomes. We need -->
<!-- visualization at multiple scales, from whole genome down to the  -->
<!-- nucleotide level. It should be easy to zoom in and out, and we may need different -->
<!-- visualization strategies for the different size scales. It can be convenient to  -->
<!-- visualize biological molecules (genomes, genes, transcripts, proteins) in a linear -->
<!-- manner, although their embedding in the physical world can matter (a great deal)." -->
<!-- [@holmes2018modern] -->
<!-- > "Visualizing the data, either 'raw' or along the various steps of processing,  -->
<!-- summarization, and inference, is one of the most important activities in applied -->
<!-- statistics and, indeed, in science. It sometimes gets short shrift in textbooks -->
<!-- since there is not much deductive theory. However, there are many good (and bad) -->
<!-- practices, and once you pay attention to it, you will quickly see whether a  -->
<!-- certain graphic is effective in conveying its message, or what -->
<!-- choices you could make to create powerful and aesthetically attractive data -->
<!-- visualizations." [@holmes2018modern] -->
<div id="applied-exercise-7" class="section level3" number="3.11.1">
<h3><span class="header-section-number">3.11.1</span> Applied exercise</h3>

</div>
</div>
<p style="text-align: center;">
<a href="3-10-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html"><button class="btn btn-default">Previous</button></a>
<a href="3-12-example-creating-a-reproducible-data-pre-processing-protocol.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>



</body>
</html>
