<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />


<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing.">

<title>Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>

<script src="libs/header-attrs-2.6/header-attrs.js"></script>
<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="toc.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li><a href="index.html#overview"><span class="toc-section-number">1</span> Overview</a>
<ul>
<li><a href="1-1-license.html#license"><span class="toc-section-number">1.1</span> License</a></li>
</ul></li>
<li><a href="2-experimental-data-recording.html#experimental-data-recording"><span class="toc-section-number">2</span> Experimental Data Recording</a>
<ul>
<li><a href="2-1-separating-data-recording-and-analysis.html#separating-data-recording-and-analysis"><span class="toc-section-number">2.1</span> Separating data recording and analysis</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#principles-and-power-of-structured-data-formats"><span class="toc-section-number">2.2</span> Principles and power of structured data formats</a></li>
<li><a href="2-3-the-tidy-data-format.html#the-tidy-data-format"><span class="toc-section-number">2.3</span> The ‘tidy’ data format</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#designing-templates-for-tidy-data-collection"><span class="toc-section-number">2.4</span> Designing templates for “tidy” data collection</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#example-creating-a-template-for-tidy-data-collection"><span class="toc-section-number">2.5</span> Example: Creating a template for “tidy” data collection</a></li>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><span class="toc-section-number">2.6</span> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a></li>
<li><a href="2-7-creating-project-templates.html#creating-project-templates"><span class="toc-section-number">2.7</span> Creating ‘Project’ templates</a></li>
<li><a href="2-8-example-creating-a-project-template.html#example-creating-a-project-template"><span class="toc-section-number">2.8</span> Example: Creating a ‘Project’ template</a></li>
<li><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#harnessing-version-control-for-transparent-data-recording"><span class="toc-section-number">2.9</span> Harnessing version control for transparent data recording</a></li>
<li><a href="2-10-enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><span class="toc-section-number">2.10</span> Enhance the reproducibility of collaborative research with version control platforms</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#using-git-and-gitlab-to-implement-version-control"><span class="toc-section-number">2.11</span> Using git and GitLab to implement version control</a></li>
</ul></li>
<li><a href="3-experimental-data-preprocessing.html#experimental-data-preprocessing"><span class="toc-section-number">3</span> Experimental Data Preprocessing</a>
<ul>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><span class="toc-section-number">3.1</span> Principles and benefits of scripted pre-processing of experimental data</a></li>
<li><a href="3-2-introduction-to-scripted-data-pre-processing-in-r.html#introduction-to-scripted-data-pre-processing-in-r"><span class="toc-section-number">3.2</span> Introduction to scripted data pre-processing in R</a></li>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><span class="toc-section-number">3.3</span> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a></li>
<li><a href="3-4-complex-data-types-in-experimental-data-pre-processing.html#complex-data-types-in-experimental-data-pre-processing"><span class="toc-section-number">3.4</span> Complex data types in experimental data pre-processing</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#complex-data-types-in-r-and-bioconductor"><span class="toc-section-number">3.5</span> Complex data types in R and Bioconductor</a></li>
<li><a href="3-6-example-converting-from-complex-to-tidy-data-formats.html#example-converting-from-complex-to-tidy-data-formats"><span class="toc-section-number">3.6</span> Example: Converting from complex to ‘tidy’ data formats</a></li>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#introduction-to-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.7</span> Introduction to reproducible data pre-processing protocols</a></li>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.8</span> RMarkdown for creating reproducible data pre-processing protocols</a></li>
<li><a href="3-9-example-creating-a-reproducible-data-pre-processing-protocol.html#example-creating-a-reproducible-data-pre-processing-protocol"><span class="toc-section-number">3.9</span> Example: Creating a reproducible data pre-processing protocol</a></li>
</ul></li>
<li><a href="4-references.html#references"><span class="toc-section-number">4</span> References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="references" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> References</h1>

<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-al2016protocol" class="csl-entry">
Al-JunDi, AzzAm, and SAlAh SAkkA. 2016. <span>“Protocol Writing in Clinical Research.”</span> <em>Journal of Clinical and Diagnostic Research: JCDR</em> 10 (11): ZE10.
</div>
<div id="ref-altarawneh2017pilot" class="csl-entry">
AlTarawneh, Ghada, and Simon Thorne. 2017. <span>“A Pilot Study Exploring Spreadsheet Risk in Scientific Research.”</span> <em>arXiv Preprint arXiv:1703.09785</em>.
</div>
<div id="ref-altman1997units" class="csl-entry">
Altman, Douglas G., and J. Martin Bland. 1997. <span>“Statistical Notes: Units of Analysis.”</span> <em>BMJ</em> 314: 1874.
</div>
<div id="ref-altschul2013anatomy" class="csl-entry">
Altschul, Stephen, Barry Demchak, Richard Durbin, Robert Gentleman, Martin Krzywinski, Heng Li, Anton Nekrutenko, et al. 2013. <span>“The Anatomy of Successful Computational Biology Software.”</span> <em>Nature Biotechnology</em> 31 (10): 894–97.
</div>
<div id="ref-anderson2007issues" class="csl-entry">
Anderson, Nicholas R, E Sally Lee, J Scott Brockenbrough, Mark E Minie, Sherrilynne Fuller, James Brinkley, and Peter Tarczy-Hornoch. 2007. <span>“Issues in Biomedical Research Data Management and Analysis: Needs and Barriers.”</span> <em>Journal of the American Medical Informatics Association</em> 14 (4): 478–88.
</div>
<div id="ref-apte2019lingua" class="csl-entry">
Apte, Poornima. 2019. <span>“The Lingua Franca of LaTeX.”</span> <em>Increment</em>. <a href="https://increment.com/open-source/the-lingua-franca-of-latex/">https://increment.com/open-source/the-lingua-franca-of-latex/</a>.
</div>
<div id="ref-RJ-2017-035" class="csl-entry">
Arnold, Taylor. 2017. <span>“<span class="nocase">A Tidy Data Model for Natural Language Processing using cleanNLP</span>.”</span> <em><span>The R Journal</span></em> 9 (2): 248–67.
</div>
<div id="ref-barga2011bioinformatics" class="csl-entry">
Barga, Roger, Bill Howe, David Beck, Stuart Bowers, William Dobyns, Winston Haynes, Roger Higdon, et al. 2011. <span>“Bioinformatics and Data-Intensive Scientific Discovery in the Beginning of the 21st Century.”</span> <em>Omics: A Journal of Integrative Biology</em> 15 (4): 199–201.
</div>
<div id="ref-biobroom" class="csl-entry">
Bass, Andrew J., David G. Robinson, Steve Lianoglou, Emily Nelson, John D. Storey, and with contributions from Laurent Gatto. 2020. <em>Biobroom: Turn Bioconductor Objects into Tidy Data Frames</em>. <a href="https://github.com/StoreyLab/biobroom">https://github.com/StoreyLab/biobroom</a>.
</div>
<div id="ref-baumer2015data" class="csl-entry">
Baumer, Ben. 2015. <span>“A Data Science Course for Undergraduates: Thinking with Data.”</span> <em>The American Statistician</em> 69 (4): 334–42.
</div>
<div id="ref-baumer2017modern" class="csl-entry">
Baumer, Benjamin S, Daniel T Kaplan, and Nicholas J Horton. 2017. <em>Modern Data Science with r</em>. Boca Raton: CRC Press.
</div>
<div id="ref-birch2018future" class="csl-entry">
Birch, David, David Lyford-Smith, and Yike Guo. 2018. <span>“The Future of Spreadsheets in the Big Data Era.”</span> <em>arXiv Preprint arXiv:1801.10231</em>.
</div>
<div id="ref-blischak2016quick" class="csl-entry">
Blischak, John D, Emily R Davenport, and Greg Wilson. 2016. <span>“A Quick Introduction to Version Control with Git and GitHub.”</span> <em>PLoS Computational Biology</em> 12 (1).
</div>
<div id="ref-brazma2006standards" class="csl-entry">
Brazma, Alvis, Maria Krestyaninova, and Ugis Sarkans. 2006. <span>“Standards for Systems Biology.”</span> <em>Nature Reviews Genetics</em> 7 (8): 593.
</div>
<div id="ref-broman2018data" class="csl-entry">
Broman, Karl W, and Kara H Woo. 2018. <span>“Data Organization in Spreadsheets.”</span> <em>The American Statistician</em> 72 (1): 2–10.
</div>
<div id="ref-brown2018git" class="csl-entry">
Brown, Zack. 2018. <span>“A Git Origin Story.”</span> <em>Linux Journal</em>.
</div>
<div id="ref-bryan2018excuse" class="csl-entry">
Bryan, Jennifer. 2018. <span>“Excuse Me, Do You Have a Moment to Talk about Version Control?”</span> <em>The American Statistician</em> 72 (1): 20–27.
</div>
<div id="ref-bryan2017data" class="csl-entry">
Bryan, Jennifer, and Hadley Wickham. 2017. <span>“Data Science: A Three Ring Circus or a Big Tent?”</span> <em>Journal of Computational and Graphical Statistics</em> 26 (4): 784–85.
</div>
<div id="ref-buffalo2015bioinformatics" class="csl-entry">
Buffalo, Vince. 2015. <em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>. " O’Reilly Media, Inc.".
</div>
<div id="ref-burns2011r" class="csl-entry">
Burns, Patrick. 2011. <em>The r Inferno</em>. Lulu. com.
</div>
<div id="ref-butler2005electronic" class="csl-entry">
Butler, Declan. 2005. <span>“Electronic Notebooks: A New Leaf.”</span> <em>Nature</em> 436 (7047): 20–22.
</div>
<div id="ref-campbell2007number" class="csl-entry">
Campbell-Kelly, Martin. 2007. <span>“Number Crunching Without Programming: The Evolution of Spreadsheet Usability.”</span> <em>IEEE Annals of the History of Computing</em> 29 (3): 6–19.
</div>
<div id="ref-chambers2006s4" class="csl-entry">
Chambers, JM. 2006. <span>“How S4 Methods Work.”</span> Technical report.
</div>
<div id="ref-chen2014edger" class="csl-entry">
Chen, Yunshun, Davis McCarthy, Mark Robinson, and Gordon K Smyth. 2014. <span>“edgeR: Differential Expression Analysis of Digital Gene Expression Data User’s Guide.”</span> <em>Bioconductor User’s Guide. Available Online: Http://Www. Bioconductor. Org/Packages/Release/Bioc/Vignettes/edgeR/Inst/Doc/edgeRUsersGuide.pdf (Accessed on 15 February 2021)</em>.
</div>
<div id="ref-creeth1985microcomputer" class="csl-entry">
Creeth, Richard. 1985. <span>“Microcomputer Spreadsheets: Their Uses and Abuses.”</span> <em>Journal of Accountancy (Pre-1986)</em> 159 (000006): 90.
</div>
<div id="ref-edwards2011science" class="csl-entry">
Edwards, Paul N, Matthew S Mayernik, Archer L Batcheller, Geoffrey C Bowker, and Christine L Borgman. 2011. <span>“Science Friction: Data, Metadata, and Collaboration.”</span> <em>Social Studies of Science</em> 41 (5): 667–90.
</div>
<div id="ref-ellis2018share" class="csl-entry">
Ellis, Shannon E, and Jeffrey T Leek. 2018. <span>“How to Share Data for Collaboration.”</span> <em>The American Statistician</em> 72 (1): 53–57.
</div>
<div id="ref-ford2015on" class="csl-entry">
Ford, Paul. 2014. <span>“On File Formats, Very Briefly.”</span> <em>The Manual</em>. <a href="https://themanual.org/read/issues/4/paul-ford/article">https://themanual.org/read/issues/4/paul-ford/article</a>.
</div>
<div id="ref-ford2015i" class="csl-entry">
———. 2015. <span>“I Dreamed of a Perfect Database.”</span> <em>New Republic</em>. <a href="https://newrepublic.com/article/124425/dreamed-perfect-database">https://newrepublic.com/article/124425/dreamed-perfect-database</a>.
</div>
<div id="ref-gatto2013msnbase" class="csl-entry">
Gatto, Laurent. 2013. <span>“MSnbase Development.”</span>
</div>
<div id="ref-gentleman2008r" class="csl-entry">
Gentleman, Robert. 2008. <em>R Programming for Bioinformatics</em>. CRC Press.
</div>
<div id="ref-ghosh2011software" class="csl-entry">
Ghosh, Samik, Yukiko Matsuoka, Yoshiyuki Asai, Kun-Yi Hsin, and Hiroaki Kitano. 2011. <span>“Software for Systems Biology: From Tools to Integrated Platforms.”</span> <em>Nature Reviews Genetics</em> 12 (12): 821.
</div>
<div id="ref-giles2012digital" class="csl-entry">
Giles, Jim. 2012. <span>“The Digital Lab: Lab-Management Software and Electronic Notebooks Are Here–and This Time, It’s More Than Just Talk.”</span> <em>Nature</em> 481 (7382): 430–32.
</div>
<div id="ref-gillespie2016efficient" class="csl-entry">
Gillespie, Colin, and Robin Lovelace. 2016. <em>Efficient r Programming: A Practical Guide to Smarter Programming</em>. " O’Reilly Media, Inc.".
</div>
<div id="ref-goodman2014ten" class="csl-entry">
Goodman, Alyssa, Alberto Pepe, Alexander W Blocker, Christine L Borgman, Kyle Cranmer, Merce Crosas, Rosanne Di Stefano, et al. 2014. <span>“Ten Simple Rules for the Care and Feeding of Scientific Data.”</span> Public Library of Science.
</div>
<div id="ref-grune1986concurrent" class="csl-entry">
Grune, Dick. 1986. <span>“Concurrent Verions System, a Method for Independent Cooperation.”</span> <em>Technical Report, Vrije Universiteit</em>.
</div>
<div id="ref-hammer2010mrna" class="csl-entry">
Hammer, Paul, Michaela S Banck, Ronny Amberg, Cheng Wang, Gabriele Petznick, Shujun Luo, Irina Khrebtukova, Gary P Schroth, Peter Beyerlein, and Andreas S Beutler. 2010. <span>“mRNA-Seq with Agnostic Splice Site Discovery for Nervous System Transcriptomics Tested in Chronic Pain.”</span> <em>Genome Research</em> 20 (6): 847–60.
</div>
<div id="ref-hermans2016spreadsheets" class="csl-entry">
Hermans, Felienne, Bas Jansen, Sohon Roy, Efthimia Aivaloglou, Alaaeddin Swidan, and David Hoepelman. 2016. <span>“Spreadsheets Are Code: An Overview of Software Engineering Approaches Applied to Spreadsheets.”</span> In <em>2016 IEEE 23rd International Conference on Software Analysis, Evolution, and Reengineering (SANER)</em>, 5:56–65. IEEE.
</div>
<div id="ref-hermans2015enron" class="csl-entry">
Hermans, Felienne, and Emerson Murphy-Hill. 2015. <span>“Enron’s Spreadsheets and Related Emails: A Dataset and Analysis.”</span> In <em>2015 IEEE/ACM 37th IEEE International Conference on Software Engineering</em>, 2:7–16. IEEE.
</div>
<div id="ref-hicks2017guide" class="csl-entry">
Hicks, Stephanie C, and Rafael A Irizarry. 2017. <span>“A Guide to Teaching Data Science.”</span> <em>The American Statistician</em> In Press. <a href="https://doi.org/10.1080/00031305.2017.1356747">https://doi.org/10.1080/00031305.2017.1356747</a>.
</div>
<div id="ref-hicks2021mbkmeans" class="csl-entry">
Hicks, Stephanie C, Ruoxi Liu, Yuwei Ni, Elizabeth Purdom, and Davide Risso. 2021. <span>“Mbkmeans: Fast Clustering for Single Cell Data Using Mini-Batch k-Means.”</span> <em>PLOS Computational Biology</em> 17 (1): e1008625.
</div>
<div id="ref-holmes2018modern" class="csl-entry">
Holmes, Susan, and Wolfgang Huber. 2018. <em>Modern Statistics for Modern Biology</em>. Cambridge University Press.
</div>
<div id="ref-hsieh2016inext" class="csl-entry">
Hsieh, TC, KH Ma, and Anne Chao. 2016. <span>“<span class="nocase">iNEXT: an R package for rarefaction and extrapolation of species diversity (Hill numbers)</span>.”</span> <em>Methods in Ecology and Evolution</em> 7 (12): 1451–56.
</div>
<div id="ref-biobase" class="csl-entry">
Huber, W., V. J. Carey, R. Gentleman, S. Anders, M. Carlson, B. S. Carvalho, H. C. Bravo, et al. 2015. <span>“<span>O</span>rchestrating High-Throughput Genomic Analysis with <span>B</span>ioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html">http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html</a>.
</div>
<div id="ref-huber2015orchestrating" class="csl-entry">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. <span>“Orchestrating High-Throughput Genomic Analysis with Bioconductor.”</span> <em>Nature Methods</em> 12 (2): 115.
</div>
<div id="ref-hunt2000pragmatic" class="csl-entry">
Hunt, Andrew, David Thomas, and Ward Cunningham. 2000. <em>The Pragmatic Programmer: From Journeyman to Master</em>. Addison-Wesley Professional.
</div>
<div id="ref-lifesciencesR" class="csl-entry">
Irizarry, Rafael A., and Michael I. Love. 2016. <em>Data Analysis for the Life Sciences with r</em>. Chapman; Hall.
</div>
<div id="ref-irving2011astonishments" class="csl-entry">
Irving, Francis. 2011. <span>“Astonishments, Ten, in the History of Version Control.”</span> <a href="https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/">https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/</a>.
</div>
<div id="ref-janssens2014data" class="csl-entry">
Janssens, Jeroen. 2014. <em>Data Science at the Command Line: Facing the Future with Time-Tested Tools</em>. " O’Reilly Media, Inc.".
</div>
<div id="ref-kaplan2018teaching" class="csl-entry">
Kaplan, Daniel. 2018. <span>“Teaching Stats for Data Science.”</span> <em>The American Statistician</em> 72 (1): 89–96.
</div>
<div id="ref-keller2017evolution" class="csl-entry">
Keller, Sallie, Gizem Korkmaz, Mark Orr, Aaron Schroeder, and Stephanie Shipp. 2017. <span>“The Evolution of Data Quality: Understanding the Transdisciplinary Origins of Data Quality Concepts and Approaches.”</span> <em>Annu. Rev. Stat. Appl</em> 4: 85–108.
</div>
<div id="ref-kernighan2011d" class="csl-entry">
Kernighan, Brian W. 2011. <em>D Is for Digital: What a Well-Informed Person Should Know about Computers and Communications</em>. CreateSpace Independent Publishing Platform.
</div>
<div id="ref-kernighan1984unix" class="csl-entry">
Kernighan, Brian W, and Rob Pike. 1984. <em>The UNIX Programming Environment</em>. Vol. 270. Prentice-Hall Englewood Cliffs, NJ.
</div>
<div id="ref-krishnan2016towards" class="csl-entry">
Krishnan, Sanjay, Daniel Haas, Michael J Franklin, and Eugene Wu. 2016. <span>“Towards Reliable Interactive Data Cleaning: A User Survey and Recommendations.”</span> In <em>Proceedings of the Workshop on Human-in-the-Loop Data Analytics</em>, 9. ACM.
</div>
<div id="ref-kwok2018lab" class="csl-entry">
Kwok, Roberta. 2018. <span>“Lab Notebooks Go Digital.”</span> <em>Nature</em> 560 (7717): 269–70.
</div>
<div id="ref-lawrence2014scalable" class="csl-entry">
Lawrence, Michael, and Martin Morgan. 2014. <span>“Scalable Genomics with r and Bioconductor.”</span> <em>Statistical Science: A Review Journal of the Institute of Mathematical Statistics</em> 29 (2): 214.
</div>
<div id="ref-leek2017toward" class="csl-entry">
Leek, Jeff. 2012. <span>“Toward Tidy Data Analysis.”</span> <em>Simply Statistics</em>.
</div>
<div id="ref-leips2010helm" class="csl-entry">
LEIPS, JEFF. 2010. <span>“At the Helm: Leading Your Laboratory.”</span> <em>Genetics Research</em> 92 (4): 325–26.
</div>
<div id="ref-levy1984spreadsheet" class="csl-entry">
Levy, Steven. 1984. <span>“A Spreadsheet Way of Knowledge.”</span> <em>Harpers</em> 269: 58–64.
</div>
<div id="ref-lowndes2017our" class="csl-entry">
Lowndes, Julia S Stewart, Benjamin D Best, Courtney Scarborough, Jamie C Afflerbach, Melanie R Frazier, Casey C O’Hara, Ning Jiang, and Benjamin S Halpern. 2017. <span>“Our Path to Better Science in Less Time Using Open Data Science Tools.”</span> <em>Nature Ecology &amp; Evolution</em> 1 (6): 160.
</div>
<div id="ref-lynch2008big" class="csl-entry">
Lynch, Clifford. 2008. <span>“Big Data: How Do Your Data Grow?”</span> <em>Nature</em> 455 (7209): 28.
</div>
<div id="ref-marwick2018packaging" class="csl-entry">
Marwick, Ben, Carl Boettiger, and Lincoln Mullen. 2018. <span>“Packaging Data Analytical Work Reproducibly Using <span>R</span> (and Friends).”</span> <em>The American Statistician</em> 72 (1): 80–88.
</div>
<div id="ref-mascarelli2014research" class="csl-entry">
Mascarelli, Amanda. 2014. <span>“Research Tools: Jump Off the Page.”</span> <em>Nature</em> 507 (7493): 523–25.
</div>
<div id="ref-mccullough2002accuracy" class="csl-entry">
McCullough, B D, and Berry Wilson. 2002. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2000 and Excel XP.”</span> <em>Computational Statistics &amp; Data Analysis</em> 40 (4): 713–21.
</div>
<div id="ref-mccullough2001does" class="csl-entry">
McCullough, BD. 2001. <span>“Does Microsoft Fix Errors in Excel?”</span> In <em>Proceedings of the 2001 Joint Statistical Meetings</em>.
</div>
<div id="ref-mccullough1999assessing" class="csl-entry">
McCullough, Bruce D. 1999. <span>“Assessing the Reliability of Statistical Software: Part II.”</span> <em>The American Statistician</em> 53 (2): 149–59.
</div>
<div id="ref-mccullough2008accuracy" class="csl-entry">
McCullough, Bruce D, and David A Heiser. 2008. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2007.”</span> <em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4570–78.
</div>
<div id="ref-mccullough1999accuracy" class="csl-entry">
McCullough, Bruce D, and Berry Wilson. 1999. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 97.”</span> <em>Computational Statistics &amp; Data Analysis</em> 31 (1): 27–37.
</div>
<div id="ref-mccullough2005accuracy" class="csl-entry">
———. 2005. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2003.”</span> <em>Computational Statistics &amp; Data Analysis</em> 49 (4): 1244–52.
</div>
<div id="ref-mcmurdie2013phyloseq" class="csl-entry">
McMurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An <span>R</span> Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> <em>PloS One</em> 8 (4): e61217.
</div>
<div id="ref-mcnamara2016state" class="csl-entry">
McNamara, Amelia. 2016. <span>“On the State of Computing in Statistics Education: Tools for Learning and for Doing.”</span> <em>arXiv Preprint arXiv:1610.00984</em>.
</div>
<div id="ref-metz2015github" class="csl-entry">
Metz, Cade. 2015. <span>“How GitHub Conquered Google, Microsoft, and Everyone Else.”</span> <em>Wired Magazine</em>.
</div>
<div id="ref-melard2014accuracy" class="csl-entry">
Mélard, Guy. 2014. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2010.”</span> <em>Computational Statistics</em> 29 (5): 1095–1128.
</div>
<div id="ref-michener2015ten" class="csl-entry">
Michener, William K. 2015. <span>“Ten Simple Rules for Creating a Good Data Management Plan.”</span> <em>PLoS Computational Biology</em> 11 (10): e1004525.
</div>
<div id="ref-murrell2009introduction" class="csl-entry">
Murrell, Paul. 2009. <em>Introduction to Data Technologies</em>. Chapman; Hall/CRC.
</div>
<div id="ref-myneni2010organization" class="csl-entry">
Myneni, Sahiti, and Vimla L Patel. 2010. <span>“Organization of Biomedical Data for Collaborative Scientific Research: A Research Information Management System.”</span> <em>International Journal of Information Management</em> 30 (3): 256–64.
</div>
<div id="ref-nardi1990spreadsheet" class="csl-entry">
Nardi, Bonnie A, and James R Miller. 1990. <span>“The Spreadsheet Interface: A Basis for End User Programming.”</span> In <em>Proceedings of the IFIP Tc13 Third Interational Conference on Human-Computer Interaction</em>, 977–83. North-Holland Publishing Co.
</div>
<div id="ref-nash2006spreadsheets" class="csl-entry">
Nash, JC. 2006. <span>“Spreadsheets in Statistical Practice—Another Look.”</span> <em>The American Statistician</em> 60 (3): 287–89.
</div>
<div id="ref-pawlik2017developing" class="csl-entry">
Pawlik, Aleksandra, Celia WG van Gelder, Aleksandra Nenadic, Patricia M Palagi, Eija Korpelainen, Philip Lijnzaad, Diana Marek, Susanna-Assunta Sansone, John Hancock, and Carole Goble. 2017. <span>“Developing a Strategy for Computational Lab Skills Training Through <span class="nocase">Software and Data Carpentry</span>: Experiences from the <span>ELIXIR Pilot</span> Action.”</span> <em>F1000Research</em> 6: ELIXIR–1040.
</div>
<div id="ref-peng2018teaching" class="csl-entry">
Peng, Roger. 2018. <span>“Teaching r to New Users—from Tapply to the Tidyverse.”</span> <em>Simply Statistics</em>.
</div>
<div id="ref-peng2016r" class="csl-entry">
Peng, Roger D. 2016. <em>R Programming for Data Science</em>. Leanpub.
</div>
<div id="ref-perez2016ten" class="csl-entry">
Perez-Riverol, Yasset, Laurent Gatto, Rui Wang, Timo Sachsenberg, Julian Uszkoreit, Felipe da Veiga Leprevost, Christian Fufezan, et al. 2016. <span>“Ten Simple Rules for Taking Advantage of Git and GitHub.”</span> <em>PLoS Computational Biology</em> 12 (7).
</div>
<div id="ref-perkel2018git" class="csl-entry">
Perkel, Jeffrey. 2018. <span>“Git: The Reproducibility Tool Scientists Love to Hate.”</span> <em>Naturejobs Blog</em>.
</div>
<div id="ref-perkel2011coding" class="csl-entry">
Perkel, Jeffrey M. 2011. <span>“Coding Your Way Out of a Problem.”</span> Nature Publishing Group.
</div>
<div id="ref-powell2012lab" class="csl-entry">
Powell, Kendall. 2012. <span>“A Lab App for That.”</span> <em>Nature</em> 484 (7395): 553–55.
</div>
<div id="ref-powell2009errors" class="csl-entry">
Powell, Stephen G, Kenneth R Baker, and Barry Lawson. 2009. <span>“Errors in Operational Spreadsheets: A Review of the State of the Art.”</span> In <em>2009 42nd Hawaii International Conference on System Sciences</em>, 1–8. IEEE.
</div>
<div id="ref-raymondunderstanding" class="csl-entry">
Raymond, Eric. 2009. <span>“Understanding Version-Control Systems (DRAFT).”</span> <a href="http://www.catb.org/ esr/writings/version-control/version-control.html">http://www.catb.org/ esr/writings/version-control/version-control.html</a>.
</div>
<div id="ref-raymond2003art" class="csl-entry">
Raymond, Eric S. 2003. <em>The Art of Unix Programming</em>. Addison-Wesley Professional.
</div>
<div id="ref-robinson2014broom" class="csl-entry">
Robinson, David. 2014. <span>“Broom: An r Package for Converting Statistical Analysis Objects into Tidy Data Frames.”</span> <em>arXiv Preprint arXiv:1412.3565</em>.
</div>
<div id="ref-robinson2017teach" class="csl-entry">
———. 2017. <span>“Teach the Tidyverse to Beginners.”</span> <em>Variance Explained</em>.
</div>
<div id="ref-edgeR" class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div id="ref-ross2017declutter" class="csl-entry">
Ross, Zev, Hadley Wickham, and David Robinson. 2017. <span>“Declutter Your <span>R</span> Workflow with Tidy Tools.”</span> <em>PeerJ Preprints</em> 5: e3180v1.
</div>
<div id="ref-sansone2012toward" class="csl-entry">
Sansone, Susanna-Assunta, Philippe Rocca-Serra, Dawn Field, Eamonn Maguire, Chris Taylor, Oliver Hofmann, Hong Fang, et al. 2012. <span>“Toward Interoperable Bioscience Data.”</span> <em>Nature Genetics</em> 44 (2): 121.
</div>
<div id="ref-schadt2010computational" class="csl-entry">
Schadt, Eric E, Michael D Linderman, Jon Sorenson, Lawrence Lee, and Garry P Nolan. 2010. <span>“Computational Solutions to Large-Scale Data Management and Analysis.”</span> <em>Nature Reviews Genetics</em> 11 (9): 647.
</div>
<div id="ref-sedgwick2014unit" class="csl-entry">
Sedgwick, Philip. 2014. <span>“Unit of Observation and Unit of Analysis.”</span> <em>BMJ</em> 348: g3840.
</div>
<div id="ref-seroul2012beginner" class="csl-entry">
Seroul, Raymond. 2012. <em>A Beginner’s Book of TEX</em>. Springer Science &amp; Business Media.
</div>
<div id="ref-silge2016tidytext" class="csl-entry">
Silge, Julia, and David Robinson. 2016. <span>“Tidytext: Text Mining and Analysis Using Tidy Data Principles in <span>R</span>.”</span> <em>The Journal of Open Source Software</em> 1 (3).
</div>
<div id="ref-silge2017text" class="csl-entry">
———. 2017. <em>Text Mining with r: A Tidy Approach</em>. Sebastopol: O’Reilly Media.
</div>
<div id="ref-smith2013lc" class="csl-entry">
Smith, Colin A. 2013. <span>“LC/MS Preprocessing and Analysis with Xcms.”</span> <em>Documentation of Bioconductor Xcms Package</em>.
</div>
<div id="ref-spraul2012think" class="csl-entry">
Spraul, V Anton. 2012. <em>Think Like a Programmer: An Introduction to Creative Problem Solving</em>. no starch press.
</div>
<div id="ref-stander2017enthusing" class="csl-entry">
Stander, Julian, and Luciana Dalla Valle. 2017. <span>“On Enthusing Students about Big Data and Social Media Visualization and Analysis Using <span class="nocase">R, RStudio, and RMarkdown</span>.”</span> <em>Journal of Statistics Education</em> 25 (2): 60–67.
</div>
<div id="ref-target2018version" class="csl-entry">
Target, Sinclair. 2018. <span>“Version Control Before Git with CVS.”</span> <a href="https://twobithistory.org/2018/07/07/cvs.html">https://twobithistory.org/2018/07/07/cvs.html</a>.
</div>
<div id="ref-teixeira2016emergence" class="csl-entry">
Teixeira, Ricardo, and Vasco Amaral. 2016. <span>“On the Emergence of Patterns for Spreadsheets Data Arrangements.”</span> In <em>Federation of International Conferences on Software Technologies: Applications and Foundations</em>, 333–45. Springer.
</div>
<div id="ref-tippmannmy2014digital" class="csl-entry">
Tippmann, Sylvia. 2014. <span>“My Digital Toolbox: Nuclear Engineer Katy Huff on Version-Control Systems.”</span> <em>Nature News</em>.
</div>
<div id="ref-topaloglou2004biological" class="csl-entry">
Topaloglou, Thodoros, Susan B Davidson, HV Jagadish, Victor M Markowitz, Evan W Steeg, and Mike Tyers. 2004. <span>“Biological Data Management: Research, Practice and Opportunities.”</span> In <em>Proceedings of the Thirtieth International Conference on Very Large Data Bases-Volume 30</em>, 1233–36. VLDB Endowment.
</div>
<div id="ref-RJ-2017-023" class="csl-entry">
Tyner, Sam, François Briatte, and Heike Hofmann. 2017. <span>“<span class="nocase">Network Visualization with ggplot2</span>.”</span> <em><span>The R Journal</span></em> 9 (1): 27–59.
</div>
<div id="ref-nih2016strategic" class="csl-entry">
U.S. Department of Health and Human Services, National Institutes of Health. 2016. <span>“<span>NIH-Wide Strategic Plan, Fiscal Years 2016-2020: Turning Discovery Into Health</span>.”</span> <a href="https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf">https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf</a>.
</div>
<div id="ref-nih2018data" class="csl-entry">
———. 2018. <span>“<span class="nocase">NIH Strategic Plan for Data Science</span>.”</span> <a href="https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf">https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf</a>.
</div>
<div id="ref-waltemath2016modeling" class="csl-entry">
Waltemath, Dagmar, and Olaf Wolkenhauer. 2016. <span>“How Modeling Standards, Software, and Initiatives Support Reproducibility in Systems Biology and Systems Medicine.”</span> <em>IEEE Transactions on Biomedical Engineering</em> 63 (10): 1999–2006.
</div>
<div id="ref-welsh2017escape" class="csl-entry">
Welsh, Eric A, Paul A Stewart, Brent M Kuenzi, and James A Eschrich. 2017. <span>“Escape Excel: A Tool for Preventing Gene Symbol and Accession Conversion Errors.”</span> <em>PloS One</em> 12 (9): e0185207.
</div>
<div id="ref-wickham2014tidy" class="csl-entry">
Wickham, Hadley. 2014. <span>“Tidy Data.”</span> <em>Journal of Statistical Software</em> 59 (10): 1–23.
</div>
<div id="ref-wickham2016ggplot2" class="csl-entry">
———. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. 2nd ed. New York: Springer.
</div>
<div id="ref-wickham2017tidy" class="csl-entry">
———. 2017. <span>“The Tidy Tools Manifesto.”</span> <em>CRAN Vignette</em>.
</div>
<div id="ref-wickham2019advanced" class="csl-entry">
———. 2019. <em>Advanced r, Second Edition</em>. CRC press.
</div>
<div id="ref-wickham2016r" class="csl-entry">
Wickham, Hadley, and Garrett Grolemund. 2016. <em>R for Data Science: Import, Tidy, Transform, Visualize, and Model Data</em>. sebastopol: O’Reilly Media.
</div>
<div id="ref-wilkinson2016fair" class="csl-entry">
Wilkinson, Mark D, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. <span>“The FAIR Guiding Principles for Scientific Data Management and Stewardship.”</span> <em>Scientific Data</em> 3.
</div>
<div id="ref-willekens2013chronological" class="csl-entry">
Willekens, Frans. 2013. <span>“Chronological Objects in Demographic Research.”</span> <em>Demographic Research</em> 28: 649–80.
</div>
<div id="ref-wilson2014software" class="csl-entry">
Wilson, G. 2014. <span>“<span>Software Carpentry</span>: Lessons Learned.”</span> <em>F1000Research</em> 3: 62–62.
</div>
<div id="ref-yin2012ggbio" class="csl-entry">
Yin, Tengfei, Dianne Cook, and Michael Lawrence. 2012. <span>“Ggbio: An <span>R</span> Package for Extending the Grammar of Graphics for Genomic Data.”</span> <em>Genome Biology</em> 13 (8): R77.
</div>
<div id="ref-zeeberg2004mistaken" class="csl-entry">
Zeeberg, Barry R, Joseph Riss, David W Kane, Kimberly J Bussey, Edward Uchio, W Marston Linehan, J Carl Barrett, and John N Weinstein. 2004. <span>“Mistaken Identifiers: Gene Name Errors Can Be Introduced Inadvertently When Using <span>Excel</span> in Bioinformatics.”</span> <em>BMC Bioinformatics</em> 5 (1): 80.
</div>
<div id="ref-ziemann2016gene" class="csl-entry">
Ziemann, Mark, Yotam Eren, and Assam El-Osta. 2016. <span>“Gene Name Errors Are Widespread in the Scientific Literature.”</span> <em>Genome Biology</em> 17 (1): 177.
</div>
</div>
</div>
<p style="text-align: center;">
<a href="3-9-example-creating-a-reproducible-data-pre-processing-protocol.html"><button class="btn btn-default">Previous</button></a>
</p>
</div>
</div>



</body>
</html>
