<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
<meta property="og:type" content="book" />


<meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
<meta name="github-repo" content="rstudio/bookdown-demo" />

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />



<meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing.">

<title>Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>




<link rel="stylesheet" href="toc.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li class="has-sub"><a href="index.html#overview"><span class="toc-section-number">1</span> Overview</a><ul>
<li><a href="1-1-license.html#license"><span class="toc-section-number">1.1</span> License</a></li>
</ul></li>
<li class="has-sub"><a href="2-experimental-data-recording.html#experimental-data-recording"><span class="toc-section-number">2</span> Experimental Data Recording</a><ul>
<li class="has-sub"><a href="2-1-separating-data-recording-and-analysis.html#separating-data-recording-and-analysis"><span class="toc-section-number">2.1</span> Separating data recording and analysis</a><ul>
<li><a href="2-1-separating-data-recording-and-analysis.html#data-recording-versus-data-analysis"><span class="toc-section-number">2.1.1</span> Data recording versus data analysis</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#hazards-of-combining-recording-and-analysis"><span class="toc-section-number">2.1.2</span> Hazards of combining recording and analysis</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#approaches-to-separate-recording-and-analysis"><span class="toc-section-number">2.1.3</span> Approaches to separate recording and analysis</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">2.1.4</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="2-2-principles-and-power-of-structured-data-formats.html#principles-and-power-of-structured-data-formats"><span class="toc-section-number">2.2</span> Principles and power of structured data formats</a><ul>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#data-recording-standards"><span class="toc-section-number">2.2.1</span> Data recording standards</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#defining-data-standards-for-a-research-group"><span class="toc-section-number">2.2.2</span> Defining data standards for a research group</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#two-dimensional-structured-data-format"><span class="toc-section-number">2.2.3</span> Two-dimensional structured data format</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#saving-two-dimensional-structured-data-in-plain-text-file-formats"><span class="toc-section-number">2.2.4</span> Saving two-dimensional structured data in plain text file formats</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#occassions-for-more-complex-data-structures-and-file-formats"><span class="toc-section-number">2.2.5</span> Occassions for more complex data structures and file formats</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#levels-of-standardizationresearch-group-to-research-community"><span class="toc-section-number">2.2.6</span> Levels of standardization—research group to research community</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">2.2.7</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="2-3-the-tidy-data-format.html#the-tidy-data-format"><span class="toc-section-number">2.3</span> The ‘tidy’ data format</a><ul>
<li><a href="2-3-the-tidy-data-format.html#what-makes-data-tidy"><span class="toc-section-number">2.3.1</span> What makes data “tidy”?</a></li>
<li><a href="2-3-the-tidy-data-format.html#why-make-your-data-tidy"><span class="toc-section-number">2.3.2</span> Why make your data “tidy”?</a></li>
<li><a href="2-3-the-tidy-data-format.html#using-tidyverse-tools-with-data-in-the-tidy-data-format"><span class="toc-section-number">2.3.3</span> Using tidyverse tools with data in the “tidy data” format</a></li>
<li><a href="2-3-the-tidy-data-format.html#practice-quiz"><span class="toc-section-number">2.3.4</span> Practice quiz</a></li>
</ul></li>
<li class="has-sub"><a href="2-4-designing-templates-for-tidy-data-collection.html#designing-templates-for-tidy-data-collection"><span class="toc-section-number">2.4</span> Designing templates for “tidy” data collection</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#creating-the-rules-for-collecting-data-in-the-same-time-each-time"><span class="toc-section-number">2.4.1</span> Creating the rules for collecting data in the same time each time</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">2.4.2</span> Subsection 1</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#dont-repeat-yourself"><span class="toc-section-number">2.4.3</span> Don’t Repeat Yourself!</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#dont-repeat-your-report-writing"><span class="toc-section-number">2.4.4</span> Don’t repeat your report-writing!</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#automating-reports"><span class="toc-section-number">2.4.5</span> Automating reports</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#scripts-and-automated-reports-as-simple-pipelines"><span class="toc-section-number">2.4.6</span> Scripts and automated reports as simple pipelines</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">2.4.7</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#example-creating-a-template-for-tidy-data-collection"><span class="toc-section-number">2.5</span> Example: Creating a template for “tidy” data collection</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">2.5.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">2.5.2</span> Subsection 2</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#example-datasets"><span class="toc-section-number">2.5.3</span> Example datasets</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#issues-with-these-data-sets"><span class="toc-section-number">2.5.4</span> Issues with these data sets</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#final-tidy-examples"><span class="toc-section-number">2.5.5</span> Final “tidy” examples</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#options-for-recording-tidy-data"><span class="toc-section-number">2.5.6</span> Options for recording tidy data</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#examples-of-how-tidy-data-can-be-easily-analyzed-visualized"><span class="toc-section-number">2.5.7</span> Examples of how “tidy” data can be easily analyzed / visualized</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">2.5.8</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><span class="toc-section-number">2.6</span> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a><ul>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#organizing-project-files-through-the-file-system"><span class="toc-section-number">2.6.1</span> Organizing project files through the file system</a></li>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#organizing-files-within-a-project-directory"><span class="toc-section-number">2.6.2</span> Organizing files within a project directory</a></li>
<li><a href="2-6-power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files.html#using-rstudio-projects-with-project-file-directories"><span class="toc-section-number">2.6.3</span> Using RStudio Projects with project file directories</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">2.6.4</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">2.6.5</span> Subsection 2</a></li>
<li><a href="2-3-the-tidy-data-format.html#practice-quiz"><span class="toc-section-number">2.6.6</span> Practice quiz</a></li>
</ul></li>
<li class="has-sub"><a href="2-7-creating-project-templates.html#creating-project-templates"><span class="toc-section-number">2.7</span> Creating ‘Project’ templates</a><ul>
<li><a href="2-7-creating-project-templates.html#making-an-existing-file-directory-an-rstudio-project"><span class="toc-section-number">2.7.1</span> Making an existing file directory an RStudio Project</a></li>
<li><a href="2-7-creating-project-templates.html#making-an-rstudio-project-template"><span class="toc-section-number">2.7.2</span> Making an RStudio Project Template</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">2.7.3</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="2-8-example-creating-a-project-template.html#example-creating-a-project-template"><span class="toc-section-number">2.8</span> Example: Creating a ‘Project’ template</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">2.8.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">2.8.2</span> Subsection 2</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">2.8.3</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#harnessing-version-control-for-transparent-data-recording"><span class="toc-section-number">2.9</span> Harnessing version control for transparent data recording</a><ul>
<li><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#what-is-version-control"><span class="toc-section-number">2.9.1</span> What is version control?</a></li>
<li><a href="2-9-harnessing-version-control-for-transparent-data-recording.html#recording-data-in-the-laboratoryfrom-paper-to-computers"><span class="toc-section-number">2.9.2</span> Recording data in the laboratory—from paper to computers</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">2.9.3</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="2-10-enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><span class="toc-section-number">2.10</span> Enhance the reproducibility of collaborative research with version control platforms</a><ul>
<li><a href="2-10-enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms.html#git-and-github-features"><span class="toc-section-number">2.10.1</span> git and GitHub features</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">2.10.2</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">2.10.3</span> Subsection 2</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">2.10.4</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#using-git-and-gitlab-to-implement-version-control"><span class="toc-section-number">2.11</span> Using git and GitLab to implement version control</a><ul>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#how-to-use-version-control"><span class="toc-section-number">2.11.1</span> How to use version control</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#leveraging-git-and-github-as-a-project-director"><span class="toc-section-number">2.11.2</span> Leveraging git and GitHub as a project director</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#leveraging-git-and-github-as-a-scientist-who-programs"><span class="toc-section-number">2.11.3</span> Leveraging git and GitHub as a scientist who programs</a></li>
<li><a href="2-11-using-git-and-gitlab-to-implement-version-control.html#notes"><span class="toc-section-number">2.11.4</span> Notes</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">2.11.5</span> Applied exercise</a></li>
</ul></li>
</ul></li>
<li class="has-sub"><a href="3-experimental-data-preprocessing.html#experimental-data-preprocessing"><span class="toc-section-number">3</span> Experimental Data Preprocessing</a><ul>
<li class="has-sub"><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><span class="toc-section-number">3.1</span> Principles and benefits of scripted pre-processing of experimental data</a><ul>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#what-is-pre-processing"><span class="toc-section-number">3.1.1</span> What is pre-processing?</a></li>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#approaches-to-simple-preprocessing-tasks"><span class="toc-section-number">3.1.2</span> Approaches to simple preprocessing tasks</a></li>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#approaches-to-more-complex-preprocessing-tasks"><span class="toc-section-number">3.1.3</span> Approaches to more complex preprocessing tasks</a></li>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#scripting-preprocessing-tasks"><span class="toc-section-number">3.1.4</span> Scripting preprocessing tasks</a></li>
<li><a href="3-1-principles-and-benefits-of-scripted-pre-processing-of-experimental-data.html#potential-quotes"><span class="toc-section-number">3.1.5</span> Potential quotes</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">3.1.6</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="3-2-introduction-to-scripted-data-pre-processing-in-r.html#introduction-to-scripted-data-pre-processing-in-r"><span class="toc-section-number">3.2</span> Introduction to scripted data pre-processing in R</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.2.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.2.2</span> Subsection 2</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">3.2.3</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><span class="toc-section-number">3.3</span> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a><ul>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#limitations-of-object-oriented-programming"><span class="toc-section-number">3.3.1</span> Limitations of object-oriented programming</a></li>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#the-tidyverse-approach"><span class="toc-section-number">3.3.2</span> The “tidyverse” approach</a></li>
<li><a href="3-3-simplify-scripted-pre-processing-through-rs-tidyverse-tools.html#how-to-tidyverse"><span class="toc-section-number">3.3.3</span> How to “tidyverse”</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.3.4</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.3.5</span> Subsection 2</a></li>
<li><a href="2-3-the-tidy-data-format.html#practice-quiz"><span class="toc-section-number">3.3.6</span> Practice quiz</a></li>
</ul></li>
<li class="has-sub"><a href="3-4-complex-data-types-in-experimental-data-pre-processing.html#complex-data-types-in-experimental-data-pre-processing"><span class="toc-section-number">3.4</span> Complex data types in experimental data pre-processing</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.4.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.4.2</span> Subsection 2</a></li>
<li><a href="3-4-complex-data-types-in-experimental-data-pre-processing.html#subsection-3"><span class="toc-section-number">3.4.3</span> Subsection 3</a></li>
<li><a href="2-3-the-tidy-data-format.html#practice-quiz"><span class="toc-section-number">3.4.4</span> Practice quiz</a></li>
</ul></li>
<li class="has-sub"><a href="3-5-complex-data-types-in-r-and-bioconductor.html#complex-data-types-in-r-and-bioconductor"><span class="toc-section-number">3.5</span> Complex data types in R and Bioconductor</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.5.1</span> Subsection 1</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#exploring-and-extracting-data-from-r-list-data-structures"><span class="toc-section-number">3.5.2</span> Exploring and extracting data from R list data structures</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#interfacing-between-object-based-and-tidyverse-workflows"><span class="toc-section-number">3.5.3</span> Interfacing between object-based and tidyverse workflows</a></li>
<li><a href="3-5-complex-data-types-in-r-and-bioconductor.html#extras"><span class="toc-section-number">3.5.4</span> Extras</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.5.5</span> Subsection 2</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">3.5.6</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="3-6-example-converting-from-complex-to-tidy-data-formats.html#example-converting-from-complex-to-tidy-data-formats"><span class="toc-section-number">3.6</span> Example: Converting from complex to ‘tidy’ data formats</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.6.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.6.2</span> Subsection 2</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">3.6.3</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#introduction-to-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.7</span> Introduction to reproducible data pre-processing protocols</a><ul>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#introducing-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.7.1</span> Introducing reproducible data pre-processing protocols</a></li>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#technique-to-create-reproducible-pre-processing-protocols"><span class="toc-section-number">3.7.2</span> Technique to create reproducible pre-processing protocols</a></li>
<li><a href="3-7-introduction-to-reproducible-data-pre-processing-protocols.html#advantages-of-reproducible-pre-processing-protocols"><span class="toc-section-number">3.7.3</span> Advantages of reproducible pre-processing protocols</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.7.4</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.7.5</span> Subsection 2</a></li>
<li><a href="2-1-separating-data-recording-and-analysis.html#discussion-questions"><span class="toc-section-number">3.7.6</span> Discussion questions</a></li>
</ul></li>
<li class="has-sub"><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><span class="toc-section-number">3.8</span> RMarkdown for creating reproducible data pre-processing protocols</a><ul>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#rmarkdown"><span class="toc-section-number">3.8.1</span> RMarkdown</a></li>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#creating-and-rendering-rmarkdown-documents"><span class="toc-section-number">3.8.2</span> Creating and rendering RMarkdown documents</a></li>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#some-helpful-additional-features-in-rmarkdown"><span class="toc-section-number">3.8.3</span> Some helpful additional features in RMarkdown</a></li>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#some-helpful-additional-features-in-rmarkdown-1"><span class="toc-section-number">3.8.4</span> Some helpful additional features in RMarkdown</a></li>
<li><a href="3-8-rmarkdown-for-creating-reproducible-data-pre-processing-protocols.html#using-rmarkdown-to-improve-reproducibility-in-data-pre-processing"><span class="toc-section-number">3.8.5</span> Using RMarkdown to improve reproducibility in data pre-processing</a></li>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.8.6</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.8.7</span> Subsection 2</a></li>
<li><a href="2-2-principles-and-power-of-structured-data-formats.html#applied-exercise"><span class="toc-section-number">3.8.8</span> Applied exercise</a></li>
</ul></li>
<li class="has-sub"><a href="3-9-example-creating-a-reproducible-data-pre-processing-protocol.html#example-creating-a-reproducible-data-pre-processing-protocol"><span class="toc-section-number">3.9</span> Example: Creating a reproducible data pre-processing protocol</a><ul>
<li><a href="2-4-designing-templates-for-tidy-data-collection.html#subsection-1"><span class="toc-section-number">3.9.1</span> Subsection 1</a></li>
<li><a href="2-5-example-creating-a-template-for-tidy-data-collection.html#subsection-2"><span class="toc-section-number">3.9.2</span> Subsection 2</a></li>
<li><a href="2-3-the-tidy-data-format.html#practice-quiz"><span class="toc-section-number">3.9.3</span> Practice quiz</a></li>
</ul></li>
</ul></li>
<li><a href="4-references.html#references"><span class="toc-section-number">4</span> References</a></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="references" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> References</h1>

<div id="refs" class="references">
<div>
<p>AlTarawneh, Ghada, and Simon Thorne. 2017. “A Pilot Study Exploring Spreadsheet Risk in Scientific Research.” <em>arXiv Preprint arXiv:1703.09785</em>.</p>
</div>
<div>
<p>Altman, Douglas G., and J. Martin Bland. 1997. “Statistical Notes: Units of Analysis.” <em>BMJ</em> 314: 1874.</p>
</div>
<div>
<p>Anderson, Nicholas R, E Sally Lee, J Scott Brockenbrough, Mark E Minie, Sherrilynne Fuller, James Brinkley, and Peter Tarczy-Hornoch. 2007. “Issues in Biomedical Research Data Management and Analysis: Needs and Barriers.” <em>Journal of the American Medical Informatics Association</em> 14 (4): 478–88.</p>
</div>
<div>
<p>Apte, Poornima. n.d. “The Lingua Franca of Latex.” <em>Increment</em>. <a href="https://increment.com/open-source/the-lingua-franca-of-latex/">https://increment.com/open-source/the-lingua-franca-of-latex/</a>.</p>
</div>
<div>
<p>Arnold, Taylor. 2017. “A Tidy Data Model for Natural Language Processing using cleanNLP.” <em>The R Journal</em> 9 (2): 248–67.</p>
</div>
<div>
<p>Barga, Roger, Bill Howe, David Beck, Stuart Bowers, William Dobyns, Winston Haynes, Roger Higdon, et al. 2011. “Bioinformatics and Data-Intensive Scientific Discovery in the Beginning of the 21st Century.” <em>Omics: A Journal of Integrative Biology</em> 15 (4): 199–201.</p>
</div>
<div>
<p>Baumer, Ben. 2015. “A Data Science Course for Undergraduates: Thinking with Data.” <em>The American Statistician</em> 69 (4): 334–42.</p>
</div>
<div>
<p>Baumer, Benjamin S, Daniel T Kaplan, and Nicholas J Horton. 2017. <em>Modern Data Science with R</em>. Boca Raton: CRC Press.</p>
</div>
<div>
<p>Birch, David, David Lyford-Smith, and Yike Guo. 2018. “The Future of Spreadsheets in the Big Data Era.” <em>arXiv Preprint arXiv:1801.10231</em>.</p>
</div>
<div>
<p>Blischak, John D, Emily R Davenport, and Greg Wilson. 2016. “A Quick Introduction to Version Control with Git and Github.” <em>PLoS Computational Biology</em> 12 (1).</p>
</div>
<div>
<p>Brazma, Alvis, Maria Krestyaninova, and Ugis Sarkans. 2006. “Standards for Systems Biology.” <em>Nature Reviews Genetics</em> 7 (8): 593.</p>
</div>
<div>
<p>Broman, Karl W, and Kara H Woo. 2018. “Data Organization in Spreadsheets.” <em>The American Statistician</em> 72 (1): 2–10.</p>
</div>
<div>
<p>Brown, Zack. 2018. “A Git Origin Story.” <em>Linux Journal</em>.</p>
</div>
<div>
<p>Bryan, Jennifer. 2018. “Excuse Me, Do You Have a Moment to Talk About Version Control?” <em>The American Statistician</em> 72 (1): 20–27.</p>
</div>
<div>
<p>Bryan, Jennifer, and Hadley Wickham. 2017. “Data Science: A Three Ring Circus or a Big Tent?” <em>Journal of Computational and Graphical Statistics</em> 26 (4): 784–85.</p>
</div>
<div>
<p>Buffalo, Vince. 2015. <em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>. " O’Reilly Media, Inc.".</p>
</div>
<div>
<p>Butler, Declan. 2005. “Electronic Notebooks: A New Leaf.” <em>Nature</em> 436 (7047): 20–22.</p>
</div>
<div>
<p>Campbell-Kelly, Martin. 2007. “Number Crunching Without Programming: The Evolution of Spreadsheet Usability.” <em>IEEE Annals of the History of Computing</em> 29 (3): 6–19.</p>
</div>
<div>
<p>Creeth, Richard. 1985. “Microcomputer Spreadsheets: Their Uses and Abuses.” <em>Journal of Accountancy (Pre-1986)</em> 159 (000006): 90.</p>
</div>
<div>
<p>Edwards, Paul N, Matthew S Mayernik, Archer L Batcheller, Geoffrey C Bowker, and Christine L Borgman. 2011. “Science Friction: Data, Metadata, and Collaboration.” <em>Social Studies of Science</em> 41 (5): 667–90.</p>
</div>
<div>
<p>Ellis, Shannon E, and Jeffrey T Leek. 2018. “How to Share Data for Collaboration.” <em>The American Statistician</em> 72 (1): 53–57.</p>
</div>
<div>
<p>Ford, Paul. 2014. “On File Formats, Very Briefly.” <em>The Manual</em>. <a href="https://themanual.org/read/issues/4/paul-ford/article">https://themanual.org/read/issues/4/paul-ford/article</a>.</p>
</div>
<div>
<p>———. 2015. “I Dreamed of a Perfect Database.” <em>New Republic</em>. <a href="https://newrepublic.com/article/124425/dreamed-perfect-database">https://newrepublic.com/article/124425/dreamed-perfect-database</a>.</p>
</div>
<div>
<p>Gentleman, Robert. 2008. <em>R Programming for Bioinformatics</em>. CRC Press.</p>
</div>
<div>
<p>Ghosh, Samik, Yukiko Matsuoka, Yoshiyuki Asai, Kun-Yi Hsin, and Hiroaki Kitano. 2011. “Software for Systems Biology: From Tools to Integrated Platforms.” <em>Nature Reviews Genetics</em> 12 (12): 821.</p>
</div>
<div>
<p>Giles, Jim. 2012. “The Digital Lab: Lab-Management Software and Electronic Notebooks Are Here–and This Time, It’s More Than Just Talk.” <em>Nature</em> 481 (7382): 430–32.</p>
</div>
<div>
<p>Goodman, Alyssa, Alberto Pepe, Alexander W Blocker, Christine L Borgman, Kyle Cranmer, Merce Crosas, Rosanne Di Stefano, et al. 2014. “Ten Simple Rules for the Care and Feeding of Scientific Data.” Public Library of Science.</p>
</div>
<div>
<p>Grune, Dick. 1986. “Concurrent Verions System, a Method for Independent Cooperation.” <em>Technical Report, Vrije Universiteit</em>.</p>
</div>
<div>
<p>Hermans, Felienne, Bas Jansen, Sohon Roy, Efthimia Aivaloglou, Alaaeddin Swidan, and David Hoepelman. 2016. “Spreadsheets Are Code: An Overview of Software Engineering Approaches Applied to Spreadsheets.” In <em>2016 Ieee 23rd International Conference on Software Analysis, Evolution, and Reengineering (Saner)</em>, 5:56–65. IEEE.</p>
</div>
<div>
<p>Hermans, Felienne, and Emerson Murphy-Hill. 2015. “Enron’s Spreadsheets and Related Emails: A Dataset and Analysis.” In <em>2015 Ieee/Acm 37th Ieee International Conference on Software Engineering</em>, 2:7–16. IEEE.</p>
</div>
<div>
<p>Hicks, Stephanie C, and Rafael A Irizarry. 2017. “A Guide to Teaching Data Science.” <em>The American Statistician</em> In Press. <a href="https://doi.org/10.1080/00031305.2017.1356747">https://doi.org/10.1080/00031305.2017.1356747</a>.</p>
</div>
<div>
<p>Hsieh, TC, KH Ma, and Anne Chao. 2016. “iNEXT: an R package for rarefaction and extrapolation of species diversity (Hill numbers).” <em>Methods in Ecology and Evolution</em> 7 (12): 1451–6.</p>
</div>
<div>
<p>Hunt, Andrew, David Thomas, and Ward Cunningham. 2000. <em>The Pragmatic Programmer: From Journeyman to Master</em>. Addison-Wesley Professional.</p>
</div>
<div>
<p>Irizarry, Rafael A., and Michael I. Love. 2016. <em>Data Analysis for the Life Sciences with R</em>. Chapman; Hall.</p>
</div>
<div>
<p>Irving, Francis. 2011. “Astonishments, Ten, in the History of Version Control.” <a href="https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/">https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/</a>.</p>
</div>
<div>
<p>Janssens, Jeroen. 2014. <em>Data Science at the Command Line: Facing the Future with Time-Tested Tools</em>. " O’Reilly Media, Inc.".</p>
</div>
<div>
<p>Kaplan, Daniel. 2018. “Teaching Stats for Data Science.” <em>The American Statistician</em> 72 (1): 89–96.</p>
</div>
<div>
<p>Keller, Sallie, Gizem Korkmaz, Mark Orr, Aaron Schroeder, and Stephanie Shipp. 2017. “The Evolution of Data Quality: Understanding the Transdisciplinary Origins of Data Quality Concepts and Approaches.” <em>Annu. Rev. Stat. Appl</em> 4: 85–108.</p>
</div>
<div>
<p>Kernighan, Brian W. 2011. <em>D Is for Digital: What a Well-Informed Person Should Know About Computers and Communications</em>. CreateSpace Independent Publishing Platform.</p>
</div>
<div>
<p>Kernighan, Brian W, and Rob Pike. 1984. <em>The Unix Programming Environment</em>. Vol. 270. Prentice-Hall Englewood Cliffs, NJ.</p>
</div>
<div>
<p>Krishnan, Sanjay, Daniel Haas, Michael J Franklin, and Eugene Wu. 2016. “Towards Reliable Interactive Data Cleaning: A User Survey and Recommendations.” In <em>Proceedings of the Workshop on Human-in-the-Loop Data Analytics</em>, 9. ACM.</p>
</div>
<div>
<p>Kwok, Roberta. 2018. “Lab Notebooks Go Digital.” <em>Nature</em> 560 (7717): 269–70.</p>
</div>
<div>
<p>Leek, Jeff. 2012. “Toward Tidy Data Analysis.” <em>Simply Statistics</em>.</p>
</div>
<div>
<p>Levy, Steven. 1984. “A Spreadsheet Way of Knowledge.” <em>Harpers</em> 269: 58–64.</p>
</div>
<div>
<p>Lowndes, Julia S Stewart, Benjamin D Best, Courtney Scarborough, Jamie C Afflerbach, Melanie R Frazier, Casey C O’Hara, Ning Jiang, and Benjamin S Halpern. 2017. “Our Path to Better Science in Less Time Using Open Data Science Tools.” <em>Nature Ecology &amp; Evolution</em> 1 (6): 160.</p>
</div>
<div>
<p>Lynch, Clifford. 2008. “Big Data: How Do Your Data Grow?” <em>Nature</em> 455 (7209): 28.</p>
</div>
<div>
<p>Marwick, Ben, Carl Boettiger, and Lincoln Mullen. 2018. “Packaging Data Analytical Work Reproducibly Using R (and Friends).” <em>The American Statistician</em> 72 (1): 80–88.</p>
</div>
<div>
<p>Mascarelli, Amanda. 2014. “Research Tools: Jump Off the Page.” <em>Nature</em> 507 (7493): 523–25.</p>
</div>
<div>
<p>McCullough, BD. 2001. “Does Microsoft Fix Errors in Excel?” In <em>Proceedings of the 2001 Joint Statistical Meetings</em>.</p>
</div>
<div>
<p>McCullough, B D, and Berry Wilson. 2002. “On the Accuracy of Statistical Procedures in Microsoft Excel 2000 and Excel Xp.” <em>Computational Statistics &amp; Data Analysis</em> 40 (4): 713–21.</p>
</div>
<div>
<p>McCullough, Bruce D. 1999. “Assessing the Reliability of Statistical Software: Part Ii.” <em>The American Statistician</em> 53 (2): 149–59.</p>
</div>
<div>
<p>McCullough, Bruce D, and David A Heiser. 2008. “On the Accuracy of Statistical Procedures in Microsoft Excel 2007.” <em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4570–8.</p>
</div>
<div>
<p>McCullough, Bruce D, and Berry Wilson. 1999. “On the Accuracy of Statistical Procedures in Microsoft Excel 97.” <em>Computational Statistics &amp; Data Analysis</em> 31 (1): 27–37.</p>
</div>
<div>
<p>———. 2005. “On the Accuracy of Statistical Procedures in Microsoft Excel 2003.” <em>Computational Statistics &amp; Data Analysis</em> 49 (4): 1244–52.</p>
</div>
<div>
<p>McMurdie, Paul J, and Susan Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PloS One</em> 8 (4): e61217.</p>
</div>
<div>
<p>McNamara, Amelia. 2016. “On the State of Computing in Statistics Education: Tools for Learning and for Doing.” <em>arXiv Preprint arXiv:1610.00984</em>.</p>
</div>
<div>
<p>Metz, Cade. 2015. “How Github Conquered Google, Microsoft, and Everyone Else.” <em>Wired Magazine</em>.</p>
</div>
<div>
<p>Mélard, Guy. 2014. “On the Accuracy of Statistical Procedures in Microsoft Excel 2010.” <em>Computational Statistics</em> 29 (5): 1095–1128.</p>
</div>
<div>
<p>Michener, William K. 2015. “Ten Simple Rules for Creating a Good Data Management Plan.” <em>PLoS Computational Biology</em> 11 (10): e1004525.</p>
</div>
<div>
<p>Murrell, Paul. 2009. <em>Introduction to Data Technologies</em>. Chapman; Hall/CRC.</p>
</div>
<div>
<p>Myneni, Sahiti, and Vimla L Patel. 2010. “Organization of Biomedical Data for Collaborative Scientific Research: A Research Information Management System.” <em>International Journal of Information Management</em> 30 (3): 256–64.</p>
</div>
<div>
<p>Nardi, Bonnie A, and James R Miller. 1990. “The Spreadsheet Interface: A Basis for End User Programming.” In <em>Proceedings of the Ifip Tc13 Third Interational Conference on Human-Computer Interaction</em>, 977–83. North-Holland Publishing Co.</p>
</div>
<div>
<p>Nash, JC. 2006. “Spreadsheets in Statistical Practice—Another Look.” <em>The American Statistician</em> 60 (3): 287–89.</p>
</div>
<div>
<p>Pawlik, Aleksandra, Celia WG van Gelder, Aleksandra Nenadic, Patricia M Palagi, Eija Korpelainen, Philip Lijnzaad, Diana Marek, Susanna-Assunta Sansone, John Hancock, and Carole Goble. 2017. “Developing a Strategy for Computational Lab Skills Training Through Software and Data Carpentry: Experiences from the ELIXIR Pilot Action.” <em>F1000Research</em> 6: ELIXIR–1040.</p>
</div>
<div>
<p>Peng, Roger. 2018. “Teaching R to New Users—from Tapply to the Tidyverse.” <em>Simply Statistics</em>.</p>
</div>
<div>
<p>Perez-Riverol, Yasset, Laurent Gatto, Rui Wang, Timo Sachsenberg, Julian Uszkoreit, Felipe da Veiga Leprevost, Christian Fufezan, et al. 2016. “Ten Simple Rules for Taking Advantage of Git and Github.” <em>PLoS Computational Biology</em> 12 (7).</p>
</div>
<div>
<p>Perkel, Jeffrey. 2018. “Git: The Reproducibility Tool Scientists Love to Hate.” <em>Naturejobs Blog</em>.</p>
</div>
<div>
<p>Perkel, Jeffrey M. 2011. “Coding Your Way Out of a Problem.” Nature Publishing Group.</p>
</div>
<div>
<p>Powell, Kendall. 2012. “A Lab App for That.” <em>Nature</em> 484 (7395): 553–55.</p>
</div>
<div>
<p>Powell, Stephen G, Kenneth R Baker, and Barry Lawson. 2009. “Errors in Operational Spreadsheets: A Review of the State of the Art.” In <em>2009 42nd Hawaii International Conference on System Sciences</em>, 1–8. IEEE.</p>
</div>
<div>
<p>Raymond, Eric. 2009. “Understanding Version-Control Systems (Draft).” <a href="http://www.catb.org/~esr/writings/version-control/version-control.html">http://www.catb.org/~esr/writings/version-control/version-control.html</a>.</p>
</div>
<div>
<p>Raymond, Eric S. 2003. <em>The Art of Unix Programming</em>. Addison-Wesley Professional.</p>
</div>
<div>
<p>Robinson, David. 2017. “Teach the Tidyverse to Beginners.” <em>Variance Explained</em>.</p>
</div>
<div>
<p>Ross, Zev, Hadley Wickham, and David Robinson. 2017. “Declutter Your R Workflow with Tidy Tools.” <em>PeerJ Preprints</em> 5: e3180v1.</p>
</div>
<div>
<p>Sansone, Susanna-Assunta, Philippe Rocca-Serra, Dawn Field, Eamonn Maguire, Chris Taylor, Oliver Hofmann, Hong Fang, et al. 2012. “Toward Interoperable Bioscience Data.” <em>Nature Genetics</em> 44 (2): 121.</p>
</div>
<div>
<p>Schadt, Eric E, Michael D Linderman, Jon Sorenson, Lawrence Lee, and Garry P Nolan. 2010. “Computational Solutions to Large-Scale Data Management and Analysis.” <em>Nature Reviews Genetics</em> 11 (9): 647.</p>
</div>
<div>
<p>Sedgwick, Philip. 2014. “Unit of Observation and Unit of Analysis.” <em>BMJ</em> 348: g3840.</p>
</div>
<div>
<p>Seroul, Raymond. 2012. <em>A Beginner’s Book of Tex</em>. Springer Science &amp; Business Media.</p>
</div>
<div>
<p>Silge, Julia, and David Robinson. 2016. “Tidytext: Text Mining and Analysis Using Tidy Data Principles in R.” <em>The Journal of Open Source Software</em> 1 (3).</p>
</div>
<div>
<p>———. 2017. <em>Text Mining with R: A Tidy Approach</em>. Sebastopol: O’Reilly Media.</p>
</div>
<div>
<p>Spraul, V Anton. 2012. <em>Think Like a Programmer: An Introduction to Creative Problem Solving</em>. no starch press.</p>
</div>
<div>
<p>Stander, Julian, and Luciana Dalla Valle. 2017. “On Enthusing Students About Big Data and Social Media Visualization and Analysis Using R, RStudio, and RMarkdown.” <em>Journal of Statistics Education</em> 25 (2): 60–67.</p>
</div>
<div>
<p>Target, Sinclair. 2018. “Version Control Before Git with Cvs.” <a href="https://twobithistory.org/2018/07/07/cvs.html">https://twobithistory.org/2018/07/07/cvs.html</a>.</p>
</div>
<div>
<p>Teixeira, Ricardo, and Vasco Amaral. 2016. “On the Emergence of Patterns for Spreadsheets Data Arrangements.” In <em>Federation of International Conferences on Software Technologies: Applications and Foundations</em>, 333–45. Springer.</p>
</div>
<div>
<p>Tippmann, Sylvia. 2014. “My Digital Toolbox: Nuclear Engineer Katy Huff on Version-Control Systems.” <em>Nature News</em>.</p>
</div>
<div>
<p>Topaloglou, Thodoros, Susan B Davidson, HV Jagadish, Victor M Markowitz, Evan W Steeg, and Mike Tyers. 2004. “Biological Data Management: Research, Practice and Opportunities.” In <em>Proceedings of the Thirtieth International Conference on Very Large Data Bases-Volume 30</em>, 1233–6. VLDB Endowment.</p>
</div>
<div>
<p>Tyner, Sam, François Briatte, and Heike Hofmann. 2017. “Network Visualization with ggplot2.” <em>The R Journal</em> 9 (1): 27–59.</p>
</div>
<div>
<p>U.S. Department of Health and Human Services, National Institutes of Health. 2016. “NIH-Wide Strategic Plan, Fiscal Years 2016-2020: Turning Discovery into Health.” <a href="https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf">https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf</a>.</p>
</div>
<div>
<p>———. 2018. “NIH Strategic Plan for Data Science.” <a href="https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf">https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf</a>.</p>
</div>
<div>
<p>Waltemath, Dagmar, and Olaf Wolkenhauer. 2016. “How Modeling Standards, Software, and Initiatives Support Reproducibility in Systems Biology and Systems Medicine.” <em>IEEE Transactions on Biomedical Engineering</em> 63 (10): 1999–2006.</p>
</div>
<div>
<p>Welsh, Eric A, Paul A Stewart, Brent M Kuenzi, and James A Eschrich. 2017. “Escape Excel: A Tool for Preventing Gene Symbol and Accession Conversion Errors.” <em>PloS One</em> 12 (9): e0185207.</p>
</div>
<div>
<p>Wickham, Hadley. 2014. “Tidy Data.” <em>Journal of Statistical Software</em> 59 (10): 1–23.</p>
</div>
<div>
<p>———. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. 2nd ed. New York: Springer.</p>
</div>
<div>
<p>———. 2017. “The Tidy Tools Manifesto.” <em>CRAN Vignette</em>.</p>
</div>
<div>
<p>———. 2019. <em>Advanced R, Second Edition</em>. CRC press.</p>
</div>
<div>
<p>Wickham, Hadley, and Garrett Grolemund. 2016. <em>R for Data Science: Import, Tidy, Transform, Visualize, and Model Data</em>. sebastopol: O’Reilly Media.</p>
</div>
<div>
<p>Wilkinson, Mark D, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. “The Fair Guiding Principles for Scientific Data Management and Stewardship.” <em>Scientific Data</em> 3.</p>
</div>
<div>
<p>Willekens, Frans. 2013. “Chronological Objects in Demographic Research.” <em>Demographic Research</em> 28: 649–80.</p>
</div>
<div>
<p>Wilson, G. 2014. “Software Carpentry: Lessons Learned.” <em>F1000Research</em> 3: 62–62.</p>
</div>
<div>
<p>Yin, Tengfei, Dianne Cook, and Michael Lawrence. 2012. “Ggbio: An R Package for Extending the Grammar of Graphics for Genomic Data.” <em>Genome Biology</em> 13 (8): R77.</p>
</div>
<div>
<p>Zeeberg, Barry R, Joseph Riss, David W Kane, Kimberly J Bussey, Edward Uchio, W Marston Linehan, J Carl Barrett, and John N Weinstein. 2004. “Mistaken Identifiers: Gene Name Errors Can Be Introduced Inadvertently When Using Excel in Bioinformatics.” <em>BMC Bioinformatics</em> 5 (1): 80.</p>
</div>
<div>
<p>Ziemann, Mark, Yotam Eren, and Assam El-Osta. 2016. “Gene Name Errors Are Widespread in the Scientific Literature.” <em>Genome Biology</em> 17 (1): 177.</p>
</div>
</div>
</div>
<p style="text-align: center;">
<a href="3-9-example-creating-a-reproducible-data-pre-processing-protocol.html"><button class="btn btn-default">Previous</button></a>
</p>
</div>
</div>



</body>
</html>
