<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Experimental Data Preprocessing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>
  <meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Experimental Data Preprocessing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Experimental Data Preprocessing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  
  <meta name="twitter:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="experimental-data-recording.html"/>
<link rel="next" href="references.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Data Visualization in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.1</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html"><i class="fa fa-check"></i><b>2</b> Experimental Data Recording</a>
<ul>
<li class="chapter" data-level="2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#separating-data-recording-and-analysis"><i class="fa fa-check"></i><b>2.1</b> Separating data recording and analysis</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-versus-data-analysis"><i class="fa fa-check"></i><b>2.1.1</b> Data recording versus data analysis</a></li>
<li class="chapter" data-level="2.1.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#hazards-of-combining-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.2</b> Hazards of combining recording and analysis</a></li>
<li class="chapter" data-level="2.1.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#approaches-to-separate-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.3</b> Approaches to separate recording and analysis</a></li>
<li class="chapter" data-level="2.1.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions"><i class="fa fa-check"></i><b>2.1.4</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#principles-and-power-of-structured-data-formats"><i class="fa fa-check"></i><b>2.2</b> Principles and power of structured data formats</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-standards"><i class="fa fa-check"></i><b>2.2.1</b> Data recording standards</a></li>
<li class="chapter" data-level="2.2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#defining-data-standards-for-a-research-group"><i class="fa fa-check"></i><b>2.2.2</b> Defining data standards for a research group</a></li>
<li class="chapter" data-level="2.2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#two-dimensional-structured-data-format"><i class="fa fa-check"></i><b>2.2.3</b> Two-dimensional structured data format</a></li>
<li class="chapter" data-level="2.2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#saving-two-dimensional-structured-data-in-plain-text-file-formats"><i class="fa fa-check"></i><b>2.2.4</b> Saving two-dimensional structured data in plain text file formats</a></li>
<li class="chapter" data-level="2.2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#occassions-for-more-complex-data-structures-and-file-formats"><i class="fa fa-check"></i><b>2.2.5</b> Occassions for more complex data structures and file formats</a></li>
<li class="chapter" data-level="2.2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#levels-of-standardizationresearch-group-to-research-community"><i class="fa fa-check"></i><b>2.2.6</b> Levels of standardization—research group to research community</a></li>
<li class="chapter" data-level="2.2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise"><i class="fa fa-check"></i><b>2.2.7</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#the-tidy-data-format"><i class="fa fa-check"></i><b>2.3</b> The ‘tidy’ data format</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-makes-data-tidy"><i class="fa fa-check"></i><b>2.3.1</b> What makes data “tidy?”</a></li>
<li class="chapter" data-level="2.3.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#why-make-your-data-tidy"><i class="fa fa-check"></i><b>2.3.2</b> Why make your data “tidy?”</a></li>
<li class="chapter" data-level="2.3.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-tidyverse-tools-with-data-in-the-tidy-data-format"><i class="fa fa-check"></i><b>2.3.3</b> Using tidyverse tools with data in the “tidy data” format</a></li>
<li class="chapter" data-level="2.3.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz"><i class="fa fa-check"></i><b>2.3.4</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#designing-templates-for-tidy-data-collection"><i class="fa fa-check"></i><b>2.4</b> Designing templates for “tidy” data collection</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#creating-the-rules-for-collecting-data-in-the-same-time-each-time"><i class="fa fa-check"></i><b>2.4.1</b> Creating the rules for collecting data in the same time each time</a></li>
<li class="chapter" data-level="2.4.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1"><i class="fa fa-check"></i><b>2.4.2</b> Subsection 1</a></li>
<li class="chapter" data-level="2.4.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#dont-repeat-yourself"><i class="fa fa-check"></i><b>2.4.3</b> Don’t Repeat Yourself!</a></li>
<li class="chapter" data-level="2.4.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#dont-repeat-your-report-writing"><i class="fa fa-check"></i><b>2.4.4</b> Don’t repeat your report-writing!</a></li>
<li class="chapter" data-level="2.4.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#automating-reports"><i class="fa fa-check"></i><b>2.4.5</b> Automating reports</a></li>
<li class="chapter" data-level="2.4.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#scripts-and-automated-reports-as-simple-pipelines"><i class="fa fa-check"></i><b>2.4.6</b> Scripts and automated reports as simple pipelines</a></li>
<li class="chapter" data-level="2.4.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-1"><i class="fa fa-check"></i><b>2.4.7</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-creating-a-template-for-tidy-data-collection"><i class="fa fa-check"></i><b>2.5</b> Example: Creating a template for “tidy” data collection</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-1"><i class="fa fa-check"></i><b>2.5.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.5.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2"><i class="fa fa-check"></i><b>2.5.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.5.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-datasets"><i class="fa fa-check"></i><b>2.5.3</b> Example datasets</a></li>
<li class="chapter" data-level="2.5.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#issues-with-these-data-sets"><i class="fa fa-check"></i><b>2.5.4</b> Issues with these data sets</a></li>
<li class="chapter" data-level="2.5.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#final-tidy-examples"><i class="fa fa-check"></i><b>2.5.5</b> Final “tidy” examples</a></li>
<li class="chapter" data-level="2.5.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#options-for-recording-tidy-data"><i class="fa fa-check"></i><b>2.5.6</b> Options for recording tidy data</a></li>
<li class="chapter" data-level="2.5.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#examples-of-how-tidy-data-can-be-easily-analyzed-visualized"><i class="fa fa-check"></i><b>2.5.7</b> Examples of how “tidy” data can be easily analyzed / visualized</a></li>
<li class="chapter" data-level="2.5.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-1"><i class="fa fa-check"></i><b>2.5.8</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><i class="fa fa-check"></i><b>2.6</b> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#organizing-project-files-through-the-file-system"><i class="fa fa-check"></i><b>2.6.1</b> Organizing project files through the file system</a></li>
<li class="chapter" data-level="2.6.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#organizing-files-within-a-project-directory"><i class="fa fa-check"></i><b>2.6.2</b> Organizing files within a project directory</a></li>
<li class="chapter" data-level="2.6.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-rstudio-projects-with-project-file-directories"><i class="fa fa-check"></i><b>2.6.3</b> Using RStudio Projects with project file directories</a></li>
<li class="chapter" data-level="2.6.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-2"><i class="fa fa-check"></i><b>2.6.4</b> Subsection 1</a></li>
<li class="chapter" data-level="2.6.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-1"><i class="fa fa-check"></i><b>2.6.5</b> Subsection 2</a></li>
<li class="chapter" data-level="2.6.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz-1"><i class="fa fa-check"></i><b>2.6.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#creating-project-templates"><i class="fa fa-check"></i><b>2.7</b> Creating ‘Project’ templates</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-an-existing-file-directory-an-rstudio-project"><i class="fa fa-check"></i><b>2.7.1</b> Making an existing file directory an RStudio Project</a></li>
<li class="chapter" data-level="2.7.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-an-rstudio-project-template"><i class="fa fa-check"></i><b>2.7.2</b> Making an RStudio Project Template</a></li>
<li class="chapter" data-level="2.7.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-2"><i class="fa fa-check"></i><b>2.7.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-creating-a-project-template"><i class="fa fa-check"></i><b>2.8</b> Example: Creating a ‘Project’ template</a>
<ul>
<li class="chapter" data-level="2.8.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-3"><i class="fa fa-check"></i><b>2.8.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.8.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-2"><i class="fa fa-check"></i><b>2.8.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.8.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-2"><i class="fa fa-check"></i><b>2.8.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#harnessing-version-control-for-transparent-data-recording"><i class="fa fa-check"></i><b>2.9</b> Harnessing version control for transparent data recording</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-is-version-control"><i class="fa fa-check"></i><b>2.9.1</b> What is version control?</a></li>
<li class="chapter" data-level="2.9.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#recording-data-in-the-laboratoryfrom-paper-to-computers"><i class="fa fa-check"></i><b>2.9.2</b> Recording data in the laboratory—from paper to computers</a></li>
<li class="chapter" data-level="2.9.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-3"><i class="fa fa-check"></i><b>2.9.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><i class="fa fa-check"></i><b>2.10</b> Enhance the reproducibility of collaborative research with version control platforms</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-are-version-control-platforms"><i class="fa fa-check"></i><b>2.10.1</b> What are version control platforms?</a></li>
<li class="chapter" data-level="2.10.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#why-use-version-control-platforms"><i class="fa fa-check"></i><b>2.10.2</b> Why use version control platforms?</a></li>
<li class="chapter" data-level="2.10.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#how-to-use-github"><i class="fa fa-check"></i><b>2.10.3</b> How to use GitHub</a></li>
<li class="chapter" data-level="2.10.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-4"><i class="fa fa-check"></i><b>2.10.4</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.11" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-git-and-gitlab-to-implement-version-control"><i class="fa fa-check"></i><b>2.11</b> Using git and GitLab to implement version control</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#how-to-use-version-control"><i class="fa fa-check"></i><b>2.11.1</b> How to use version control</a></li>
<li class="chapter" data-level="2.11.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#leveraging-git-and-github-as-a-project-director"><i class="fa fa-check"></i><b>2.11.2</b> Leveraging git and GitHub as a project director</a></li>
<li class="chapter" data-level="2.11.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#leveraging-git-and-github-as-a-scientist-who-programs"><i class="fa fa-check"></i><b>2.11.3</b> Leveraging git and GitHub as a scientist who programs</a></li>
<li class="chapter" data-level="2.11.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#notes"><i class="fa fa-check"></i><b>2.11.4</b> Notes</a></li>
<li class="chapter" data-level="2.11.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-3"><i class="fa fa-check"></i><b>2.11.5</b> Applied exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html"><i class="fa fa-check"></i><b>3</b> Experimental Data Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><i class="fa fa-check"></i><b>3.1</b> Principles and benefits of scripted pre-processing of experimental data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#what-is-pre-processing"><i class="fa fa-check"></i><b>3.1.1</b> What is pre-processing?</a></li>
<li class="chapter" data-level="3.1.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#approaches-to-simple-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.2</b> Approaches to simple preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#approaches-to-more-complex-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.3</b> Approaches to more complex preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#scripting-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.4</b> Scripting preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#potential-quotes"><i class="fa fa-check"></i><b>3.1.5</b> Potential quotes</a></li>
<li class="chapter" data-level="3.1.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#discussion-questions-5"><i class="fa fa-check"></i><b>3.1.6</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-to-scripted-data-pre-processing-in-r"><i class="fa fa-check"></i><b>3.2</b> Introduction to scripted data pre-processing in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-4"><i class="fa fa-check"></i><b>3.2.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.2.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-3"><i class="fa fa-check"></i><b>3.2.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.2.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-4"><i class="fa fa-check"></i><b>3.2.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><i class="fa fa-check"></i><b>3.3</b> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#limitations-of-object-oriented-programming"><i class="fa fa-check"></i><b>3.3.1</b> Limitations of object-oriented programming</a></li>
<li class="chapter" data-level="3.3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-tidyverse-approach"><i class="fa fa-check"></i><b>3.3.2</b> The “tidyverse” approach</a></li>
<li class="chapter" data-level="3.3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#how-to-tidyverse"><i class="fa fa-check"></i><b>3.3.3</b> How to “tidyverse”</a></li>
<li class="chapter" data-level="3.3.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-5"><i class="fa fa-check"></i><b>3.3.4</b> Subsection 1</a></li>
<li class="chapter" data-level="3.3.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-4"><i class="fa fa-check"></i><b>3.3.5</b> Subsection 2</a></li>
<li class="chapter" data-level="3.3.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-2"><i class="fa fa-check"></i><b>3.3.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-data-types-in-experimental-data-pre-processing"><i class="fa fa-check"></i><b>3.4</b> Complex data types in experimental data pre-processing</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-6"><i class="fa fa-check"></i><b>3.4.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.4.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-versus-simple-structures-for-storing-data"><i class="fa fa-check"></i><b>3.4.2</b> Complex versus simple structures for storing data</a></li>
<li class="chapter" data-level="3.4.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#advantages-of-complex-structures-for-storing-data"><i class="fa fa-check"></i><b>3.4.3</b> Advantages of complex structures for storing data</a></li>
<li class="chapter" data-level="3.4.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#limitations-of-complex-structures-for-storing-data"><i class="fa fa-check"></i><b>3.4.4</b> Limitations of complex structures for storing data</a></li>
<li class="chapter" data-level="3.4.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-5"><i class="fa fa-check"></i><b>3.4.5</b> Subsection 2</a></li>
<li class="chapter" data-level="3.4.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-3"><i class="fa fa-check"></i><b>3.4.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-data-types-in-r-and-bioconductor"><i class="fa fa-check"></i><b>3.5</b> Complex data types in R and Bioconductor</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-7"><i class="fa fa-check"></i><b>3.5.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.5.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#exploring-and-extracting-data-from-r-list-data-structures"><i class="fa fa-check"></i><b>3.5.2</b> Exploring and extracting data from R list data structures</a></li>
<li class="chapter" data-level="3.5.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#interfacing-between-object-based-and-tidyverse-workflows"><i class="fa fa-check"></i><b>3.5.3</b> Interfacing between object-based and tidyverse workflows</a></li>
<li class="chapter" data-level="3.5.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#extras"><i class="fa fa-check"></i><b>3.5.4</b> Extras</a></li>
<li class="chapter" data-level="3.5.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-6"><i class="fa fa-check"></i><b>3.5.5</b> Subsection 2</a></li>
<li class="chapter" data-level="3.5.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-5"><i class="fa fa-check"></i><b>3.5.6</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#example-converting-from-complex-to-tidy-data-formats"><i class="fa fa-check"></i><b>3.6</b> Example: Converting from complex to ‘tidy’ data formats</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-8"><i class="fa fa-check"></i><b>3.6.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.6.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-biobroom-package"><i class="fa fa-check"></i><b>3.6.2</b> The <code>biobroom</code> package</a></li>
<li class="chapter" data-level="3.6.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-ggbio-package"><i class="fa fa-check"></i><b>3.6.3</b> The <code>ggbio</code> package</a></li>
<li class="chapter" data-level="3.6.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-7"><i class="fa fa-check"></i><b>3.6.4</b> Subsection 2</a></li>
<li class="chapter" data-level="3.6.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-6"><i class="fa fa-check"></i><b>3.6.5</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-to-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.7</b> Introduction to reproducible data pre-processing protocols</a></li>
<li class="chapter" data-level="3.8" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introducing-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.8</b> Introducing reproducible data pre-processing protocols</a></li>
<li class="chapter" data-level="3.9" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#from-pre-processing-scripts-to-pre-processing-protocols"><i class="fa fa-check"></i><b>3.9</b> From pre-processing scripts to pre-processing protocols</a>
<ul>
<li class="chapter" data-level="3.9.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#technique-to-create-reproducible-pre-processing-protocols"><i class="fa fa-check"></i><b>3.9.1</b> Technique to create reproducible pre-processing protocols</a></li>
<li class="chapter" data-level="3.9.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#advantages-of-reproducible-pre-processing-protocols"><i class="fa fa-check"></i><b>3.9.2</b> Advantages of reproducible pre-processing protocols</a></li>
<li class="chapter" data-level="3.9.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#what-knitted-documents-are-and-why-to-use-them"><i class="fa fa-check"></i><b>3.9.3</b> What knitted documents are and why to use them</a></li>
<li class="chapter" data-level="3.9.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#how-knitted-documents-work"><i class="fa fa-check"></i><b>3.9.4</b> How knitted documents work</a></li>
<li class="chapter" data-level="3.9.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#quotes"><i class="fa fa-check"></i><b>3.9.5</b> Quotes</a></li>
<li class="chapter" data-level="3.9.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#discussion-questions-6"><i class="fa fa-check"></i><b>3.9.6</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.10</b> RMarkdown for creating reproducible data pre-processing protocols</a>
<ul>
<li class="chapter" data-level="3.10.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#creating-knitted-documents-in-r"><i class="fa fa-check"></i><b>3.10.1</b> Creating knitted documents in R</a></li>
<li class="chapter" data-level="3.10.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#creating-and-exploring-an-rmarkdown-file"><i class="fa fa-check"></i><b>3.10.2</b> Creating and exploring an Rmarkdown file</a></li>
<li class="chapter" data-level="3.10.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#more-advanced-rmarkdown-functionality"><i class="fa fa-check"></i><b>3.10.3</b> More advanced Rmarkdown functionality</a></li>
<li class="chapter" data-level="3.10.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#learning-more-about-rmarkdown."><i class="fa fa-check"></i><b>3.10.4</b> Learning more about Rmarkdown.</a></li>
<li class="chapter" data-level="3.10.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-7"><i class="fa fa-check"></i><b>3.10.5</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#example-creating-a-reproducible-data-pre-processing-protocol"><i class="fa fa-check"></i><b>3.11</b> Example: Creating a reproducible data pre-processing protocol</a>
<ul>
<li class="chapter" data-level="3.11.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-and-example-data"><i class="fa fa-check"></i><b>3.11.1</b> Introduction and example data</a></li>
<li class="chapter" data-level="3.11.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#advice-on-designing-a-pre-processing-protocol"><i class="fa fa-check"></i><b>3.11.2</b> Advice on designing a pre-processing protocol</a></li>
<li class="chapter" data-level="3.11.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#writing-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.11.3</b> Writing data pre-processing protocols</a></li>
<li class="chapter" data-level="3.11.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-4"><i class="fa fa-check"></i><b>3.11.4</b> Practice quiz</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>4</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Improving the Reproducibility of Experimental Data Recording and Pre-Processing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="experimental-data-preprocessing" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Experimental Data Preprocessing</h1>
<div id="principles-and-benefits-of-scripted-pre-processing-of-experimental-data" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Principles and benefits of scripted pre-processing of experimental data</h2>
<p>The experimental data collected for biomedical research often requires
pre-processing before it can be analyzed (e.g., gating of flow cytometry data,
feature finding / quantification for mass spectrometry data). Use of
point-and-click software can limit the transparency and reproducibility of this
analysis stage and is time-consuming for repeated tasks. We will explain how
scripted pre-processing, especially using open source software, can improve
transparency and reproducibility.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Define ‘pre-processing’ of experimental data</li>
<li>Describe an open source code script and explain how it can increase
reproducibility of data pre-processing</li>
</ul>
<div id="what-is-pre-processing" class="section level3" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> What is pre-processing?</h3>
<p>Some data collected through laboratory experiments is very straightforward and
requires little or no pre-processing before it’s used in analysis. For example,
[example]. Other data may require some minimal pre-processing. For example, if
you plate bacteria from a sample at a variety of dilutions, you might count each
plate and determine a measure of Colony Forming Units from the set of plates
with different dilutions by deciding which dilution provides the clearest count
and then back-calculating based on its dilution to get the total number of
colony-forming units in the original sample.</p>
<p>This step of pre-processing data can become much more complex with data that was
collected using complex equipment, like a flow cytometer or a mass spectrometer.
In these cases, there are often steps required to extract from the machine’s
readings a biologically-relevant measurement. For example, the data output from
a mass spectrometer must be processed to move from measurements of mass and
retention time to estimates of concentrations of different molecules in the
sample. If you want to compare across multiple samples, then the preprocessing
will also involve steps to align the different samples (in terms of …), as
well as to standardize the measurements for each sample, to make the
measurements from the different samples comparable. For data collected from a
flow cytometer, preprocessing may include steps to disentangle the florescence
from different markers to ensure that the read for one marker isn’t inflated by
spillover florescence from a different marker.</p>
</div>
<div id="approaches-to-simple-preprocessing-tasks" class="section level3" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Approaches to simple preprocessing tasks</h3>
<p>There are several approaches for tackling this type of data preprocessing, to
get from the data that you initial observe (or that is measured by a piece of
laboratory equipment) to meaningful biological measurements that can be analyzed
and presented to inform explorations of a scientific hypothesis. While there are
a number of approaches that don’t involve writing code scripts for this
preprocessing, there are some large advantages to scripting preprocessing any
time you are preprocessing experimental data prior to including it in figures or
further analysis. In this section, we’ll describe some common non-scripted
approaches and discuss the advantages that would be brought by instead using a
code script. In the next module, we’ll walk through an example of how scripts
for preprocessing can be created and applied in laboratory research.</p>
<p>In cases where the pre-processing is mathematically straightforward and the
dataset is relatively small, many researchers do the preprocessing by hand in a
laboratory notebook or through an equation or macro embedded in a spreadsheet.
For example, if you have plated samples at different dilutions and are trying to
calculate from these the CFUs in the original sample, this calculation is simple
enough that it could be done by hand. However, there are advantages to instead
writing a code script to do this simple preprocessing.</p>
<p>When you write a script to do a task with data, it is like writing a recipe that
can be applied again and again. By writing a script, you encode the process a
single time, so you can take the time to check and recheck to make sure that
you’ve encoded the process correctly. This helps in avoiding small errors when
you do the preprocessing—if you are punching numbers into a calculator over
and over, it’s easy to mistype a number or forget a step every now and then,
while the code will ensure that the same process is run every time and that it
faithfully uses the numbers saved in the data for each step, rather than relying
on a person correctly entering each number in the calculation.</p>
<p>Scripts can be used across projects, as well, and so they can ensure consistency
in the calculation across projects. If different people do the calculation in
the lab for different projects or experiments, and they are doing the
calculations by hand, they might each do the calculation slightly differently,
even if it’s only in small details like how they report rounded numbers. A
script will do the exact same thing every time it is applied. You can even share
your script with colleagues at other labs, if you want to ensure that your data
preprocessing is comparable for experiments conducted in different research
groups, and many scientific journals will allow supplemental material with
code used for data preprocessing and analysis, or links within the manuscript
to a repository of this code posted online.</p>
<p>There are also gains in efficiency when you use a script. For small
pre-processing steps, these might seem small for each experiment, and certainly
when you first write the script, it will likely take longer to write and test
the script than it would to just do the calculation by hand (even more if
you’re just starting to learn how to write code scripts). However, since the
script can be applied again and again, with very little extra work to apply it
to new data, you’ll save yourself time in the future, and over a lot of
experiments and projects, this can add up. This makes it particularly useful to
write scripts for preprocessing tasks that you find yourself doing again and
again in the lab.</p>
</div>
<div id="approaches-to-more-complex-preprocessing-tasks" class="section level3" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Approaches to more complex preprocessing tasks</h3>
<p>Other preprocessing tasks can be much more complex, particularly those that need
to conduct a number of steps to extract biologically meaningful measurements
from the measurements made by a complex piece of laboratory equipment, as well
as steps to make sure these measurements can be meaningfully compared across
samples.</p>
<p>For these more complex tasks, the equipment manufacturer will often provide
software that can be used for the preprocessing. This software might conduct
some steps using defaults, and others based on the user’s specifications. These
are often provided through “GUIs” (graphical user interfaces), where the user
does a series of point-and-click steps to process the data. In some software,
this series of point-and-click steps is recorded as the user does them, so that
these steps can be “re-run” later or on a different dataset.</p>
<p>For many types of biological data, including output from equipment like flow
cytometers and mass spectrometers, open-source software has been developed
that can be used for this preprocessing. Often, the most cutting edge methods
for data preprocessing are first available through open-source software packages,
if the methods are developed by researchers rather than by the companies, and
often many of the algorithms that are made available through the equipment
manufacturer’s proprietary software are encoded versions of an algorithm
first shared by researchers as open-source software.</p>
<p>It can take a while to develop a code script for preprocessing the raw data from
a piece of complex equipment like a mass spectrometer. However, the process of
developing this script requires a thoughtful consideration of the steps of
preprocessing, and so this is often time well-spent. Again, this initial time
investment will pay off later, as the script can then be efficiently applied to
future data you collect from the equipment, saving you time in pointing and
clicking through the GUI software. Further, it’s easier to teach someone else
how to conduct the preprocessing that you’ve done, and apply it to future
experiments, because the script serves as a recipe.</p>
<p>When you conduct data preprocessing in a script, this also gives you access to
all the other tools in the scripting language. For example, as you work through
preprocessing steps for a dataset, if you are doing it through an R script, you
can use any of the many visualization tools that are available through R. By
contrast, in GUI software, you are restricted to the visualization and other
tools included in that particular set of software, and those software developers
may not have thought of something that you’d like to do. Open-source scripting
languages like R, Python, and Julia include a huge variety of tools, and once
you have loaded your data in any of these platforms, you can use any of these
tools.</p>
<p>If you have developed a script for preprocessing your raw data, it also becomes
much easier to see how changes in choices in preprocessing might influence your
final results. It can be tricky to guess whether your final results are sensitive,
for example, to what choice you make for a particular tranform for part of your
data, or in how you standardize data in one sample to make different samples
easier to compare. If the preprocessing is in a script, then you can test making
these changes and running all preprocessing and analysis scripts, to see if it
makes a difference in the final conclusions. If it does, then it helps you
identify parts of preprocessing that need to be deeply thought through for the
type of data you’re collecting, and you may want to explore the documentation on
that particular step of preprocessing to determine what choice is best for your
data, rather than relying on defaults.</p>
</div>
<div id="scripting-preprocessing-tasks" class="section level3" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Scripting preprocessing tasks</h3>
<p>Code scripts can be developed for any open-source scripting languages, including
Python, R, and Julia. These can be embedded in or called from literate programming
documents, like RMarkdown and Julia, which are described in other modules. The
word “script” is a good one here—it really is as if you are providing the script
for a play. In an interactive mode, you can send requests to run in the programming
language step by step using a console, while in a script you provide the whole list
of all of your “lines” in that conversation, and the programming language will run
them all in order without you needing to interact from the console.</p>
<p>For preprocessing the data, the script will have a few predictible parts. First,
you’ll need to read the data in. There are different functions that can be used
to read in data from different file formats. For example, data that is stored in
an Excel spreadsheet can be loaded into R using functions in a package called
<code>readxl</code>. Data that is stored in a plain-text delimited format (like a csv file)
can be loaded into R using functions in the <code>readr</code> package.</p>
<p>When preprocessing data from complex equipment, you can determine how to read the
data into R by investigating the file type that is output by the equipment.
Fortunately, many types of scientific equipment follow standardized file formats.
This means that open-source developers can develop a single package that can
load data from equipment from multiple manufacturers. For example, flow cytometry
data is often stored in [file format]. Other biological datasets use file
formats that are appropriate for very large datasets and that allow R to work
with parts of the data at a time, without loading the full data in. [netCDF?]
In these cases, the first step in a script might not be to load in all the data,
but rather to provide R with a connection to the larger datafile, so it can
pull in data as it needs it.</p>
<p>Once the data is loaded or linked in the script, the script can proceed through
steps required to preprocess this data. These steps will often depend on the type
of data, especially the methods and equipment used to collect it. For example, for
mass spectrometry data, these steps will include … . For flow cytometry data,
these steps would include … .</p>
<p>The functions for doing these steps will often come from extensions that
different researchers have made for R. Base R is a simpler collection of data
processing and statistics tools, but the open-source framework of R has allowed
users to make and share their own extensions. In R, these are often referred to
as “packages.” Many of these are shared through the Comprehensive R Archive
Network (CRAN), and packages on CRAN can be directly installed using the
<code>install.packages</code> function in R, along with the package’s names. While CRAN
is the common spot for sharing general-purpose packages, there is a specialized
repository that is used for many genomics and other biology-related R packages
called Bioconductor. These packages can also be easily installed through a call
in R, but in this case it requires an installation function from the <code>BiocManager</code>
package. Many of the functions that are useful for preprocessing biological
data from laboratory experiments are available through Bioconductor.</p>
<p>Table [x] includes some of the primary R packages on Bioconductor that can be
used in preprocessing different types of biological data. There are often
multiple choices, developed by different research groups, but this list provides
a starting point of several of the standard choices that you may want to
consider as you start developing code.</p>
<p>Much of the initial preprocessing might use very specific functions that are
tailored to the format that the data takes once it is loaded. Later in the
script, there will often be a transfer to using more general-purpose tools in
that coding language. For example, once data is stored in a “dataframe” format
in R, it can be processed using a powerful set of general purpose tools
collected in a suite of packages called the “tidyverse.” This set of packages
includes functions for filtering to specific subsets of the data, merging
separate datasets, adding new measurements for each observation that are
functions of the initial measurements, summarizing, and visualizing. The
tidyverse suite of R tools is very popular in general R use and is widely
taught, including through numerous free online resources. By moving from
specific tools to these more general tools as soon as possible in the script, a
researcher can focus his or her time in learning these general purpose tools
well, as these can be widely applied across many types of data.</p>
<p>By the end of the script, data will be in a format that has extracted
biologically relevant measurements. Ideally, this data will be in a general
purpose format, like a dataframe, to make it easier to work with using general
purpose tools in the scripting language when the data is used in further data
analysis or to create figures for reports, papers, and presentations. Often, you
will want to save a version of this preprocessed version of the data in your
project files, and so the last step of the script might be to write out the
cleaned data in a file that can be loaded in later scripts for analysis and
visualization. This is especially useful if these data preprocessing steps are
time consuming, as is often the case for the large raw datasets output by
laboratory equipment like flow cytometers and mass spectrometers.</p>
<p>Figure [x] gives an example of a data preprocessing script, highlighting these
different common areas that often show up in these scripts.</p>
<p>[Some data may be incorporated into the preprocessing by downloading it from
databases or other online sources. These data downloads can be automated and
recorded by using scripted code for the download in many cases, as long as the
database or online source offers web services or another API for this type of
scripted data access. In this case, you can incorporate the script in a
RMarkdown document to record the date the data was downloaded, as well as the
code used to download it. R is able to run system calls, and one of these
will provide the current date, so this can be included in an RMarkdown file
to record the date the file is run. Further, there may be a call that can
be made to the online data source’s API that returns the working version of
the database or source, and if so this can also be included in the RMarkdown
code used to access the data.]</p>
<p>RMarkdown files can be used to combine both code and more manual document
(for example, a record of which collaborator provided each type of data file).
While traditionally this more manual documentation was recommended to be
recorded in plain-text README files in a project’s directory and subdirectories
<span class="citation">(Buffalo 2015)</span>, RMarkdown files provide some advantages over
this traditional approach. First, RMarkdown files are themselves in plain
text, and so they offer the advantages of simple plain text documentation
files (e.g., ones never rendered to another format) in terms of being able
to use script-based tools to search them. Further, they can be rendered into
attractive formatted documents that may be easier to share with project
team members who do not code.</p>
<p>[Example of a function: recipe for making a vinaigrette. There will be a
“basic” way that the function can run, which uses its default parameters.
However, you can also specify and customize certain inputs (for example,
using walnut oil instead of olive oil, or adding mustard) to tweak the
recipe in slight ways each time you use it, and to get customized outputs.]</p>
<p>[History of the mouse—enable GUIs, before everything was from the terminal.]</p>
</div>
<div id="potential-quotes" class="section level3" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Potential quotes</h3>
<blockquote>
<p>For bioinformatics, “all too often the software is developed without
thought toward future interoperability with other software products. As a
result, the bioinformatics software landscape is currently characterized
by fragmentation and silos, in which each research group develops and uses
only the tools created within their lab.” <span class="citation">(Barga et al. 2011)</span></p>
</blockquote>
<blockquote>
<p>“The group also noted the lack of agility. Although they may be aware of
a new or better algorithm they cannot easily integrate it into their
analysis pipelines given the lack of standards across both data formats
and tools. It typically requires a complete rewrite of the code in order
to take advntge of a new technique or algorithm, requiring time and often
funding to hire developers.” <span class="citation">(Barga et al. 2011)</span></p>
</blockquote>
<blockquote>
<p>“The benefit of working with a programming language is that you have the code in
a file. This means that you can easily reuse that code. If the code has
parameters it can even be applied to problems that follow a similar pattern.”
<span class="citation">(Janssens 2014)</span></p>
</blockquote>
<blockquote>
<p>“Data exploration in spreadsheet software is typically conducted via menus and
dialog boxes, which leaves no record of the steps taken.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“One reason Unix developers have been cool toward GUI interfaces is that, in their
designers’ haste to make them ‘user-friendly’ each one often becomes frustratingly
opaque to anyone who has to solve user problems—or, indeed, interact with it anywhere
outside the narrow range predicted by the user-interface designer.” <span class="citation">(E. S. Raymond 2003)</span></p>
</blockquote>
<blockquote>
<p>“Many operating systems touted as more ‘modern’ or ‘user friendly’ than Unix achieve their
surface glossiness by locking users and developers into one interface policy, and offer an
application-programming interface that for all its elaborateness is rather narrow and rigid.
On such systems, tasks the designers have anticipated are very easy—but tasks they have
not anticipated are often impossible or at best extremely painful. Unix, on the other hand, has
flexibility in depth. The many ways Unix provides to glue together programs means that components
of its basic toolkit can be combined to produce useful effects that the designers of the individual
toolkit parts never anticipated.” <span class="citation">(E. S. Raymond 2003)</span></p>
</blockquote>
<blockquote>
<p>“The good news is that a computer is a general-purpose machine, capable of performing
any computation. Although it only has a few kinds of instructions to work with, it can
do them blazingly fast, and it can largely control its own operation. The bad news is
that it doesn’t do anything itself unless someone tells it what to do, in excruciating
detail. A computer is the ultimate sorcere’s apprentice, able to follow instructions
tirelessly and without error, but requiring painstaking accuracy in the
specification of what to do.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“<em>Software</em> is the general term for sequences of instructions that make a computer
do something useful. It’s ‘soft’ in contrast with ‘hard’ hardware, because it’s
intangible, not easy to put your hands on. Hardware is quite tangible: if you drop
a computer on your foot, you’ll notice. Not true for software.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“Modern system increasingly use general purpose hardware—a processor, some memory,
and connections to the environment—and create specific behaviors by software. The
conventional wisdom is that software is cheaper, more flexible, and easier to change than
hardware is (especially once some device has left the factory).” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“An algorithm is a precise and unambiguous recipe. It’s expressed in terms of a fixed
set of basic operations whose meanings are completely known and specified; it spells out
a sequence of steps using those operations, with all possible situations covered; it’s
guaranteed to stop eventually. On the other hand, a <em>program</em> is the opposite of
abstract—it’s a concrete statement of the steps that a real computer must perform to
accomplish a task. The distinction between an algorithm and a program is like the difference
between a blueprint and a building; one is an idealization and the other is the real thing.”
<span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“One way to view a program is as one or more algorithms expressed in a form that a computer
can process directly. A program has to worry about practical problems like inadequate memory,
limited processor speed, invalid and even malicious input data, faulty hardware, broken
network connections, and (in the background and often exacerbating the other problems)
human frailty. So if an algorithm is an idealized recipe, a program is the instructions for
a cooking robot preparing a month of meals for an army while under enemy attack.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“During the late 1950s and early 1960s, another step was taken towards getting the
computer to do more for programmers, arguably the most important step in the history of
programming. This was the development of ‘high-level’ programming languages that were
independent of any particular CPU architecture. High-level languages make it possible to
express computations in terms that are closer to the way a person might express them.”
<span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“Programming in the real world tends to happen on a large scale. The strategy is similar
to what one might use to write a book or undertake any other big project: figure out what
to do, starting with a broad specification that is broken into smaller and smaller pieces,
then work on the pieces separately, while making sure that they hang together. In programming,
pieces tend to be of a size such that one person can write the precise computational steps
in some programming language. Ensuring that the pieces written by different programmers
work together is challenging, and failing to get this right is a major source of errors.
For instance, NASA’s Mars Climate Orbiter failed in 1999 because the flight system software
used metric units for thrust, but course correction data was entered in English units,
causing an erroneous trajectory that brought the Orbiter too close to the planet’s
surface.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“If you’re going to build a house today, you don’t start by cutting down trees to make
lumber and digging clay to make your own bricks. Instead, you buy prefabricated pieces like
doors, windows, plumbing fixtures, a furnace, and a water heater. House construction is still
a big job, but it’s manageable because you can build on the work of many others and rely
on an infrastructure, indeed an entire industry, that will help. The same is true of
programming. Hardly any significant program is created from nothing. Many components written
by others can be taken off the shelf and used. For instance, if you’re writing a program for
Windows or a Mac, there are libraries of prefabricated menus, buttons, text editors, graphics,
network connections, database access, and so on. Much of the job is understanding the components
and gluing them together in your own way. Of course, many of these components in turn rest on
other simpler and more basic ones, often for several layers. Below that, everything runs on
the operating system, a program that manages the hardware and controls everything that happens.”
<span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“At the simplest level, programming languages provide a mechanism called functions that make
it possible for one programmer to write code that performs a useful a useful task, then package
it in a form that other programmers can use in their programs without having to know how it
works.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“A function has a name and a set of input data values that it needs to do its job; it does
a computation and returns a result to the part of the program that called it. … Functions
make it possible to create a program by building on components that have been created separately
and can be used as necessary by all programmers. A collection of related functions is usually
called a <em>library</em>. … The services that a function library provides are described to programmers
in terms of an <em>Application Programming Interface</em>, or <em>API</em>, which lists the functions, what
they do, how to use them in a program, what input data they require, and what values they
produce. The API might also describe data structures—the organization of data that is passed
back and forth—and various other bits and pieces that all together define what a programmer
has to do to request services and what will be computed as a result. This specification must
be detailed and precise, since in the end the program will be interpreted by a dumb literal
computer, not by a friendly and accomodating human.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“The code that a programmer writes, whether in assembly language or (much more likely) in
a high-level language, is called <em>source code</em>. … Source code is readable by other programmers,
though perhaps with some effort, so it can be studied and adapted, and any innovations or ideas
it contains are visible.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“In early times, most software was developed by companies and most source code was
unavailable, a trade secret of whoever developed it.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“An <em>operating system</em> is the software underpinning that manages the hardware of a
computer and makes it possible to run other programs, which are called <em>applications</em>.
… It’s a clumsy but standard terminology for programs that are more or less self-contained
and focused on a single task.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“Software, like many other things in computing, is organized into layers, analogous to
geological strata, that separate one concern from another. Layering is one of the important
ideas that help programmers to manage complexity.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“I think that it’s important for a well-informed person to know something about
programming, perhaps only that it can be surprisingly difficult to get very simple
programs working properly. There is nothing like doing battle with a computer to teach
this lesson, but also to give people a taste of the wonderful feeling of accomplishment
when a program does work for the first time. It may also be valuable to have enough
programming experience that you are cautious when someone says that programming is easy,
or that there are no errors in a program. If you have trouble making 10 lines of code
work after a day of struggle, you might be legitimately skeptical of someone who claims
that a million-line program will be delivered on time and bug-free.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“Programming languages share certain basic ideas, since they are all notations for spelling
out a computation as a sequence of steps. Every programming language thus will provide ways
to get input data upon which to compute; do arithmetic; store and retrieve intermediate
values as computation proceeds; display results along the way; decide how to proceed on the basis
of previous computations; and save results when the computation is finished. Languages have
<em>syntax</em>, that is, rules that define what is grammatically legal and what is not.
Programming languages are picky on the grammatical side: you have to say it right or there
will be a complaint. Languages also have <em>semantics</em>, that is, a defined meaning for every
construction in the language.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“In programming, a <em>library</em> is a collection of related pieces of code. A library typically
includes the code in compiled form, along with needed source code declarations [for C++].
Libraries can include stand-alone functions, classes, type declarations, or anything else that
can appear in code.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“One way to write R code is simply to enter it interactively at the command line… This
interactivity is beneficial for experimenting with R or for exploring a data set in a casual
manner. … However, interactively typing code at the R command line is a very bad approach from
the perspective of recording and documenting code because the code is lost when R is shut down.
A superior approach in general is to write R code in a file and get R to read the code from the file.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“The features of R are organized into separate bundles called <em>packages</em>. The standard R
installation includes about 25 of those packages, but many more can be downloaded from CRAN and
installed to expand the things that R can do. … Once a package is installed, it must be
<em>loaded</em> within an R session to make the extra features available. … Of the 25 packages
that are installed by default, nine packages are <em>loaded</em> by default when we start a new
R session; these provide the basic functionality of R. All other packages must be loaded
before the relevant features can be used.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“The R environment is the software used to run R code.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“<em>Document your methods and workflows.</em> This should include full command lines (copied
and pasted) that are run through the shell that generate data or intermediate results.
Even if you use the default values in software, be sure to write these values down;
later versions of the program may use different default values. Scripts naturally
document all steps and parameters …, but be sure to document any command-line options
used to run this script. In general, any command that produces results in your work needs
to be documented somewhere.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“<em>Document the version of the software that you ran.</em> This may seem unimportant, but
remember the example from ‘Reproducible Research’ on page 6 where my colleagues and I
traced disagreeing results down to a single piece of software being updated. These
details matter. Good bioinformatices software usually has a command-line option to
return the current version. Software managed with a version control system such as
Git has explicit identifiers to every version, which can be used to document the
precise version you ran… If no version information is available, a release date,
link to the software, and download date will suffice.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“<em>Document when you downloaded data.</em> It’s important to include when the data was downloaded,
as the external data source (such as a website or server) might change in the future. For example,
a script that downloads data directly from a database might produce different results if
rerun after the external database is updated. Consequently, it’s important to document
when data came into your repository.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“All of this [documentation] information is best stored in plain-text README files.
Plain text can easily be read, searched, and edited directly from the command line,
making it the perfect choice for portable and accessible README files. It’s also available
on all computer systems, meaning you can document your steps when working directly on
a server or computer cluster. Plain text also lacks complex formatting, which can create
issues when copying and pasting commands from your documentation back into the command
line.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“The computer is a very flexible and powerful tool, and it is a tool that is ours
to control. Files and documents, especially those in open standard formats, can be
manipulated using a variety of software tools, not just one specific piece of software.
A programming lanuage is a tool that allows us to manipulate data stored in files and
to manipulate data held in RAM in unlimited ways. Even with a basic knowledge of
programming, we can perform a huge variety of data processing tasks.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Computer code is the preferred approach to communicating our instructions to the
computer. The approach allows us to be precise and expressive, it provides a complete
record of our actions, and it allows others to replicate our work.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Programming in R is carried out, primarily, by manipulating and modifying data structures.
These different transformations are carried out using functions and operators. In R,
virtually every operation is a function call, and though we separate our discussion into
operators and function calls, the distinction is not strong … The R evaluator and
many functions are written in C but most R functions are written in R itself.”
<span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Many biologists are first exposed to the R language by following a cookbook-type
approach to conduct a statistical analysis like a t-test or an analysis of
variance (ANOVA). ALthough R excels at these and more complicated statistical
tests, R’s real power is as a data programming lanugage you can use to explore and
understand data in an open-ended, highly interactive, iterative way. Learning R as a
data programming language will give you the freedom to experiment and problem solve
during data analysis—exactly what we need as bioinformaticians.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Popularized by statistician John W. Tukey, EDA is an approach that emphasizes
understanding data (and its limitations) through interactive investigation
rather than explicit statitical modeling. In his 1977 book <em>Exploratory Data
Analysis</em>, Tukey described EDA as ‘detective work’ involved in ‘finding and
revealing the clues’ in data. As Tukey’s quote emphasizes, EDA is much more an approach
to exploring data than using specific statistical methods. In the face of rapidly
changing sequencing technologies, bioinformatics software, and statistical methods,
EDA skills are not only widely applicable and comparatively stable—they’re also
essential to making sure that our analyses are robust to these new data and methods.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Developing code in R is a back-and-forth between writing code in a rerunnable script
and exploring data interactively in the R interpreter. To be reproducible, all steps
that lead to results you’ll use later must be recorded in the R script that accompanies
your analysis and interactive work. While R can save a history of the commands you’ve
entered in the interpreter during a session (with the command <code>savehistory()</code>),
storing your steps in a well-commented R script makes your life much easier when you
need to backtrack to understand what you did or change your analysis.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“It’s a good idea to avoid referring to specific dataframe rows in your analysis code.
This would produce code fragile to row permutations or new rows that may be generated
by rerunning a previous analysis step. In every case in which you might need to refer
to a specific row, it’s avoidable by using subsetting… Similarly, it’s a good idea
to refer to columns by their column name, <em>not</em> their position. While columns may be
less likely to change across dataset versions than rows, it still happens. Column names
are more specific than positions, and also lead to more readable code.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“In bioinformatics, we often need to extract data from strings. R has several
functions to manipulate strings that are handy when working with bioinformatics data in
R. Note, however, that for most bioinformatics text-processing tasks, R is <em>not</em>
the preferred language to use for a few reasons. First, R works with all data stored
in memory; many bioinformatics text-processing tasks are best tackled with the
stream-based approaches…, which explicityly avoid loading all data in memory at
once. Second, R’s string processing functions are admittedly a bit clunky compared
to Python’s.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Versions fo R and any R pakcages installed change over time. This can lead to
reproducibility headaches, as the results of your analyses may change with the
changing version of R and R packages. … you should always record the versions
of R and any packages you use for analysis. R actually makes this incrediably
easy to do—just call the <code>sessionInfo()</code> function.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor is an open source R software project focused on developing tools
for high-throughput genomics and molecular biology data.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor’s pakcage system is a bit different than those on the Comprehensive R
Archive Network (CRAN). Bioconductor packages are released on a set schedule, twice
a year. Each release is coordinated with a version of R, making Bioconductor’s versions
tied to specific R versions. The motivation behind this strict coordination is that it
allows for packages to be thoroughly tested before being released for public use.
Additionally, because there’s considerable code re-use within the Bioconductor project,
this ensures that all package versions within a Bioconductor release are compatible
with one another. For users, the end result is that packages work as expected and
have been rigorously tested before you use it (this is good when your scientific
results depend on software reliability!). If you need the cutting-edge version of a
package for some reason, it’s always possible to work with their development branch.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“When installing Bioconductor packages, we use the <code>biocLite()</code> function. <code>biocLite()</code>
installs the correct version of a package for your R version (and its corresponding
Bioconductor version).” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“In addition to a careful release cycle that fosters package stability, Bioconductor
also has extensive, excellent documentation. The best, most up-to-date documentation
for each package will always be a Bioconductor [web address]. Each package has a full
reference manual covering all functions and classes included in a package,
as well as one or more in-depth vignettes. Vignettes step through many examples and
common workflows using packages.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Quite often, users don’t appreciate the opportunities. Noncomputational
biologists don’t know when to complain about the status quo. With modest amounts
of computational consulting, long or impossible jobs can become much shorter or
richer.” — Barry Demchak in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“People not doing the computational work tend to think that you can write a
program very fast. That, I think, is frankly not true. It takes a lot of time to
implement a prototype. Then it actually takes a lot of time to really make it
better.” — Heng Li in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“There is also a problem with discovering software that exists; often people
reinvent the wheel just because they don’t know any better. Good repositories
for software and best practice workflows, especially if citable, would be a
start.” — James Taylor in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>" Now there are a lot of strong, young, faculty members who label themselves
as computational analysts, yet very often want wet-lab space. They’re not
content just working off data sets that come from other people. They want to be
involved in data generation and experimental design and mainstreaming
computation as a valid research tool. Just as the boundaries of biochemistry and
cell biology have kind of blurred, I think the same will be true of
computational biology. It’s going to be alongside biochemistry, or molecular
biology or microscopy as a core component." — Richard Durbin in
<span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“I would say that computation is now as important to biology as chemistry is.
Both are useful background knowledge. Data manipulation and use of information
are part of the technology of biology research now. Knowing how to program also
gives people some idea about what’s going on inside data analysis. It helps them
appreciate what they can and can’t expect from data analysis software.” —
Richard Durbin in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Does every new biology PhD student need to learn how to program?</strong> To some,
the answer might be “no” because that’s left to the experts, to the people
downstairs who sit in front of a computer. But a similar question would be: does
every graduate student in biology need to learn grammar? Clearly, yes. Do they
all need to learn to speak? Clearly, yes. We just don’t leave it to the
literature experts. That’s because we need to communicate. Do students need to
tie their shoes? Yes. It has now come to the point where using a computer is as
essential as brushing your teeth. If you want some kind of a competitive edge,
you’re going to want to make as much use of that computer as you can. The
complexity of the task at hand will mean that canned solutions don’t exist. It
means that if you’re using a canned solution, you’re not at the edge of
research." — Martin Krzywinski in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“Although we are tackling many different types of data, questions, and
statistical methods hands-on, we maintain a consistent computational approach
by keeping all the computation under one roof: the R programming language and
statistical environment, enhanced by the biological data infrastructure and
specialized method packages from the Bioconductor project.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“The availablility of over 10,000 packages [in R] ensures that almost all
statistical methods are available, including the most recent developments.
Moreover, there are implementations of or interfaces to many methods from
computer science, mathematics, machine learning, data management, visualization
and internet technologies. This puts thousands of person-years of work by
experts at your fingertips.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor packages support the reading of many of the data types and formats
produced by measurement instruments used in modern biology, as well as the
needed technology-specific ‘preprocessing’ routines. This community is
actively keeping these up-to-date with the rapid developments in the
instrument market.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“An equivalent to the laboratory notebook that is standard good practice in
labwork, we advocate the use of a computational diary written in the R markdown
format. … Together with a version control system, R markdown helps with
tracking changes.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“There are (at least) two types of data visualization. The first enables
a scientist to explore data and make discoveries about the complex processes
at work. The other type of visualization provides informative, clear and
visually attractive illustrations of her results that she can show to others
and eventually include in a publication.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“A common task in biological data analysis is comparison between several
samples of univariate measurements. … As an example, we’ll use the intensities
of a set of four genes… A popular way to display [this] is through
barplots [and boxplots, violin plots, dot plots, and beeswarm plots].” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“At different stages of their development, immune cells express unique
combinations of proteins on their surfaces. These protein-markers are called
CDs (clusters of differentiation) and are collected by flow cytometry
(using fluorescence…) or mass cytometry (using single-cell atomic mass
spectrometry of heavy metal reporters). An example of a commonly used CD is
CD4; this protein is expressed by helper T cells that are referred to as
being ‘CD4+.’ Note, however, that some cells express CD4 (thus are CD4+)
but are not actually helper T cells. We start by loading some useful Bioconductor
packages for flow cytometry, <code>flowCore</code> and <code>flowViz</code>. …
First we load the table data that reports the mapping between isotopes and
markers (antibodies), and then we replace the isotope names in the column
names … with the marker names. Changing the column names makes the subsequent
analysis and plotting easier to read. … Plotting the data in two dimensions…
already shows that the cells can be grouped into subpopulations. Sometimes just
one of the markers can be used to define populations on its own; in that case,
simple rectangular gating is used to separate the populations. For instance,
CD4+ populations can be gating by taking the subpopulation with high values
for the CD4 marker. Cell clustering can be improved by carefully choosing
transformations of the data. … [Such a transformation] reveals
bimodality and the existence of two cell populations… It is standard
to transform both flow and mass cytometry data using one of several special
functions. We take the example of the inverse hyperbolic arcsine (asinh) …
for large values of x, asinh(x) behaves like the log and is practically
equal to log(x) + log(2); for small x the function is close to linear in
x. … This is another example of a variance-stabilizing transformation.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Consider a set of measurements that reflect some underlying true values
(say, species represented by DNA sequences from their genomes) but
have been degraded by technical noise. Clustering can be used to remove
such noise.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“In the bacterial 16SrRNA gene there are so-called variable regions that are
taxa-specific. These provide fingerprints that enable taxon identification. The
raw data are FASTQ-files with quality scored sequences of PCR-amplified DNA
regions. We use an iterative alternating approach to build a probabilistic noise
mode from the data. We call this a de novo method, because we use clustering,
and we use the cluster centers as our denoised sequence variants… After
finding all the denoised variants, we create contingency tables of their counts
across the different samples. … these tables can be used to infer properties
of the underlying bacterial communities using networks and graphs. <strong>In order to
improve data quality, we often have to start with the raw data and model all the
sources of variation carefully.</strong> We can think of this as an example of cooking
from scratch. … The DADA method … uses a parameterized model of substitution
errors that distinguishes sequencing errors from real biological variation. …
The dereplicated sequences are read in, and then divisive denoising and
estimation is run with the <code>dada</code> function… In order to verify that the error
transition rates have been reasonably well estimated, we inspect the fit between
the observed error rates … and the fitted error rates .. Once the errors have
been estimated, the algorithm is rerun on the data to find the sequence
variants. … Sequence inference removes nearly all substition and indel errors
from the data. [Footnote: ’The term indel stands for insertion-deletion; when
comparing two sequences that differ by a small stretch of characters, it is a
matter of viewpoint whether this is an insertion or a deletion, hence the name].
We merge the inferred forward and reverse sequences while removing paired
sequences that do not perfectly overlap, as a final control against residual
errors.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Chimera are sequences that are artificially created during the PCR
amplification by the melding of two (or, in rare cases, more) of the
original sequences. To complete our denoising workflow, we remove them
with a call to the function <code>removeBimeraDenovo</code>, leaving us with a
clean contingency table that we will use later.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“We load up the RNA-Seq dataset <code>airway</code>, which contains gene expression
measurements (gene-level counts) of four primary human airway smooth muscle
cell lines with and without treatment with dexamethasone, a synthetic
glucocorticoid. We’ll use the <code>DESeq2</code> method … it performs a test
for differential expression for each gene.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“In many cases, different variables are measured in different units, so they
have different baselines and different scales. [Footnote: ‘Common measures of
scale are the range and the standard deviation…’] For PCA and many other
methods, we therefore need to transform the numeric values to some common scale
in order to make comparisons meaningful. Centering means subtracting the mean,
so that the mean of the centered data is at the origin. Scaling or standardizing
then means dividing by the standard deviation, so that the new standard
deviation is 1. … To perform these operations, there is the R function
<code>scale</code>, whose default behavior when given a matrix or a dataframe is to make
each column have a mean of 0 and a standard deviation of 1. … We have already
encountered other data transformation choices in Chapters 4 and 5, where we used
the log and asinh functions. The aim of these transformations is (usually)
variance stabilization, i.e., to make the variances of the replicate
measurements of one and the same variable in different parts of the dynamic
range more similar. In contrast, the standardizing transformation described
above aims to make the scale (as measured by mean and standard deviation) of
<em>different</em> variables the same. Sometimes it is preferable to leave variables at
different scales because they are truely of different importance. If their
original scale is relevant, then we can (and should) leave the data alone. In
other cases, the variables have different precisions known a priori. We will see
in Chapter 9 that there are several ways of weighting such variables. After
preprocessing the data, we are ready to undertake data simplification through
dimension reduction.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>With data that give the number of reads for each gene in a sample, “The
data have a large dynamic range, starting from zero up to millions. The
variance and, more generally, the distribution shape of the data in different
parts of the dynamic range are very different. We need to take this
phenomenon, called heteroscedascity, into account. The data are non-negative
integers, and their distribution is not symmetric—thus normal or log-normal
distribution models may be a poor fit. We need to understand the systematic
sampling biases and adjust for them. Confusingly, such adjustment is often
called normalization. Examples are the total sequencing depth of an experiment
(even if the true abundance of a gene in two libraries is the same, we expect
different numbers of reads for it depending on the total number of reads
sequenced) and differing sampling probabilities (even if the true abundance of
two genes within a biological sample is the same, we expect different numbers
of reads for them if they have differing biophysical properties, such as length,
GC content, secondary structure, binding partners).” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Often, systematic biases affect the data generation and are worth taking
into account. Unfortunately, the term normalization is commonly used for that
aspect of the analysis, even though it is misleading; it has nothing to do
with the normal distribution, nor does it involve a data transformation.
Rather, what we aim to do is identify the nature and magnitude of
systematic biases and take them into account in our model-based analysis of the
data. The most important systematic bias [for count data from high-throughput
sequencing applications like RNA-Seq] stems from variations in the total number
of reads in each sample. If we have more reads for one library than for another,
then we might assume that, everything else being equal, the counts are
proportional to each other with some proportionality factor <em>s</em>. Naively,
we could propose that a decent estimate of <em>s</em> for each sample is simply
given by the sum of the counts of all genes. However, it turns out that we
can do better…” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“When testing for differential expression, we operate on raw counts and
use discrete distributions. For other downstream analyses—e.g., for
visualization or clustering—it can be useful to work with transformed versions
of the count data. Maybe the most obvious choice of transformation is the
logarithm. However, since count values for a gene can be zero, some analysts
advocate the use of pseudocounts, i.e., transformations of the form
y = log2(n + 1) or more generally y = log2(n + n0).” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“The data sometimes contain isolated instances of very large counts that
are apparently unrelated to the experimental or study design and may be
considered outliers. Outliers can arise for many reasons, including rare
technical or experimental artifacts, read mapping problems in the case of
genetically differing samples, and genuine but rare biological events. In
many cases, users appear primarily interested in genes that show consistent
behavior, and this is the reason why, by default, genes that are affected by such
outliers are set aside by <code>DESeq</code>. The function calculates, for every gene
and for every sample, a diagnostic test for outliers called Cook’s distance.
Cook’s distance is a measure of how much a single sample is influencing the
fitted coefficients for a gene, and a large value of Cook’s distance is
intended to indicate an outlier count. <code>DESeq2</code> automatically flags genes
with Cook’s distance above a cutoff and sets their p-values and adjusted
p-values to NA. … With many degrees of freedom—i.e., many more samples
than number of parameters to be estimated—it might be undesirable to remove
entire genes from the analysis just becuase their data include a single
count outlier. An alternative strategy is to replace the outlier counts with
the trimmed mean over all sample, adjusted by the size factor for that
sample. This approach is conservative: it will not lead to false positives,
as it replaces the outlier value with the value predicted by the null hypothesis.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Since the sampling depthy is typically different for different sequencing
runs (replicates), we need to estimate the effect of this variable parameter
and take it into account in our model. … Often this part of the analysis
is called normalization (the term is not particularly descriptive, but
unfortunately it is now well established in the literature).” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Our experimental interventions and our measurement instruments have limited
precision and accuracy; often we don’t know these limitations at the outset and
have to collect preliminary data to estimate them.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Our treatment conditions may have undesired but hard-to-avoid side effects;
our measurements may be overlaid with interfering signals or ‘background noise.’”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Sometimes we explicitly know about factors that cause bias, for instance, when
different reagent batches were used in different phases of the experiment. We
call these batch effects (Leek et al., 2010). At other times, we may expect that
such factors are at work but have no explicit record of them. We call these
latent factors. We can treat them as adding to the noise, and in Chapter
4 we saw how to use mixture models to do so. But this may not be enough; with
high-dimensional data, noise caused by latent factors tends to be correlated,
and this can lead to faulty inference (Leek et al., 2010). The good news is that
these same correlations can be exploited to estimate latent factors from
the data, model them as bias, and thus reduce the noise (Leek and Storey 2007;
Stegle et al. 2010).” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Regular noise can be modeled by simple probability models such as independent
normal distributions or Poissons, or by mixtures such as the gamma-Poisson or
Laplace. We can use relatively straightforward methods to take such noise into
account in our data analyses and to compute the probability of extraordinarily
large or small values. In the real world, this is only part of the story:
measurements can be completely off-scale (a sample swap, a contamination, or
a software bug), and they can all go awry at the same time (a whole microtiter
plate went bad, affecting all data measured from it). Such events are hard to model
or even correct for—our best chance of dealing with them is data quality
assessment, outlier detection, and documented removal.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“In Chapters 4 and 8 we saw examples of data transformations that compress or
stretch the space of quantitative measurements in such a way that the
measurements’ variance is more similar throughout. Thus the variance between
replicated measurements is no longer highly dependent on the mean value. The
mean-variance relationship of our data before transformation can, in
principle, be any function, but in many cases, the following prototypic relationships
are found, at least approximately: 1. Constant: the variance is independent
of the mean…; 2. Poisson: the variance is proproational to the mean…;
3. Quadratic: the standard deviation is proportional to the mean; therefore
the variance grows quadratically… The mean-variance relationship in real data
can also be a combination of these basic types. For instance, with DNA microarrays,
the fluorescence intensities are subject to a combination of background noise that
is largely independent of the signal, and multiplicative noise whose standard
deviation is proportional to the signal (Rocke and Durbin 2001). …
What is the point of applying a variance-stabilizing transformation? Analyzing the
data on the transformed scale tends to: 1. Improve visualization, since the
physical space on the plot is used more ‘fairly’ throughout the range of the
data. A similar argument applies to the color space in the case of a heatmap.
2. Improve the outcome of ordination methods such as PCA or clustering based
on correlation, as the results are not so much dominated by the signal from a few
very highly expressed genes, but more uniformly from many genes throughout the
dynamic range. 3. Improve the estimates and inference from statistical models
that are based on the assumption of identially distributed (and, hence,
homoscedastic) noise.”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“We distinguis between data quality assessment (QA)—steps taken to measure
and monitor data quality—and quality control—the removal of bad data.
These activities pervade all phases of an analysis, from assembling the raw
data over transformation, summarization, model fitting, hypothesis testing or
screening for ‘hits’ to interpretation. QA-related questions include:
1. How do the marginal distributions of the variables look (histograms,
ECDF plots)? 2. How do their joint distributions look (scatterplots, pair plots)?
3. How well do replicated agree (as compared to different biological conditions)?
Are the magnitudes of different between several conditions plausible?
4. Is there evidence of batch effects? These could be of a categorical (stepwise)
or continuous (gradual) nature, e.g., due to changes in experimental reagents,
protocols or environmental factors. Factors associated with such effects may
be explicitly known, or unknown and latent, and often they are somewhere in
between (e.g., when a measurement apparatus slowly degrades over time, and
we have recorded the times, but don’t really know exactly when the degradation
becomes bad). For the last two sets of questions, heatmaps, principal components
plots, and other ordination plots (as we have seen in Chapters 7 and 9) are
useful.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“It’s not easy to define quality, and the word is used with many meanings. The
most pertinent for us is fitness for purpose, and this contrasts with other
definitions that are based on normative specifications. For instance, in
differential expression analysis with RNA-Seq data, our purpose may be the
detection of differentially expressed genes between two biological conditions.
We can check specificiations such as the number of reads, read length, base
calling quality and fraction of aligned reads, but ultimately these measures in
isolation have little bearing on our purpose. More to the point will be the
identification of samples that are not behaving as expected, e.g., because of
a sample swap or degradation, or genes that were not measured properly.
… Useful plots include ordination plots … and heatmaps …
A quality metric is any value that we use to measure quality, and having
explicit quality metrics helps in automating QA/QC.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Use literate programming tools.</strong> Examples are Rmarkdown and Jupyter. This
makes code more readable (for yourself and for others) than burying
explanations and usage instructions in comments in the source code or in
separate README files. In addition, you can directly embed figures and tables
in these documents. Such documents are good starting points for the supplementary
material of your paper. Moreover, they’re great for reporting analyses to your
collaborators.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Use functions.</strong> It’s better than copy-pasting (or repeatedly source-ing)
stretches of code.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p><strong>Use the R package system.</strong> Soon you’ll note recurring function or variable
definitions that you want to share between your different scripts. It is fine to
use the R function <code>source</code> to manage them initially, but it is never too early
to move them into your own package—at the latest when you find yourself staring
to write emails or code comments explaining to others (or to yourself) how to use
some functionality. Assembling existing code into an R package is not hard, and it
offers you many goodies, including standardized ways of composing documentation,
showing code usage examples, code testing, versioning and provision to others.
And quite likely you’ll soon appreciate the benefits of using namespaces."
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Think in terms of cooking recipes and try to automate them.</strong> When developing
downstream analysis ideas that bring together several different data types, you
don’t want to do the conversion from data-type-specific formats into a
representation suitable for machine learning or a generic statistical method
each time anew, on an ad hoc basis. Have a recipe script that assembles the
different ingredients and cooks them up as an easily consumable [footnote:
In computer science, the term data warehouse is sometimes used for such a concept]
matrix, dataframe or Bioconductor <code>SummarizedExperiment</code>.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Centralize the location of the raw data files and automate the derivation of
intermediate data.</strong> Store the input data on a centralized file server that is
profesionally backed up. Mark the files as read-only. Have a clear and linear
workflow for computing the derived data (e.g., normalized, summarized, transformed,
etc.) from the raw files, and store these in a separate directory. Anticipate that
this workflow will need to be run several times, and version it. Use the
<code>BiocFileCache</code> package to mirror these files on your personal computer.
[footnote: A more basic alternative is the rsync utility. A popular solution offered
by some organizations is based on ownCloud. Commercial options are Dropbox,
Google Drive and the like].” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Keep a hyperlinked webpage with an index of all analyses.</strong> This is helpful
for collaborators (especially if the page and the analysis can be accessed via
a web browser) and also a good starting point for the methods part of your paper.
Structure it in chronological or logical order, or a combination of both.”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Getting data ready for analysis or visualization often involves a lot of
shuffling until they are in the right shape and format for an analytical
algorithm or a graphics routine.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Data analysis pipelines in high-throughput biology often work as ‘funnels’
that successively summarize and compress the data. In high-throughput
sequencing, we may start with individual sequencing reads, then align them to
a reference, then only count the aligned reads for each position, summarize
positions to genes (or other kinds of regions), then ‘normalize’ these numbers
by library size to make them comparable across libraries, etc. At each step,
we lose information, yet it is important to make sure we still have enough
information for the task at hand. [footnote: For instance, for the RNA-Seq
differential experession analysis we saw in Chapter 8, we needed the actual
read counts, not ‘normalized’ versions; for some analyses, gene-level summaries
might suffice, for others, we’ll want to look at the exon or isoform level.]
The problem is particularly acute if we build our data pipeline with a series
of components from separate developers. Statisticians have a concept for
whether certain summaries enable the reconstruction of all the relevant
information in the data: sufficiency. … Iterative approaches akin to what
we saw when we used the EM algorithm can sometimes help to avoid information
loss. For instance, when analyzing mass spectroscopy data, a first run
guesses at peaks individually for each sample. After this preliminary
spectrum-spotting, another iteration allows us to borrow strength from the
other samples to spot spectra that may have been overlooked (or looked like
noise) before.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<p>Try to avoid adding in supplementary / meta data “by hand.” For example, if you
need to add information about the ID of each sample, and this information is
included somewhere in the filename, it will be more robust to use regular
expressions to extract this information from the file names, rather than
entering it by hand. If you later add new files, the automated approach will
be robust to this update, while errors might be introduced for information
added by hand.</p>
<p>Example of quality control functionality in <code>xcms</code>:
“Below we create boxplots representing the distribution of total ion currents per file. Such plots can be very useful to spot problematic or failing MS runs.
… Also, we can cluster the samples based on similarity of their base peak chromatogram. This can also be helpful to spot potentially problematic samples in an experiment or generally get an initial overview of the sample grouping in the experiment. Since the retention times between samples are not exactly identical, we use the bin function to group intensities in fixed time ranges (bins) along the retention time axis. In the present example we use a bin size of 1 second, the default is 0.5 seconds. The clustering is performed using complete linkage hierarchical clustering on the pairwise correlations of the binned base peak chromatograms.”
<span class="citation">(Smith 2013)</span></p>
<p>After some quality checks on the data from LCMS, the next step is to
detect the chromatographic peaks in the samples. This requires some
specifications to the algorithm that will depend on the settings used
on the equipment when the samples were run.
“Next we perform the chromatographic peak detection using the centWave algorithm [2]. Before running the peak detection it is however strongly suggested to visually inspect e.g. the extracted ion chromatogram of internal standards or known compounds to evaluate and adapt the peak detection settings since the default settings will not be appropriate for most LCMS experiments. The two most critical parameters for centWave are the peakwidth (expected range of chromatographic peak widths) and ppm (maximum expected deviation of m/z values of centroids corresponding to one chromatographic peak; this is usually much larger than the ppm specified by the manufacturer) parameters. To evaluate the typical chromatographic peak width we plot the EIC for one peak.”
<span class="citation">(Smith 2013)</span></p>
<blockquote>
<p>“Peak detection will not always work perfectly leading to peak detection artifacts, such as overlapping peaks or artificially split peaks. The refineChromPeaks function allows to refine peak detection results by either removing identified peaks not passing a certain criteria or by merging artificially split chromatographic peaks.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“The time at which analytes elute in the chromatography can vary between samples (and even compounds). Such a difference was already observable in the extracted ion chromatogram plot shown as an example in the previous section. The alignment step, also referred to as retention time correction, aims at adjusting this by shifting signals along the retention time axis to align the signals between different samples within an experiment. A plethora of alignment algorithms exist (see [3]), with some of them being implemented also in xcms. The method to perform the alignment/retention time correction in xcms is adjustRtime which uses different alignment algorithms depending on the provided parameter class. … In some experiments it might be helpful to perform the alignment based on only a subset of the samples, e.g. if QC samples were injected at regular intervals or if the experiment contains blanks. Alignment method in xcms allow to estimate retention time drifts on a subset of samples (either all samples excluding blanks or QC samples injected at regular intervals during a measurement run) and use these to adjust the full data set.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“The final step in the metabolomics preprocessing is the correspondence that matches detected chromatographic peaks between samples (and depending on the settings, also within samples if they are adjacent). The method to perform the correspondence in xcms is groupChromPeaks. We will use the peak density method [5] to group chromatographic peaks. The algorithm combines chromatographic peaks depending on the density of peaks along the retention time axis within small slices along the mz dimension.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“The performance of peak detection, alignment and correspondence should always be evaluated by inspecting extracted ion chromatograms e.g. of known compounds, internal standards or identified features in general.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<p>Normalization can help adjust for technical bias across the samples:</p>
<blockquote>
<p>“At last we perform a principal component analysis to evaluate the grouping of the samples in this experiment. Note that we did not perform any data normalization hence the grouping might (and will) also be influenced by technical biases. … We can see the expected separation between the KO and WT samples on PC2. On PC1 samples separate based on their ID, samples with an ID &lt;= 18 from samples with an ID &gt; 18. This separation might be caused by a technical bias (e.g. measurements performed on different days/weeks) or due to biological properties of the mice analyzed (sex, age, litter mates etc).” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“Normalizing features’ signal intensities is required, but at present not (yet) supported in xcms (some methods might be added in near future). It is advised to use the SummarizedExperiment returned by the quantify method for any further data processing, as this type of object stores feature definitions, sample annotations as well as feature abundances in the same object. For the identification of e.g. features with significant different intensities/abundances it is suggested to use functionality provided in other R packages, such as Bioconductor’s excellent limma package.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
</div>
<div id="discussion-questions-5" class="section level3" number="3.1.6">
<h3><span class="header-section-number">3.1.6</span> Discussion questions</h3>

</div>
</div>
<div id="introduction-to-scripted-data-pre-processing-in-r" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Introduction to scripted data pre-processing in R</h2>
<p>We will show how to implement scripted pre-processing of experimental data
through R scripts. We will demonstrate the difference between interactive coding
and code scripts, using R for examples. We will then demonstrate how to create,
save, and run an R code script for a simple data cleaning task.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Describe what an R code script is and how it differs from interactive
coding in R</li>
<li>Create and save an R script to perform a simple data pre-processing task</li>
<li>Run an R script</li>
<li>List some popular packages in R for pre-processing biomedical data</li>
</ul>
<div id="subsection-1-4" class="section level3" number="3.2.1">
<h3><span class="header-section-number">3.2.1</span> Subsection 1</h3>
</div>
<div id="subsection-2-3" class="section level3" number="3.2.2">
<h3><span class="header-section-number">3.2.2</span> Subsection 2</h3>
</div>
<div id="applied-exercise-4" class="section level3" number="3.2.3">
<h3><span class="header-section-number">3.2.3</span> Applied exercise</h3>

</div>
</div>
<div id="simplify-scripted-pre-processing-through-rs-tidyverse-tools" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</h2>
<p>The R programming language now includes a collection of ‘tidyverse’ extension
packages that enable user-friendly yet powerful work with experimental data,
including pre-processing and exploratory visualizations. The principle behind
the ‘tidyverse’ is that a collection of simple, general tools can be joined
together to solve complex problems, as long as a consistent format is used for
the input and output of each tool (the ‘tidy’ data format taught in other
modules). In this module, we will explain why this ‘tidyverse’ system is so
powerful and how it can be leveraged within biomedical research, especially for
reproducibly pre-processing experimental data.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Define R’s ‘tidyverse’ system</li>
<li>Explain how the ‘tidyverse’ collection of packages can be both user-friendly
and powerful in solving many complex tasks with data</li>
<li>Describe the difference between base R and R’s ‘tidyverse.’</li>
</ul>
<div id="limitations-of-object-oriented-programming" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Limitations of object-oriented programming</h3>
<p>In previous sections, we described how the R programming language allows for
object-oriented programming, and how customized objects are often used in
preprocessing for biological data. This is a helpful approach for preprocessing,
because it can handle complexities in biological data at its early stages of
preprocessing, when R must handle complex input formats from equipment like
flow cytometers or mass spectrometers, and data sizes that are often very large.</p>
<p>However, once you have preprocessed your data, it is often possible to work with it
in a smaller, more consistent object type. This will give you a lot of flexibility
and power. While object-oriented approaches can handle complex data, it can be a
little hard to write and work with code that is built on an object oriented approach.
Working with this type of code requires you to keep track of what object type your
data is in at each stage of a code pipeline, as well as which functions can work with
that type of object.</p>
<p>Further, this type of coding, in practice at least, can be a bit inflexible.
Often, specific functions only work with a single or few types of functions. In
theory, object-oriented programming allows for <em>methods</em> that work in customized
ways with different types of objects to apply customized code to that type of
object for similar, common-sense results. For example, there are often <code>summary</code>
and <code>plot</code> methods for most types of objects, and these apply code that is
customized to that object type and output, respectively, summarized information
about the data in the object and a plot of the data in the object. However, when
you want to do more with the object that summarize it or create its default
plot, you often end up needing to move to more customized functions that work
only with a single or few object types. When you get to this point, you find that
you have to remember which functions work with which object type, and you have to
use different functions at different stages of your code pipeline, as your code
changes from one object class to another.</p>
<p>Further, many of these functions input one object type and output a different
one. This evolution of object types for storing data can be difficult to navigate
and keep track of. Different object types store data in different ways, and so
this evolution of data object types for storage can make it tricky to figure out
how to extract and explore data along the pipeline. It makes it hard to write your
own code to explore and visualize the data along the way, as well, and so users
are often restricted to the visualization and analysis functions pre-made and
shared in packages when working with data in complex object types, especially
until the user becomes very comfortable with coding in R.</p>
<p>Overall, what does this all mean? Object-oriented approaches offer real advantages
early in the process of pre-processing biological data, especially complex and
large data output from complex laboratory equipment. However, once this pre-processing
is completed, there is a big advantage in moving the data into a simple format
and then continuing coding, data analysis, and visualization using tools that
work with this simple format. This is the approach taken by a suite of R packages
called the “tidyverse,” as well as extensions that build off the approach that
this suite of tools embraces. This “tidyverse” approach is described in the
next section.</p>
</div>
<div id="the-tidyverse-approach" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> The “tidyverse” approach</h3>
<p>The term “elegance” often captures styles and approaches that are beautiful and
functional without unneeded extras or complexity. Engineers and scientists
sometimes use this term to capture approaches that achieve a desired result with
minimal complexity and friction. A coding problem, for example, could be solved
by an average coder with a hundred lines of code that get the job done, but a
very good coder might be able to solve the same problem with five lines of code
that are easy to follow. The second approach would be applauded as the “elegant”
solution. In mathematics, similarly, proofs can be complex and unwieldy, or they
can be simple and elegant—this idea was beautifully captured by the Hungarian
mathematician Paul Erdos, who famously described very elegant mathematical proofs
as being from “The Book”—that is, God’s own version of the proof of the
mathematical idea.</p>
<blockquote>
<p>“Paul Erdos liked to talk about The Book, in which God maintains the perfect
proofs for mathematical theorems, following the dictum of G. H. Hardy that
there is no permanent place for ugly mathematics. Erdos also said that you
need not believe in God but, as a mathematician, you should believe in
The Book.” [Proofs from the Book, Third Edition, Preface]</p>
</blockquote>
<p>The “tidyverse” approach in R is elegant. It is powerful, and gives you immense
flexibility once you’ve gotten the hang of it, but it’s also so straightforward
that the basics can be quickly taught to and applied by beginning coders. It
focuses on keeping data in a simple, standard format called “tidy” dataframes.
By keeping data in this format while working with it, common tools can be applied
that work with the data at any stage of a “tidy” coding pipeline. These tools take
a “tidy” dataframe as their input, and they also output a “tidy” dataframe, with
whatever change the function implements applied. Because each of these “tidyverse”
tools input and output data in the same standard format, they can be strung together
in order you want. By contrast, when functions input and output data in different
object types, they can only be joined in a specified order, because you can only
apply certain functions to certain object types.</p>
<p>Since the “tidyverse” tools can be strung together in any order, they can be
used very flexibly to build up to do interesting tasks. The tidyverse tools
generally each do very small and simple things. For example, one function
(<code>select</code>) just limits the data to a subset of its original columns; another
(<code>mutate</code>) adds or changes values in columns of the dataset, while another
(<code>distinct</code>) limits the dataframe to remove any rows that are duplicates. These
small, simple steps can be combined together in different patterns to add up to
complex operations on the data, while keeping each step very simple and clear.
Since the data stays in a standard and simple object type, it is easy to check
in on your data at any stage, as the common visualization tools for this
approach (from the <code>ggplot2</code> package and its extensions) can be always be
applied to data stored in a tidy dataframe.</p>
<p>The centralizing principal of the tidyverse approach is the format in which data
is stored throughout “tidyverse” coding—the tidy dataframe. We’ve described
this data type, including its rules and advantages, in an earlier module of this
book. Briefly, you can think of this format in two parts. First, there’s the R
object type that the data should be stored in—a basic “dataframe” object. The
dataframe object type is a very basic two-dimensional format for storing data in
R. When you print it out, it will remind you of looking at data in a
spreadsheet. The two dimensions—rows and columns—allow you to include data
for one or more observations, with different values that were measured for each.
For example, if you were conducting a study of children’s BMI and blood sugar,
you might have an observation for each child in the study, and values measured
for each child of height, weight, a blood sugar measure, study ID, and date of
the observation.</p>
<p>The two-dimensional structure of a dataframe keeps the values
measured for each observation lined up with each other, and lets you keep them
aligned as you work with the data. You could also store data for each value as
separate objects, in one-dimensional vectors, which you can visualize as strings
of values of the same data type, like the dates that each observation was made,
or the weight of each study subject. However, when the data is in separate
vectors, it is easy to make coding mistakes, and coding is often less efficient.
If you want to remove one observation, for example, because you find it is a
duplicate, you would need to carefully make sure you remove it correctly from
each vector. When data are stored in a dataframe, you can remove the row for
that observation with one command, and you can be sure that you’ve removed the
value you meant to from each of the measured values.</p>
<p>Sometimes, you’ll see that data in a tidyverse approach are stored in a special
type of dataframe called a “tibble”—this isn’t very different from a
dataframe, and in fact is a special type of dataframe. It’s only differences in
practice are that it has a slightly different <code>print</code> method. The <code>print</code> method
is the method that’s run, by default, when you just type the R object’s name
at the console. A tibble prints more nicely than a basic dataframe. By default,
it will only print the first few lines. By contrast, a dataframe will, by default,
print everything—if you have a lot of data, this can create an overwhelming
amount of output when you just want to check out what the data looks like. The
printout of a tibble will also include some interesting annotations to help you
see what’s in the data, including the dimensions of the full dataframe and the
data type of each column in the data.</p>
<p>The R object class—dataframe, and more specifically, tibble—of the standard
format for data for a tidyverse approach is just the first part of the standard
data format for the tidyverse approach. The second part of the standard format is
how you organize your data in this format. To easily work with tidyverse functions,
you’ll want to make sure that your data is stored within that dataframe following
“tidy” data principals. These are fully described in an earlier module in this
book [which module]. If you use this data format to initially collect your
data, as described in an earlier module, you will find it very easy to read the
data into R and work within the tidyverse approach. When working with larger and
more complex data collected from laboratory equipment, you may find you need to
do some preprocessing of the data using an object-oriented approach before you
can move the data into this tidy format, but at that point, you can continue with
analysis and visualization of your data using a tidyverse approach.</p>
</div>
<div id="how-to-tidyverse" class="section level3" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> How to “tidyverse”</h3>
<p>Once data are in the “tidy” data format, you can create a pipeline of code that
uses small tools, each of which does one simple thing, to work with the data.
This work can include cleaning the data, adding values that are functions of the
original values for each observation (e.g., adding a column with BMI based on
values for each observation on height and weight), applying statistical models to
test hypotheses, summarizing data to create tables, and visualizing the data.</p>
<p>The tidyverse approach is now widely taught, both in in-person courses at
universities and through a variety of online resources. One key resource for
learning the tidyverse approach for R is the book <em>R for Data Science</em> by Hadley
Wickham (the primary developer of the tidyverse) and Garrett Grolemund. This
book is available as a print edition through O’Reilly Media. It is also freely
available online at <a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a>. This book is geared to beginners in
R, moving through to get readers to an intermediate stage of coding expertise,
which is a level that will allow most scientific researchers to powerfully
work with their experimental data. The book includes exercises for practicing the
concepts, and a separate online book is available with solutions for the
exercises (<a href="https://jrnold.github.io/r4ds-exercise-solutions/" class="uri">https://jrnold.github.io/r4ds-exercise-solutions/</a>).</p>
<p>[More on other resources for learning the tidyverse.]</p>
<p>Since there are so many excellent resources available—many for free—to learn
how to code in R using the tidyverse approach, we consider it beyond the scope
of these modules to go more deeply into these instructions. However, we do think
it is critical that biological researchers learn how to connect this approach to
the type of coding that is often necessary for pre-processing large and complex
data that is output from laboratory equipment. Through many of the modules in this book,
we provide advice on how to make these connections, so that data from different
sources—including different types of laboratory equipment and hand-recorded data
collected by personnel in the lab, like colony forming units measured from plating
samples—can all be connected in a tidyverse pipeline by recording hand-recorded
data following a tidy format and by pre-processing data with the aim of moving
data toward a tidy dataframe that can be integrated with other “tidy” data for
analysis and visualization.</p>
</div>
<div id="subsection-1-5" class="section level3" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Subsection 1</h3>
<blockquote>
<p>“There is a now-old trope in the world of programming. It’s called the ‘worse is
better’ debate; it seeks to explain why the Unix operating systems (which include
Mac OS X these days), made up of so many little interchangeable parts, were so much
more successful in the marketplace than LISP systems, which were ideologically pure,
based on a single languagae (again, LISP), which itself was exceptionally simple,
a favorite of ‘serious’ hackers everywhere. It’s too complex to rehash here, but one
of the ideas inherent within ‘worse is better’ is thata systems made up of many
simple pieces that can be roped together, even if those pieces don’t share a consistent
interface, are likely to be more successful than systems that are designed with consistency
in every regard. And it strikes me that this is a fundamental drama of new technologies.
Unix beat out the LISP machines. If you consider mobile handsets, many of which run
descendants of Unit (iOS and Andriod), Unix beat out Windows as well. And HTML5 beat out
all of the various initiatives to create a single unified web. It nods to accessibility:
it doesn’t get in the way of those who want to make something huge and interconnected.
But it doesn’t enforce; it doesn’t seek to change the behavior of page creators in the
same way that such lost standards as XHTML 2.0 (which eremged from the offices of
the World Wide Web Consortium, and then disappeared under the weight of its own
intentions) once did. It’s not a bad place to end up. It means that there is no
single framework, no set of easy rules to lear, no overarching principles that,
once learned, can make the web appear like a golden statue atop a mountain. There
are just components: HTML to get the words on the page, forms to get people to
write in, videos and images to put up pictures, moving or otherwise, and
JavaScript to make everything dance.” <span class="citation">(Ford 2014)</span></p>
</blockquote>
<blockquote>
<p>“One of the fundamental contributions of the Unix system [is] the idea of a <em>pipe</em>.
A pipe is a way to connect the output of one program to the input of another program
without any temporary file; a <em>pipeline</em> is a connection of two or more programs through
pipes. … Any program that reads from a terminal can read from a pipe instead; any program
that writes on the terminal can write to a pipe. … The programs in a pipeline actually
run at the same time, not one after another. This means that the programs in a pipeline
can be interactive; the kernel looks after whatever scheduling and synchronization is needed
to make it all work. As you probably suspect by now, the shell arranges things when you
ask for a pipe; the individual programs are oblivious to the redirection.” <span class="citation">(Kernighan and Pike 1984)</span></p>
</blockquote>
<blockquote>
<p>“Even though the Unix system introduces a number of innovative programs and techniques,
no single program or idea makes it work well. Instead, what makes it effective is an approach
to programming, a philosophy of using the computer. Although that philosophy can’t be written
down in a single sentence, at its heart is the idea that the power of a system comes more from
the relationships among programs than from the programs themselves. Many Unix programs do
quite trivial things in isolation, but, combined with other programs, become general and
useful tools.” <span class="citation">(Kernighan and Pike 1984)</span></p>
</blockquote>
<blockquote>
<p>“What is ‘Unix?’ In the narrowest sense, it is a time-sharing operating system <em>kernel</em>:
a program that controls the resources of a computer and allocates them among its users.
It lets users run their programs; it controls the peripheral devices (discs, terminals,
printers, and the like) connected to the machine; and it provides a file system that
manages the long-term storage of information such as programs, data, and documents.
In a broader sense, ‘Unix’ is often taken to include not only the kernel, but also
essential programs like compiles, editors, command languages, programs for copying and
printing files, and so on. Still more broadly, ‘Unix’ may even include programs
develpoed by you or others to be run on your system, such as tools for document
preparation, routines for statistical analysis, and graphics packages.” <span class="citation">(Kernighan and Pike 1984)</span></p>
</blockquote>
<blockquote>
<p>“A common observation is that more of the data scientist’s time is occupied
with data cleaning, manipulation, and ‘munging’ than it is with actual
statistical modeling (Rahm and Do, 2000; Dasu and Johnson, 2003). Thus, the
development of tools for manipulating and transforming data is necessary for
efficient and effective data analysis. One important choice for a data scientist
working in R is how data should be structured, particularly the choice of
dividing observations across rows, columns, and multiple tables. The concept of
‘tidy data,’ introduced by Wickham (2014a), offers a set of guidelines for
organizing data in order to facilitate statistical analysis and visualization. … This framework makes it easy for analysts to reshape, combine, group and otherwise manipulate data. Packages such as ggplot2, dplyr, and many built-in R modeling and plotting functions require the input to be in a tidy form, so keeping the data in this form allows multiple tools
to be used in sequence in a seamless analysis pipeline (Wickham, 2009; Wickham and Francois,
2014).”
<span class="citation">(D. Robinson 2014)</span></p>
</blockquote>
</div>
<div id="subsection-2-4" class="section level3" number="3.3.5">
<h3><span class="header-section-number">3.3.5</span> Subsection 2</h3>
</div>
<div id="practice-quiz-2" class="section level3" number="3.3.6">
<h3><span class="header-section-number">3.3.6</span> Practice quiz</h3>

</div>
</div>
<div id="complex-data-types-in-experimental-data-pre-processing" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Complex data types in experimental data pre-processing</h2>
<p>Raw data from many biomedical experiments, especially those that use
high-throughput techniques, can be very large and complex. Because of the scale
and complexity of these data, software for pre-processing the data in R often
uses complex, ‘untidy’ data formats. While these formats are necessary for
computational efficiency, they add a critical barrier for researchers wishing to
implement reproducibility tools. In this module, we will explain why use of
complex data formats is often necessary within open source pre-processing
software and outline the hurdles created in reproducibility tool use among
laboratory-based scientists.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Explain why R software for pre-processing biomedical data often stores
data in complex, ‘untidy’ formats</li>
<li>Describe how these complex data formats can create barriers to
laboratory-based researchers seeking to use reproducibility tools for
data pre-processing</li>
</ul>
<div id="subsection-1-6" class="section level3" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Subsection 1</h3>
<p>In previous modules, we have gone into a lot of detail about all of the advantages of the tidyverse approach. However as you work with biomedical data, particularly complex data from complex research equipment like mass spectrometers and Flow cytometers, you may find that it is unreasonable to start with a tiny burst approach from the first steps of pre processing the data. In this module we will explain why the tidyverse approach is currently not appropriate throughout all steps of pre-processing, analysis, and visualization of the types of data that you may collect through a biomedical research experiment. We will present some of the approaches and data storage methods used in passages in the bioconductor project, as well as explain some more about the purpose an approach of bioconductor. This will include an explanation of the more complex structures that are used to store data through many of the packages in by a conductor. These largely leverage a system in our call BF4 object-oriented system. We will cover several of the most popular object classes that are used to store data for bioconductor packages. Any leader module we will explain how early steps in data preprocessing, which use the bioconductor approach and bioconductor data storage objects, can be combined and transferred during it works slow to convert to using a tidyverse approach leader in the workflow when it is appropriate to store data in simpler structures like dataframes.</p>
<p>When you process data using a programming language, there will be different
structures that you can use to store data as you work with it. In other modules,
we’ve discussed the “tidyverse” approach to processing data in R—this approach
emphasizes the dataframe as a way to store data while you’re working with it.
In fact, its use of this data structure for data storage is one of the defining
features of the “tidyverse” approach.</p>
<p>Data in R can be stored in a variety of other formats, too. When you are working
with biological data—in particular, complex or large data output from laboratory
equipment—there can be advantages to using data structures besides dataframes.
In this section, we’ll discuss some of the complex characteristics of biomedical
data that recommend the use of data structures in R beyond the dataframe. We’ll
also discuss how the use of these other data structures can complicate the use of
“tidyverse” functions and principles that you might learn in beginning R programming
courses and books. In later modules, we’ll discuss how to connect your work in R
to clean and analyze data by performing earlier pre-processing steps using more
complex data structures and then transferring when possible to dataframes for
storing data, to allow you to take advantage of the power and ease of the
“tidyverse” approach as early as possible in your pipeline.</p>
</div>
<div id="complex-versus-simple-structures-for-storing-data" class="section level3" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Complex versus simple structures for storing data</h3>
<p>The R programming language offers a wide variety of structures that can be
used to store data as you work with it, including steps of preprocessing
and analysis of the data. Some of these structures are defined through the
base R language that you first install, while other structures are specially
defined through the extension R packages you add as you continue to work
with R. These packages are specific to the tasks you aim to do, and if they
define their own data storage structures, those structures are typically
customized to that task.</p>
<p>For example, there are packages—including the <code>xcms</code> package, for
example—that allow you to load and preprocess data from LC-MS experiments.
These packages include functionality to load data from a specialized format
output by mass spectometry equipment, as well as identify and align peaks within
the data that might indicate, for example, metabolite features for a
metabolomics analysis. The <code>xcms</code> package defines its own structures that are
used to store data during this preprocessing, and also draws on specialized data
structures defined in other R extension packages, including the <code>OnDiskMSnExp</code>
data object class that is defined by the <code>MSnbase</code> package.</p>
<p>Complex data structures like these can be very precise in defining what types of
data they contain and where each component of the data goes. Later in this and
other modules, we will provide more details about the advantages and
disadvantages of these types of specialized data storage formats, especially in
the context of improving transparency, rigor, and reproducibility across the
steps of preprocessing experimental biomedical data.</p>
<p>By contrast to these more complex data formats, there are a number of simple,
general purpose data structures that are often used to store data in R. These
include <strong>vectors</strong>, which are used to store one-dimensional strings of data of
a single type (e.g., all numeric, or all character strings), <strong>matrices</strong>, which
are also used to store data of a single type, but with a two-dimensional
structure, and <strong>dataframes</strong>, which are used to store multiple vectors of the
same length, and so allow for storing measurements of different data types for
multiple observations.</p>
<p>[Figure: examples of these three structures]</p>
<p>As you learn R, you will almost certainly learn how to create and work with
these more general data formats, including how to explore the data stored in
each of them. By contrast, you may never learn many of the more complex data
storage formats, especially if you are not using packages from Bioconductor.
However, there are a number of good reasons why R packages—especially those
shared through Bioconductor—define and use more complex data formats. In this
and following modules, we will explain the advantages and disadvantages of
complex versus simpler data storage formats in R. We will also explain how these
advantages and disadvantages weigh out differently in different stages of a data
preprocessing and analysis workflow. Finally, we will describe how you can
leverage both to your advantage, and in particular the tools and approaches that
you can use to shift from a Bioconductor-style approach—with heavy use of
complex data storage formats—early in your preprocessing pipeline to a
tidyverse approach—centered on storing data in a simple, tidy dataframe
object—at later stages, when the data are more suitable to this simpler
storage format, which allows you to leverage the powerful and widely-taught
tidyverse approach in later steps of analysis and visualization.</p>
<p>In these modules, we will focus on explaining these ideas within the R
programming language. This language is a very popular one for both biomedical
data sets and also for more general tasks in data management and analysis.
However, these principles also apply to other programming languages,
particularly those that can be used in an interactive format, including Python
and Julia.</p>
</div>
<div id="advantages-of-complex-structures-for-storing-data" class="section level3" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Advantages of complex structures for storing data</h3>
<p>There are two main features of biomedical data—in particular, data collected
from laboratory equipment like flow cytometers and mass spectrometers—that
make it useful to use more complex data structures in R in the earlier stages of
preprocessing the data. First, the data are often very large, in some cases so
large that it is difficult to read them into R. Second, the data might combine
various elements, each with their own natural structures, that you’d like to
keep together as you move through the steps of preprocessing the data.</p>
<p>[Data size, on-disk backends for files, like HDF5 and netCDF—used for flow
cytometry file format?]</p>
<p>[Potential future direction—developments of tidyverse based front ends for
data stored in databases or on-disk file formats—<code>sergeant</code> package is one
example, also running tidyverse commands on data in database, <code>matter</code> package?,
<code>disk.frame</code> package?]</p>
<blockquote>
<p>“Reading in a large dataset for which you do not have enough RAM is one easy way to freeze up your computer (or at least your R session). This is usually an unpleasant experience that usually requires you to kill the R process, in the best case scenario, or reboot your computer, in the worst case.” <span class="citation">(R. D. Peng 2016)</span></p>
</blockquote>
<blockquote>
<p>“Input/output (I/O) is the technical term for reading and writing data: the process of getting information into a particular computer system (in this case R) and then exporting it to the ‘outside world’ again (in this case as a file format that other software can read). Data I/O will be needed on projects where data comes from, or goes to, external sources. However, the majority of R resources and documentation start with the optimistic assumption that your data has already been loaded, ignoring the fact that importing datasets into R, and exporting them to the world outside the R ecosystem, can be a time-consuming and frustrating process. Tricky, slow or ultimately unsuccessful data I/O can cripple efficiency right at the outset of a project. Conversely, reading and writing your data efficiently will make your R projects more likely to succeed in the outside world.” <span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“There are circumstances when datasets become too large to read directly into
R. Reading in a 4 GB text file using the functions tested above, for example,
consumes all available RAM on a 16 GB machine. To overcome this limitation,
external stream processing tools can be used to preprocess large text files. The
following command, using the Linux command line ‘shell’ (or Windows based Linux
shell emulator Cygwin) command split, for example, will break a large multi GB
file into many chunks, each of which is more manageable for R: <code>split -b100m bigfile.csv</code> The result is a series of files, set to 100 MB each with the
<code>-b100m</code> argument in the above code. By default these will be called xaa, xab
and can be read in one chunk at a time (e.g. using read.csv(), fread() or
read_csv(), described in the previous section) without crashing most modern
computers. Splitting a large file into individual chunks may allow it to be read
into R. This is not an efficient way to import large datasets, however, because
it results in a non-random sample of the data this way. A more efficient, robust
and scalable way to work with large datasets is via databases, covered in
Section 6.6 in the next chapter.” <span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“Random access memory (RAM) is a type of computer memory that can be accessed
randomly: any byte of memory can be accessed without touching the preceding
bytes. RAM is found in computers, phones, tablets and even printers. The amount
of RAM R has access to is incredibly important. Since R loads objects into RAM,
the amount of RAM you have available can limit the size of data set you can
analyse.” <span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“A rough rule of thumb is that your RAM should be three times the size of your
data set.” <span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“RAM is cheap and thinking hurts.” – Uwe Ligges (about memory requirements in
R) R-help (June 2007) <code>fortunes::fortune(192)</code>, also quoted in
<span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“R comes in two versions: 32 -bit and 64 -bit. Your operating system also
comes in two versions, 32 -bit and 64 -bit. Ideally you want 64 -bit versions of
both R and the operating system. Using a 32 -bit version of either has severe
limitations on the amount of RAM R can access. So when we suggest that you
should just buy more RAM, this assumes that you are using a 64 -bit operating
system, with a 64 -bit version of R. [Note: If you are using an OS version from
the last five years, it is unlikely to be 32 -bit OS.] A 32 -bit machine can
access at most only 4 GB of RAM. Although some CPUs offer solutions to this
limitation, if you are running a 32 -bit operating system, then R is limited to
around 3 GB RAM. If you are running a 64 -bit operating system, but only a 32
-bit version of R, then you have access to slightly more memory (but not much).
Modern systems should run a 64 -bit operating system, with a 64 -bit version of
R. Your memory limit is now measured as 8 terabytes for Windows machines and 128
TB for Unix-based OSs.” <span class="citation">(Gillespie and Lovelace 2016)</span></p>
</blockquote>
<blockquote>
<p>“The OnDiskMSnExp class extends MSnExp and inherits all of its functionality
but is aimed to use as little memory as possible based on a balance between
memory demand and performance. Most of the spectrum-specific data, like
retention time, polarity, total ion current are stored within the object’s
featureData slot. The actual M/Z and intensity values from the individual
spectra are, in contrast to MSnExp objects, not kept in memory (in the assayData
slot), but are fetched from the original files on-demand. Because mzML files are
indexed, using the mzR package to read the relevant spectrum data is fast and
only moderately slower than for in-memory MSnExp.” <span class="citation">(Gatto 2013)</span></p>
</blockquote>
<blockquote>
<p>“[For OnDiskMSnExp:] To keep track of data manipulation steps that are applied
to spectrum data (such as performed by methods removePeaks or clean) a lazy
execution framework was implemented. Methods that manipulate or subset a
spectrum’s M/Z or intensity values can not be applied directly to a OnDiskMSnExp
object, since the relevant data is not kept in memory. Thus, any call to a
processing method that changes or subset M/Z or intensity values are added as
ProcessingStep items to the object’s spectraProcessingQueue. When the spectrum
data is then queried from an OnDiskMSnExp, the spectra are read in from the file
and all these processing steps are applied on-the-fly to the spectrum data
before being returned to the user. The operations involving extracting or
manipulating spectrum data are applied on a per-file basis, which enables
parallel processing. Thus, all corresponding method implementations for
OnDiskMSnExp objects have an argument BPPARAM and users can set a
PARALLEL_THRESH option flag 2 that enables to define how and when parallel
processing should be performed (using the BiocParallel package). Note that all
data manipulations that are not applied to M/Z or intensity values of a spectrum
(e.g. sub-setting by retention time etc) are very fast as they operate directly
to the object’s featureData slot.” <span class="citation">(Gatto 2013)</span></p>
</blockquote>
<blockquote>
<p>“The distinction between MSnExp and OnDiskMSnExp is often not explicitly
stated as it should not matter, from a user’s perspective, which data structure
they are working with, as both behave in equivalent ways. Often, they are
referred to as in-memory and on-disk MSnExp implementations.”
<span class="citation">(Gatto 2013)</span></p>
</blockquote>
<blockquote>
<p>“Big data is encountered in genomics for two reasons: the size of the genome
and the heterogeneity of populations. Complex organisms, such as plants and
animals, have genomes on the order of billions of base pairs (the human genome
consists of over three billion base pairs). The diversity of populations,
whether of organisms, tissues or cells, means we need to sample deeply to detect
low frequency events. To interrogate long and/or numerous genomic sequences,
many measurements are necessary. For example, a typical whole genome sequencing
experiment will consist of over one billion reads of 75–100 bp each. The reads
are aligned across billions of positions, most of which have been annotated in
some way. This experiment may be repeated for thousands of samples. Such a data
set does not fit within the memory of a current commodity computer, and is not
processed in a timely and interactive manner. To successfully wrangle a large
data set, we need to intimately understand its structure and carefully consider
the questions posed of it.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“To compare data across samples, we often summarize experimental annotations
over a set of reference features to yield a feature-by-sample matrix. For
example, we might count read alignments overlapping a common set of genes across
a number of samples. Larger matrices often arise in genetics, where thousands of
samples are compared over millions of SNPs, positions that are known to vary
within a population. In every case, the summaries are tied to a genomic range.”
<span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“The strengths of R are also its weaknesses: the R API encourages users to
store entire data sets in memory as vectors. These vectors are implicitly and
silently copied to achieve copy-on-write semantics, contribuing to high memory
usage and poor performance.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“There are general strategies for handling large genomic data that are well
suited to R programs. Sometimes the analyst is only interested in one aspect of
the data, such as that overlapping a single gene. In such cases, restricting the
data to that subset is a valid and effective means of data reduction. However,
once our interests extend beyond a single region or the region becomes too
large, resource constraints dictate that we cannot load the entire data set into
memory at once, and we need to iterate over the data to reduce them to a set of
interpretable summaries. Iteration lends itself to parallelism, that is,
computing on multiple parts of the same problem simultaneously. Thus, in
addition to meeting memory constraints, iteration lets us leverage additional
processing resources to reduce overall computation time. Investing in additional
hardware is often more economical than investment in software optimization. This
is particularly relevant in scientific computing, where we are faced with a
diverse, rapidly evolving set of unsolved problems, each requiring specialized
software. The costs of investment in general purpose hardware are amortized over
each problem, rather than paid each time for software optimization. This also
relates to maintainability: optimization typically comes at a cost of increased
code complexity. Many types of summary and filter operations are cheap to
implement in parallel because the data partitions can be processed
independently. We call this type of operation embarrassingly parallel. For
example, the counting of reads overlapping a gene does not depend on the
counting for a different gene.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“Our ultimate goal is to process and summarize a large data set in its
entirety, and iteration enables this by limiting the resource commitment at a
given point in time. Limiting resource consumption generalizes beyond iteration
and is a fundamental technique for computing with big data. In many cases, it
may render iteration unnecessary. Two effective approaches for being frugal with
data are restriction and compression. Restriction means controlling which data
are loaded and lets us avoid wasting resources on irrelevant or excessive data.
Compression helps by representing the same data with fewer resources.”
<span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“A special mode of restriction is to randomly generate a selection of records.
Down-sampling can address many questions, especially during quality assessment
and data exploration. For example, short reads are initially summarized in FASTQ
files containing a plain text representation of base calls and corresponding
quality scores. Basic statistics of quality assessment such as the nucleotide
count as a function of sequencing cycle or overall GC content are very well
characterized by random samples of a million reads, which might be 1% of the
data. This sample fits easily in memory. Computations on this size of data are
very nimble, enabling interactive exploration on commodity computers. An
essential requirement is that the data represent a random sample. The ShortRead
package is designed for the QA and exploratory analysis of the output from
high-througput sequencing instruments. It defines the FastqSampler object, which
draws random samples from FASTQ files.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“An example of a situation where random sampling does not work is when
prototyping a statistical method that depends on a significant amount of data to
achieve reasonable power. Variant calling is a specific example: restricting the
number of reads would lead to less coverage, less power and less meaningful
results. Instead, we need to restrict the analysis to a particular region and
include all of the reads falling within it. To optimize range-based queries, we
often sort and index our data structures by genomic coordinates. We should
consider indexing an investment because an index is generally expensive to
generate but cheap to query. The justification is that we will issue a
sufficient number of queries to outweigh the initial generation cost. Three
primary file formats follow this pattern: BAM, Tabix and BigWig [7, 10]. Each
format is best suited for a particular type of data. The BAM format is specially
designed for sequence alignments and stores the complex alignment structure, as
well as the aligned sequence. Tabix is meant for indexing general rangebased
annotations stored in tabular text files, such as BED and GFF. Finally, BigWig
is optimized for storing genome-length vectors, such as the coverage from a
sequencing experiment. BAM and Tabix compress the primary data with block-wise
gzip compression and save the index as a separate file. BigWig files are
similarly compressed but are self-contained. The Rsamtools package is an
interface between R and the samtools library, which implements access to BAM,
Tabix and other binary file formats. Rsamtools enables restriction of BAM
queries through the ScanBamParam object. This object can be used as an argument
to all BAM input functions, and enables restriction to particular fields of the
BAM file, to specific genomic regions of interest and to properties of the
aligned reads (e.g., restricting input to paired-end alignments that form proper
pairs).” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“One common scenario in high-throughput sequencing is the calculation of
statistics such as coverage (the number of short sequence reads overlapping each
nucleotide in the genome). The data required for this calculation usually come
from very large BAM files containing alignment coordinates (including the
alignment “cigar”), sequences and quality scores for tens of millions of short
reads. Only the smallest element of these data, the alignment coordinates, is
required for calculation of coverage. By restricting input to alignment
coordinates, we transform the computational task from one of complicated memory
management of large data to simple vectorized operations on in-memory objects."
<span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“Some vectors, in particular, the coverage, have long stretches of repeated
values, often zeroes. An efficient compression scheme for such cases is
run-length encoding. Each run of repeated values is reduced to two values: the
length of the run and the repeated value. This scheme saves space and also
reduces computation time by reducing computation size. For example, the vector
0, 0, 0, 1, 1, 5, 5, 5 would have run-values 0, 1, 5 and run-lengths 3, 2, 3.
The data have been reduced from a size of 8 to a size of 6 (3 values plus 3
lengths). The IRanges Rle class is a run-length encoded vector that supports the
full R vector API on top of the compressed representation. Operations on an Rle
gain efficiency by taking advantage of the compression. For example, the sum
method computes a sum of the run values, using the run lengths as weights. Thus,
the time complexity is on the order of the number of runs, rather than the
length of the vector.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“The Biostrings package [12] provides XStringViews for views on top of DNA,
RNA and amino acid sequences. XString is a reference, rather than a value as is
typical in R, so we can create multiple XStringViews objects without copying the
underlying data. This is an application of the <em>fly-weight</em> design pattern:
multiple objects decorate the same primary data structure, which is stored only
once in memory.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“Iterative summarization of data may be modeled as three separate steps:
split, apply and combine [15]. The split step is typically the only one that
depends on the size of the input data. The apply step operates on data of
restricted size, and it should reduce the data to a scale that facilitates
combination. Thus, the most challenging step is the first: splitting the data
into chunks small enough to meet resource constraints. Two modes of splitting
are particularly applicable to genomic data: sequential chunking and genomic
partitioning. Sequential chunking is a popular and general technique that simply
loads records in fixedcount chunks, according to the order in which they are
stored. Genomic partitioning iterates over a disjoint set of ranges that cover
the genome. Typical partitioning schemes include one range per chromosome and
sub-chromosomal ranges of some uniform size. Efficient range-based iteration,
whether over a partitioning or list of interesting regions, depends on data
structures, file formats and algorithms that are optimized for range-based
queries.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“As an alternative to streaming over chunks, we can iterate over a
partitioning of the genome or other domain. Genomic partitioning can be
preferable to streaming when we are only interested in certain regions. The
tileGenome function is a convenience for generating a set of ranges that
partition a genome. … A caveat with partitioning is that since many query
algorithms return ranges with any overlap of the query, care must be taken to
intersect the results with each partition, so that reads are not double counted,
for example.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“The Bioconductor project distributes the software as a number of different R
packages, including Rsamtools, IRanges, GenomicRanges, GenomicAlignments,
Biostrings, rtracklayer, biovizBase and BiocParallel. The software enables the
analyst to conserve computational resources, iteratively generate summaries and
visualize data at arbitrary levels of detail. These advances have helped to
ensure that R and Bioconductor remain relevant in the age of high-throughput
sequencing. We plan to continue in this direction by designing and implementing
abstractions that enable user code to be agnostic to the mode of data storage,
whether it be memory, files or databases. This will bring much needed agility to
resource allocation and will enable the user to be more resourceful, without the
burden of increased complexity.” <span class="citation">(Lawrence and Morgan 2014)</span></p>
</blockquote>
<blockquote>
<p>“The most widely used partitional clustering algorithm is k-means [6–8]. The
algorithm partitions N cells into k clusters each represented by a centroid, or
mean profile, for the cells in the kth cluster. This algorithm is commonly used
not only on its own, but also as a component of ensemble clustering [9, 10].
While k-means is easy to implement, it assumes that the user has enough
computational resources (specifically RAM) to store the data and all
intermediate calculations into memory. However, file sizes generated from
scRNA-seq experiments can be on the order of tens to hundreds of gigabytes. For
large enough data, k-means can be slow or completely fail if a user lacks
sufficient computational resources. Ensemble clustering approaches that depend
on the use of k-means [9, 10] run it multiple times (e.g., with different
parameter values or on a different data subset) limiting the usability of these
packages for large scRNA-seq datasets [11]. We note that our goal here is not to
debate the relative merits of k-means as a clustering algorithm—k-means is a
well-established method, which has been thoroughly investigated [12]—but to
provide users with the ability to use the popular k-means algorithm on large
single-cell datasets. To address the problems of using k-means with large data,
two solutions are (1) parallelization and (2) subsampling. Parallelization
approaches typically leverage some combination of (i) MapReduce [13] concepts to
handle a large volume of data over a distributed computing environment [14, 15],
(ii) k-dimensional (k-d) trees to either optimize for the nearest centroid [16]
or to partition datasets into subsets, representative of the larger dataset
[17], and (iii) leverage multi-core processors [18]. While these approaches do
improve the speed of k-means, they can be limited to the number of reducers for
each centroid and can often require extensive computational resources. In
contrast, subsampling approaches, such as the mini-batch k-means algorithm [19]
work on small, random subsamples of data (“mini batches”) that can fit into
memory on standard computers. We would emphasize, however, that while mini-batch
k-means only operates on small subsamples of the data at any one time, the
algorithm still minimizes the same global objective function evaluated over all
samples as in traditional implementations of k-means. Current implementations of
the mini-batch k-means algorithm [19] are available in standard programming
languages such as in the scikit-learn machine learning Python library [20] or in
the ClusterR R package [21]. However, these implementations either implicitly or
explicitly require all the data to be read into memory, and therefore do not
leverage the potential of the algorithm to provide a low memory footprint. To
address the described problems, we implemented the mini-batch k-means clustering
algorithm in the open-source mbkmeans R package [22], providing fast, scalable,
and memory-efficient clustering of scRNA-seq data in the Bioconductor framework
[5, 23]. Like existing implementations, our package can be applied to in-memory
data input for smaller datasets, but also to on-disk data, such as from the HDF5
file format [24], which is widely used for distributing single-cell sequencing
data. For on-disk input, mbkmeans leverages the subsampling structure of the
algorithm to read into memory only the current ‘mini batch’ of data at any given
point, thereby greatly reducing the required memory (RAM) needed. … Our
contribution is two-fold: we implement a mini-batch k-means algorithm for
on-disk data, and we benchmark the performance of a non-trivial algorithm for
HDF5 against its in-memory counterpart." <span class="citation">(Hicks et al. 2021)</span></p>
</blockquote>
<blockquote>
<p>“The mbkmeans software package implements the mini-batch k-means clustering
algorithm described above and works with matrix-like objects as input.
Specifically, the package works with standard R data formats that store the data
in memory, such as the standard matrix class in base R and sparse and dense
matrix classes from the Matrix R package [29], and with file-backed matrices,
e.g., by using the HDF5 file format [24]. In addition, the package provides
methods to interface with standard Bioconductor data containers such as the
SummarizedExperiment [30] and SingleCellExperiment [31] classes. We implemented
the computationally most intensive steps of our algorithm in C++, leveraging the
Rcpp [32] and beachmat [33] packages. Furthermore, we make use of Bioconductor’s
DelayedArray [34] framework, and in particular the HDF5Array [35] package to
interface with HDF5 files. The mbkmeans package was built in a modular format
that would allow it to easily operate on alternative on-disk data
representations in the future. To initialize the k centroids, the mbkmeans
package uses the k-means++ initialization algorithm [36] with a random subset of
b observations (the batch size), by default. Finally, to predict final cluster
labels, we use block processing through the DelayedArray [34] package to avoid
working with all the data at once.” <span class="citation">(Hicks et al. 2021)</span></p>
</blockquote>
<blockquote>
<p>“A major challenge in the analysis of scRNA-seq data is the scalability of analysis methods as datasets increase in size over time. This is particularly problematic as experiments now frequently produce millions of cells [50–53], possibly across multiple batches, making it challenging to even load the data into memory and perform downstream analyses including quality control, batch correction and dimensionality reduction. Providing analysis methods, such as unsupervised clustering, that do not require data to be loaded into memory is an imperative step for scalable analyses. While large-scale scRNA-seq data are now routinely stored in on-disk data formats (e.g. HDF5 files), the methods to process and analyze these data are lagging.” <span class="citation">(Hicks et al. 2021)</span></p>
</blockquote>
<blockquote>
<p>“Unlike other existing implementations of mini-batch k-means, our algorithm harnesses the structure of the mini-batch k-means algorithm to only read in the data needed for each batch, controlling memory usage for large datasets. This makes our implementation truly scalable and applicable to both standard in-memory matrix objects, including sparse matrix representations, and on-disk data representations that do not require all the data to be loaded into memory at any one time, such as HDF5 matrices.” <span class="citation">(Hicks et al. 2021)</span></p>
</blockquote>
<blockquote>
<p>“If you use too much memory, R will complain. The key issue is that R holds
all the data in RAM. This is a limitation if you have huge datasets. The up-side
is flexibility—in particular, R imposes no rules on what data are like.” <span class="citation">(Burns 2011)</span>
“Another way of reducing memory use is to store your data in a database and
only extract portions of the data into R as needed. While this takes some time
to set up, it can become quite a natural way to work.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“Are tomorrow’s bigger computers going to solve the problem? For some people,
yes—their data will stay the same size and computers will get big enough to
hold it comfortably. For other people it will only get worse—more powerful
computers means extraordinarily larger datasets. If you are likely to be in this
latter group, you might want to get used to working with databases now.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“Traditionally, the assay data are stored in-memory as an ordinary array object3. Storing the data in-memory becomes a real pain with the ever-growing size of ’omics datasets. It is now not uncommon to collect 10,000–100,000,000 measurements on 100–1,000,000 samples, which would occupy 10–1,000 gigabytes (Gb) if stored in-memory as ordinary R arrays.
The DelayedArray framework offers a solution to this problem. Wrapping an array-like object (typically an on-disk object) in a DelayedArray object allows one to perform common array operations on it without loading the object in memory. In order to reduce memory usage and optimize performance, operations on the object are either delayed or executed using a block processing mechanism.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“The DelayedArray framework enables the analysis of datasets that are too large to be stored or processed in-memory. This has become particularly relevant with the advent of large single-cell RNA-sequencing (scRNA-seq) studies containing tens of thousands to millions of cells.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“The data contained in an HDF5Matrix is actually stored on disk in a Hierarchical Data Format (HDF5) file. Consequently, the tenx_counts object takes up very little space in memory.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“The subsetting operation has been registered in what is termed a ‘delayed operation.’ Registering a delayed operation does not modify the underlying data. Instead, the operation is recorded and only performed when the DelayedArray object is ‘realized.’ Realization of a DelayedArray triggers the execution of the delayed operations carried by the object and returns the result as an ordinary array.
This allows us to chain together multiple operations and only perform them as required. … To <em>realize</em> a DelayedArray object is to trigger execution of the delayed operations carried by the object and return the result as an ordinary array. … A large DelayedArray object is preferably realized on disk, which is most commonly an HDF5 file.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“Hopefully you can now begin to see the general pattern, a strategy which the DelayedArray package calls ‘block-processing’: 1. Load a ‘block’ of the data into memory.
2. Compute a summary statistic.
3. Combine the block-level statistics in an appropriate way to get the final result.”
<a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“Let’s move onto something a little more computationally challenging. Proper normalization is essential for all analyses of gene expression data. We apply the deconvolution method of Lun, Bach, and Marioni (2016) to compute size factors for all cells.
For highly heterogeneous datasets, like this sample of PBMCs, it is advisable to perform a rough clustering of the cells to better satisfy the assumptions of this normalization method. Namely, we want to avoid normalizing together cells with a large number of differentially expressed genes between them.
We will use the scran::quickCluster() function to perform a clustering based on the principal component scores generated from the log-expression matrix. This principal component analysis (PCA) in turn uses an approximate singular value decomposition (SVD) with the augmented implicitly restarted Lanczos bidiagonalization algorithm (irlba). If that all sounds rather complicated, then don’t worry: that’s the point of this example! We are able to apply these cutting edge techniques to our HDF5-backed data.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“A HDF5-backed SummarizedExperiment, like the 10x PBMC dataset we analysed in Real world encounter with DelayedArray analysing scRNA-seq data, is a light-weight shell (the SummarizedExperiment) around a large disk-backed data matrix (the HDF5Matrix).” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“A SummarizedExperiment derivative can have one or more of its assays that point to datasets (one per assay) in an HDF5 file. … These objects have 2 parts: one part is in memory, and one part is on disk. The 1st part is sometimes called the object shell and is generally thin (i.e. it has a small memory footprint). The 2nd part is the data and is typically big. The object shell and data are linked together via some kind of pointer stored in the shell (e.g. an SQLite connection, or a path to a file, etc.). Note that this is a one way link in the sense that the object shell ‘knows’ where to find the on-disk data but the on-disk data knows nothing about the object shell (and is completely agnostic about what kind of object shell could be pointing to it). Furthermore, at any given time on a given system, there could be more than one object shell pointing to the same on-disk data. These object shells could exist in the same R session or in sessions in other languages (e.g. Python). These various sessions could be run by the same or by different users.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“For example, a normalized scRNA-seq dataset carries around two matrices: the raw counts and the normalized expression values. You might have enough RAM to load one of these a time but not both at once. With a HDF5-backed SingleCellExperiment you can easily just load into memory the matrix you actually need at a given step in the analysis.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<blockquote>
<p>“Don’t use a DelayedArray if you can avoid it! … If you can load your data into memory and still compute on it then you’re always going to have a better time doing it that way. Analyses will be faster, simpler, and you will have more options available to you. But when this isn’t an option then the DelayedArray framework is a powerful set of packages to help you get your work done. I find it pretty remarkable that a first-class single-cell analysis workflow can so seamlessly support the use of in-memory and disk-backed data.” <a href="http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/" class="uri">http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/DelayedArrayWorkshop__Effectively_using_the_DelayedArray_framework_for_users/</a></p>
</blockquote>
<p>[Different elements in the data]
Most laboratory equipment can output a raw data file that you can then read into R.
For many types of laboratory equipment, these raw data files follow a strict format.
For example [flow cytometery format…]…</p>
<p>The file formats will often have different pieces of data stored in specific spots.
For example, the equipment might record not only the measurements taken for the
sample, but also information about the setting that were applied to the equipment
while the measurements were taken, the date of the measurements, and other metadata
that may be useful to access when preprocessing the data. Each piece of data may
have different “dimensions.” For example, the measurements might provide one
measurement per metabolite feature or per marker. Some metadata might also be
provided with these dimensions (e.g., metadata about the markers for flow
cytometry data), but other metadata might be provided a single time per sample
or even per experiment—for example, the settings on the equipment when the
sample or samples were run.</p>
<p>When it comes to data structures, dataframes and other two-dimensional data storage
structures (you can visualize these as similar to the format of data in a spreadsheet,
with rows and columns) work well to store data where all data conform to a common
dimension. For example, a dataframe would work well to store the measurements
for each marker in each sample in a flow cytometry experiment. In this case,
each column could store the values for a specific marker and each row could
provide measurements for a sample. In this way, you could read the measurements
for one marker across all samples by reading down a column, or read the measurements
across all markers for one sample by reading across a row.</p>
<p>When you have data that doesn’t conform to these common dimensions [unit of
measurement?] however, a dataframe may work poorly to store the data. For
example, if you have measurements taken at the level of the equipment settings
for the whole experiment, these don’t naturally fit into the dataframe format.
In the “tidyverse” approach, one approach to handling data with different units
of measurement is to store data for each unit of measurement in a different
dataframe and to include identifiers that can be used to link data across the
dataframes. More common, however, in R extensions for preprocessing biomedical
data is to use more complex data structures that can store data with different
units of measurement in different slots within the data structure, and use these
in conjunction with specific functions that are built to work with that specific
data structure, and so know where to find each element within the data
structure.</p>
<p>[Validation of data as its entered in an S4 class]</p>
</div>
<div id="limitations-of-complex-structures-for-storing-data" class="section level3" number="3.4.4">
<h3><span class="header-section-number">3.4.4</span> Limitations of complex structures for storing data</h3>
<blockquote>
<p>“There is a cost to the free lunch. That print is generic means that what
you see is not what you get (sometimes). In the printing of an object you may
see a number that you want—an R-squared for example—but don’t know how
to grab that number.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
</div>
<div id="subsection-2-5" class="section level3" number="3.4.5">
<h3><span class="header-section-number">3.4.5</span> Subsection 2</h3>
</div>
<div id="practice-quiz-3" class="section level3" number="3.4.6">
<h3><span class="header-section-number">3.4.6</span> Practice quiz</h3>

</div>
</div>
<div id="complex-data-types-in-r-and-bioconductor" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Complex data types in R and Bioconductor</h2>
<p>Many R extension packages for pre-processing experimental data use complex
(rather than ‘tidy’) data formats within their code, and many output data in
complex formats. Very recently, the <em>broom</em> and <em>biobroom</em> R
packages have been developed to extract a ‘tidy’ dataset from a complex data
format. These tools create a clean, simple connection between the complex data
formats often used in pre-processing experimental data and the ‘tidy’ format
required to use the ‘tidyverse’ tools now taught in many introductory R courses.
In this module, we will describe the ‘list’ data structure, the common backbone
for complex data structures in R and provide tips on how to explore and extract
data stored in R in this format, including through the <em>broom</em> and
<em>biobroom</em> packages.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Describe the structure of R’s ‘list’ data format</li>
<li>Take basic steps to explore and extract data stored in R’s complex, list-based
structures</li>
<li>Describe what the <em>broom</em> and <em>biobroom</em> R packages can do</li>
<li>Explain how converting data to a ‘tidy’ format can improve reproducibility</li>
</ul>
<div id="subsection-1-7" class="section level3" number="3.5.1">
<h3><span class="header-section-number">3.5.1</span> Subsection 1</h3>
<p>When you are writing scripts in R to work with your code, if you are at a
point in your pipeline when you can use a “tidyverse” approach, then you will
“keep” your data in a dataframe, as your data structure, throughout your
work. However, at earlier stages in your preprocessing, you may need to use
tools that use other data structures. It’s helpful to understand the
basic building blocks of R data structures, so you can find elements of your
data in these other, more customized data structures.</p>
<p>Many R data structures are built on a general structure called a “list.” This
data structure is a useful basic general data structure, because it is
extraordinarily flexible. The list data structure is flexible in two
important ways: it allows you to include data of different <em>types</em> in the
same data structure, and it allows you to include data with different
dimensions—and data stored hierarchically, including various other data
structures—within the list structure. We’ll cover each of these points a bit
more below and describe why they’re helpful in making the list a very good
general purpose data structure.</p>
<p>In R, your data can be stored as different <em>types</em> of data: whole numbers can be
stored as an <em>integer</em> data type, continuous [?] numbers through a few types of
<em>floating</em> data types, character strings as a <em>character</em> data type, and logical
data (which can only take the two values of “TRUE” and “FALSE”) as a <em>logical</em>
data type. More complex data types can be built using these—for example,
there’s a special data type for storing dates that’s based on a combination of
an [integer?] data type, with added information counting the number of days [?]
from a set starting date (called the [Unix epoch?]), January 1, 1970. (This
set-up for storing dates allows them to be printed to look like dates, rather
than numbers, but at the same time allows them to be manipulated through
operations like finding out which date comes earliest in a set, determining the
number of days between two dates, and so on.) R uses these different data types
for several reasons. First, by using different data types, R can improve its
efficiency [?] in storing data. Each piece of data must—as you go deep in the
heart of how the computer works—as a series of binary digits (0s and 1s).
Some types of data can be stored using fewer of these <em>bits</em> (<em>bi</em>nary dig<em>its</em>).
Each measurement of logical data, for example, can be stored in a single bit,
since it only can take one of two values (0 or 1, for FALSE and TRUE, respectively).
For character strings, these can be divided into each character in the string
for storage (for example, “cat” can be stored as “c,” “a,” “t”). There is a set
of characters called the ASCII character set that includes the lowercase and
uppercase of the letters and punctuation sets that you see on a standard
US keyboard [?], and if the character strings only use these characters, they
can be stored in [x] bits per character. For numeric data types, integers can
typically be stores in [x] bits per number, while continuous [?] numbers,
stored in single or double floating point notation [?], are stored in [x]
and [x] bits respectively. When R stores data in specific types, it can be
more memory efficient by packing the types of data that can be stored in less
space (like logical data) into very compact structures.</p>
<p>The second advantage of the list structure in R is that it has enormous
flexibility in terms of storing lots of data in lots of possible places. This
data can have different types and even different substructures. Some data
structures in R are very constrained in what type of data they can store and
what structure they use to store it. For example, one of the “building block”
data structures in R is the vector. This data structure is one dimensional and
can only contain data that have the same data type—you can think of this as a
bead string of values, each of the same type. For example, you could have a
vector that gives a series of names of study sites (each a character string), or
a vector that gives the dates of time points in a study (each a date data type),
or a vector that gives the weights of mice in a study (each a numeric data
type). You cannot, however, have a vector that includes some study site names
and then some dates and then some weights, since these should be in different
data types. Further, you can’t arrange the data in any structure except a
straight, one-dimensional series if you are using a vector. The dataframe
structure provides a bit more flexibility—you can expand into two dimensions,
rather than one, and you can have different data types in different columns of
the dataframe (although each column must itself have a single data type).</p>
<p>The list data structure is much more flexible. It essentially allows you to
create different “slots,” and you can store any type of data in each of these
slots. In each slot you can store any of the other types of data structures in
R—for example, vectors, dataframes, or other lists. You can even store unusual
things like R <em>environments</em> [?] or <em>pointers</em> that give the directions to where
data is stored on the computer without reading the data into R (and so saving
room in the RAM memory, which is used when data is “ready to go” in R, but which
has much more limited space than the mass [?] storage on your computer).</p>
<p>Since you can put a list into the slot of a list, it allows you to create deep,
layered structures of data. For example, you could have one slot in a list where
you store the metadata for your experiment, and this slot might itself be a list
where you store one dataframe with some information about the settings of the
laboratory equipement you used to collect the data, and another dataframe that
provides information about the experimental design variables (e.g., which animal
received which treatment). Another slot in the larger list then might have
experimental measurements, and these might either be in a dataframe or, if the
data are very large, might be represented through pointers to where the data is
stored in memory, rather than having the data included directly in the data
structure.</p>
<p>Given all these advantages of the list data structure, then, why not use it all
the time? While it is a very helpful building block, it turns out that its flexibility
can have some disadvantages in some cases. This flexibility means that you can’s
always assume that certain bits of data are in a certain spot in each instance
of a list in R. Conversely, if you have data stored in a less flexible structure,
you can often rely on certain parts of the data always being in a certain part of
the data structure. In a “tidy” dataframe, for example, you can always assume
that each row represents the measurements for one observation at the unit of
observation for that dataframe, and that each column gives values across all
observations for one particular value that was measured for all the observations.
For example, if you are conducting an experiment with mice, where a certain number
of mice were sacrificed at certain time points, with their weight and the bacteria
load in their lungs measured when the mouse was sacrificed, then you could store
the data in a dataframe, with a row for each mouse, and columns giving the
experimental characteristics for each mouse (e.g., treatment status, time point
when the mouse was sacrificed), the mouse’s weight, and the mouse’s bacteria
load when sacrificed. You could store all of this information in a list, as
well, but the defined, two-dimensional structure of the dataframe makes it much
more clearly defined where all the data goes in the dataframe structure, while
you could order the data in many ways within a list.</p>
<p>There is a big advantage to having stricter standards for what parts of data go
where when it comes to writing functions that can be used across a lot of data.
You can think of this in terms of how cars are set up versus how kitchens are
set up. Cars are very standardized in the “interface” that you get when you sit
down to drive them. The gas and brakes are typically floor pedals, with the gas
to the right of the brake. The steering is almost always provided through a wheel
centered in front of the driver’s torso. The mechanism for shifting gears (e.g.,
forward, reverse) is typically to the right of the steering wheel, while
mecahnisms for features like lights and windshield wipers, are typically to the
left of the steering wheel. Because this interface is so standardized, you can
get into a car you’ve never driven before and typically figure out how to
drive it very quickly. You don’t need a lot of time exploring where everything
is or a lot of directions from someone familiar with the car to figure out where
things are. Think of the last time that you drove a rental car—within five minutes,
at most, you were probably able to orient yourself to figure out where everything
you needed was. This is like a dataframe in R—you can pretty quickly figure out
where everything you might need is stored in the data structure, and people can
write functions to use with these dataframes that work well generally across lots
of people’s data because they can assume that certain pieces of data are in
certain places.</p>
<p>By contrast, think about walking into someone else’s kitchen and orienting yourself
to use that. Kitchen designs do tend to have some general features—most will have
a few common large elements, like a stove somewhere, a refrigerator somewhere,
a pantry somewhere, and storage for pots, pans, and utensils somewhere. However,
there is a lot of flexibility in where each of these are in the kitchen design,
and further flexibility in how things are organized within each of these structures.
If you cook in someone else’s kitchen, it is easy to find yourself disoriented in the
middle of cooking a recipe, where a utensil that you can grab almost without
thinking in your own kitchen requires you to stop and search many places in
someone else’s kitchen. This is like a list in R—there are so many places that
you can store data in a list, and so much flexibility, that you often find yourself
having to dig around to find a certain element in a list data structure that someone
else has created, and you often can’t assume that certain pieces are in certain
places if you are writing your own functions, so it becomes hard to write
functions that are “general purpose” for generic list structures in R.</p>
<p>There is a way that list structures can be used in R in a way that retains some
of their flexibility while also leveraging some of the benefits of
standardization. This is R’s system for creating <em>objects</em>. These object
structures are built on the list data structure, but each object is constrained
to have certain elements of data in certain structures of the data. These
structures cannot be used as easily as dataframes in a “tidyverse” approach,
since the tidyverse tools are built based on the assumption that data is stored
in a tidy dataframe. However, they are used in many of the Bioconductor
approaches that allow powerful tools for the earlier stages in preprocessing
biological data. The types of standards that are imposed in the more specialized
objects include which slots the list can have, the names they have, what order
they’re in (e.g., in a certain object, the metadata about the experiment might
always be stored in the first slot of the list), and what structures and/or data
types the data in each slot should have.</p>
<p>R programmers get a lot of advantages from using these classes because they can
write functions under the assumption that certain pieces of the data will always
be in the same spot for that type of object. There is still flexibility in the
object, in that it can store lots of different types of data, in a variety of
different structures. While this “object oriented” approach in R data structures
does provide great advantages for programmers, and allow them to create powerful
tools for you to use in R, it does make it a little trickier in some cases for
you to explore your data by hand as you work through preprocessing. This is
because there typically are a variety of these object classes that your data
will pass through as you go through different stages of preprocessing, because
different structures are suited to different stages of analysis. Functions often
can only be used for a single class of objects, and so you have to keep track of
which functions pair up with which classes of data. Further, it can be a bit
tricky—at least in comparison to when you have data in a dataframe—to
explore your data by hand, because you have to navigate through different slots
in the object. By contrast, a dataframe always has the same two-dimensional,
rectangular structure, and so it’s very easy to navigate and explore data in
this structure, and there are a large number of functions that are built to be
used with dataframes, providing enormous flexibility in what you can do with
data stored in this structure.</p>
<p>While it is a bit trickier to explore your data when it is stored in a
list—either a general list you created, or one that forms the base for a
specialized class structure through functions from a Bioconductor package—you
can certainly learn how to do this navigation. This is a powerful and critical
tool for you to learn as you learn to preprocess your data in R, as you should
<em>never</em> feel like you data is stored in a “black box” structure, where you can’t
peek in and explore it. You should <em>always</em> feel like you can take a look at any
part of your data at any step in the process of preprocessing, analyzing, and
visualizing it.</p>
<hr />
<blockquote>
<p>“There are four primary types of atomic vectors: logical, integer, double,
and character (which contains strings). Collectively, integer and double
vectors are known as numeric vectors.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“… the most important family of data types in base R [is] vectors. …
Vectors come in two flavours: atomic vectors and lists. They differ in terms of
their elements’ types: for atomic vectors, all elements must have the
same type; for lists, elements can have different types. …
Each vector can also have <strong>attributes</strong>, which you can think of as a
named list of arbitrary metadata. Two attributes are particularly important.
The <strong>dimension</strong> attribute turns vectors into matrices and arrays and
the <strong>class</strong> attribute powers the S3 object system.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“A few places in R’s documentation call lists generic vectors to emphasise
their difference from atomic vectors.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Some of the most important S3 vectors [are] factors, dates and times,
data frames, and tibbles.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<p>You can use <code>typeof</code> to determine the data type and <code>is.[x]</code> (<code>is.logical</code>,
<code>is.character</code>, <code>is.double</code>, and <code>is.integer</code>) to test if data has a certain
type <span class="citation">(Wickham 2019)</span>.</p>
<blockquote>
<p>“You may have noticed that the set of atomic vectors does not include a
number of important data structures like matrices, arrays, factors, or
date-times. These types are all built on top of atomic vectors by
adding attributes.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Adding a <code>dim</code> attribute to a vector allows it to behave like a 2-dimensional
<strong>matrix</strong> or a multi-dimensional <strong>array</strong>. Matrices and arrays are primarily
mathematical and statistical tools, not programming tools…” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“One of the most important vector attributes is <code>class</code>, which underlies the
S3 object system. Having a class attribute turns an object into a <strong>S3 object</strong>,
which means it will behave differently from a regular vector when passed to
a <strong>generic</strong> function. Every S3 object is built on top of a base type, and often
stores additional information in other attributes.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Lists are a step up in complexity from atomic vectors: each element can be
any type, not just vectors.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Lists are sometimes called <strong>recursive</strong> vectors because a list can contain
other lists. This makes them fundamentally different from atomic vectors.”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The two most important S3 vectors built on top of lists are data frames and
tibbles. If you do data analysis in R, you’re going to be using data frames.
A data frame is a named list of vectors with attributes for (column) <code>names</code>,
<code>row.names</code>, and its class,”data.frame“… In contrast to a regular list, a
data frame has an additional constraint: the length of each of its vectors must
be the same. This gives data frames their rectangular structure…” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Data frames are one of the biggest and most important ideas in R, and one of the
things that makes R different from other programming languages. However, in the
over 20 years since their creation, the ways that people use R have changed, and
some of the design decisions that made sense at the time data frames were created
now cause frustration. This frustration led to the creation of the tibble
[Muller and Wickham, 2018], a modern reimagining of the data frame. Tibbles are
meant to be (as much as possible) drop-in replacements for data frames that fix
those frustrations. A concise, and fun, way to summarise the main differences is
that tibbles are lazy and surly: they do less and complain more.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Tibbles are provided by the tibble package and share the same structure as data
frames. The only difference is that the class vector is longer, and includes <code>tbl_df</code>.
This allows tibbles to behave differentlyin [several] key ways. … tibbles never
coerce their input (this is one feature that makes them lazy)… Additionally, while
data frames automatically transform non-syntactic names (unless <code>check.names = FALSE</code>),
tibbles do not… While every element of a data frame (or tibble) must have the
same length, both <code>data.frame()</code> and <code>tibble()</code> will recycle shorter inputs. However,
while data frames automatically recycle columns that are an integer multiple of the
longest column, tibbles will only recycle vectors of length one. … There is one
final different: <code>tibble()</code> allows you to refer to variables created during
construction. … [Unlike data frames,] tibbles do not support row names. … One
of the most obvious differences between tibbles and data frames is how they
print… Tibbles tweak [a data frame’s subsetting] behaviours so that a <code>[</code> always
returns a tibble, and a <code>$</code> doesn’t do partial matching and warns if it can’t find
a variable (this is what makes tibbles surly). … List columns are easier to use
with tibbles because they can be directly included inside <code>tibble()</code> and they
will be printed tidily.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Since the elements of lists are references to values, the size of a list might
be much smaller than you expect.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“[The] behavior [of environments] is different from that of other objects:
environments are always modified in place. This property is sometimes described
as <strong>reference semantics</strong> because when you modify an environment all existing
bindings to that environment continue to have the same reference. … This
basic idea can be used to create functions that ‘remember’ their previous
state… This property is also used to implement the R6 object-oriented
programming system…” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
</div>
<div id="exploring-and-extracting-data-from-r-list-data-structures" class="section level3" number="3.5.2">
<h3><span class="header-section-number">3.5.2</span> Exploring and extracting data from R list data structures</h3>
<p>To feel comfortable exploring your data at any stage during the preprocessing
steps, you should learn how to investigate and explore data that’s stored
in a list structure in R. Because the list structure is the building block
for complex data structures, including Bioconductor class structures, this will
serve you well throughout your work. You should get in the habit of checking
the structure and navigating where each piece of data is stored in the data
structure at each step in preprocessing your data. Also, by checking your data
throughout preprocessing, you might find that there are bits of information
tucked in your data at early stages that you aren’t yet using. For example,
many file formats for laboratory equipment include slots for information about
the equipment and its settings during when running the sample. This information
might be read in from the file into R, but you might not know it’s there for you
to use if you’d like, to help you in creating reproducible reports that include
this metadata about the experimental equipment and settings.</p>
<p>First, you will want to figure out whether your data is stored in a generic
list, or if it’s stored in a specific class-based data structure, which means it
will have a bit more of a standardized structure. To do this, you can run the
<code>class</code> function on your data object. The output of this might be a single value
(e.g., “list” [?]) or a short list. If it’s a short list, it will include both
the specific class of the object and, as you go down the list, the more
general data structure types that this class is built on. For example, if the
<code>class</code> function returns this list:</p>
<pre><code>[Example list of data types---maybe some specific class, then &quot;list&quot;?]</code></pre>
<p>it means that the data’s in a class-based structure called … which is built on
the more general structure of a list. You can apply to this data any of the functions
that are specifically built for … data structures, but you can also apply
functions built for the more general list data structure.</p>
<p>There are several tools you can use to explore data structured as lists in R.
R lists can sometimes be very large—in terms of the amount of data stored in
them—particularly for some types of biomedical data. With some of the tools
covered in this subsection, that will mean that your first look might seem
overwhelming. We’ll also cover some tools, therefore, that will let you peel
away levels of the data in a bit more manageable way, which you can use when
you encounter list-structured data that at first feels overwhelming.</p>
<p>First, if your data is stored in a specific class-based data structure, there
likely will also be help files specifically for the class structure that can
help you navigate it and figure out where things are. [Example]</p>
<p>[More about exploring data in list structures.]</p>
<hr />
<blockquote>
<p>“Use <code>[</code> to select any number of elements from a vector. … <strong>Positive
integers</strong> return elements at the specified positions. … <strong>Negative
integers</strong> exclude elements at the specified positions…
<strong>Logical vectors</strong> select elements where the corresponding logical vector
is <code>TRUE</code>. This is probably the most useful type of subsetting
because you can write an expression that uses a logical vector… If the
vector is named, you can also use <strong>character vectors</strong> to return elements
with matching names.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Subsetting a list works in the same way as subsetting an atomic vector.
Using <code>[</code> always returns a list; <code>[[</code> and <code>$</code> … lets you pull out
elements of a list.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“<code>[[</code> is most important when working with lists because subsetting a list
with <code>[</code> always returns a smaller list. … Because <code>[[</code> can return only
a single item, you must use it with either a single positive integer or
a single string.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“<code>$</code> is a shorthand operator: <code>x$y</code> is roughly equivalent to <code>x[["y"]]</code>.
It’s often used to access variables in a data frame… The one important
difference between <code>$</code> and <code>[[</code> is that <code>$</code> does (left-to-right) partial
matching [which you likely want to avoid to be safe].” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“There are two additional subsetting operators, which are needed for
S4 objects: <code>@</code> (equivalent to <code>$</code>), and <code>slot()</code> (equivalent to <code>[[</code>).”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The environment is the data structure that powers scoping. … Understanding
environments is not necessary for day-to-day use of R. But they are important to
understand because they power many important features like lexical scoping,
name spaces, and R6 classes, and interact with evaluation to give you powerful
tools for making domain specific languages, like <code>dplyr</code> and <code>ggplot2</code>.”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The job of an environment is to associate, of <strong>bind</strong>, a set of names to a
set of values. You can think of an environment as a bag of names, with no
implied order (i.e., it doesn’t make sense to ask which is the first element
in an environment).” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“… environments have reference semantics: unlike most R objects, when you
modify them, you modify them in place, and don’t create a copy.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“As well as powering scoping, environments are also useful data structures in
their own right because they have reference semantics. There are three common
problems that they can help solve: <strong>Avoiding copies of large data.</strong> Since
environments have reference semantics, you’ll never accidentally create a copy.
But bare environments are painful to work with, so I instead recommend using R6
objects, which are built on top of environments. …” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Generally in R, functional programming is much more important than object-oriented
programming, because you typically solve complex problems by decomposing them
into simple functions, not simple objects. Nevertheless, there are important reasons
to learn each of the three [object-oriented programming] systems [S3, R6, and S4]:
S3 allows your functions to return rich results with user-friendly display and
programmer-friendly internals. S3 is used throughout base R, so it’s important to
master if you want to extend base R functions to work with new types of input.
R6 provides a standardised way to escape R’s copy-on-modify semantics. This is
particularly important if you want to model objects that exist independently
of R. Today, a common need for R6 is to model data that comes from a web API,
and where changes come from inside or outside R. S4 is a rigorous system that
forces you to thing carefully about program design. It’s particularly well-suited
for building large systems that evolve over time and will receive contributions
from many programmers. This is why it’s used by the Bioconductor project, so another
reason to learn S4 is to equip you to contribute to that project.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The main reason to use OOP is <strong>polymorphism</strong> (literally: many shapes).
Polymorphism means that a developer can consider a function’s interface
separately from its implementation, making it possible to use the same function
form for different types of input. This is closely related to the idea of
<strong>encapsulation:</strong> the user doesn’t need to worry about details of an object
because they are encapsulated behind a standard interface. … To be more precise,
<strong>OO</strong> systems call the type of an object its <strong>class</strong>, and an implementation for
a specific class is called a <strong>method</strong>. Roughly speaking, a class defines what an
object <em>is</em> and methods describe what that object can <em>do</em>. The class defines
the <strong>fields</strong>, the data possessed by every instance of that class. Classes
are organised in a hierarchy so that if a method does not exist for one
class, its parent’s method is used, and the child is said to <strong>inherit</strong> behaviour.
… The process of finding the correct method given a class is called
<strong>method dispatch</strong>.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“There are two main paradigms of object-oriented programming which differ in how
methods and classes are related. In this book, we’ll borrow the terminology of
<em>Extending R</em> [Chambers 2016] and call these paradigms encapsulated and
functional: In <strong>encapsulated</strong> OOP, methods belong to objects or classes, and
method calls typically look like <code>object.method(arg1, arg2)</code>. This is called
encapsulated because the object encapsulates both data (with fields) and
behaviour (with methods), and is the paradigm found in most popular languages.
In <strong>functional</strong> OOP, methods belong to <strong>generic</strong> functions, and method calls
look like ordinary function calls: <code>generic(object, arg2, arg3)</code>. This is called
functional because from the outside it looks like a regular function call, and
internally the components are also functions.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“<strong>S3</strong> is R’s first OOP system… S3 is an informal implementation of functional
OOP and relies on common conventions rather than ironclad guarantees.
This makes it easy to get started with, providing a low cost way of solving many
simple problems. … <strong>S4</strong> is a formal and rigorous rewrite of S3…
It requires more upfront work than S3, but in return provides more guarantees and greater
encapsulation. S4 is implemented in the base <strong>methods</strong> package, which is
always installed with R.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“While everything <em>is</em> an object, not everything is object-oriented. This confusion
arises because the base objects come from S, and were developed before anyone
thought that S might need an OOP system. The tools and nomenclature evolved
organically over many years without a single guiding principle. Most of the time,
the distinction between objects and object-oriented objects is not important. But
here we need to get into the nitty gritty details so we’ll use the terms
<strong>base objects</strong> and <strong>OO objects</strong> to distinguish them. … Techinally, the difference
between base and OO objects is that OO objects have a ‘class attribute.’”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“An S3 object is a base type with at least a <code>class</code> attribute (other attributes
may be used to store other data). … An S3 object behaves differently from its
underlying base type whenever it’s passed to a <strong>generic</strong> (short for generic
function). … A generic function defines an interface, which uses a different
implementation depending on the class of an argument (almost always the first
argument). Many base R functions are generic, including the important <code>print</code>…
The generic is a middleman: its job is to define the interface (i.e., the arguments)
then find the right implements for the job. The implementation for a specific class
is called a <strong>method</strong>, and the generic finds the method by performing <strong>method
dispatch</strong>.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“If you have done object-oriented programming in other languages, you may be
surprised to learn that S3 has no formal definition of a class: to make an object
an instance of a class, you simply set the <strong>class attribute</strong>. … You can determine
the class of an S3 object with <code>class(x)</code>, and see if an object is an instance of
a class using <code>inherits(x, "classname")</code>.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The job of an S3 generic is to perform method dispatch, i.e., find the specific
implementation for a class.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“An important new componenet of S4 is the <strong>slot</strong>, a named component of the
object that is accessed using the specialised subsetting operator <code>@</code> (pronounced
‘at’). The set of slots, and their classes, forms an important part of the
definition of an S4 class.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Given an S4 object you can see its class with <code>is()</code> and access slots with <code>@</code>
(equivalent to <code>$</code>) and <code>slot()</code> (equivalent to <code>[[</code>) … Generally, you should only
use <code>@</code> in your methods. If you’re working with someone else’s class, look for
<strong>accessor</strong> functions that allow you to safely set and get slot values. … Accessors
are typically S4 generics allowing multiple classes to share the same external
interface.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“If you’re using an S4 class defined in a package, you can get help on it
with <code>class?Person</code>. To get help for a method, put <code>?</code> in front of a call (e.g.,
<code>?age(john)</code>) and <code>?</code> will use the class of the arguments to figure out which help
file you need.” <span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“Slots [in S4 objects] should be considered an internal implementation detail:
they can change without warning and user code should avoid accessing them
directly. Instead, all user-accessible slots should be accompanied by a pair
of <strong>accessors</strong>. If the slot is unique to the class, this can just be a function…
Typically, however, you’ll define a generic so that multiple classes can used
the same interface” <span class="citation">(<strong>wickham2019advancedr?</strong>)</span></p>
</blockquote>
<blockquote>
<p>“The strictness and formality of S4 make it well suited for large teams. Since more
structure is provided by the system itself, there is less need for convention, and new
contributors don’t need as much training. S4 tends to require more upfront design
than S3, and this investment is more likely to pay off on larger projects where
greater resources are available. One large team where S4 is used to good effect is
Bioconductor. Bioconductor is similar to CRAN: it’s a way of sharing packages
amongst a wider audient. Bioconductor is smaller than CRAN (~1,300 versus ~10,000 packages,
July 2017) and the packages tend to be more tightly integrated because of the
shared domain and because Bioconductor has a stricter review process. Bioconductor
packages are not required to use S4, but most will because the key data structures
(e.g., Summarized Experiment, IRanges, DNAStringSet) are built using S4.”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
<blockquote>
<p>“The biggest challenge to using S4 is the combination of increased complexity and
absence of a single source of documentation. S4 is a complex system and it can be
challenging to use effectively in practice. This wouldn’t be such a problem if
S4 wasn’t scattered through R documentation, books, and websites. S4 needs a
book length treatment, but that book does not (yet) exist. (The documentation for
S3 is no better, but the lack is less painful because S3 is much simpler.)”
<span class="citation">(Wickham 2019)</span></p>
</blockquote>
</div>
<div id="interfacing-between-object-based-and-tidyverse-workflows" class="section level3" number="3.5.3">
<h3><span class="header-section-number">3.5.3</span> Interfacing between object-based and tidyverse workflows</h3>
<p>The tidyverse approach in R is based on keeping data in a dataframe structure.
By keeping this common structure, the tidyverse allows for straightforward but
powerful work with your data by chaining together simple, single-purpose
functions. This approach is widely covered in introductory R programming courses
and books. A great starting point is the book <em>R Programming for Data Science</em>,
which is available both in print and freely online at [site]. Many excellent
resources exist for learning this approach, and so we won’t recover that
information here. Instead, we will focus on how to interface between this
approach and the object-based approach that’s more common with Bioconductor
packages. Bioconductor packages often take an object-based approach, and with
good reason because of the complexity and size of many early versions of
biomedical data in the preprocessing process. There are also resources for
learning to use specific Bioconductor packages, as well as some general
resources on Bioconductor, like <em>R Programming for Bioinformatics</em> [ref].
However, there are fewer resources available online that teach how to coordinate
between these two approaches in a pipeline of code, so that you can leverage the
needed power of Bioconductor approaches early in your pipeline, as you
preprocess large and complex data, and then shift to use a tidyverse approach
once your data is amenible to this more straightforward approach to analysis and
visualization.</p>
<p>The heart of making this shift is learning how to convert data, when possible,
from a more complex, class-type data structure (built on the flexible list
data structure) to the simpler, more standardized two-dimensional dataframe
structure that is required for the tidyverse approach. In this subsection, we’ll
cover approaches for converting your data from Bioconductor data structures to
dataframes.</p>
<p>If you are lucky, this might be very straightforward. A pair of packages called
<code>broom</code> and <code>biobroom</code> have been created specifically to facilitate the conversion
of data from more complex structures to dataframes. The <code>broom</code> package was
created first, by David Robinson, to convert the data stored in the objects that
are created by fitting statistical models into tidy dataframes. Many of the functions
in R that run statistical tests or fit statistical models output results in a
more complex, list-based data structure. These structures have nice “print” methods,
so if fitting the model or running the test is the very last step of your pipeline,
you can just read the printed output from R. However, often you want to include
these results in further code—for example, creating plots or tables that show
results from several statistical tests or models. The <code>broom</code> package includes
several functions for pulling out different bits of data that are stored in the
complex data structure created by fitting the model or running the test and convert
those pieces of data into a tidy dataframe. This tidy dataframe can then be
easily used in further code using a tidyverse approach.</p>
<p>The <code>biobroom</code> package was created to meet a similar need with data stored in some
of the complex structures commonly used in Bioconductor packages. [More about
<code>biobroom</code>.]</p>
<p>[How to convert data if there isn’t a <code>biobroom</code> method.] If you are unlucky,
there may not be a <code>broom</code> or <code>biobroom</code> method that you can use for the particular
class-based data structure that your data’s in, or it might be in a more general
list, rather than a specific class with a <code>biobroom</code> method. In this case, you’ll
need to extract the data “by hand” to move it into a dataframe once your data is
simple enough to work with using a tidyverse approach. If you’ve mastered how to
explore data stored in a list (covered in the last subsection), you’ll have a headstart
on how to do this. Once you know where to find each element of the data in the
structure of the list, you can assign these specific pieces to their own R objects
using typical R assignment (e.g., with the <em>gets arrow</em>, <code>&lt;-</code>, or with <code>=</code>, depending
on your preferred R programming style). …</p>
</div>
<div id="extras" class="section level3" number="3.5.4">
<h3><span class="header-section-number">3.5.4</span> Extras</h3>
<p>[Comparison of complexity of biological systems versus complexity of code and
algorithms for data pre-processing—for the later, nothing is unknowable or even
unknown. Someone somewhere is guaranteed to know exactly how it works, what it’s
doing, and why. By contrast, with biological systems, there are still things
that noone anywhere completely understands. It’s helpful to remember that all
code and algorithms for data pre-processing is knowable, and that the details
are all there if and when you want to dig to figure out what’s going on.]</p>
<p>[There are ways to fully package up and save the computer environment used to
run a pipeline of pre-processing and analysis, including any system settings,
all different software used in analysis steps, and so on. Some of the approaches
that are being explored for this include the use of “containers,” including
Docker containers. This does allow, typically, for full reproducibility of
the workflow. However, this approach isn’t very proactive in emphasizing the
robustness of a workflow or its comprehensibility to others—instead, it
makes the workflow reproducible by putting everything in a black box that must
be carefully unpackaged and explored if someone wants to understand or adapt
the pipeline.]</p>
<blockquote>
<p>“Object-oriented design doesn’t have to be over-complicated design, but we’ve
observed that too often it is. Too many OO designs are spaghetti-like tangles of
is-a and has-a relationships, or feature thick layers of glue in which many of the
objects seem to exist simply to hold places in a steep-sided pyramid of abstractions.
Such designs are the opposite of transparent; they are (notoriously) opaque and
difficult to debug.” <span class="citation">(E. S. Raymond 2003)</span></p>
</blockquote>
<blockquote>
<p>“Unix programmers are the original zealots about modularity, but tend to go about it
in a quiter way [that with OOP]. Keeping glue layers thin is part of it; more generally,
our tradition teaches us to build lower, hugging the ground with algorithms and structures
that are designed to be simple and transparent.” <span class="citation">(E. S. Raymond 2003)</span></p>
</blockquote>
<blockquote>
<p>“A <em>standard</em> is a precise and detailed description of how some artifact is built
or is supposed to work. Examples of software standards include programming
languages (the definition of syntax and semantics), data formats (how information is
represented), algorithmic processing (the steps necessary to do a computation), and
the like. Some standards, like the Word <code>.doc</code> file format, are <em>de facto</em> standards—they
have no official standing but everyoen uses them. The word ‘standard’ is best reserved for
formal descriptions, often developed and maintained by a quasi-neutral party like a
government or a consortium, that define how something is built or operated. The definition
is sufficiently complete and precise that separate entities can interact or provide independent
implementations. We benefit from hardware standards all the time, though we may not notice
how many there are. If I buy a new television set, I can plug it inot the electrical outlets
in my home thanks to standards for the size and shape of plugs and the voltage they provide.
The set itself will receive signals and display pictures because of standards for broadcast
and cable television. I can plug other devices into it through standard cables and connectors
like HDMI, USB, S-video and so on. But every TV needs its own remote control and every cell
phone needs a different charger because those have not been standardized. Computing has plenty
of standards as well, including character sets like ASCII and Unicode, programming languages
like C and C++, algorithms for encryption and compression, and protocols for exchanging
information over networks.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“Standards are important. They make it possible for independently created things to cooperate,
and they open an area to competition from multiple suppliers, while proprietary systems tend
to lock everyone in. … Standards have disadvantages, too—a standard can impede progress if
it is inferior or outdated yet everyone is forced to use it. But these are modest drawbacks
compared to the advantages.” <span class="citation">(Kernighan 2011)</span></p>
</blockquote>
<blockquote>
<p>“A <em>class</em> is a blueprint for constructing a particular package of code and data; each
variable created according to a class’s blueprint is known as an <em>object</em> of that class.
Code outside of a class that creates and uses an object of that class is known as a <em>client</em>
of the class. A <em>class declaration</em> names the class and lists all of the <em>members</em>, or
items inside that class. Each item is either a <em>data member</em>—a variable declared within the
class—or a <em>method</em> (also known as a <em>member function</em>), which is a function declared within
the class. Member functions can include a special type called a <em>constructor</em>, which has the
same name as the class and is invoked implicitly when an object of the class is declared.
In addition to the normal attributes of a variable or function declaration (such as type, and
for functions, the parameter list), each member has an <em>access specifier</em>, which indicates
what functions can access the member. A <em>public member</em> can be accessed by any code using the
object: code inside the class, a client of the class, or code in a <em>subclass</em>, which is a class
that ‘inherits’ all the code and data of an existing class. A <em>private member</em> can be
accessed only by the code inside the class. <em>Protected members</em> … are similar to private
members, except that methods in subclasses can also reference them. Both private and
protected members, though, are inaccessible from client code.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“An object should be a meaningful, closely knit collection of data and code that operates
on the data.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“Recognizing a situation in which a class would be useful is essential to reaching the
higher levels of programming style, but it’s equally important to recognize situations in
which a class is going to make things worse.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“The word <em>encapsulation</em> is a fancy way of saying that classes put multiple pieces of
data and code together in a single package. If you’ve ever seen a gelatin medicine capsule
filled with little spheres, that’s a good analogy: The patient takes one capsule and swallows
all the individual ingredient spheres inside. … From a problem-solving standpoint,
encapsulation allows us to more easily reuse the code from previous problems to solve current
problems. Often, even though we have worked on a problem similar to our current project,
reusing what we learned before still takes a lot of work. A fully encapsulated class can
work like an external USB drive; you just plug it in and it works. FOr this to happen,
though, we must design the class correctly to make sure that the code and data is truly
encapsulated and as independent as possible from anything outside of the class. For example,
a class that references a global variable can’t be copied into a new project without
copying the global variable, as well.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“Beyond reusing classes from one program to the next, classes offer the potential for
a more immediate form of code reuse: inheritance. … Using inheritance, we create parent
classes with methods common to two or more child classes, thereby ‘factoring out’ not
just a few lines of code [as with helper functions in procedural code] but whole methods.”
<span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“One technique we’re returned to again and again is dividing a complex problem into
smaller, more manageable pieces. Classes are great at dividing programs up into functional
units. Encapsulation not only holds data and code together in a reusable package; it also
cordons off that data and code from the rest of the program, allowing us to work on that
class, and everything else separately. The more classes we make in a program, the greater
the problem-dividing effect.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“Some people use the terms <em>information hiding</em> and <em>encapsulation</em> interchangeable, but
we’ll separate the ideas here. As described previously …, encapsulation is
packaging data and code together. Information hiding means separating the interface of a
data structure—the definition of the operations and their parameters—from the implementation
of a data structure, or the code inside the functions. If a class has been written with
information hiding as a goal, then it’s possible to change the implementation of the methods
without requiring any changes in the client code (the code that uses the class). Again, we
have to be clear on the term <em>interface</em>; this means not only the name of the methods and
their parameter list but also the explanation (perhaps expressed in code documentation) of
what the different methods do. When we talk about changing the implementation without
changing the interface, we mean that we change <em>how</em> the class methods work but not
<em>what</em> they do. Some programming authors have referred to this as a kind of implicit contract
between the class and the client: The class agrees never to change the effects of
existing operations, and the client agrees to use the class strictly on the basis of its
interface and to ignore any implementation details.” <span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“So how does information hiding affect problem solving? The principle of information hiding
tells the programmer to put aside the class implementation details when working on the
client code, or more broadly, to be concerned about a particular class’s implementation
only when working inside that class. When you can put implementation details out of your
mind, you can eliminate distracting thoughts and concentrate on solving the problem at hand.”
<span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“A final goal of a well-designed class is expressiveness, or what might be broadly called
writability—the ease with which code can be written. A good class, once written, makes
the rest of the code simpler to write in the way that a good function makes code simpler to
write. Classes effectively extend a language, becoming high-level counterparts to basic
low-level features such as loops, if statements, and so forth. … With classes, programming
actions that previously took many steps can be done in just a few steps or just one.”
<span class="citation">(Spraul 2012)</span></p>
</blockquote>
<blockquote>
<p>“Right now, in labs across the world, machines are sequencing the genomes of the life
on earth. Even with rapidly decreasing costs and huge technological advancements in
genome sequencing, we’re only seeing a glimpse of the biological information contained
in every cell, tissue, organism, and ecosystem. However, the smidgen of total biological
information we’re gathering amounts to mountains of data biologists need to work with. At
no other point in human history has our ability to understand life’s complexities been so
dependent on our skills to work with and analyze data.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioinformaticians are concerned with deriving biological understanding from large
amounts of data with specialized skills and tools. Early in biology’s history, the
datasets were small and manageable. Most biologists could analyze their own data after
taking a statistics course, using Microsoft Excel on a personal desktop computer.
However, this is all rapidly changing. Large sequencing datasets are widespread, and will
only become more common in the future. Analyzing this data takes different tools, new skills,
and many computers with large amounts of memory, processing power, and disk space.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“In a relatively short period of time, sequencing costs dropped drastically, allowing
researchers to utilize sequencing data to help answer important biological questions.
Early sequencing was low-throughput and costly. Whole genome sequencing efforts were
expensive (the human genome cost around $2.7 billion) and only possible through large
collaborative efforts. Since the release of the human genome, sequencing costs have
decreased explonentially until about 2008 … With the introduction of next-generation
sequencing technologies, the cost of sequencing a megabase of DNA dropped even more
rapidly. At this crucial point, a technology that was only affordable to large collaborative
sequencing efforts (or individual researchers with very deep pockets) became affordable
to researchers across all of biology. … What was the consequence of this drop in
sequencing costs due to these new technologies? As you may have guessed, lots and lots
of data. Biological databases have swelled with data after exponential growth. Whereas once
small databases shared between collaborators were sufficient, now petabytes of useful
data are sitting on servers all over the world. Key insights into biological questions are
stored not just in the unanalyzed experimental data sitting on your hard drive, but also
spinning around a disk in a data center thousands of miles away.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“To make matters even more complicated, new tools for analyzing biological data are
continually being created, and their underlying algorithms are advancing. A 2012 review
listed over 70 short-read mappers … Likewise, our approach to genome assembly has
changed considerably in the past five years, as methods to assemble long sequences
(such as overlap-layout-consensus algorithms) were abandoned with the emergence of
short high-throughput sequencing reads. Now, advances in sequencing chemistry are
leading to longer sequencing read lengths and new algorithms are replacing others that
were just a few years old. Unfortunately, this abundance and rapid development of
bioinformatics tools has serious downsides. Often, bioinformatics tools are not adequately
benchmarked, or if they are, they are only benchmarked in one organism. This makes it
difficult for new biologists to find and choose the best tool to analyze their data.
To make matters more difficult, some bioinformatics programs are not actively developed
so that they lose relevance or carry bugs that could negatively affect results. All of
this makes choosing an appropriate bioinformatics program in your own research difficult.
More importantly, it’s imperative to critically assess the output of bioinformatics
programs run on your own data.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“With the nature of biological data changing so rapidly, how are you supposed to
learn bioinformatics? With all of the tools out there and more continually being
created, how is a biologist supposed to know whether a program will work appropriately
on her organism’s data? The solution is to approach bioinformatics as a bioinformatician
does: try stuff, and assess the results. In this way, bioinformatics is just about having
the skills to experiment with data using a computer and understanding your results.
The experimental part is easy: this comes naturally to most scientists. The limiting
factor for most biologists is having the data skills to freely experiment and work with
large data on a computer.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Unfortunately, many of the biologist’s common computational tools can’t scale to the
size and complexity of modern biological data. Complex data formats, interfacing
numerous programs, and assessing software and data make large bioinformatics datasets
difficult to work with.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“In 10 years, bioinformaticians may only be using a few of the bioinformatics
software programs around today. But we most certainly will be using data skills and
experimentation to assess data and methods of the future.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Biology’s increasing use of large sequencing datasets is changing more that the tools
and skills we need: it’s also changing how reproducible and robust our scientific
findings are. As we utilize new tools and skills to analyze genomics data, it’s
necessary to ensure that our approaches are still as reproducible and robust as
any other experimental approaches. Unfortunately, the size of our data and the complexity
of our analysis workflows make these goals especially difficult in genomics.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“The requisite of reproducibility is that we share our data and methods. In the pre-genomics
era, this was much easier. Papers coule include detailed method summaries and entire
datasets—exactly as Kreitman’s 1986 paper did with a 4,713bp Adh gene flanking sequence
(it was embedded in the middle of the paper). Now papers have long supplementary methods,
code, and data. Sharing data is no longer trivial either, as sequencing projects can include
terabytes of accompanying data. Reference genomes and annotation datasets used in analyses are
constantly updated, which can make reproducibility tricky. Links to supplemental materials,
methods, and data on journal websites break, materials on faculty websites disappear when
faculty members move or update their sites, and software projects become stale when
developers leave and don’t update code. … Additionally, the complexity of bioinformatics
analyses can lead to findings being susceptible to errors and technical confounding.
Even fairly routine genomics projects can use dozens of different programs, complicated
input paramter combinations, and many sample and annotation datasets; in addition, work
may be spread across servers and workstations. All of these computational data-processing
steps create results used in higher-level analyses where we draw our biological conclusions.
The end result is that research findings may rest on a rickety scaffold of numerous
processing steps. To make matters worse, bioinformatics workflows and analyses are usually
only run once to produce results for a publication, and then never run or tested again.
These analyses may rely on very specific versions of all software used, which can make it
difficult to reproduce on a different system. In learning bioinformatics data skills, it’s
necessary to concurrently learn reproducibility and robust best practices.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“When we are writing code in a programming language, we work most of the time with RAM,
combining and restructuring data values to produce new values in RAM. … The computer memory
in RAM is a series of 0’s and 1’s, just like the computer memory used to store files in mass
storage. In order to work with data values, we need to get those values into RAM in some
format. At the basic level of representing a single number or a single piece of text, the
solution is the same as it was in Chapter 5 [on file formats for mass storage]. Everything
is represented as a pattern of bits, using various numbers of bytes for different sorts of
values. In R, in an English locale, and on a 32-bit operating system, a single character
usually takes up one byte, an integer takes up four bytes, and a real number 8 bytes.
Data values are stored in different ways depending on the <strong>data type</strong>—whether the values
are numbers or texts.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“ALthough we do not often encounter the details of the memory representation, except
when we need a rough estimate of how much RAM a data set might require, it is
important to keep in mind what sort of data type we are working with because the computer
code that we will produce different results for different data types. For example,
we can only calculate an average if we are dealing with values that have been stored
as text.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Another important issue is how <em>collections</em> of values are stored in memory. The
tasks that we will consider will typically involve working with an entire data set,
or an entire variable from a data set, rather than just a single value, so we need
to have a way to represent several related values in memory. This is similar to the
problem of deciding on a storage format for a data set… However, rather than talking
about different file formats, [in this case] we will talk about different <strong>data
structures</strong> for storing a collection of data values in RAM. … It will be important
to always keep close in our minds what data type we are working with and what sort
of data structure we are working with.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Every individual data value has a data type that tells us what sort of value it
is. The most common data types are numbers, which R calls <strong>numeric values</strong>, and
text, which R calls <strong>character values</strong>.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Vectors:</strong> A collection of values that all have the same data type. The
<strong>elements</strong> of a vector are all numbers, giving a <strong>number vector</strong>, or all
character values, giving a <strong>character vector.</strong>” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Data frames:</strong> A collection of vectors that all have the same length. This is
like a matrix, except that each column can contain a different data type.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Lists:</strong> A collection of data structures. The <strong>components</strong> of a list can
be simply vectors—similar to a data frame, but with each column allowed to have
a different length. However, a list can also be a much more complicated
structure. This is a very flexible data structure. Lists can be used to store
any combination of data values together.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Notice the way that lists are displayed. The first component of the list
starts with the component indes, <code>[[1]]</code>, followed by the contents of this
component…The second component of the list starts with the component
index <code>[[2</code>]] followed by the contenets of this component…” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“A list is a very flexible data structure. It can have any number of <strong>components</strong>,
each of which can be any data structure of any length or size. A simple
example is a data-frame-like structure where each column can have a different
length, but much more complex structures are also possible. For example, it is
possible for a component of a list to be another list.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Anyone who has worked with a computer should be familiar with the idea of
a list containing another list because a directory or folder of files has this sort
of structure: a folder contains multiple files of different kinds and sizes and
a folder can contain other folders, which can contain more files or even more
folders, and so on. Lists allow for this kind of hierarchical structure.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“One of the most basic ways that we can manipulate data structures is to
<strong>subset</strong> them—select a smaller portion from a larger data structure. This
is analogous to performing a query on a database. … R has very powerful
mechanisms for subsetting… A subset from a vector may be obtained by
appending an <strong>index</strong> within square brackets to the end of a symbol name. …
The index can be a vector of any length … The index does not have to be
a contiguous sequence, and it can include repetitions… As well as using
integers for indices, we can use logical values… A data frame can also
be indexed using square brackets, though slightly differently because we have
to specify both which rows <em>and</em> which columns we want … When a data structure
has named components, a subset may be selected using those names.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Single square bracket subsetting on a data frame is like taking an egg container
that contains a dozen eggs and chopping up the <em>container</em> so that we are left
with a smaller egg container that contains just a few eggs. Double square bracket
subsetting on a data frame is like selecting just one <em>egg</em> from an egg container.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“We can often get some idea of what sort of data structure we are working with
by simply viewing how the data are displayed on screen. However, a more definitive
answer can be obtained by calling the <code>class()</code> function. … Many R functions
return a data structure that is not one of the basic data structures that we have
already seen [like the ‘xtabs’ and ‘table’ classes]. … We have not seen either
of these data structures before. However, much of what we know about working with
the standard data structures … will work with any new class that we encounter.
For example, it is usually possible to subset any class using the standard
square bracket syntax. … Where appropriate, arithmetic and comparisons will
also generally work… Furthermore, if necessary, we can ofter resort to coercing
a class to something more standard and familiar.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Dates are an important example of a special data structure. Representing
dates as just text is convenient for humans to view, but other representations
are better for computers to work with. … Having a special class for dates
means that we can perform tasks with dates, such as arithmetic and comparisons,
in a meaningful way, something we could not do if we stored the date as just
a character value.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“The Date class stores date values as integer values, representing the number
of days since January 1st 1970, and automatically converts the numbers to
a readable text value to display the dates on the screen.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“When working with anything but tiny data sets, basic reatures of the data
set cannot be determined by just viewing the data values. [There are] a number of
functions that are useful for obtaining useful summary features from a data structure.
The <code>summary()</code> function produces summary information for a data structure… The
<code>length()</code> function is useful for determining the number of values in a vector or
the number of components in a list. … The <code>str()</code> function (short for ‘structure’)
is useful when dealing with large objects because it only shows a sample of the values
in each part of the object, although the display is very low-level so it may not always
make things clearer. … Another function that is useful for inspecting a large
object is the <code>head()</code> function. This shows just the first few elemeents of an
object, so we can see the basic structure without seeing all of the values.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Generic functions … will accept many different data structures as
arguments. … a generic function adapts itself to the data structure it is
given. Generic functions do different things when given different data structures.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“An example of a generic function is the <code>summary()</code> function. The result of a
call to <code>summary()</code> sill depend on what sort of data structure we provide.”
<span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“Generic functions are another reason why it is easy to work with data in R; a
single function will produce a sensible result no matter what data structure
we provide. However, generic functions are also another reason why it is so
important to be aware of what data structures we are working with. Without
knowing what sort of data we are using, we cannot know what sort of result to
expect from a generic function.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“R has become very popular and is now being used for projects that require
substantial software engineering as well as its continued widespread use as
an interactive environment for data analysis. This essentially means that
there are two masters—<em>reliability</em> and <em>ease of use</em>. S3 is indeed easy to
use, but can be made unreliable through nothing other than bad luck, or
a poor choice of names, and hence is not a suitable paradigm for constructing
large systems. S4, on the other hand, is better suited for developing large
software projects but has an increased complexity of use.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Object-oriented programming has become a widely used and valuable tool for
software engineering. Much of its value derives from the fact that it is often
easier to design, write, and maintain software when there is some clear
separation of the data representation from the operations that are to be
performed on it. In an OOP system, real physical things … are generally
represented by classes, and methods (functions) are written to handle the
different manipulations that need to be performed on the objects.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“The views that many people have of OOP have been based largely on
exposure to languages like Java, where the system can be described as class-centric.
In a a class-centric system, classes define objects and are repositories for
the methods that act on those objects. In contrast, languages such as … R
separate the class specification from the specification of generic functions,
and could be described as function-centric systems.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“The genome of every organism is encoded in chromosomes that consist of either
DNA or RNA. High throughput sequencing technology has made it possible to determine
the sequence of the genome for virtually any organism, and there are many that are
currently available. … However, in many cases, either the exact nucleotide at
any location is unknown, or is variable, and the International Union of Pure and
Applied Chemistry (IUPAC) has provided a standard nomenclature suitable for
representing such sequences. The alphabet for dealing with protein sequences
is based on the 20 amino acids. … The basic class used to hold strings [in the
<strong>Biostrings</strong> package] is the <em>BString</em> class, which has been designed to be
efficient in its handling of large character strings. Subclasses include
<em>DNAString</em>, <em>RNAString</em>, and <em>AAString</em> (for holding amino acid sequences).
The <em>BStringViews</em> class holds a set of <em>views</em> on a single <em>BString</em> instance;
each view is essentially a substring of the underlying <em>BString</em> instance. Alignments
are stored using the <em>BStringAlign</em> class.” <span class="citation">(Gentleman 2008)</span> [More on functions that
work with these classes on p. 171]</p>
</blockquote>
<blockquote>
<p>“A number of complete genomes, represented as <em>DNAString</em> objects, are provided
through the Bioconductor project. They rely on the infrastructure in the <strong>BSgenome</strong>
package, and all such packages have names that begin with <code>BSgenome</code>. You can find the
list of available genomes using the <code>available.genomes</code> function.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Atomic vectors are the most basic of all data structures. An atomic vector
contains some number of values of the same type; that number could be zero.
Atomic vectors can contain integers, doubles, logicals or character strings.
Both complex numbers and raw (pure bytes) have atomic representations …
Character vectors in the S language are vectors of character strings, not the
vectors of characters. For example, the string ‘super’ would be represented as
a character vector of length one, not of lenth five…” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Lists can be used to store items that are not all of the same type. …
Lists are also referred to as generic vectors since they share many of the properties
of vectors, but the elements are allowed to have different types.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Lists can be of any length, and the elements of a list can be named, or not.
Any R object can be an element of a list, including another list…” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“A <code>data.frame</code> is a special kind of list. Data frames were created to provide a
common structure for storing rectangular data sets and for passing them to
different functions for modeling and visualization. In many cases a data set can be
thought of as a rectangular structure with rows corresponding to cases and columns
corresponding to the different variables that were measured on each of the cases. One
might think that a matrix would be the appropriate representation, but that is only
true if all of the variables are of the same type, and this is seldom the case.”
<span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“[Data frames] are essentially a list of vectors, with one vector for each variable.
It is an error if the vectors are not all of the same length.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Sometimes it will be helpful to find out about an object. Obvious functions to
try are <code>class</code> and <code>typeof</code>. But many find that both <code>str</code> and <code>object.size</code> are
more useful. … The functions <code>head</code> and <code>tail</code> are convenience functions that list
the first few, or the last few, rows of a matrix.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“The S langauge has its roots in the Algol family of languages and has adopted some
of the general vector subsetting and subscripting techniques that were available in
languages such as APL. This is perhaps one area wehre programmers more familiar with
other languages fail to make appropriate use of the available functionality. …
There are slight differents between subsetting of vectors, arrays, lists, data.frames,
and enviroments that can sometimes catch the unwary. But there are also many
commonalities. … Subsetting can be carried out by three different operators:
the single square bracket <code>[</code>, the double square bracket <code>[[</code>, and the dollar, <code>$</code>.
We note that each of these three operators are actually generic functions and users
can write methods that extend and override them… One way of describing the behavior
of the single bracket operator is that the type of the return value matches the type
of the value it is applied to. Thus, a single bracket subset of a list is a list
itself. … Both <code>[[</code> and <code>$</code> extract a single value. There are some differences
between the two; <code>$</code> does not evaluate its second argument while <code>[[</code> does, and hence
one can use expressions. The <code>$</code> operator uses partial matching when extracting
named elements but <code>[</code> and <code>[[</code> do not.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“Subsetting plays two roles in the S language. One is an extraction role, where a subset
of a vector is identified by a set of supplied indices and the resulting subset is
returned as a value. Venables and Ripley (2000) refer to this as <em>indexing</em>. The
second purpose is subset assignment, where the goal is to identify a subset of
values that should have their values changed; we call this subset assignment.”
<span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“There are four basic types of subscript indices: positive integers, negative
integers, logical vectors, and character vectors. These four types cannot be mixed…
For matrix and array subscripting, one can use different types of subscripts for
the different dimensions. Not all vectors, or recursive objects, support all types
of subscripting indices. For example, atomic vectors cannot be subscripted using
<code>$</code>, while environments cannot be subscripted using <code>[</code>.” <span class="citation">(Gentleman 2008)</span></p>
</blockquote>
<blockquote>
<p>“In bioinformatics, the plain-text data we work with is often encoded in <em>ASCII</em>.
ASCII is a character encoding scheme that uses 7 bits to represent 128 different
values, including letters (upper- and lowercase), numbers, and special nonvisible
characters. While ASCII only uses 8 bits, nowadays computers use an 8-bit <em>byte</em>
(a unit representing 8 bits) to store ASCII characters. More information about
ASCII is available in your terminal through <code>man ascii</code>.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Some files will have non-ASCII encoding schemes, and may contain special characters.
The most common character encoding scheme is UTF-8, which is a superset of ASCII
but allows for special characters.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioinformatics data is often text—for example, the As, Cs, Ts, and Gs in
sequencing read files or reference genomes, or tab-delimited files fo gene coordinates.
The text data in bioinformatics is often large, too (gigabytes or more that can’t
fit into your computer’s memory at once). This is why Unix’s philosophy of
handling text streams is useful to bioinformatics: text streams allow us to
do processing on a <em>stream</em> of data rather than holding it all in memory.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Exploratory data analysis plays an integral role throughout an entire bioinformatics
project. Exploratory data analysis skills are just as applicable in analyzing intermediate
bioinformatics data (e.g., are fewer reads from this sequencing lane aligning?) as
they are in making sense of results from statistical analyses (e.g., what’s the distribution
of these p-values, and do they correlate with possible confounders like gene length?).
These exploratory analyses need not be complex or exceedingly detailed (many patterns
are visible with simple analyses and visualization); it’s just about wanting to look
into the data and having the skill set to do so.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Functions like <code>table()</code> are generic—they are designed to work with objects of
all kinds of classes. Generic functions are also designed to do the right thing
depending on the class of the object they’re called on (in programming lingo, we
say that the function is <em>polymorphic</em>).” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“It’s quite common to encounter genomics datasets that are difficult to load into
R because they’re large files. This is either because it takes too long to load the
entire dataset into R, or your machine simply doesn’t have enough memory. In many
cases, the best strategy is to reduce the size of your data somehow: summarizing data
in earlier processing steps, omitting unnecessary columns, splitting your data into
chunks (e.g., working with a chromosome at a tiem), or working on a random subset
of your data. Many bioinformatics analyses do not require working on an entire
genomic dataset at once, so these strategies can work quite well. These approaches
are also the only way to work with data that is truly too large to fit in your
machine’s memory (apart from getting a machine with more memory).” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“If your data is larger than the available memory on your machine, you’ll need to
use a strategy that keeps the bulk of your data out of memory, but still allows for
each access from R. A good solution for moerately large data is to use SQLite and
query out subsets for computation using the R package <code>RSQLite</code>. … Finally …
many Unix data tools have versions that work on gzipped files: <code>zless</code>, <code>zcat</code> (<code>gzcat</code>
on BSD-derived systems like Mac OS X), and others. Likewise, R’s data-reading
functions can also read gzipped files directly—there’s some slight performance
gains in reading in gzipped files, as there are fewer bytes to read off of
(slow) hard disks.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Quite often, data we load in to R will be in the wrong <em>shape</em> for what we want to
do with it. Tabular data can come in two different formats: <em>long</em> and <em>wide</em>. …
In many cases, data is recorded by humans in wide format, but we need data in long
format when working with and plotting statistical modeling functions.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Exploratory data analysis emphasizes visualization as the best tool to understand
and explore our data—both to learn what the data says and what its limitations are.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“R vectors require all elements to have the same data type (that is, vectors are
<em>homogeneous</em>). They only support the six data types discussed earlier (integer, double,
character, logical, complex, and raw). In contrast, R’s lists are more versatile: Lists can
contain elements of different types (they are <em>heterogenesou</em>); Elements can be <em>any</em>
object in R (vectors with different types, other lists, environments, dataframes,
matrices, functions, etc.); Because lists can store other lists, they allow for storing
data in a recursive way (in contrast, vectors cannot contain other vectors.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“The versatility of lists make them indispensable in programming and data analysis with
R.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“As with R’s vectors, we can extract subsets of a list or change values of
specific elements using indexing. However, accessing elements from an R list is
slightly different than with vectors. Because R’s list can contain objects with
different types, a subset containing multiple list elements could contain objects with
different types. Consequently, the only way to return a subset of more than one
list element is with another list. As a result, there are two indexing operators for lists:
one for accessing a subset of multiple elements as a list (the single bracket…) and
one for accessing an element within a list (the double bracket…).”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Because R’s lists can be nested and contain any type of data, list-based data
structures can grow to be quite complex. In some cases, it can be difficult to
understand the overall structure of some lists. The function <code>str()</code> is a convenient
R function for inspecting complex data structures. <code>str()</code> prints a line for each
contained data structure, complete with its type, length (or dimensions), and the
first few elements it contains. … For deeply nested lists, you can simplify <code>str()''s output by specifying the maximum depth of nested structure to return with</code>str()<code>'s  second argument,</code>max.level<code>. By default,</code>max.level<code>is</code>NA`, which returns all
nested structures.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Understanding R’s data structures and how subsetting works are fundamental to
having the freedom in R to explore data any way you like.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Some of Bioconductor’s core packages: <strong>GenomicRanges:</strong> Used to represent and
work with genomic ranges; <strong>GenomicFeatures:</strong> used to represent and work with ranges
that represent gene models and other features of a genome (genes, exons, UTRs,
transcripts, etc.); <strong>Biostrings</strong> and <strong>BSgenome:</strong> Used for manipulating genome
sequence data in R… <strong>rtracklayer:</strong> Used for reading in common bioinformatics formats
like BED, GTF/GFF, and WIG.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“The <code>GenomicRanges</code> package introduces a new class called <code>GRanges</code> for storing
genomic ranges. The <code>GRanges</code> builds off of <code>IRanges</code>. <code>IRanges</code> objects are used
to store ranges of genomic regions on a single sequence, and <code>GRanges</code> objects contain
the two other pieces of information necessary to specify a genomic location:
sequence name (e.g., which chromosome) and strand. <code>GRanges</code> objects also have
<em>metadata columns</em>, which are the data linked to each genomic range.”
<span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“All metadata attached to a <code>GRanges</code> object are stored in a <code>DataFrame</code>, which
behaves identically to R’s base <code>data.frame</code> but supports a wider variety of
column types. For example, <code>DataFrames</code> allow for run-length encoded vectors
to save memory … in practice, we can store any type of data: identifiers
and names (e.g., for genes, transcripts, SNPs, or exons), annotation data
(e.g., conservation scores, GC content, repeat content, etc.), or experimental
data (e.g., if ranges correspond to alignments, data like mapping quality and the
number of gaps). … the union of genomic location with any type of data is what makes
<code>GRanges</code> so pwoerful.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<p>Some object classes in BioConductor:</p>
<ul>
<li><code>eSet</code> from <code>Biobase</code></li>
<li><code>Sequence</code> from <code>IRanges</code></li>
<li><code>MAlist</code> from <code>limma</code></li>
<li><code>ExpressionSet</code> from <code>Biobase</code></li>
</ul>
<p>You can use the <code>getSlots</code> function with S4 objects to see all the
slots within the object.</p>
<blockquote>
<p>“Methods and classes in the S language are essentially programming concepts to
enable good organization of functions and of general objects, respectively.”
<span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“Programming in R starts out usually as writing functions, at least once we
get past the strict cut-and-paste stage. Functions are the actions of the
language; calls to them express what the user wants to happen. The arguments to
the functions and the values returned by function calls are the objects. These
objects represent everything we deal with. Actions create new objects (such as
summaries and models) or present the information in the objects (by plots,
printed summaries, or interfaces to other software). R is a functional,
object-based system where users program to extend the capacity of the system in
terms of new functionality and new kinds of objects.” <span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“Languages to which the object-oriented programming (OOP) term is typically
applied mostly support what might better be called <em>class</em>-oriented programming,
well-known examples being C++ and Java. In these languages the essential
programming unit is the class definition. Objects are generated as instances of
a class and computations on the objects consist of <em>invoking methods on</em> that
object. Depending on how strict the language is, all or most of the computations
must be expressed in this form. Method invocation is an operator, operating on
an instance of a class. Software organization is essentially simple and
hierarchical, in the sense that all methods are defined as part of a particular
class. That’s not how S works; as mentioned, the first and most important
programming unit is the function. From the user’s perspective, it’s all done by
calling functions (even if some of the functions are hidden in the form of
operators). Methods and classes provide not class-oriented programming but
function- and class-oriented programming. It’s a richer view, but also a more
complicated one.” <span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“A generic function will collect or <em>cache</em> all the methods for that function
belonging to all the R packages that have been loaded in the session. When the
function is called, the R evaluator then <em>selects</em> a method from those available,
by examining how well different methods match the actual arguments in the call.”
<span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“From the users’ view, the generic function has (or at least should have) a
natural definition in terms of what it is intended to do: <code>plot()</code> displays
graphics to represent an object or the relation between two objects; arithmetic
operators such as ‘+’ carry out the corresponding intuitive numerical
computations or extensions of those. Methods should map those intuitive notions
naturally and reliably into the concepts represented by the class definitions.”
<span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“The class definition contains a definition of the slots in objects from the
class and other information of various kinds, but the most important information
for the present discussion defines what other classes this class extends; that
is, the inheritance or to use the most common term, the <em>superclasses</em> of this
class. In R, the names of the superclasses can be seen as the value of
<code>extends(thisClass)</code>. By definition, an object from any class can be used in a
computation designed for any of the superclasses of that class. Therefore, it’s
precisely the superclasses of the class of an argument that define candidate
methods in a particular function call.” <span class="citation">(Chambers 2006)</span></p>
</blockquote>
<blockquote>
<p>“Conceptually, a generic function extends the idea of a function in R by
allowing different methods to be selected corresponding to the classes of the
objects supplied as arguments in a call to the function.” <span class="citation">(Chambers 2006)</span></p>
</blockquote>
<p>The code for different implementations of a method (in other words, different
ways it will run with new object classes) can come in different R packages.
This allows a developer to add his or her own applications of methods, suited
for object classes he or she creates.</p>
<p>A class defines the structure for a way of storing data. When you create
an object that follows this structure, it’s an instance of that class.
The <code>new</code> function is used to create new instances of a class.</p>
<p>When a generic function determines what code to run based on the class of the
object, it’s called method dispatch.</p>
<p>By using the accessor function, instead of <code>@</code>, your code will be more robust
to changes that the developers make. They will be sensitive to insuring that
the accessor function for a particular part of the data continues to work
regardless of changes they make to the structure that is used to store data in
objects in that class. They will be less committed, however, to keeping the
same slots, and in the same positions, as they develop the software. The
“contract” with the user is through the accessor function, in other words,
rather than through the slot name in the object.</p>
<blockquote>
<p>“Bioconductor is an open-source, open-development software project for the
analysis and comprehension of high-throughput data in genomics and molecular
biology. The project aims to enable interdisciplinary research, collaboration
and rapid development of scientific software. Based on the statistical
programming language R, Bioconductor comprises 934 interoperable packages
contributed by a large, diverse community of scientists. Packages cover a range
of bioinformatic and statistical applications. They undergo formal initial
review and continuous automated testing.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor provides core data structures and methods that enable
genome-scale analysis of highthroughput data in the context of the rich
statistical programming environment offered by the R project. It supports many
types of high-throughput sequencing data (including DNA, RNA, chromatin
immunoprecipitation, Hi-C, methylomes and ribosome profiling) and associated
annotation resources; contains mature facilities for microarray analysis; and
covers proteomic, metabolomic, flow cytometry, quantitative imaging,
cheminformatic and other high-throughput data. Bioconductor enables the rapid
creation of workflows combining multiple data types and tools for statistical
inference, regression, network analysis, machine learning and visualization at
all stages of a project from data generation to publication.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor is also a flexible software engineering environment in which to
develop the tools needed, and it offers users a framework for efficient learning
and productive work. The foundations of Bioconductor and its rapid coevolution
with experimental technologies are based on two motivating principles. The first
is to provide a compelling user experience. Bioconductor documentation comes at
three levels: workflows that document complete analyses spanning multiple tools;
package vignettes that provide a narrative of the intended uses of a particular
package, including detailed executable code examples; and function manual pages
with precise descriptions of all inputs and outputs together with working
examples. In many cases, users ultimately become developers, making their own
algorithms and approaches available to others. The second is to enable and
support an active and open scientific community developing and distributing
algorithms and software in bioinformatics and computational biology. The support
includes guidance and training on software development and documentation, as
well as the use of appropriate programming paradigms such as unit testing and
judicious optimization. A primary goal is the distributed development of
interoperable software components by scientific domain experts. <strong>In part we
achieve this by urging the use of common data structures that enable workflows
integrating multiple data types and disciplines.</strong> To facilitate research and
innovation, we employ a high-level programming language. This choice yields
rapid prototyping, creativity, flexibility and reproducibility in a way that
neither point-and-click software nor a general-purpose programming language can.
We have embraced R for its scientific and statistical computing capabilities,
for its graphics facilities and for the convenience of an interpreted language.
R also interfaces with low-level languages including C and C++ for
computationally intensive operations, Java for integration with enterprise
software and JavaScript for interactive web-based applications and reports.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Case study: high-throughput sequencing data analysis.</strong> Analysis of
large-scale RNA or DNA sequencing data often begins with aligning reads to a
reference genome, which is followed by interpretation of the alignment patterns.
Alignment is handled by a variety of tools, whose output typically is delivered
as a BAM file. The Bioconductor packages Rsamtools and GenomicAlignments provide
a flexible interface for importing and manipulating the data in a BAM file, for
instance for quality assessment, visualization, event detection and
summarization. The regions of interest in such analyses are genes, transcripts,
enhancers or many other types of sequence intervals that can be identified by
their genomic coordinates. Bioconductor supports representation and analysis of
genomic intervals with a ‘Ranges’ infrastructure that encompasses data
structures, algorithms and utilities including arithmetic functions, set
operations and summarization (Fig. 1). It consists of several packages
including IRanges, GenomicRanges, GenomicAlignments, GenomicFeatures,
VariantAnnotation and rtracklayer. The packages are frequently updated for
functionality, performance and usability. <strong>The Ranges infrastructure was designed
to provide tools that are convenient for end users analyzing data while
retaining flexibility to serve as a foundation for the development of more
complex and specialized software. We have formalized the data structures to the
point that they enable interoperability, but we have also made them adaptable to
specific use cases by allowing additional, less formalized userdefined data
components such as application-defined annotation.</strong> Workflows can differ vastly
depending on the specific goals of the investigation, but a common pattern is
reduction of the data to a defined set of ranges in terms of quantitative and
qualitative summaries of the alignments at each of the sites. Examples include
detecting coverage peaks or concentrations in chromatin
immunoprecipitation–sequencing, counting the number of cDNA fragments that match
each transcript or exon (RNA-seq) and calling DNA sequence variants (DNA-seq).
Such summaries can be stored in an instance of the class GenomicRanges.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“To facilitate the analysis of experiments and studies with multiple samples,
Bioconductor defines the SummarizedExperiment class. The computed summaries for
the ranges are compiled into a rectangular array whose rows correspond to the
ranges and whose columns correspond to the different samples .. . For a typical
experiment, there can be tens of thousands to millions of ranges and from a
handful to hundreds of samples. The array elements do not need to be single
numbers: the summaries can be multivariate. The SummarizedExperiment class also
stores metadata on the rows and columns. Metadata on the samples usually include
experimental or observational covariates as well as technical information such
as processing dates or batches, file paths, etc. Row metadata comprise the start
and end coordinates of each feature and the identifier of the containing
polymer, for example, the chromosome name. Further information can be inserted,
such as gene or exon identifiers, references to external databases, reagents,
functional classifications of the region (e.g., from efforts such as the
Encyclopedia of DNA Elements (ENCODE)) or genetic associations (e.g., from
genome-wide association studies, the study of rare diseases, or cancer
genetics). The row metadata aid integrative analysis, for example, when matching
two experiments according to overlap of genomic regions of interest. Tight
coupling of metadata with the data reduces opportunities for clerical errors
during reordering or subsetting operations.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“The integrative data container SummarizedExperiment. Its assays component is
one or several rectangular arrays of equivalent row and column dimensions. Rows
correspond to features, and columns to samples. The component rowData stores
metadata about the features, including their genomic ranges. The colData
component keeps track of samplelevel covariate data. The exptData component
carries experiment-level information, including MIAME (minimum information about
a microarray experiment)-structured metadata. The R expressions exemplify how to
access components. For instance, provided that these metadata were recorded,
rowData(se)<span class="math inline">\(entrezId returns the NCBI Entrez Gene identifiers of the features, and se\)</span>tissue returns the tissue descriptions for the samples. Range-based
operations, such as %in%, act on the rowData to return a logical vector that
selects the features lying within the regions specified by the data object CNVs.
Together with the bracket operator, such expressions can be used to subset a
SummarizedExperiment to a focused set of genes and tissues for downstream
analysis.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“A genomics-specific visualization type is plots along genomic coordinates.
There are several packages that create attractive displays of along-genome data
tracks, including Gviz and ggbio … <strong>These packages operate directly on common
Bioconductor data structures and thus integrate with available data manipulation
and modeling functionality.</strong> A basic operation underlying such visualizations
is computing with genomic regions, and the biovizBase package provides a bridge
between the Ranges infrastructure and plotting packages.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Genomic data set sizes sometimes exceed what can be managed with standard
in-memory data models, and then tools from high performance computing come into
play. An example is the use of rhdf5—an interface to the HDF5 large data
management system (<a href="http://www.hdfgroup.org/HDF5" class="uri">http://www.hdfgroup.org/HDF5</a>)—by the h5vc package to slice
large, genome-size data cubes into chunks that are amenable for rapid
interactive computation and visualization. Both ggbio and Gviz issue
range-restricted queries to file formats including BAM, BGZIP/Tabix and BigWig
via Rsamtools and rtracklayer to quickly integrate data from multiple files over
a specific genomic region.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Developers are constantly updating their packages to extend capabilities, improve performance, fix bugs and enhance documentation. These changes are introduced into the development branch of
Bioconductor and released to end users every 6 months; changes are
tracked using a central, publicly readable Subversion software repository, so details of all changes are fully accessible. Simultaneously,
R itself is continually changing, typically around performance
enhancements and increased functionality. Owing to this dynamic
environment, all packages undergo a daily testing procedure. Testing
is fully automated and ensures that all code examples in the package documentation, as well as further unit tests, run without error.
Successful completion of the testing will result in the package being
built and presented to the community.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Interoperability between software components for different stages and types
of analysis is essential to the success of Bioconductor. Interoperability is
established through the definition of common data structures that package
authors are expected to use … Technically, Bioconductor’s common data
structures are implemented as classes in the S4 object-oriented system of the R
language. In this manner, useful software concepts including encapsulation,
abstraction of interface from implementation, polymorphism, inheritance and
reflection are directly available. It allows core tasks such as matching of
sample data and metadata to be adopted across disciplines, and it provides a
foundation on which community development is based. It is instructive to compare
such a representation to popular alternatives in bioinformatics: file-based data
format conventions and primitive data structures of a language such as matrices
or spreadsheet tables. With file-based formats, operations such as subsetting or
data transformation can be tedious and error prone, and the serialized nature of
files discourages operations that require a global view of the data. In either
case, validity checking and reflection cannot rely on preformed or standardized
support and need to be programmed from scratch again for every convention—or are
missing altogether. As soon as the data for a project are distributed in
multiple tables or files, the alignment of data records or the consistency of
identifiers is precarious, and interoperability is hampered by having to
manipulate disperse, loosely coordinated data collections.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<p>Some of the most important data structures in Bioconductor are <span class="citation">(Wolfgang Huber et al. 2015)</span> (from Table 2 in this reference):</p>
<ul>
<li><code>ExpressionSet</code> (<code>Biobase</code> package)</li>
<li><code>SummarizedExperiment</code> (<code>GenomicRanges</code> package)</li>
<li><code>GRanges</code> (<code>GenomicRanges</code> package)</li>
<li><code>VCF</code> (<code>VariantAnnotation</code> package)</li>
<li><code>VRanges</code> (<code>VariantAnnotation</code> package)</li>
<li><code>BSgenome</code> (<code>BSgenome</code> package)</li>
</ul>
<blockquote>
<p>“For Bioconductor, which provides tools in R for analyzing genomic data,
interoperability was essential to its success. We defined a handful of data
structures that we expected people to use. For instance, if everybody puts their
gene expression data into the same kind of box, it doesn’t matter how the data
came about, but that box is the same and can be used by analytic tools. Really,
I think it’s data structures that drive interoperability.” — Robert Gentlemen
in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“I have found that real hardcore software engineers tend to worry about
problems that are just not existent in our space. They keep wanting to write
clean, shiny software, when you know that the software that you’re using today
is not the software you’re going to be using this time next year.” — Robert
Gentlemen in <span class="citation">(Altschul et al. 2013)</span></p>
</blockquote>
<blockquote>
<p>“Biology, formerly a science with sparse, often only qualitative data, has
turned into a field whose production of quantitative data is on par with high
energy physics or astronomy and whose data are wildly more heterogeneous and
complex.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Any biological system or organism is composed of tens of thousands of
components, which can be in different states and interact in multiple ways.
Modern biology aims to understand such systems by acquiring comprehensive—and
this means high dimensional—data in their temporal and spatial context, with
multiple covariates and interactions.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Biological data come in all sorts of shapes: nucleic acid and protein sequences,
rectagular tables of counts, multiple tables, continuous variables, batch factors,
phenotypic images, spatial coordinates. Besides data measured in lab experiments,
there are clinical data, longitudinal information, environmental measurements,
networks, lineage trees, annotation from biological databases in free text or
controlled vocabularies, …” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor packages support the reading of many of the data types and formats
produced by measurement instruments used in modern biology, as well as the
needed technology-specific ‘preprocessing’ routines. This community is
actively keeping these up-to-date with the rapid developments in the
instrument market.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“The Bioconductor project has defined specialized data containers to represent
complex biological datasets. These help to keep your data consistent, safe and
easy to use.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Bioconductor in particular contains packages from diverse authors that cover
a wide range of functionalities but still interoperate because of the common
data containers.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<code>IRanges</code> is a general container for mathematical intervals. We create the
biological context with the next line [which uses <code>GRanges</code>]. [Footnote:
‘The ’I in IRanges stands for ’interval,’ the ‘G’ in GRanges for ‘genomic’].”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Here we had to assemble a copy of the expression data (<code>exprs(x)</code>) and
the sample annotation data (<code>pData(x)</code>) all together into the dataframe
<code>dftx</code>—since this is the data format that ggplot2 functions most easily take
as input.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>GRanges is “a specialized class from the Bioconductor project for storing
data that are associated with genomic coordinated. The first three columns are
obligatory: <code>seqnames</code>, the name of the containing biopolymer (in our case, the
names of human chromosomes); <code>ranges</code>, the genomic coordinates of the intervals
(in this case, the intervals all have lengths 1, as they refer to a single
nucleotide), and the DNA <code>strand</code> from which the RNA is transcribed. You can
find out more on how to use this class and its associated infrastructure in
the documentation, e.g., the vignette of the <code>GenomicRanges</code> package. Learning
it is worth the effort if you want to work with genome-associated datasets, as
it enables convenient, efficient and safe manipulation of these data and provides
many powerful utilities.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>ChiP-Seq data “are sequences of pieces of DNA that are obtained from chromatin
immunoprecipitation (ChIP). This technology enables the mapping of the locations
along genomic data of transcription factors, nucleosomes, histone modifications,
chromatin remodeling enzymes, chaperones, polymerases and other proteins. It was
the main technology used by the Encyclopedia of DNA Elements (ENCODE) project.
Here we use an example (Kuan et al., 2011) from the <code>mosaicsExample</code> package,
which shows data measured on chromosome 22 from a ChIP-Seq of antibodies for the
STAT1 protein and the H3K4me3 histone modification applied to the GM12878 cell
line. Here we do not show the code used to construct the <code>binTFBS</code> object that
contains the binding sites for one chromosome (22) [in a <code>BinData</code> class, it
looks like from the <code>mosaics</code> package perhaps].” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“At different stages of their development, immune cells express unique
combinations of proteins on their surfaces. These protein-markers are called
CDs (clusters of differentiation) and are collected by flow cytometry
(using fluorescence…) or mass cytometry (using single-cell atomic mass
spectrometry of heavy metal reporters). An example of a commonly used CD is
CD4; this protein is expressed by helper T cells that are referred to as
being ‘CD4+.’ Note, however, that some cells express CD4 (thus are CD4+)
but are not actually helper T cells. We start by loading some useful Bioconductor
packages for flow cytometry, <code>flowCore</code> and <code>flowViz</code>.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Many datasets consist of several variables measured on the same set of
subjects: patients, samples or organisms. For instance, we may have biometric
characteristics such as height, weight, age as well as clinical variables such
as blood pressure, blood sugar, heart rate and genetic data for, say, a
thousand patients. The raison d’etre for multivariate analysis is the
investigation of connections or associations between the different variables
measured. Usually the data are reported in a tabular data structure, with one
row for each subject and one column for each variable. … in the special
case where each of the variables is numeric, … we can represent the data
structure as a matrix in R. If the columns of the matrix are independent of
each other (unrelated), we can simply study each column separately and do
standard ‘univariate’ statistics on them one by one; there would be no
benefit in studying them as a matrix. More often, there will be patterns and
dependencies. For instance, in the biology of cells, we know that the
proliferation rate will influence the expression of many genes simultaneously.
Studying the expression of 25,000 genes (columns) on many samples (rows) of
patient-derived cells, we notice that many of the genes act together; either
they are positively correlated or they are anti-correlated. We would miss a
lot of important information if we were to only study each gene separately.
Important connections between genes are detectable only if we consider the
data as a whole, each row representing the many measurements made on the
same observational unit. However, having 25,000 dimensions of variation to
consider at once is daunting; [you can] reduce our data to a smaller number
of the most important dimensions without losing too much information.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“RNA-Seq transcriptome data report the number of sequence reads matching
each gene [or sub-gene structure, such as exons] in each of several biological
samples… It is customary in the RNA-Seq field … to report genes in rows and
samples in columns. Compared with the other matrices we look at here, this
is transposed: rows and columns swapped. Such different conventions easily
lead to errors, so they are worth paying attention to. [Footnote: ‘The
Bioconductor project tries to help users and developers to avoid such
ambiguities by defining data containers in which such conventions are
explicitly fixed…’]” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Proteomic profiles:</strong> Here the columns are aligned mass spectroscopy peaks
or molecules identified through their m/z ratios; the entries in the matrix
are measured intensities.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“… unlike regression, PCA treats all variables equally (to the extent that
they were preprocessed to have equivalent standard deviations). However, it is
still possible to map other continuous variables or categorical factors onto
plots in order to help interpret the results. Often we have supplementary
information on the samples, for example diagnostic lables in the diabetes
data or cell types in the T-cell gene expression data. Here we see how we
can use such extra variables to inform our interpretation. The best place to
store such so-called <strong>metadata</strong> is in approapriate slots of the data object
(such as in the Bioconductor <code>SummarizedExperiment</code> class); the second best
is in additional columns of the data frame that also contains the numeric
data. In practice, such information is often stored in a more or less
cryptic manner in the row names of the matrix.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Multivariate data anlayses require ‘conscious’ preprocessing. After
consulting all the means, variances, and one-dimensional histograms, we
saw how to rescale and center the data.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Many measurement devices in biotechnology are based on massively parallel
sampling and counting of molecules. One example is high-throughput DNA sequencing.
It’s applications fall broadly into two main classes of data output. In the
first case, the outputs of interest are the sequences themselves, perhaps
also their polymorphisms or differences from other sequences seen before. In
the second case, the sequences themselves are more or less well understood
(if, say, we have a well-assembled and annotated genome) and our interest
is the abundance of different sequence regions in our sample. For instance,
in RNA-Seq…, we sequence the RNA molecules found in a population of cells
or in a tissue. In ChIP-Seq, we sequence DNA regions that are bound to a particular
RNA-binding protein. In DNA-Seq, we sequence genomic DNA and are interested in
the prevalence of genetic variants in heterogeneous populations of cells, for
instance the clonal composition of a tumor. In high-throughput chromatin
conformation capture (HiC) we aim to map the 3D spatial arrangement of
DNA. In genetic screens (using, say, RNAi or CRISPR-Cas9 libraries for
perturbation and high-throughput sequencing for readout), we’re interested
in the proliferation or survival of cells upon gene knockdown,
knockout, or modification. In microbiome analysis, we study the abundance of
different microbial species in complex microbial habitats. Ideally, we
might want to sequence and count <em>all</em> molecules of interest in the sample.
Generally this is not possible; the biochemical protocols are not 100%
efficient, and some molecules or intermediates get lost along the way.
Moreover, it’s often also not even necessary. Instead, we sequence and
count a <em>statistical sample</em>. The sample size will depend on the complexity of the
sequence pool assayed; it can go from tens of thousands to billions. This
<em>sampling</em> nature of the data is important when it comes to analyzing
them. We hope that the sampling is sufficiently representative for us
to identify interesting trends and patterns.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<code>DESeq2</code> uses a specialized data container, called <code>DESeqDataSet</code>, to
store the datasets it works with. Such use of specialized containers—or,
in R terminology, classes—is a common principle of the Bioconductor
project, as it helps users keep related data together. While this way of
doing things requires users to invest a little more time up front to
understand the classes, compared with just using basic R data types like
matrix and dataframe, it helps in avoiding bugs due to loss of
synchronization between related parts of the data. It also enables the
abstraction and encapsulation of common operations that could be quite
wordy if always expressed in basic terms [footnote: Another advantage is that
classes can contain validity methods, which make sure that the data always
fulfill certain expectations, for instance, that the counts are positive
integers, or that the columns of the counts matrix align with the rows
of the sample annotation dataframe.] <code>DESeqDataSet</code> is an extension of the
class <code>SummarizedExperiment</code> in Bioconductor. The <code>SummarizedExperiment</code> class
is also used by many other packages, so learning to work with it will
enable you to use a large range of tools. We will use the constructor function
<code>DESeqDataSetFromMatrix</code> to create a <code>DESeqDataSet</code> from the count data matrix
… and the sample annotation dataframe … The <code>SummarizedExperiment</code> class—and
therefore <code>DESeqDataSet</code>—also contains facilities for storing annotations
of the rows of the count matrix.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“We introduced the R data.frame class, which allows us to combine
heterogeneous data types: categorical factors and continuous measurements.
Each row of the dataframe corresponds to an object, or a record, and the
columns are the different variables or features. Extra information about
sample batches, dates of measurement and different protocols is often
misnamed metadata. This information is actually real data that needs to
be integrated into analyses. Here we show an example of an analysis that was
done by Holmes et al. (2011) on bacterial abundance data from Phylochip microarrays.
The experiment was designed to detect differences between a group of healthy
rats and a group that had irritable bowel disease. This example shows how
nuisance batch effects can become apparent in the analysis of experimental
data. It illustrates why best practices in data analysis are sequential and
why it is better to analyze data as they are collected—to adjust for severe
problems in the experimental design as they occur—instead of trying to deal
with deficiencies post mortem. When data collection started on this project, data
for days 1 and 2 were delivered and we made the plot … This shows a definite
day effect. When investigating the source of this effect, we found that both the
protocol and the array were different on days 1 and 2. This leads to uncertainty
about the source of variation; we call this confounding of effects.”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“Many programs and workflows in biological sequence analysis or assays
separate the environmental and contextual information they call metadata
from the assays or sequence read numbers; we discourage this practice, as the
exact connections between the samples and covariates are important. The lost
connections between the assays and covariates makes later analyses impossible.
Covariates such as clinical history, time, batch and location are important
and should be considered components of the data.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“The data provide and example of an awkward way of combining bach information
from the actual data. The day information has been combined with the array
data and encoded as a number and could be confused with a continuous variable.
We will see in the next section a better practice for storing and manipulating
heterogeneous data using a Bioconductor container called <code>SummarizedExperiment</code>”
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“A more rational way of combining the batch and treatment information into
compartments of a composite object is to use <code>SummarizedExperiment</code> classes.
These include special slots for the assay(s) where rows represent features
of interest (e.g., genes, transcripts, exons, etc.) and columns represent samples.
Supplementary information about the features can be stored in a <code>DataFrame</code>
object, accessible using the function <code>rowData</code>. Each row of the <code>DataFrame</code>
provides informaiton on the feature in the corresponding row of the
<code>SummarizedExperiment</code> object. … This is the best way to keep all the
relevant data together. It will also enable you to quickly filter the
data while keeping all the information aligned properly. …
Columns of the <code>DataFrame</code> represent different attributes of the features
of interest, e.g., gene or transcript IDs. This is an example of a hybrid
data container from a single-cell experiment…” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“The success of the tidyverse attests to the power of its underlying ideas and
the quality of its implementation. … Nevertheless, dataframes in the long
format are not a panacea. … When we write a function that expects to work on
an object like <code>xldf</code>, we have no guarantee that the column <code>probe</code> does indeed
contain valid probe identifiers, or that such a column even exists. There is not
even a proper way to express programmatically what ‘an object like <code>xldf</code> means
in the tidyverse. Object-oriented (OO) programming, and its incarnation S4 in R,
solves such questions. For instance, the above-mentioned checks could be
performed by a <code>validObject</code> method for a suitably defined class, and the class
definition would formalize the notion of ’an object like <code>xldf</code>.’ Addressing
such issues is behind the object-oriented design of the data structures in
Bioconductor, such as the <code>SummarizedExperiment</code> class. Other potentially useful
features of OO data representations include: 1. Abstraction of interface from
implementation and encapsulation: the user accesses the data only through
defined channels and does not need to see how the data are stored
‘inside’—which means that inside can be changed and optimized without breaking
user-level code. 2. Polymorphism: you can have different functions with the same
name, such as plot or filter, for different classes of objects, and R figures
out for you which to call. 3. Inheritance: you can build up more complex data
representations from simpler ones. 4. Reflection and self-documentation: you can
send programmatic queries to an object to ask for more information about itself.
All of these make it easier to write high-level code that focuses on the ‘big
picture’ functionality rather than on implementation details of the building
blocks—albeit at the cost of more initial investment in infrastructure and
‘bureaucracy.’” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Data provenance and metadata.</strong> THere is no obvious place in an object like
<code>xldf</code> to add information about data provenance: e.g., who performed the
experiment, where it was published, where the data were downloaded from, or
which version of the data we’re looking at (data bugs exist …). Neither are
there any explanations of the columns, such as units and assay type. Again, the
data classes in Bioconductor try to address this.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p><strong>Matrix-like data.</strong> Many datasets in biology have a natural matrix-like
structure, since a number of features (e.g., genes: conventionally, the rows of
the matrix) are assayed on several samples (conventionally, the columns of the
matrix). Unrolling the matrix into a long form like <code>xldf</code> makes some operations
(say, PCA, SVD, clustering of features or samples) more awkward."
<span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Out-of memory data and chunking.</strong> Some datasets are too big to load
into random access memory (RAM) and manipulate all at once. Chunking means
splitting the data into manageable portions (‘chunks’) and then sequentially
loading each portion, computing on it, storing the results and removing it
from memory before loading the next portion. R also offers infrastructure
for working with large datasets that are stored on disk in a relational
database management systems (the DBI package) or in HDF5 (the rhdf5 package).
The Bioconductor project provides the class <code>SummarizedExperiment</code>, which
can store big data matrices either in RAM or in an HDF5 backend in a manner
that is transparent to the user of objects of this class.” <span class="citation">(Holmes and Huber 2018)</span></p>
</blockquote>
<p>Many of the statistical algorithms rely on matrices—these store data all of
the same data type (e.g., numeric or counts). If you store extra variables, like
binary outcome classifications (sick/well; alive/dead) or categorical variables,
it will complicate these operations. Further, if these aren’t to be used in things
like dimension reduction and clustering, then you will continuously need to
subset as you perform those matrix-based processes. Conversely, once you move
to using ggplot to visualize your data and other tidy tools to create summary
tables and other output for reports, it’s handy to have all the information
revlevant to each of your observations handy within a dataframe—a structure
than can hold and align data of many different types in its different columns.
It therefore makes sense to evolve from more complex object types, in which
different types of variables for each observation are stored in their own places,
and where variables with similar types can be collected in a matrix that is
ready for statistical processing, to the simpler dataframe at later stages in
the pipeline, when working on publication-ready tables and figures. This requires
a switch as some point in the pipeline from a coding approach that stores data
in more complex Bioconductor S4 objects to one that stores data in a simple
and straightforward tidy dataframe.</p>
<p>One file format called a <em>fasta</em> file is used to store DNA sequence data. The
<code>Biostrings</code> package has a function for reading these data in from a fasta file.
It stores the data in an instance of the <code>DNAStringSet</code> class from that package.
Within this class are <code>DNAString</code> objects for each sample.</p>
<hr />
<p>Bioconductor is used for many of the R packages for working with genomic and
other bioinformatic data. One characteristic of packages on Bioconductor is that
they make heavy use of a system for object-oriented programming in R. There
are several systems for object-oriented programming in R. Bioconductor relies
heavily on one called S4.</p>
<p>Object-oriented programming allows developers to create <em>methods</em>. These are
functions in R that first check the class of the object that is input, and then
run different code for different functions. For example, <code>summary</code> is one of these
method-style functions. If you call the summary function with the input as a
numeric vector, one set of code will be run: you will get numeric summaries of
the values in that vector, including the minimum, maximum and median. However,
if you run the same function, <code>summary</code>, on a dataframe with columns of factor
data, the output will be a small summary for each column, giving the levels of
the factor in each column and the number of column values in the most common
of those levels.</p>
<p>With this system of writing methods, the same function call can be used for
many different object types. By contrast, other approaches to programming might
constrain a function to work with a single class of object—for example, a
function might work on a dataframe and only a dataframe, not a vector, matrix,
or other more complex types of objects.</p>
<p>These methods often have very names. Examples of these method-style functions
include <code>plot</code>, <code>summary</code>, <code>head</code> [?], [others]. You can try running these
method-style functions on just about any object class that you’re using to
store your data, and chances are good that it will work on the object and
output something interesting.</p>
<p>The S4 system of object-oriented programming in R allows for something
called <em>inheritance</em>. [More on this.]</p>
<p>As you use R with the Bioconductor packages, you often might not notice
how much S4 objects are being used “under the hood,” as you pre-process
and analyze data. By contrast, you may have learned the “tidyverse” approach
in R, which is a powerful general approach for working with data. The
tidyverse approach is centered on the object class is predominated uses,
the dataframe, and so a lot of attention is given to thinking about that
style of data storage in an object when learning the approach.</p>
<p>A pre-processing pipeline in Bioconductor might take the data through a
number of different object classes over the course of the pipeline.
Different functions within a Bioconductor package may manage this
progression without you being very aware of it. For example, one function
may read the data from the file format for the equipment and move the
data directly into a certain complex object class, which a second function
might input this object class, do some actions on the data, and then output
the result in a different class.</p>
<p>Generally, if the functions in a pipeline handle these object transitions
gracefully, you may not feel the need to dig more deeply into the object
types. However, ideally you should feel comfortable taking a peek at your
data at any step in the process. This can include seeing snippets of the data
in its object (e.g., the first few elements in each component of the data
at that stage) and also feel comfortable visualizing parts of the data in
simple ways.</p>
<p>This is certainly possible even when data are stored in complex or unfamiliar
object classes. However, it’s a bit less natural than exploring your data when
it’s stored in an object class that you feel very comfortable with. For
example, most R programmers have several go-to strategies for checking any
data they have stored in a dataframe. You can develop these same go-to
strategies for data in more complex object classes once you understand
a few basics about the S4 system and the object classes created using this
system.</p>
<p>First, there are a few methods you can use to figure out what’s in a data
object. [More on this. <code>str</code>, some on interactive ways to look at objects?]
Further, most S4 objects will have their own helpfiles [doublecheck], and you
can use this resource to learn more about what it’s storing and where it puts
each piece of data. [More on accessing these help files. <code>?ExpressionSet</code>,
for example.]</p>
<p>Once you know what’s in your object, there are a few ways that you can pull
out different elements of the data. One basic way (it’s a bit heavy-handed,
and best to save for when you’re struggling with other methods) is to
extract <em>slots</em> from the object using the <code>@</code> symbol. If you have worked much
with base R, you will be familiar with pulling out elements of more basic object
classes using <code>$</code>. For example, if you wanted to extract a column named <code>weight</code>
from a dataframe object called <code>experiment_1</code>, you could do so using
the syntax <code>experiment_1$weight</code>. The <code>$</code> operator does not work in the same
way with S4 objects. With these, we say that the elements are stored in different
<em>slots</em> of the object, and each slot can be extracted using <code>@</code>. So if you had
an S4 object with data on animal weights stored in a slot called <code>weight</code>, you
could extract it from an S4 object instanced named <code>experiment_2</code> with
<code>experiment_2@weight</code>.</p>
<p>A more elegant approach is to access elements stored in the object using a
special type of function called an <strong>accessor</strong> function.</p>
<p>One important object class in Bioconductor is <code>ExpressionSet</code>. This object class
helps to keep different elements of data from an experiment aligned—for
example, it helps ensure that higher-level data about each sample is kept
well-aligned with data on more specific values—like measurements from each
metabolite feature [? better example? gene expression values for each gene?]
specific to each sample. The three slots in this object class are <code>pData</code>,
<code>exprs</code>, and <code>fData</code>. The data in these three slots can be accessed using the
accessor functions of <code>pData</code>, <code>exprs</code>, and <code>fData</code>.</p>
<p>Often, the contents of the slots within a Bioconductor class will be a more
generic object type that you’re familiar with, like a matrix or vector.</p>
<p>S4 objects can be set to check that the inputs are valid for that object
class when someone creates a new object of that class. This helps with
quality control in creating new objects, where these issues can be caught
early, before functions are run on the object that assume certain characteristics
of its data.</p>
<p>Methods are also referred to as generic functions within the S4 system?</p>
<hr />
<blockquote>
<p>“the whole point of OOP is not to have to worry about what is inside an object.
Objects made on different machines and with different languages should be able to talk to each other”
— Alan Kay</p>
</blockquote>
<p>This idea in object-oriented programming may be very helpful for large,
multi-developer programming, since different people, or even whole teams could
develop their parts independently. As long as the teams have all agreed on
the way that messages will be passed between different objects and parts of the
code, they could have independence in how they conduct work on their own
objects. There are rules for how things connect, and independence in how
each part works. If the rules for connecting different objects are set, then
this approach allows for immense flexibility in how the code to work with the
objects on their own can be written and changed, without breaking the whole
system of code.</p>
<p>However, the idea of not worrying about what’s inside an object is at odds with
some basic principles for working with experimental data. Exploratory data
analysis is a key principle for improving quality control, rigor, and even
creativity in working with scientific data sets. [More on EDA, including from
Tukey] EDA requires a researcher to be able to explore the data stored inside
an object, ideally at any stage along a pipeline of pre-processing and then
analyzing those data. Therefore, there’s a bit of tension in the S4 approach in
R, between using a system that allows for powerful development of tools to
explore data and the fundamental needs of the researcher to access and explore
their data as they work with it—to “see inside” the objects storing their
data at every step.</p>
<p>Objects store data. They are data structures, with certain rules for where they
store different elements of the data. They also are associated with specific
functions that work with the way they store the data.</p>
<blockquote>
<p>“A programming language serves two related purposes: it provides a vehicle for the programmer to specify actions to be executed and a set of concepts for the programmer to use when thinking about what can be done. The first aspect ideally requires a language that is ‘close to the machine,’ so that all important aspects of a machine are handled simply and efficiently in a way that is reasonably obvious to the programmer. The C language was primarily designed with this in mind. The second aspect ideally requires a language that is ‘close to the problem to be solved’ so that the concepts of a solution can be expressed directly and concisely. The facilities added to C to create C++ were primarily designed with this in mind.”
— Bjarne Stroustrup, <em>The C++ Programming Language</em>, Addison-Wesley, 1986</p>
</blockquote>
<blockquote>
<p>“The basis for OOP started in the early 1960s. A breakthrough involving
instances and objects was achieved at MIT with the PDP-1, and the first
programming language to use objects was Simula 67. It was designed for the
purpose of creating simulations, and was developed by Kristen Nygaard and
Ole-Johan Dahl in Norway. They were working on simulations that deal with
exploding ships, and realized they could group the ships into different
categories. Each ship type would have its own class, and the class would
generate its unique behavior and data. Simula was not only responsible for
introducing the concept of a class, but it also introduced the instance of a
class. The term ‘object oriented programming’ was first used by Xerox PARC in
their Smalltalk programming language. The term was used to refer to the process
of using objects as the foundation for computation. The Smalltalk team was
inspired by the Simula 67 project, but they designed Smalltalk so that it would
be dynamic. The objects could be changed, created, or deleted, and this was
different from the static systems that were commonly used. Smalltalk was also
the first programming language to introduce the inheritance concept. It is this
feature that allowed Smalltalk to surpass both Simula 67 and the analog
programming systems. While these systems were advanced for their time, they did
not use the inheritance concept.” —
<a href="http://www.exforsys.com/tutorials/oops-concepts/the-history-of-object-oriented-programming.html" class="uri">http://www.exforsys.com/tutorials/oops-concepts/the-history-of-object-oriented-programming.html</a></p>
</blockquote>
<blockquote>
<p>“Object-oriented programming is first and foremost about objects. Initially
object-oriented languages were geared toward modeling real world objects so the
objects in a program corresponded to real world objects. Examples might include:
1. Simulations of a factory floor–objects represent machines and raw materials
2. Simulations of a planetary system–objects represent celestial bodies such as
planets, stars, asteroids, and gas clouds 3. A PC desktop–objects represent
windows, documents, programs, and folders 4. An operating system–objects
represent system resources such as the CPU, memory, disks, tapes, mice, and
other I/O devices”
<a href="https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/" class="uri">https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/</a></p>
</blockquote>
<blockquote>
<p>“The idea with an object is that it advertises the types of data that it will store and the types of operations that it allow to manipulate that data. However, it hides its implementation from the user. For a real world analogy, think of a radio. The purpose of a radio is to play the program content of radio stations (actually translate broadcast signals into sounds that humans can understand). A radio has various dials that allow you to control functions such as the station you are tuned to, the volume, the tone, the bass, the power, and so on. These dials represent the operations that you can use to manipulate the radio. The implementation of the radio is hidden from you. It could be implemented using vacuum tubes or solid state transistors, or some other technology. The point is you do not need to know. The fact that the implementation is hidden from you allows radio manufacturers to upgrade the technology within radios without requiring you to relearn how to use a radio.”
<a href="https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/" class="uri">https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/</a></p>
</blockquote>
<blockquote>
<p>“The set of operations provided by an object is called its <strong>interface</strong>. The
interface defines both the names of the operations and the behavior of these
operations. In essence the interface is a contract between the object and the
program that uses it. The object guarantees that it will provide the advertised
set of operations and that they will behave in a specified fashion. Any object
that adheres to this contract can be used interchangeably by the program. Hence
the implementation of an object can be changed without affecting the behavior of
a program.”
<a href="https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/" class="uri">https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/</a></p>
</blockquote>
<blockquote>
<p>“An object is not much good if each one must be custom crafted. For example,
radios would not be nearly as prevalent if each one was handcrafted. What is
needed is a way to provide a blueprint for an object and a way for a ‘factory’
to use this blueprint to mass produce objects. Classes provide this mechanism in
object-oriented programming. A <strong>class</strong> is a factory that is able to mass
produce objects. The programmer provides a class with a blueprint of the desired
type of object. A ‘blueprint’ is actually composed of: 1. A declaration of a set
of variables that the object will possess, 2. A declaration of the set of
operations that the object will provide, and 3. A set of function definitions
that implements each of these operations. The set of variables possessed by each
object are called <strong>instance variables</strong>. The set of operations that the object
provides are called <strong>methods</strong>. For most practical purposes, a method is like a
function. When a program wants a new instance of an object, it asks the
appropriate class to create a new object for it. The class allocates memory to
hold the object’s instance variables and returns the object to the program. Each
object knows which class created it so that when an operation is requested for
that object, it can look up in the class the function that implements that
operation and call that function.”
<a href="https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/" class="uri">https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/</a></p>
</blockquote>
<blockquote>
<p>“In object-oriented programming, <strong>inheritance</strong> means the inheritance of
another object’s interface, and possibly its implementation as well. Inheritance
is accomplished by stating that a new class is a <strong>subclass</strong> of an existing
class. The class that is inherited from is called the <strong>superclass</strong>. The
subclass always inherits the superclass’s complete interface. It can extend the
interface but it cannot delete any operations from the interface. The subclass
also inherits the superclass’s implementation, or in other words, the functions
that implement the superclass’s operations. However, the subclass is free to
define new functions for these operations. This is called <strong>overriding</strong> the
superclass’s implementation. The subclass can selectively pick and choose which
functions it overrides. Any functions that are not overridden are inherited.”
<a href="https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/" class="uri">https://www.ephemeralobjects.org/2014/02/03/a-brief-history-of-object-oriented-programming/</a></p>
</blockquote>
<p>For <code>xcms</code>, basic object class is now <code>SCMSnExp</code>. <span class="citation">(Holmes and Huber 2018)</span>
This is the container the data is stored in while pre-processing
LCMS data with the <code>xcms</code> package.</p>
<blockquote>
<p>“xcms supports analysis of LC/MS data from files in (AIA/ANDI) NetCDF,
mzML/mzXML and mzData format. For the actual data import Bioconductor’s mzR is used.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“Subsequently we load the raw data as an OnDiskMSnExp object using the readMSData method from the MSnbase package. The MSnbase provides based structures and infrastructure for the processing of mass spectrometry data.
… The resulting OnDiskMSnExp object contains general information about the number of spectra, retention times, the measured total ion current etc, but does not contain the full raw data (i.e. the m/z and intensity values from each measured spectrum). Its memory footprint is thus rather small making it an ideal object to represent large metabolomics experiments while allowing to perform simple quality controls, data inspection and exploration as well as data sub-setting operations. The m/z and intensity values are imported from the raw data files on demand, hence the location of the raw data files should not be changed after initial data import.”
<span class="citation">(Smith 2013)</span></p>
</blockquote>
<p>Important skills for working with data in Bioconductor classes:</p>
<ul>
<li>How to get to the help file specific for that type of class for
generic functions. For example, how to see specific parameters that
can be included in a plot call.</li>
<li>How to find/list all the functions / methods available for that
class of object.</li>
<li>How to access a help file that explains the structure of that object
class. What slots are included? What’s in each slot? How can you
create a new instance of the class? <code>?`Chromatogram-class`</code>
accesses the helpfile that describes the Chromatogram class in the
MSnbase package. It includes info on what is typically stored in
objects of this class (renetion time-intensity value pairs for chromatographic
mass spectroscopy data). It tells how to create a new object of that
class using its constructor function. It lists accessor functions for
objects in that class: <code>rtime</code> to get retention times, <code>intensity</code> to
get the intensities, <code>mz</code> to get the range of the chromatogram, etc.
It also lists some functions, including generic functions like <code>length</code>,
that can be used with objects in that class, as well as some details on
how the class’s method for that generic function works (in terms of
what it will return). It provides the usage, and defines the parameters,
for functions that work with this object class.</li>
<li>How to figure out and use accessor functions to access specific
pieces of data from objects of that class.</li>
</ul>
<p>Often, you’ll have a class that stores data for one sample (e.g.,
<code>Chromatogram</code> from the <code>MSnbase</code> package, which stores chromatographic
mass spectrometry data), and then another class that will collectively
store these sample-specific data in a larger object (e.g.,
<code>Chromatograms</code> class, also from the <code>MSnbase</code> package, which stores
multiple <code>Chromatogram</code> objects, from different samples, in a
structure derived from the matrix structure). j</p>
<p>You can use the pipe operator from <code>magrittr</code> in Bioconductor workflows,
too. It works by “piping” the output from one function call as the
input into the next function call (typically, the parameter in the
first position among parameters to that function call).</p>
<p>Calling the object name at the R console will run the print method for
that object’s class on the object. Often, this will provide a print out of
useful metadata, descriptions, and summaries for the data stored in that
object. If you want a more granular look at what’s contained in the
object, you can use the <code>str</code> function.</p>
<p>Object classes are often set up to inherit from another class. This means
that a method that works for one class might also work for a similar
class, if the second inherits from the first.
“The results are returned as an XCMSnExp object which extends the OnDiskMSnExp object by storing also LC/GC-MS preprocessing results. This means also that all methods to sub-set and filter the data or to access the (raw) data are inherited from the OnDiskMSnExp object and can thus be re-used. Note also that it is possible to perform additional rounds of peak detection (e.g. on MS level &gt; 1 data) on the xdata object by calling findChromPeaks with the parameter add = TRUE.” <span class="citation">(Smith 2013)</span></p>
<p>Sometimes there will be a class just for storing the parameters for
running an algorithm, for example, the “CentWaveParam” and
“MergeNeighboringPeaksParam” classes in the <code>xcms</code> package.
Presumably this is to allow validity checking before using them
in the algorithm?</p>
<p>Moving into a more general object class after pre-processing:</p>
<blockquote>
<p>“Results from the xcms-based preprocessing can be summarized into a SummarizedExperiment object from the SummarizedExperiment package with the quantify method. This object will contain the feature abundances as the assay matrix, the feature definition (their m/z, retention time and other metadata) as rowData (i.e. row annotations) and the sample/phenotype information as colData (i.e. column annotations). All the processing history will be put into the object’s metadata. This object can then be used for any further (xcms-independent) processing and analysis.” <span class="citation">(Smith 2013)</span></p>
</blockquote>
<blockquote>
<p>“The concept in R of attributes of an object allows an exceptionally rich
set of data objects. S3 methods make the class attribute the driver of an
object-oriented system. It is an optional system. Only if an object has a class
attribute do S3 methods really come into effect.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“There are some functions that are generic. Examples include print, plot,
summary. These functions look at the class attribute of their first argument. If
that argument does have a class attribute, then the generic function looks for a
method of the generic function that matches the class of the argument. If such a
match exists, then the method function is used. If there is no matching method
or if the argument does not have a class, then the default method is used.
Let’s get specific. The lm (linear model) function returns an object of class
‘lm.’ Among the methods for print are print.lm and print.default. The
result of a call to lm is printed with print.lm. The result of 1:10 is printed
with print.default.”
<span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“S3 methods are simple and powerful. Objects are printed and plotted and
summarized appropriately, with no effort from the user. The user only needs to
know print, plot and summary.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“If your mystery number is in obj, then there are a few
ways to look for it:
<code>print.default(obj)</code>
<code>print(unclass(obj))</code>
<code>str(obj)</code>
The first two print the object as if it had no class, the last prints an outline of
the structure of the object. You can also do:
<code>names(obj)</code>
to see what components the object has—this can give you an overview of the
object.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“median is a generic
function as evidenced by the appearance of UseMethod. What the new user
meant to ask was, ‘How can I find the default method for median?’
The most sure-fire way of getting the method is to use getS3method:
getS3method(’median’, ’default’).” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“The methods function lists the methods of a generic function [for classes loaded in the current session]. Alternatively
given a class it returns the generic functions that have methods for the class.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<pre><code>## [1] &quot;print.acf&quot;     &quot;print.AES&quot;     &quot;print.anova&quot;   &quot;print.aov&quot;    
## [5] &quot;print.aovlist&quot; &quot;print.ar&quot;</code></pre>
<pre><code>## [1] ExpressionSet,environment-method ExpressionSet,matrix-method     
## [3] ExpressionSet,missing-method    
## see &#39;?methods&#39; for accessing help and source code</code></pre>
<blockquote>
<p>“Inheritance should be based on similarity of the structure of the objects,
not similarity of the concepts for the objects. Matrices and data frames have
similar concepts. Matrices are a specialization of data frames (all columns of the
same type), so conceptually inheritance makes sense. However, matrices and
data frames have completely different implementations, so inheritance makes
no practical sense. The power of inheritance is the ability to (essentially) reuse
code.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“S3 methods are simple and powerful, and a bit ad hoc. S4 methods remove
the ad hoc—they are more strict and more general. The S4 methods technology
is a stiffer rope—when you hang yourself with it, it surely will not break. But
that is basically the point of it—the programmer is restricted in order to make
the results more dependable for the user. That’s the plan anyway, and it often
works.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“S4 is quite strict about what an object of a specific class looks like. In contrast
S3 methods allow you to merely add a class attribute to any object—as long
as a method doesn’t run into anything untoward, there is no penalty. A key
advantage in strictly regulating the structure of objects in a particular class is
that those objects can be used in C code (via the .Call function) without a
copious amount of checking.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>“Along with the strictures on S4 objects comes some new vocabulary. The
pieces (components) of the object are called slots. Slots are accessed by the <code>@</code>
operator.” <span class="citation">(Burns 2011)</span></p>
</blockquote>
<blockquote>
<p>"By now you will have noticed that S4 methods are driven by the class
attribute just as S3 methods are. This commonality perhaps makes the two
systems appear more similar than they are. In S3 the decision of what method
to use is made in real-time when the function is called. In S4 the decision is
made when the code is loaded into the R session—there is a table that charts
the relation. <span class="citation">(Burns 2011)</span></p>
</blockquote>
</div>
<div id="subsection-2-6" class="section level3" number="3.5.5">
<h3><span class="header-section-number">3.5.5</span> Subsection 2</h3>
</div>
<div id="applied-exercise-5" class="section level3" number="3.5.6">
<h3><span class="header-section-number">3.5.6</span> Applied exercise</h3>
<ul>
<li>[Example data in a basic list]</li>
<li>[Example data in a Bioconductor list-based class]</li>
<li>[Explore each example dataset. What slots do each have? What are the names of each slot?
What data structures / data types are in each slot?]</li>
<li>[Extract certain elements from each dataset by hand. Assign to its own object name so
you can use it by itself.]</li>
<li>[Use <code>biobroom</code> to extract pieces of data in the Bioconductor dataset as tidy dataframes.
Try using this with further tidyverse code to create a nice table/visualization.]</li>
</ul>

</div>
</div>
<div id="example-converting-from-complex-to-tidy-data-formats" class="section level2" number="3.6">
<h2><span class="header-section-number">3.6</span> Example: Converting from complex to ‘tidy’ data formats</h2>
<p>We will provide a detailed example of a case where data pre-processing in R
results in a complex, ‘untidy’ data format. We will walk through an example of
applying automated gating to flow cytometry data. We will demonstrate the
complex initial format of this pre-processed data and then show trainees how a
‘tidy’ dataset can be extracted and used for further data analysis and
visualization using the popular R ‘tidyverse’ tools. This example will use real
experimental data from one of our Co-Is research on the immunology of
tuberculosis.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Describe how tools like  were used in this real research
example to convert from the complex data format from pre-processing to a format
better for further data analysis and visualization</li>
<li>Understand how these tools would fit in their own research pipelines</li>
</ul>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>## ✓ ggplot2 3.3.3     ✓ purrr   0.3.4
## ✓ tibble  3.0.6     ✓ dplyr   1.0.4
## ✓ tidyr   1.1.2     ✓ stringr 1.4.0
## ✓ readr   1.4.0     ✓ forcats 0.5.1</code></pre>
<pre><code>## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## x dplyr::combine()    masks Biobase::combine(), BiocGenerics::combine()
## x tidyr::extract()    masks magrittr::extract()
## x dplyr::filter()     masks stats::filter()
## x dplyr::lag()        masks stats::lag()
## x ggplot2::Position() masks BiocGenerics::Position(), base::Position()
## x purrr::set_names()  masks magrittr::set_names()</code></pre>
<div id="subsection-1-8" class="section level3" number="3.6.1">
<h3><span class="header-section-number">3.6.1</span> Subsection 1</h3>
<p>In the previous modules, we have talked about two topics. First we have talked about the convenience and power of tidyverse tools. These tools can be used at points in your workflow when the data can be stored in a simple standard format: the Tidy dataframe format. We have also talked about reasons why there are advantages to using more complex data storage formats earlier in the process. In discussing this, we explained why this leads to early steps of the data preprocessing often being conducted outside of the tidyverse approach, instead using tools available through the bioconductor collection of packages. Once your workflow has advanced to a stage where it is straightforward to store the data in a simpler form at like a dataframe there are a large advantages to shifting into the tidyverse approach as compared to using more complex object-oriented classes for storing the DOT. In this module we will talk through approaches to make the shift from having data stored in bioconductor classes, including very specific classes for some data, into a tidy data format. This will allow an easy transition into using the tiny verse approach for data analysis and visualization at later stages in your workflow.</p>
<p>There are two key tools that have been developed as our packages that facilitate the shift of data from being stored in a more customized object-oriented class, for example one of the S4 type classes that we discussed when talking about complex data formats for bioconductor. These packages move data from one of those storage containers into a tidy dataframe format. By doing this it moves the data into a format that is very easy to use in conjunction with the Tidy barstools and the tidyverse approach.</p>
<p>In this module we will focus specifically on the biobroom package. Of the two
packages this focuses specifically on moving data out of many of the common
bioconductor classes and into tidy dataframes. this package drawers and an
object-oriented approach in that it provides generic functions for extracting
data from many different object classes that are coming in by a conductor. You
will call the same function regardless of the class that the dad is in. If that
object class has a bio broom method for that generic function, then the function
will be able to extract parts of the data into a tidy data frame.</p>
<p>In this module we will also discuss another tool from the tidyverse, or rather a
tool that draws on the tiny verse approach, that can be easily used in
conjunction with biomedical data that has been processed using bioconductor
tools. This is a package called <code>ggbio</code> that facilitates the visualization of
biomedical data. It includes functions and Specialized gian’s or geometrical
objects that are customized for some of the tasks that you might want to conduct
in visualizing biomedical data in r. by drawing on tools and an approach from
ggplot which is part of the tidyverse approach, these tools allow you to work
with this data while still leveraging the powerful visualization tools and
philosophy underlying the ggplot package.</p>
<p>Finally it is quite likely better purchase will continue to evolve through are,
and that in the future there might be tidy data frame format that are adaptable
enough to handle earlier stages in the data preprocessing. Tidy first dataframe
have already been adapted to enable them to include more complex types of data
within certain columns of the data frame any special list type column. This
functionality is being leveraged through the ffs package to an evil a tidy
approach to working with geographical data. This allows those who are working
with geographical data, for example data from shapefiles for creating Maps, to
use the standard tidyverse approaches while still containing complex data needed
for this geographical information. It seems very possible that similar
approaches may be adapted in the near future to allow for biomedical or genomic
data to be stored in a way that both accounts for complexity early and
pre-processing of these data but also allows for a more natural integration with
the wealth of powerful tools available through the tidyverse approach.</p>
</div>
<div id="the-biobroom-package" class="section level3" number="3.6.2">
<h3><span class="header-section-number">3.6.2</span> The <code>biobroom</code> package</h3>
<p>The <code>biobroom</code> package includes three main generic functions (methods), which
can be used on a number of Bioconductor object classes. When applied to object
stored in one of these Bioconductor classes, these functions will extract part
of the data into a tidy dataframe format. In this format, it is easy to use the
tools from the tidyverse to further explore, analyze, and visualize the data.</p>
<p>The three generic functions of <code>biobroom</code> are the functions <code>tidy</code>, <code>augment</code>,
and <code>glance</code>. These function names mimic the names of the three main functions
in the <code>broom</code> package, which is a more general purpose package for extracting
tidy datasets from more complex R object containers. The <code>broom</code> package
focuses on the output from functions in R for statistical testing and modeling,
while the newer <code>biobroom</code> package replicates this idea, but for many of the
common object classes used to store data through Bioconductor packages and
workflows.</p>
<p>The <code>biobroom</code> package includes methods for the following object classes
<span class="citation">(Bass et al. 2020)</span>:</p>
<ul>
<li><code>qvalue</code> objects, which are used …</li>
<li><code>DESeqDataSet</code> objects, which are used …</li>
<li><code>DGEExact</code> objects, which are used …</li>
<li>[limma objects]</li>
<li>[ExpressoinSet objects]</li>
<li><code>MSnSet</code> objects, which are used …</li>
</ul>
<p>As an example, we can look at how the <code>biobroom</code> package can be used to
convert output generated by functions in the <code>edgeR</code> package into a tidy
dataframe, and how that output can then be explored and visualized using
functions from the tidyverse.</p>
<p>The <code>edgeR</code> package is a popular Bioconductor package that can be used on gene
expression data to explore which genes are expressed differently across
experimental groups (<em>differential expression analysis</em>) <span class="citation">(M. D. Robinson, McCarthy, and Smyth 2010)</span>. Before using
the functions in the package, the data must be preprocessed to align sequence
reads from the raw data and then to create a table with the counts of each read
at each gene across each sample. The <code>edgeR</code> package includes functions for
pre-processing through its own functions, as well, including capabilities for
filtering out genes with low read counts across all samples and model-based
normalization across samples to help handle technical bias, including
differences in sequencing depth <span class="citation">(Chen et al. 2014)</span>.</p>
<p>The <code>edgeR</code> package operates on data stored in a special object class
defined by the package, the <code>DGEList</code> object class <span class="citation">(Chen et al. 2014)</span>.
This object class includes areas for storing the table of read counts,
in the form of a matrix appropriate for analysis by other functions in
the package, as well as other spots for storing information about each
sample and, if needed, a space to store annotations of the genes
<span class="citation">(Chen et al. 2014)</span>.</p>
<p>[Example from the <code>biobroom</code> help documentation—uses the <code>hammer</code> data
that comes with the package. These data are stored in an <code>ExpressionSet</code>
object, an object class defined by the <code>Biobase</code> package. You can see how the <code>tidy</code> function extracts these data in a
tidy format. Then, the data are put in a <code>DGEList</code> class so they are in
the right container for operations from <code>edgeR</code>. Then functions from the
<code>edgeR</code> package are run to perform differential expression analysis on the
data. The result is an object in the <code>DGEExact</code> class, which is defined
by the <code>edgeR</code> package. To extract data from this class in a tidy format,
you can use the <code>tidy</code> and <code>glance</code> functions from <code>biobroom</code>.]</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="experimental-data-preprocessing.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biobroom)</span>
<span id="cb8-2"><a href="experimental-data-preprocessing.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biobase)</span>
<span id="cb8-3"><a href="experimental-data-preprocessing.html#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(edgeR)</span>
<span id="cb8-4"><a href="experimental-data-preprocessing.html#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="experimental-data-preprocessing.html#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(hammer)</span>
<span id="cb8-6"><a href="experimental-data-preprocessing.html#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="experimental-data-preprocessing.html#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(hammer)</span></code></pre></div>
<pre><code>## [1] &quot;ExpressionSet&quot;
## attr(,&quot;package&quot;)
## [1] &quot;Biobase&quot;</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="experimental-data-preprocessing.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(hammer)</span></code></pre></div>
<pre><code>## # A tibble: 236,128 x 3
##    gene               sample    value
##    &lt;chr&gt;              &lt;chr&gt;     &lt;int&gt;
##  1 ENSRNOG00000000001 SRX020102     2
##  2 ENSRNOG00000000007 SRX020102     4
##  3 ENSRNOG00000000008 SRX020102     0
##  4 ENSRNOG00000000009 SRX020102     0
##  5 ENSRNOG00000000010 SRX020102    19
##  6 ENSRNOG00000000012 SRX020102     7
##  7 ENSRNOG00000000014 SRX020102     0
##  8 ENSRNOG00000000017 SRX020102     4
##  9 ENSRNOG00000000021 SRX020102     7
## 10 ENSRNOG00000000024 SRX020102    86
## # … with 236,118 more rows</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="experimental-data-preprocessing.html#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Example from `biobroom` help documentation</span></span>
<span id="cb12-2"><a href="experimental-data-preprocessing.html#cb12-2" aria-hidden="true" tabindex="-1"></a>hammer.counts <span class="ot">&lt;-</span> <span class="fu">exprs</span>(hammer)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb12-3"><a href="experimental-data-preprocessing.html#cb12-3" aria-hidden="true" tabindex="-1"></a>hammer.treatment <span class="ot">&lt;-</span> <span class="fu">phenoData</span>(hammer)<span class="sc">$</span>protocol[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb12-4"><a href="experimental-data-preprocessing.html#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="experimental-data-preprocessing.html#cb12-5" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">DGEList</span>(<span class="at">counts=</span>hammer.counts,<span class="at">group=</span>hammer.treatment)</span>
<span id="cb12-6"><a href="experimental-data-preprocessing.html#cb12-6" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">calcNormFactors</span>(y)</span>
<span id="cb12-7"><a href="experimental-data-preprocessing.html#cb12-7" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">estimateCommonDisp</span>(y)</span>
<span id="cb12-8"><a href="experimental-data-preprocessing.html#cb12-8" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">estimateTagwiseDisp</span>(y)</span>
<span id="cb12-9"><a href="experimental-data-preprocessing.html#cb12-9" aria-hidden="true" tabindex="-1"></a>et <span class="ot">&lt;-</span> <span class="fu">exactTest</span>(y)</span>
<span id="cb12-10"><a href="experimental-data-preprocessing.html#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="experimental-data-preprocessing.html#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(et)</span></code></pre></div>
<pre><code>## [1] &quot;DGEExact&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="experimental-data-preprocessing.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(et)</span></code></pre></div>
<pre><code>## # A tibble: 29,516 x 4
##    gene               estimate  logCPM    p.value
##    &lt;chr&gt;                 &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
##  1 ENSRNOG00000000001   2.65    1.49   0.00000131
##  2 ENSRNOG00000000007  -0.409  -0.226  1         
##  3 ENSRNOG00000000008   2.22   -0.407  0.129     
##  4 ENSRNOG00000000009   0      -1.31   1         
##  5 ENSRNOG00000000010   0.0331  1.79   1         
##  6 ENSRNOG00000000012  -3.39    0.0794 0.00375   
##  7 ENSRNOG00000000014   3.65   -0.854  0.252     
##  8 ENSRNOG00000000017   2.42    1.11   0.0000638 
##  9 ENSRNOG00000000021  -2.02    0.211  0.0373    
## 10 ENSRNOG00000000024   0.133   3.97   0.508     
## # … with 29,506 more rows</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="experimental-data-preprocessing.html#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(et)</span></code></pre></div>
<pre><code>##   significant     comparison
## 1        6341 control/L5 SNL</code></pre>
<p>The creator of the <code>broom</code> package listed some of the common ways that
statistical model output objects—the focus on ‘tidying’ in <code>broom</code>—tend to
be untidy. These include that important information is stored in the row names,
where it is harder to access, that the names of some columns can be tricky to
work with because they use non-standard conventions (i.e., they don’t follow the
rules for naming objects in R), that some desired information is not available
in the object, but rather is typically computed with later methods for the
object, like when <code>summary</code> is run on the object, or are only available as the
result of a <code>print</code> method run on the object, and vectors that a user may want
to explore in tandem are stored in different places in the object.
<span class="citation">(D. Robinson 2014)</span></p>
<p>[Examples of these in Bioconductor objects?]</p>
<p>These ‘messy’ characteristics show up in the data stored in Bioconductor
objects, as well, in terms of characteristics that impede working with
data stored in these formats easily using tools from the tidyverse.
As an example, one common class for storing data in Bioconductor work is
the <code>ExpressionSet</code> object class, defined in the <code>Biobase</code> package <span class="citation">(W. Huber et al. 2015)</span>.
This object class can be used to store the data from high-throughput
assays. It includes slots for the assay data, as well as slots
for storing metadata about the experiment, which could include information
like sampling time points or sample strains, as well as the experimental
group of each sample (control versus treated, for example).</p>
<p>Data from the assay for the experiment—for example, gene expression
or intensity [?] measurements for each gene and each sample [?]—can be
extracted from an <code>ExpressionSet</code> object using an extractor function
called <code>exprs</code>. Here is an example using the <code>hammer</code> example dataset
available with the <code>biobroom</code> package. The code call here extracts the
assay data from the <code>hammer</code> R object, which is an instance of the
<code>ExpressionSet</code> object class. It uses indexing (<code>[1:10, 1:3]</code>) to limit
printing to the first ten rows and first three columns of the output, so
we can investigate a small snapshot of the data:</p>
<pre><code>##                    SRX020102 SRX020103 SRX020104
## ENSRNOG00000000001         2         4        18
## ENSRNOG00000000007         4         1         3
## ENSRNOG00000000008         0         1         4
## ENSRNOG00000000009         0         0         0
## ENSRNOG00000000010        19        10        19
## ENSRNOG00000000012         7         5         1
## ENSRNOG00000000014         0         0         2
## ENSRNOG00000000017         4         1        12
## ENSRNOG00000000021         7         5         2
## ENSRNOG00000000024        86        53        86</code></pre>
<p>These data are stored in a matrix format. The gene identifiers [?] are given in
the rownames and the samples in the column names. Each cell of the matrix
provides the expression level (number of reads [?]) of a specific gene in a
specific sample.</p>
<p>These data are structured and stored in such a way that they have some of the
characteristics that can make data difficult to work with the data using
tidyverse tools. For example, they store gene identifiers in the rownames,
rather than in a separate column where they can be easily accessed when using
tidyverse functions. Also, there are phenotype / meta data that are stored
in other parts of the <code>ExpressionSet</code> data but that may be interesting to
explore in conjunction with these assay data, including the experimental
group of each sample (control versus animals in which chronic neuropathic
pain was induced, in these example data).</p>
<p>The <code>tidy</code> function from the <code>biobroom</code> package extracts these data and
restructures them into a ‘tidy’ format, ready to use easily with tidyverse
tools.</p>
<pre><code>## # A tibble: 236,128 x 3
##    gene               sample    value
##    &lt;chr&gt;              &lt;chr&gt;     &lt;int&gt;
##  1 ENSRNOG00000000001 SRX020102     2
##  2 ENSRNOG00000000007 SRX020102     4
##  3 ENSRNOG00000000008 SRX020102     0
##  4 ENSRNOG00000000009 SRX020102     0
##  5 ENSRNOG00000000010 SRX020102    19
##  6 ENSRNOG00000000012 SRX020102     7
##  7 ENSRNOG00000000014 SRX020102     0
##  8 ENSRNOG00000000017 SRX020102     4
##  9 ENSRNOG00000000021 SRX020102     7
## 10 ENSRNOG00000000024 SRX020102    86
## # … with 236,118 more rows</code></pre>
<p>This output is a tidy dataframe object, with three columns providing the
gene name, the sample identifier, and the expression level. In this format,
the data can easily be explored and visualized with tidyverse tools. For example,
you could easily create a set of histograms, one per sample, showing the
distribution of expression levels across all genes in each sample:
[better example visualization here?]</p>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="improve_repro_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>You can also incorporate data that are stored in the <code>phenoData</code> slot of
the <code>ExpressionSet</code> object by specifying <code>addPheno = TRUE</code>:</p>
<pre><code>## # A tibble: 236,128 x 8
##    gene         sample   sample.id num.tech.reps protocol strain    Time   value
##    &lt;chr&gt;        &lt;chr&gt;    &lt;fct&gt;             &lt;dbl&gt; &lt;fct&gt;    &lt;fct&gt;     &lt;fct&gt;  &lt;int&gt;
##  1 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     2
##  2 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     4
##  3 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     0
##  4 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     0
##  5 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…    19
##  6 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     7
##  7 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     0
##  8 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     4
##  9 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…     7
## 10 ENSRNOG0000… SRX0201… SRX020102             1 control  Sprague … 2 mon…    86
## # … with 236,118 more rows</code></pre>
<p>With this addition, visualizations can easily be changed to also show the
experimental group of each sample:</p>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="improve_repro_files/figure-html/unnamed-chunk-12-1.png" width="672" />
You can also do things like look at differences in values for
specific genes, pairing tools for exploring data with tools for
visualization, both from the tidyverse:</p>
<p><img src="improve_repro_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>The data for a subset of the sample can be analyzed using functions from the
<code>edgeR</code> package, to complete needed pre-processing (for example, calculating
normalizing factors with <code>calcNormFactors</code>, to reduce impacts from technical
bias [?]), estimate dispersion using conditional maximum likelihood and
empirical Bayesian methods (<code>estimateCommonDisp</code> and <code>estimateTagwiseDisp</code>), and
then perform a statistical analysis, conducting a differential expression
analysis (<code>exactTest</code>) <span class="citation">(Chen et al. 2014)</span>.</p>
<p>The data are stored in a special Bioconductor class, as an instance of
<code>DGEList</code>, throughout most of this process. This special class can be initialized
with data from the original <code>ExpressionSet</code> object, specifically, assay
data with the counts per gene in each sample and data on the experimental
phenotypes for the experiment—specifically, the protocol for each sample,
in terms of whether it was a control or if the sample was from an animal
in which chronic neuropathic pain was induced <span class="citation">(Hammer et al. 2010)</span>.</p>
<pre><code>## [1] &quot;DGEList&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<p>Once the data are stored in this special <code>DGEList</code> class, different
steps of the preprocessing can be conducted. In each case, the results are
stored in special slots of the <code>DGEList</code> object. In this way, the original
data and results from preprocessing are all kept together in a single
object, each in a special slot within the object’s structure.</p>
<pre><code>## [1] &quot;DGEList&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<pre><code>## [1] &quot;DGEList&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<pre><code>## [1] &quot;DGEList&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<p>After the preprocessing, the data can be analyzed using the <code>exactText</code> function.
This inputs the data stored in a <code>DGEList</code> object and outputs results into
a different object class, a <code>DGEExact</code> class.</p>
<pre><code>## [1] &quot;DGEExact&quot;
## attr(,&quot;package&quot;)
## [1] &quot;edgeR&quot;</code></pre>
<p>The <code>DGEExact</code> class is defined by the <code>edgeR</code> package and was created
specifically to store the results from a differential expression analysis
<span class="citation">(Chen et al. 2014)</span>. It has slots for a dataframe giving the estimates of
differential change in expression across the experimental groups for each
gene, within a <code>table</code> slot. Again, in this output, gene identifiers are
stored as rownames—which makes them hard to access with tidyverse tools—rather
than in their own column:</p>
<pre><code>##                          logFC      logCPM       PValue
## ENSRNOG00000000001  2.64635814  1.49216267 1.309933e-06
## ENSRNOG00000000007 -0.40869816 -0.22616605 1.000000e+00
## ENSRNOG00000000008  2.22296029 -0.40665547 1.288756e-01
## ENSRNOG00000000009  0.00000000 -1.31347471 1.000000e+00
## ENSRNOG00000000010  0.03307909  1.79448965 1.000000e+00
## ENSRNOG00000000012 -3.39210151  0.07939132 3.745676e-03</code></pre>
<p>The <code>DGEExact</code> object also has a slot that contains a vector with identifiers
for the two experimental groups that are being compared in the
differential expression analysis, under the slot <code>comparison</code>:</p>
<pre><code>## [1] &quot;control&quot; &quot;L5 SNL&quot;</code></pre>
<p>There is also a space in this object class where information about each
gene can be stored, if desired.</p>
<p>Two <code>biobroom</code> methods are defined for the <code>DGEExact</code> object class, <code>glance</code>
and <code>tidy</code>. The <code>tidy</code> method extracts the results from the differential
experssion analysis, but moves these results into a dataframe where the
gene names are given their own column, rather than being stored in the
hard-to-access rownames:</p>
<pre><code>## # A tibble: 29,516 x 4
##    gene               estimate  logCPM    p.value
##    &lt;chr&gt;                 &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
##  1 ENSRNOG00000000001   2.65    1.49   0.00000131
##  2 ENSRNOG00000000007  -0.409  -0.226  1         
##  3 ENSRNOG00000000008   2.22   -0.407  0.129     
##  4 ENSRNOG00000000009   0      -1.31   1         
##  5 ENSRNOG00000000010   0.0331  1.79   1         
##  6 ENSRNOG00000000012  -3.39    0.0794 0.00375   
##  7 ENSRNOG00000000014   3.65   -0.854  0.252     
##  8 ENSRNOG00000000017   2.42    1.11   0.0000638 
##  9 ENSRNOG00000000021  -2.02    0.211  0.0373    
## 10 ENSRNOG00000000024   0.133   3.97   0.508     
## # … with 29,506 more rows</code></pre>
<p>Now that the data are in this tidy format, tools from the tidyverse can
be easily applied. For example, you could use functions from the <code>dplyr</code>
package to see the genes for which the differential expression analysis
resulted in both a very low p-value and a large difference in expression
across the experimental groups:</p>
<pre><code>## # A tibble: 803 x 4
##    gene               estimate logCPM   p.value
##    &lt;chr&gt;                 &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt;
##  1 ENSRNOG00000013496     6.39   3.41 5.95e- 46
##  2 ENSRNOG00000001338     6.01   2.25 1.37e- 22
##  3 ENSRNOG00000020136     5.17   3.89 3.19e- 52
##  4 ENSRNOG00000018808     4.78   6.46 1.75e-169
##  5 ENSRNOG00000006151     4.43   2.99 2.39e- 30
##  6 ENSRNOG00000009768     4.40   7.83 5.57e-293
##  7 ENSRNOG00000030927    -4.29   2.45 6.32e- 23
##  8 ENSRNOG00000001476     4.25   3.76 1.45e- 47
##  9 ENSRNOG00000004805     4.05   6.64 2.51e-185
## 10 ENSRNOG00000014327     3.85   4.51 5.39e- 63
## # … with 793 more rows</code></pre>
<p>Other exploratory analysis will also be straightforward with the data
using tidyverse tools, now that they are in a “tidy” format.</p>
<p>The <code>glance</code> method can also be applied to data that are stored in a
<code>DGEExact</code> class. In this case, the method will extract the names of the
experimental groups being compared (from the <code>comparison</code> slot of the
object) as well as count the number of genes with statistically
significant differences in expression level, based on the values in the
<code>table</code> slot of the object.</p>
<pre><code>##   significant     comparison
## 1        6341 control/L5 SNL</code></pre>
<pre><code>##   significant     comparison
## 1        4225 control/L5 SNL</code></pre>
<p>As another example, you can now use tools from <code>ggplot2</code>, as well as
extensions built on this package, to do things like create a volcano
plot of the data with highlighting of noteworthy genes on the plot:</p>
<p><img src="improve_repro_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
<p>If you wanted to access the help files for these <code>biobroom</code> methods for this
object class, you could do so by calling help in R (<code>?</code>) using the name of the
method, a dot, and then the name of the object class, e.g., <code>?tidy.DGEExact</code>.</p>
</div>
<div id="the-ggbio-package" class="section level3" number="3.6.3">
<h3><span class="header-section-number">3.6.3</span> The <code>ggbio</code> package</h3>
</div>
<div id="subsection-2-7" class="section level3" number="3.6.4">
<h3><span class="header-section-number">3.6.4</span> Subsection 2</h3>
<blockquote>
<p>“The biobroom package contains methods for converting standard objects in Bioconductor into a ‘tidy format.’ It serves as a complement to the popular broom package, and follows the same division (tidy/augment/glance) of tidying methods.”
<span class="citation">(Bass et al. 2020)</span></p>
</blockquote>
<blockquote>
<p>“Tidying data makes it easy to recombine, reshape and visualize bioinformatics analyses. Objects that can be tidied include: ExpressionSet object,
GRanges and GRangesList objects, RangedSummarizedExperiment object, MSnSet object,
per-gene differential expression tests from limma, edgeR, and DESeq2, qvalue object for multiple hypothesis testing.” <span class="citation">(Bass et al. 2020)</span></p>
</blockquote>
<blockquote>
<p>“We are currently working on adding more methods to existing Bioconductor objects.” <span class="citation">(Bass et al. 2020)</span></p>
</blockquote>
<blockquote>
<p>“All biobroom tidy and augment methods return a tbl_df by default (this prevents them from printing many rows at once, while still acting like a traditional data.frame).” <span class="citation">(Bass et al. 2020)</span></p>
</blockquote>
<blockquote>
<p>“The concept of ‘tidy data’ offers a powerful framework for structuring data
to ease manipulation, modeling and visualization. However, most R functions,
both those builtin and those found in third-party packages, produce output that
is not tidy, and that is therefore difficult to reshape, recombine, and
otherwise manipulate. Here I introduce the broom package, which turns the output
of model objects into tidy data frames that are suited to further analysis,
manipulation, and visualization with input-tidy tools.” <span class="citation">(D. Robinson 2014)</span></p>
</blockquote>
<blockquote>
<p>“Tools are classified as ‘messy-output’ if their output does not fit into this
[tidy] framework. Unfortunately, the majority of R modeling tools, both from the
built-in stats package and those in common third party packages, are
messy-output. This means the data analyst must tidy not only the original data,
but the results at each intermediate stage of an analysis.” <span class="citation">(D. Robinson 2014)</span></p>
</blockquote>
<blockquote>
<p>“The broom package is an attempt to solve this issue, by bridging the gap from
untidy outputs of predictions and estimations to create tidy data that is easy
to manipulate with standard tools. It centers around three S3 methods, tidy,
augment, and glance, that each take an object produced by R statistical
functions (such as lm, t.test, and nls) or by popular third-party packages (such
as glmnet, survival, lme4, and multcomp) and convert it into a tidy data frame
without rownames (Friedman et al., 2010; Therneau, 2014; Bates et al., 2014;
Hothorn et al., 2008). These outputs can then be used with input-tidy tools such
as dplyr or ggplot2, or downstream statistical tests. broom should be
distinguished from packages such as reshape2 and tidyr, which rearrange and
reshape data frames into different forms (Wickham, 2007b, 2014b). Those packages
perform essential tasks in tidy data analysis but focus on manipulating data
frames in one specific format into another. In contrast, broom is designed to
take data that is not in a data frame (sometimes not anywhere close) and convert
it to a tidy data frame.” <span class="citation">(D. Robinson 2014)</span></p>
</blockquote>
<blockquote>
<p>“<code>tidy</code> constructs a data frame that summarizes the model’s statistical
components, which we refer to as the component level. In a regression such as
the above it may refer to coefficient estimates, p-values, and standard errors
for each term in a regression. The tidy generic is flexible- in other models it
could represent per-cluster information in clustering applications, or per-test
information for multiple comparison functions. … <code>augment</code> add columns to the
original data that was modeled, thus working at the observation level. This
includes predictions, residuals and prediction standard errors in a regression,
and can represent cluster assignments or classifications in other applications.
By convention, each new column starts with . to ensure it does not conflict with
existing columns. To ensure that the output is tidy and can be recombined,
rownames in the original data, if present, are added as a column called
.rownames. … Finally, <code>glance</code> constructs a concise one-row summary of the
model level values. In a regression this typically contains values such as R2 ,
adjusted R2 , residual standard error, Akaike Information Criterion (AIC), or
deviance. In other applications it can include calculations such as cross
validation accuracy or prediction error that are computed once for the entire
model. … These three methods appear across many analyses; indeed, the fact
that these three levels must be combined into a single S3 object is a common
reason that model outputs are not tidy. Importantly, some model objects may have
only one or two of these methods defined. (For example, there is no sense in
which a Student’s T test or correlation test generates information about each
observation, and therefore no augment method exists).” <span class="citation">(D. Robinson 2014)</span></p>
</blockquote>
<blockquote>
<p>“While model inputs usually require tidy inputs, such attention to detail
doesn’t carry over to model outputs. Outputs such as predictions and estimated
coefficients aren’t always tidy. For example, in R, the default representation
of model coefficients is not tidy because it does not have an explicit variable
that records the variable name for each estimate, they are instead recorded as
row names. In R, row names must be unique, so combining coefficients from many
models (e.g., from bootstrap resamples, or subgroups) requires workarounds to
avoid losing important information. This knocks you out of the flow of analysis
and makes it harder to combine the results from multiple models.”
<span class="citation">(Wickham 2014)</span></p>
</blockquote>
<blockquote>
<p>“edgeR can be applied to differential expression at the gene, exon, transcript
or tag level. In fact, read counts can be summarized by any genomic feature.
edgeR analyses at the exon level are easily extended to detect differential
splicing or isoform-specific differential expression.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“edgeR provides statistical routines for assessing differential expression in
RNA-Seq experiments or differential marking in ChIP-Seq experiments. The package
implements exact statistical methods for multigroup experiments developed by
Robinson and Smyth [33, 34]. It also implements statistical methods based on
generalized linear models (glms), suitable for multifactor experiments of any
complexity, developed by McCarthy et al. [22], Lund et al. [20], Chen et al. [5]
and Lun et al. [19]. … A particular feature of edgeR functionality, both
classic and glm, are empirical Bayes methods that permit the estimation of
gene-specific biological variation, even for experiments with minimal levels of
biological replication.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“edgeR performs differential abundance analysis for pre-defined genomic
features. Although not strictly necessary, it usually desirable that these
genomic features are non-overlapping. For simplicity, we will hence-forth refer
to the genomic features as ‘genes,’ although they could in principle be
transcripts, exons, general genomic intervals or some other type of feature. For
ChIP-seq experiments, abundance might relate to transcription factor binding or
to histone mark occupancy, but we will henceforth refer to abundance as in terms
of gene expression.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“edgeR stores data in a simple list-based data object called a DGEList. This
type of object is easy to use because it can be manipulated like any list in R.
… The main components of an DGEList object are a matrix counts containing the
integer counts, a data.frame samples containing information about the samples or
libraries, and a optional data.frame genes containing annotation for the genes
or genomic features. The data.frame samples contains a column lib.size for the
library size or sequencing depth for each sample. If not specified by the user,
the library sizes will be computed from the column sums of the counts. For
classic edgeR the data.frame samples must also contain a column group,
identifying the group membership of each sample.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“Genes with very low counts across all libraries provide little evidence for
differential expression. In the biological point of view, a gene must be
expressed at some minimal level before it is likely to be translated into a
protein or to be biologically important. In addition, the pronounced
discreteness of these counts interferes with some of the statistical
approximations that are used later in the pipeline. These genes should be
filtered out prior to further analysis. As a rule of thumb, genes are dropped if
they can’t possibly be expressed in all the samples for any of the conditions.
Users can set their own definition of genes being expressed. Usually a gene is
required to have a count of 5-10 in a library to be considered expressed in that
library. Users should also filter with count-per-million (CPM) rather than
filtering on the counts directly, as the latter does not account for differences
in library sizes between samples.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“The most obvious technical factor that affects the read counts [in data for
edgeR and so requires normalizations], other than gene expression levels, is the
sequencing depth of each RNA sample. edgeR adjusts any differential expression
analysis for varying sequencing depths as represented by differing library
sizes. This is part of the basic modeling procedure and flows automatically into
fold-change or p-value calculations. It is always present, and doesn’t require
any user intervention.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
<blockquote>
<p>“In edgeR, normalization takes the form of correction factors that enter into
the statistical model. Such correction factors are usually computed internally
by edgeR functions, but it is also possible for a user to supply them. The
correction factors may take the form of scaling factors for the library sizes,
such as computed by calcNormFactors, which are then used to compute the
effective library sizes. Alternatively, gene-specific correction factors can be
entered into the glm functions of edgeR as offsets. In the latter case, the
offset matrix will be assumed to account for all normalization issues, including
sequencing depth and RNA composition. Note that normalization in edgeR is
model-based, and the original read counts are not themselves transformed. This
means that users should not transform the read counts in any way before inputing
them to edgeR.” <span class="citation">(Chen et al. 2014)</span></p>
</blockquote>
</div>
<div id="applied-exercise-6" class="section level3" number="3.6.5">
<h3><span class="header-section-number">3.6.5</span> Applied exercise</h3>

</div>
</div>
<div id="introduction-to-reproducible-data-pre-processing-protocols" class="section level2" number="3.7">
<h2><span class="header-section-number">3.7</span> Introduction to reproducible data pre-processing protocols</h2>
<p>Reproducibility tools can be used to create reproducible data pre-processing
protocols—documents that combine code and text in a ‘knitted’ document, which
can be re-used to ensure data pre-processing is consistent and reproducible
across research projects. In this module, we will describe how reproducible data
pre-processing protocols can improve reproducibility of pre-processing
experimental data, as well as to ensure transparency, consistency, and
reproducibility across the research projects conducted by a research team.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Define a ‘reproducible data pre-processing protocol’</li>
<li>Explain how such protocols improve reproducibility at the data pre-processing
phase</li>
<li>List other benefits, including improving efficiency and consistency of data
pre-processing</li>
</ul>
</div>
<div id="introducing-reproducible-data-pre-processing-protocols" class="section level2" number="3.8">
<h2><span class="header-section-number">3.8</span> Introducing reproducible data pre-processing protocols</h2>
<p><strong>Data pre-processing</strong></p>
<p>When we take measurements of experimental samples, we do so with the goal of
using the data we collect to gain scientific knowledge. The data are direct
measurement of something, but need to be interpreted to gain knowledge.
Sometimes direct measurements line up very closely with a research
question—for example if you are conducting a study that investigates the
mortality status of each test subject then whether or not each subject to dies
is a data point that is directly related to the research question you are aiming
to answer. In this case these data may go directly into a statistical analysis
model without extensive pre-processing. However, there are often cases where we
collect data that are not as immediately linked to the scientific question.
Instead, these data may require pre-processing before they can be used to test
meaningful scientific hypotheses. This is often the case for data extracted
using complex equipment. Equipment like mass spectrometers and flow cytometers
leverage physics, chemistry, and biology in clever ways to help us derive more
information from samples, but one tradeoff is that the data from such equipment
often require a bit of work to move into a format that is useful for answering
scientific questions.</p>
<p>One example if the data collected through liquid chromatography-mass
spectrometry (LC-MS). This is a powerful and useful technique for chemical
analysis, including analysis of biochemical molecules like metabolites and
proteins. However, when using this technique, the raw data require extensive
pre-processing before they can be used to answer scientific questions.</p>
<p>First, the data that are output by the mass spectrometer are often stored in a
specialized file format, like a netCDF or mzML file format. While these file
formats are standardized, they are likely formats you don’t regularly use in
other contexts, and so you may need to find special tools to read the data into
programs to analyze it. In some cases, the data are very large, and so it may be
necessary to use analysis tools that allow most of the data to stay “on disk”
while you analyze it, bringing only small parts into your analysis software at a
time.</p>
<p>Once the data are read in, they must be pre-processed in a number of ways. For
example, these data can be translated into features that are linked to the
chemical composition of the sample, with each feature showing up as a “peak” in
the data that are output from the mass spectrometer. A peak can be linked to a
specific metabolite feature based on its mass-to-charge ratio (m/z) and its
retention time. However, the exact retention time for a metabolite feature may
vary a bit from sample to sample. Pre-processing is required both to identify
peaks in the data and also to align the peaks from the same metabolite feature
across all samples from your experiment. There may also be technical bias across
samples, resulting in differences in the typical expression levels of all peaks
from one sample to the next. For example it may be the case that all intensities
measured for one sample tend to be higher than for another sample because of
technical bias in terms of the settings used for the equipment when the two
samples were run. These biases must also be corrected through pre-processing
before you can use the data within statistical tests or models to explore scientific hypotheses.</p>
<p>[Image of identifying and aligning peaks in LC-MS data]</p>
<p>In the research process, these pre-processing steps should be done before
the data are used for further analysis. There are the first step in working
with the data after they are collected by the equipment (or by laboratory
personal, in the case of data from simpler process, like plating samples
and counting colony-forming units). After the data are appropriately
pre-processed, you can use them for statistical tests—for example, to
determine if metabolite profiles are different between experimental groups—and
also combine them with other data collected from the experiment—for example,
to see whether certain metabolite levels are correlated with the bacterial
load in a sample.</p>
<p><strong>Approaches for pre-processing data.</strong></p>
<p>There are two main approaches for pre-processing experimental data in this
way. First, when data are the output of complex laboratory equipment, there
will often be proprietary software that is available for this pre-processing.
This software may be created by the same company that made the equipment, or
it may be created and sold by other companies. The interface will typically
be a graphical-user interface (GUI), where you will use pull-down menus and
point-and-click interfaces to work through the pre-processing steps. You
often will be able to export a pre-processed version of the data in a
common file format, like a delimited file or an Excel file, and that version
of the data can then be read into more general data analysis software, like
Excel or R.</p>
<p>[Include a screenshot of this type of software in action.]</p>
<p>The second approach is to conduct the pre-processing directly within general
data analysis software like R or Python. These programs are both open-source,
and include extensions that were created and shared by users around the world.
Through these extensions, there are often powerful tools that you can use to
pre-process complex experimental data. In fact, the algorithms used in
proprietary software are sometimes extended from algorithms first shared through
R or Python. With this approach, you will read the data into the program (R,
for example) directly from the file output from the equipment. You can
record all the code that you use to read in and pre-process the data in a
code script, allowing you to reproduce this pre-processing work. You can
also go a step further, and incorporate your code into a pre-processing
protocol, which combines nicely formatted text with executable code, and
which we’ll describe in much more detail later in this module and in the
following two modules.</p>
<p>There are advantages to taking the second approach—using scripted code in an
open-source program—rather than the first—using proprietary software with a
GUI interface. The use of codes scripts ensures that the steps of pre-processing
are reproducible. This means both that you will be able to re-do all the steps
yourself in the future, if you need to, but that also that other researchers can
explore and replicate what you do. You may want to share your process with
others in your laboratory group, for example, so they can understand the choices
you made and steps you took in pre-processing the data. You may also want to
share the process with readers of the articles you publish, and this may in fact
be required by the journal. Further, the use of a code script encourages you to
document this code and this process, even moreso when you move beyond a script
and include the code in a reproducible pre-processing protocol. Well-documented
code makes it much easier to write up the method section later in manuscripts
that leveraged the data collected in the experiment.</p>
<p>Also, when you use scripted code to pre-process biomedical data, you will find
that the same script can often be easily adapted and re-used in later projects
that use the same type of data. You may need to change small elements, like the
file names of files with data you want to use, or some details about the methods
used for certain pre-processing steps. However, often almost all of the
pre-processing steps will repeat over different experiments that you do. By
extending to write a pre-processing protocol, you can further support the
ease of adapting and re-using the pre-processing steps you take with one
experiment when you run later experiments that are similar.</p>
</div>
<div id="from-pre-processing-scripts-to-pre-processing-protocols" class="section level2" number="3.9">
<h2><span class="header-section-number">3.9</span> From pre-processing scripts to pre-processing protocols</h2>
<p>In the previous subsection, we mentioned that you can move a step beyond scripts
when pre-processing, and instead create a pre-processing protocol that includes
formatted text and executable code. This type of protocol can be easily created
using R, and it can serve as a reference for your laboratory group both of what
you did to pre-process data for the current experiment, but also as a starting
point for how to do similar analyses in the future. In module 3.9, we will walk
in detail through an example of this type of pre-processing protocol—you can
take a look now to get an idea by downloading the example
<a href="https://github.com/geanders/improve_repro/raw/master/data/bactcountr_example_data/example_protocol.pdf">here</a>.</p>
<p>This protocol includes all the code that was used for pre-processing, but it
isn’t as limited as a simple code script with comments. Instead,</p>
<p>If you have used open-source software tools, like Bioconductor packages, you
are likely familiar with the <em>vignettes</em> that come with the packages. These
provide tutorial guides showing you how to work with the package. They often
leverage example data that you can download so that you can try all the
example code yourself, before you move on to adapting the code to use with
your own data.</p>
<p>You can create your own version of these types of documents. This can use
real data from your research group, and you can create customized instructions
and code examples showing how to use open-source tools to pre-process a
certain type of biomedical data for experiments in your research group.
You can use this document the next time you need to pre-process that type
of data yourself, and you can also share it with others in your research
group. This can help in teaching new laboratory members how to work with
this type of data in your research group. It can also help ensure that
different members of the research group are all using the same steps to
pre-process data, so that there is greater consistency across results from
the group.</p>
<p>You may already create something similar to this, using a general word
processing program like Google Docs or Word. There are two key differences,
however, between how vignettes are created compared to a similar tutorial
created in Word or Google Docs. First, the vignettes are created using
a document compiling program that ensures that any code uses only ASCII
characters. This means that you can copy and paste code from the tutorial
into your R session and it will work. By contrast, programs like Word often
try to “correct” some of the characters when you paste in or type in code.
For example, when you have an apostrophe mark in your code (for example,
when you’re quoting to create a character string), the computer code needs
to have this character as a very basic ASCII version of an apostrophe. Word,
by contrast, will often try to convert the character to use an apostrophe
character that looks smoother—and so is nice for a word processed document
that humans will read—but that R cannot recognize. Hyphens can have similar
problems. [Other examples?]</p>
<p>When you create a reproducible pre-processing protocol using the techniques that
are used to create vignettes—which we’ll teach you how to do in the next few
sections—you will avoid this autocorrection of characters, and so someone
reading the protocol will be able to directly copy and paste example code from
the protocol into their own scripts. This will avoid hard-to-diagnose errors
that come from this character conversion in programs like Word.</p>
<p>The second difference is that the tools that are used to create vignettes
contain code that is not just copied and pasted from a script, but that is
actually, in essence, <em>still in a script</em>. The code, in other words, is
executable and, unless you change the default settings, is re-run every
time you compile the document. This means that you will quickly determine if
there are any typos or other errors in the code, because the document will
not run and render correctly unless the code works. This means that you can
guarantee, when you first create the document, that the code runs, and also
that you can regularly check to see if the example code still works at
later time points. This allows you to, for example, see if changes in the
version of R or of specific packages that you’re using has created problems
with the code running correctly over time.</p>
<p>Finally, these documents can be separated, allowing you to extract solely the
script part of the document, into a classic R script. You can use this directly
to run (or adapt) the pre-processing code for further research.</p>
<div id="technique-to-create-reproducible-pre-processing-protocols" class="section level3" number="3.9.1">
<h3><span class="header-section-number">3.9.1</span> Technique to create reproducible pre-processing protocols</h3>
<p>The vignettes that come with Bioconductor packages are created using a system
for “knitting” documents. These documents “knit” together text with executable
code. Once you have written the document, you can render it, which executes the
code, adds to the document results from this execution (figures, tables, and
code output, for example), and formats all text using the formatting choices
you’ve specified. The end result is a nicely format document, which can be in
one of several output formats, including pdf, Word, or HTML. Since the code
was executed to create the document, you can ensure that all the code
is worked as intended.</p>
<p>There are several techniques and principles that come together to make these
knitteed documents work. First are the tools that allow you to write text
in plain text, include formatting specifications in that plain text, and
render this to an attractive output document in pdf, Word, or HTML. This
part of the process uses a tool from a set of tools called <em>Markup languages</em>.
[A bit more on history / development of Markup languages.]</p>
<p>Here, we will use a markup language called <em>Markdown</em>. It is one of the easiest
markup languages to learn, as it has a fairly small set of formatting indicators
that can be used to “markup” the formatting in a document. This small set,
however, covers much of the formatting you might want to do, and so this
language provides an easy introduction to markup languages while still providing
adequate functionality for most purposes.</p>
<p>The Markdown markup languages evolved starting in spaces where people could
communicate in plain text only, without point-and-click methods for adding
formatting like bold or italic type <span class="citation">(Buffalo 2015)</span>. For example,
early versions of email only allowed users to write using plain text. These
users eventually evolved some conventions for how to “mark-up” this plain text,
to serve the purposes served by things like italics and bold in formatted text
(e.g., emphasis, highlighting). For example, to emphasize a word, a user could
surround it with asterisks, like:</p>
<pre><code>I just read a *really* interesting article!</code></pre>
<p>In this early prototype for a markup language, the reader’s mind was doing
the “rendering,” interpreting these markers as a sign that part of the text
was emphasized. In Markdown, the text can be rendered into more attractive
output documents, like pdf, where the rendering process has actually
changed the words between asterisks to print in italics.</p>
<p>The Markdown language has developed a set of these types of marks—like
asterisks—that are used to “mark up” the plain text with the formatting
that should be applied when the text is rendered. There are marks that you
can use for a number of formatting specifications, including: italics,
bold, underline, strike-through, bulleted lists, numbered lists, web links,
headers of different levels (e.g., to mark off sections and subsections),
block quotes, horizontal rules, and block quotes. Details and examples of
the Markdown syntax can be found on the Markdown Guide page at
<a href="https://www.markdownguide.org/basic-syntax/" class="uri">https://www.markdownguide.org/basic-syntax/</a>. We’ll cover more examples of
using Markdown in the next two modules, as we move more specifically into
how RMarkdown can be used to created knitted documents with R and provide
an example of creating a reproducible protocol with this system.</p>
<p>The other technique that’s needed to create knitted documents is the ability to
include executable code within the plain text version of the document, and to
execute that code and incorporate its results before moving on to render the
document to its final format.</p>
<p>The idea here is that you can use special markers to indicate in the document
where code starts and where it ends. With these markings, a computer program can
figure out the lines of the document that it should run as code, and the ones it
should ignore when it’s looking for executable code. With these markings in
place, the document will be run through two separate programs as it is rendered.
The first program will look for code to execute and ignore any other lines of
the file. It will execute this code and then place any results, like figures,
tables, or code output, into the document right after that piece of code. The
output from running this program will then be input into a more traditional
markup renderer, which will format the document based on any of the mark up
indications and will output an attractive document in a format like pdf,
Word, or HTML.</p>
<p>This technique comes from an idea that you could include code to be executed in
a document that is otherwise easy for humans to read. This is an incrediably
powerful idea. It originated with a famous computer scientist named Donald
Knuth, who realized that one key to making computer code sound is to make sure
that it is clear to humans what the code is doing. Computers will faithfully do
exactly what you tell them to do, so they will do what you’re hoping they will
as long as you provide the correct instructions. The greatest room for error,
then, comes from humans not giving the right instructions to computers. To
write sound code, and code that is easy for yourself and others to maintain and
extend, you must make sure that you and other humans understand what it is
asking the computer to do. Donald Knuth came up with a system called <em>literate
programming</em> that allows programmers to write code in a way that focuses on
documenting the code for humans, while also allowing the computer to easily
pull out just the parts that it needs to execute, while ignoring all the text
meant for humans. This process flips the idea of documenting code by including
plain text comments in the code—instead of the code being the heart of the
document, the documentation of the code is the heart, with the code provided
to illustrate the implementation. When used well, this technique results in
beautiful documents that clearly and comprehensively document the intent and
the implementation of computer code. The knitted documents that we can build
with R or Python through systems like RMarkdown and Jupyter Notebooks build
on these literate programming ideas, applying them in ways that complement
programming languages that can be run interactively, rather than needing to
be compiled before they’re run.</p>
<p>You can visualize the full process of creating and rendering a knitted document
in the following way. Imagine that you write a document by hand on sheets of
paper. There are parts where you need a team member to add their data or to run
a calculation, so you include notes in square brackets telling your team member
where to do these things. Then, you use traditional editing marks to show where
text should be italicized and which text should be section a header:</p>
<pre><code># Results

We measured the bacterial load of 
*Mycobacterium tuberculosis* for each 
sample. 

[Kristina: Calculate bacterial loads for 
each sample based on dilutions and
add table with results here.]</code></pre>
<p>This is analogous to writting up a knitted document in plain text with appropriate
“executable” sections, designated with special markings, and with other markings
used to show how the text should be formatted in its final version.</p>
<p>You send the document to your team member first, and she does his calculations
and adds the results at the indicated spot in the paper. She focuses on the notes to
her in square brackets and ignores the rest of the document. This is analogous
to the first stage of rendering a knitted document, where the document is passed
through software that looks for executable code and ignores everything else,
executing that code and adding in results in the right place.</p>
<p>Next, your research team member sends the document, with her additions, to an
assistant to type up the document. The assistant types the full document, paying
attention to any indications that are included for formatting. For example, he
sees that “Results” is meant to be a section heading, since it is on a line that
starts with “#,” your team’s convention for section headings. He therefore
types this on a line by itself in larger font. He also sees that “Mycobacterium
tuberculosis” is surrounded by asterisks, so he types this in italics. This
step is analogous to the second stage of rendering a formatted document, when
a software program takes the output of the first stage and formats the full
document into an attractive, easy-to-read final document, using any markings you
include to format the document.</p>
</div>
<div id="advantages-of-reproducible-pre-processing-protocols" class="section level3" number="3.9.2">
<h3><span class="header-section-number">3.9.2</span> Advantages of reproducible pre-processing protocols</h3>
<p>With point-and-click software, even if you are doing the same process from one
experiment to another to pre-process your data, you will still have to go through
each step of preprocessing, re-selecting each choice along the way. For example,
if you are using software to gate flow cytometry data, someone in your research
group must typically go through the gating step-by-step, even if they are trying to
gate the data using the same rules and approach that they’ve applied to gate data
in previous experiments. This approach therefore has two key limitations.</p>
<p>First, it takes a lot of time for someone in the research group to go through
the same series of selection / point-and-click steps over and over each time
research data needs to be pre-processed. If steps do indeed need to be
customized extensively from one experiment to the next, there may be no way to
avoid this time-consuming work. However, if the same choices for pre-processing
apply from one experiment to the next, then there’s not a good reason for
someone in the research group to need to spend a lot of time with this process.</p>
<p>Using a reproducible pre-processing protocol therefore helps make the
pre-processing more <strong>efficient</strong>. While developing this type of protocol will
take more time the first time or two that you do the pre-processing, as you
create, refine, and check the document, this time investment will pay off as you
continue to re-use the protocol in later experiments. Since the code can be
extracted as a script, you may find that you can often use that script as a
direct starting point for later experiments, needing only to change a few areas,
like the name of the input data files.</p>
<p>The second issue with a point-and-click approach is that it’s hard to be sure
that you’re being completely consistent from experiment to experiment if you’re
going through the pre-processing “by hand,” going through different steps and
selections using point-and-click software. Even if you’ve written down the
choices you plan to make from time to time, there may be subtle small choices
that you forget to write down. Further, there are some choices that might not be
as easy to make consistent from time to time. For example, when you gate flow
cytometry data using point and click software, you are often visual adjusting a
threshold or a box to select certain data points in the sample to gate. These
visual choices can be subjective from day to day, so you might gate the data
slightly different from one day to the next. Even if the same person does the
preprocessing from time to time, there will likely be subtle variations in the
process; these are likely to expand quite a bit when different people in the
research group do the pre-processing from one experiment to another.</p>
<p>By creating a reproducible pre-processing protocol, and starting from the
embedded code each time you pre-process data from a new experiment, you can
make your pre-processing more <strong>reproducible</strong> and more <strong>consistent</strong> from
one experiment to the next. When pre-processing is done using a code script,
the choices are based on an algorithm that can be reproduced faithfully from
experiment to experiment. For example, coding tools for gating flow cytometry
data use consistent algorithms, based on clear rules for fitting gates based
on the distribution of the data [?], and so remove the subjectivity of
gating the data by hand, and the differences in gating that can result from
person to person or even from day to day in gating done by the same person.</p>
<p>In scientific research, there are numerous factors that can affect the
measurements that result from an experiment. These include differences across
experimental animals or subjects, differences in the set-up of different
laboratories, and so on. We often try to control as many extraneous factors as
we can, so that we can focus very precisely on how a single element affects an
outcome. By controlling extraneous factors, we can reduce the noise that might
obscure a signal in the relationship we care about and are trying to measure.
Subjectivity in data pre-processing is a potential source of noise in the
experimental process, especially for more complex biomedical data that requires
extensive pre-processing. Just as strong control to prevent variation in factors
like laboratory conditions and experimental animals can add power and clarity to
detect important signals in biomedical research, so can enforced consistency in
data pre-processing, including through the explicit use of consistent, objective
algorithms for pre-processing steps through the use of scripted code for as much
of the pre-processing as possible.</p>
</div>
<div id="what-knitted-documents-are-and-why-to-use-them" class="section level3" number="3.9.3">
<h3><span class="header-section-number">3.9.3</span> What knitted documents are and why to use them</h3>
<p>If you have coded using a scripting language like R or Python, you likely have
already seen many examples of knitted documents. For both these languages, there
are many tutorials available that are created as knitted documents. Figure
<a href="experimental-data-preprocessing.html#fig:xcmsexample">3.1</a> shows an example from the start of a vignette for the
<code>xcms</code> package in R. This is a package that helps with pre-processing and
analyzing data from LC-MS experiments. You can see that this document includes
text to explain the package and also example code and the output from that code.</p>
<div class="figure"><span id="fig:xcmsexample"></span>
<img src="figures/vignette_example_annotated.png" alt="An example of a knitted document. This shows a section of the online vignette for the `xcms` package from Bioconductor. The two types of content are highlighted: formatted text for humans to read, and executable computer code." width="\textwidth" />
<p class="caption">
Figure 3.1: An example of a knitted document. This shows a section of the online vignette for the <code>xcms</code> package from Bioconductor. The two types of content are highlighted: formatted text for humans to read, and executable computer code.
</p>
</div>
<p>As a larger example, all the modules in this online book were written as knitted
documents. In this module we will describe the characteristics of knitted
documents and how the technology behind them works. We will then talk about how
you can create them using Rmarkdown. In the next module, we’ll provide an
example of doing this for a set of research data from a common laboratory
experiment.</p>
<p>The defining characteristic of a knitted document is that it interweaves two
types of content: first, executable code, and, second, documentation that is
formatted in a way that is easy to read. In the figure with the example from the
<code>xcms</code> vignette, we have highlighted the areas that demonstrate these two types
of content (Figure <a href="experimental-data-preprocessing.html#fig:xcmsexample">3.1</a>). Later in this module we will
describe how a knitted document can incorporate these two elements. First,
however, we will explain why you might want to use knitted documents to
document your own research code, especially for pre-processing protocols.</p>
<p>There are several advantages to using knitted documents when writing code to
pre-process or analyze research data. These include improvements in terms of
reliability, efficiency, transparency, and reproducibility.</p>
<p>First, when you have written your code within a knitted document, this code is
checked every time you render the document. In other words, you are checking
your code to ensure it operates as you intend throughout the process of writing
and editing your document, checking the code each time you render the document
to its formatted version. This helps to increase the <strong>reliability</strong> of the code
that you have written. Open-source software evolves over time, and by continuing
to check code as you work on protocols and reports with your data, you can
ensure that you will quickly identify and adapt to any such changes. Further,
you can quickly identify if updates to your research data introduce any issues
with the code. Again, by checking the code frequently, you can identify any
issues quickly, and this often will allow you to easily pinpoint and fix these
issues. By contrast, if you only identify a problem after writing a lot of code,
it is often difficult to identify the source of the issue. By including code
that is checked each time of document is rendered, you can quickly identify when
a change an open source software effects the analysis that you were conducting
or the pre-processing and work 2 adapt to any changes quickly.</p>
<p>Second, when you write a document that includes executable code, it allows you
to easily rerun the code as you update your research data set, or adopt the code
to work with a new data set. If you are not using a knitted document to write
pre-processing protocols and research reports, then your workflow is probably to
run all your code—either from a script or the command line—and copy the
results into a document in a word processing program like Word or Google Docs.
If you do that, you must recopy all your results every time you adapt any part
of the code or add new data. By contrast, when you use a knitted document, the
rendering process executes the code and incorporates the results directly and
automatically into a nicely formatted final document. The use of knitted
documents therefore can substantially improve the <strong>efficiency</strong> of
pre-processing and analyzing your data and generating the reports that summarize
this process.</p>
<p>Third, documents that are created in knitted format are created using plain
text. Plain text files can easily be tracked well and clearly using version
control tools like <em>git</em>, and associated collaboration tools like GitHub, as
discussed in earlier modules. This substantially increases the <strong>transparency</strong>
of the data pre-processing and analysis. It allows you to clearly document
changes you or others make in the document, step-by-step. You can document who
made the change, and that person can include a message about why they made the
change. This full history of changes is recorded and can be searched to explore
how the document has evolved and why.</p>
<p>The final advantage of using knitted documents, especially for pre-processing
research data, is that it allows the code to be clearly and thoroughly
documented. This can help increase the <strong>reproducibility</strong> of the process. In
other words, it can help ensure that another researcher could repeat the same
process, making adaptations as appropriate for their own data set, or ensuring
they arrive at the same results if using the original data. It also ensures that
you can remember exactly what you did, which is especially useful if you plan to
reuse or adopt the code to work with other data sets, as will often be the case
for a pre-processing protocol. If you are not using a knitted document, but are
using code for preprocessing, then as an alternative you may be documenting your
code through comments in a code script. A code script does allow you to include
documentation about the code through these code comments, which are demarcated
from code in the script through a special symbol (<code>#</code> in R). However these code
comments are much less expressive and harder to read than nicely formatted text,
and it is hard to include elements like mathematical equations and literature
citations in code comments. A knitted document allows you to write the
documentation in a format that is clear and attractive for humans to read, while
including code that is clear and easy for a computer to execute.</p>
</div>
<div id="how-knitted-documents-work" class="section level3" number="3.9.4">
<h3><span class="header-section-number">3.9.4</span> How knitted documents work</h3>
<p>Next, we will describe how knitted documents work. There are seven components of
how these documents work. It is helpful to understand these to understand these
to begin creating and adapting knitted documents. Knitted documents can be
created through a number of programs, and while we will later focus on
Rmarkdown, these seven components are in play regardless of the exact system
used to create a knitted document, and therefore help in gaining a general
understanding of this type of document. We have listed the seven components here
and in the following paragraphs will describe each more fully:</p>
<ol style="list-style-type: decimal">
<li>Knitted documents start as plain text;</li>
<li>A special section at the start of the document (<strong>preamble</strong>) gives some
overall directions about the document;</li>
<li>Special combinations of characters indicate where the executable code starts;</li>
<li>Other special combinations show where the regular text starts (and the
executable code section ends);</li>
<li>Formatting for the rest of the document is specified with a <strong>markup
language</strong>;</li>
<li>You create the final document by <strong>rendering</strong> the plain text document. This
process runs through two software programs; and</li>
<li>The final document is attractive and <strong>read-only</strong>—you should never make
edits to this output, only to your initial plain text document.</li>
</ol>
<p>First, a knitted document should be written in plain text. In an earlier module,
we described some of the advantages of using plain text file formats, rather
than proprietary and/or binary file formats, especially in the context of saving
research data (e.g., using csv file formats rather than Excel file formats).
Plain text can also be used to write documentation, including through knitted
documents. Figure <a href="experimental-data-preprocessing.html#fig:xcmsexampleplain">3.2</a> shows an example of what the plan text might look like for the
start of the <code>xcms</code> tutorial shown in Figure <a href="experimental-data-preprocessing.html#fig:xcmsexample">3.1</a>.</p>
<div class="figure"><span id="fig:xcmsexampleplain"></span>
<img src="figures/plaintext_vignette_example.png" alt="An example of a the plain text used to write a knitted document. This shows a section of the plain text used to write the online vignette for the `xcms` package from Bioconductor. The full plain text file used for the vignette can be viewed on GitHub [here](https://github.com/sneumann/xcms/blob/master/vignettes/xcms.Rmd)." width="\textwidth" />
<p class="caption">
Figure 3.2: An example of a the plain text used to write a knitted document. This shows a section of the plain text used to write the online vignette for the <code>xcms</code> package from Bioconductor. The full plain text file used for the vignette can be viewed on GitHub <a href="https://github.com/sneumann/xcms/blob/master/vignettes/xcms.Rmd">here</a>.
</p>
</div>
<p>There are a few things to keep in mind when writing plain text. First, you
should always use a text editor rather than a word processor when you are
writing a document in plain text. Text editors can include software programs
like Notepad on Microsoft operating systems and TextEdit on Mac operating
systems. You can also use a more advanced text editor, like vi/vim or emacs.
Rstudio can also serve as a text editor, and if you are doing other work in
Rstudio, this is often the most obvious option as a text editor to use to write
knitted documents.</p>
<p>You must use a text editor to write plain text for knitted documents for the
same reasons that you must use one to write code scripts. Word processors often
introduce formatting that is saved through underlying code rather than clearly
evident on the document that you see as you type. This hidden formatting can
complicate the written text. Conversely, text written in a text editor will not
introduce such hard-to-see formatting. Word processing programs also tend to
automatically convert some symbols into slightly fancier versions of the symbol.
For example, they may change a basic quotation symbol into one with shaping,
depending on whether the mark comes at the beginning or end of a quotation. This
subtle change in formatting can cause issues in both the code and the formatting
specifications that you include in a knitted document.</p>
<p>Further, when are writing plain text, typically you should only use characters
from the American Standard Code for Information Interchange, or ASCII. This is a
character set from early in computing that includes 128 characters. Such a small
character set enforces simplicity: this character set mostly includes what you
can see on your keyboard, like the digits 0 to 9, the lowercase and uppercase
alphabet, some symbols, including punctuation symbols like the exclamation point
and quotation marks, some mathematical symbols like plus, minus, and division,
and some control codes, including ones for a new line, a tab, and even ringing a
bell. The full set of characters included in ASCII can be found in a number of
sources including a very thorough Wikipedia page on this character set (<a href="https://en.wikipedia.org/wiki/ASCII" class="uri">https://en.wikipedia.org/wiki/ASCII</a>).</p>
<p>Because the character set available for plain text files is so small, you will
find that it becomes important to leverage the limited characters that are
available. One example is <strong>white space</strong>. White space can be created in ASCII
with both the space character and with the new line command. It is an important
component that can be used to make plain text files clear for humans to read. As
we begin discussing the convention for markdown languages, we will find that
white space is often used to help specify formatting as well.</p>
<p>The second component of how knitted documents work is that each knitted document
will have a special section at its start called the <strong>preamble</strong>. This preamble
will give some overall directions regarding the document, like its title and
authors and the format to which is should be rendered. Knitted documents are
created using a <strong>markup language</strong> to specify formatting for the document, and
there are a number of different markup languages including HTML, LaTeX, and
Markdown. The specifications for the document’s preamble will depend on the
markup language being used.</p>
<p>In Rmarkdown, we will be focusing on Markdown, for which the preamble is
specified using something called YAML (short for YAML Ain’t Markup Language).
Here is an example of the YAML for a sample pre-processing protocol created
using RMarkdown:</p>
<pre><code>  ---
  title: &quot;Preprocessing Protocol for LC-MS Data&quot;
  author: &quot;Jane Doe&quot;
  date: &quot;1/25/2021&quot;
  output: pdf_document
  ---</code></pre>
<p>This YAML preamble specifies information about the document with <strong>keys</strong> and
<strong>values</strong>. For example, the title is specified using the YAML key <code>title</code>,
followed by a colon and a space, and then the desired value for that
component of the document, <code>"Preprocessing Protocol for LC-MS Data"</code>.
Similarly, the author is specified with the <code>author</code> key and the desired
value for that component, and the date with the <code>date</code> key and associated
component.</p>
<p>Different keys can take different types of values in the YAML
(this is similar to how different parameters in a function can take different values). For example, the keys of <code>author</code>, <code>title</code>, and <code>date</code> all take
a character string with any desired character combination, and the quotation
marks surrounding the values for each of these keys denote those character strings. By contrast, the <code>output</code> key—which specifies the format that
that the knitted document should be rendered to—can only take one of a
few set values, each of which is specified without surrounding
quotation marks (<code>pdf_document</code> in this case, to render the document
as a PDF report).</p>
<p>The rules for which keys can be included in the preamble will depend on the
markup language being used. Here, we are showing an example in Markdown, but you
can also use other markup languages like LaTeX and HTML, and these will have
their own convention for specifying the preamble. Later in this module, when we
talk more specifically about Rmarkdown, we will give some resources where you
can find more about how to customize the preamble in Rmarkdown specifically. If
you are using a different markup language, there are numerous websites,
cheatsheets, and other resources you can use to find which keywords are
available for the preamble in that markup language, as well as the possible
values those keywords can take.</p>
<p>The next characteristic of knitted documents is that they need to clearly
demarcate where executable code starts and where regular formatted text starts
(in other words, where the executable code section ends). To do this, knitted
documents have two special combination of characters, one that can be used in
the plain text to indicate where executable code starts and and one to indicate
where it ends. For example, Figure <a href="experimental-data-preprocessing.html#fig:demarcatecode">3.3</a> shows the plain text
that could be used in an Rmarkdown document to write some regular text, then
some executable code, and then indicate the start of more regular text:</p>
<div class="figure"><span id="fig:demarcatecode"></span>
<img src="figures/demarcating_code.png" alt="An example of how special combinations of characters are used to demarcate code in an RMarkdown file. The color formatting here is applied automatically by RStudio; all the text in this example is written in plain text." width="\textwidth" />
<p class="caption">
Figure 3.3: An example of how special combinations of characters are used to demarcate code in an RMarkdown file. The color formatting here is applied automatically by RStudio; all the text in this example is written in plain text.
</p>
</div>
<p>The combination that indicates the start of executable code will vary depending
on the markup language being. In Rmarkdown, the following combination indicates
the start of executable code:</p>
<p><code>```{r}</code></p>
<p>while this combination indicates the end of executable code (in other
words the start of regular text):</p>
<p><code>```</code></p>
<p>In Figure <a href="experimental-data-preprocessing.html#fig:demarcatecode">3.3</a> and above, we have shown the most basic
version of the markup character combination used to specify the start of
executable code (<code>```{r}</code>). This character combination can be expanded,
however, to include some specifications for how you want the code in the section
following it to be run, as well as how you want output to be shown. For example,
you could use the following indications to specify that the code should be run,
but the code itself should not be printed in the final document, by specifying
<code>echo = FALSE</code>, as well as that the created figure should be centered on the
page, by specifying <code>fig.align = "center"</code>:</p>
<p><code>```{r echo = FALSE, fig.align = "center"}</code></p>
<p>There are numerous options that can be used to specify how the code will be run,
and later when we talk specifically about Rmarkdown, we will point toward
resources that clarify and outline all the possible choices that can be used for
these specifications.</p>
<p>You may have noticed that these markers, which indicate the beginning and end of
executable code, seem like very odd character combination. There is a good
reason for this. By making this character combination unusual, there will be
less of a chance that it shows up in regular text. This way there are fewer
cases where the writer unintentionally indicate the start of a new section of
executable code when trying to write regular text in the knitted document.</p>
<p>The next characteristic of knitted documents is that formatting for the regular
text in the document—that is, everything that is not executable code—is
specified using what is called a <strong>markup language</strong>. When you were writing in
plain text, you do not have buttons to click on for formatting, for example, to
specify words or phrases that should be in bold or italics, font size, headings,
and so on. Instead you use special characters or character combinations to
specify formatting in the final document. These character combinations are
defined based on the markup language you use. As mentioned earlier, Rmarkdown
uses the Markdown language; other knitted documents can be created using LaTeX
or HTML. As an example of how these special character combinations work, in
Markdown, you place two asterisks around a word or phrase to make it bold. To
write “<strong>this</strong>” in the final document, in other words, you’ll write
<code>"**this**"</code> in the plain text in the initial document.</p>
<p>You can start to see how this works by looking at the example of the <code>xcms</code>
vignette shown earlier in Figures <a href="experimental-data-preprocessing.html#fig:xcmsexample">3.1</a> and
<a href="experimental-data-preprocessing.html#fig:xcmsexampleplain">3.2</a>. In Figure <a href="experimental-data-preprocessing.html#fig:xcmsbothversions">3.4</a>, we’ve
recreated these two parts side-by-side, so they’re easier to compare.</p>
<div class="figure"><span id="fig:xcmsbothversions"></span>
<img src="figures/originalandfinal.png" alt="The original plain text for a knitted document and the final output, side by side. These examples are from the `xcms` package vignette, a package available on Bioconductor. The left part of the figure shows the plain text that was written to create the output, which is shown in the left part of the figure. You can see how elements like sections headers and different font styles are indicated in the original plain text through special characters or combinations of charaters, using the Markdown language syntax." width="\textwidth" />
<p class="caption">
Figure 3.4: The original plain text for a knitted document and the final output, side by side. These examples are from the <code>xcms</code> package vignette, a package available on Bioconductor. The left part of the figure shows the plain text that was written to create the output, which is shown in the left part of the figure. You can see how elements like sections headers and different font styles are indicated in the original plain text through special characters or combinations of charaters, using the Markdown language syntax.
</p>
</div>
<p>You can look for several formatting elements here. First, the section is headed
“Initial data inspection.” You can that in the original plain text document,
this is marked using a <code>#</code> to start the line with the text for the header. You
can also see that words or phrases that are formatted in a computer-style font
in the final document—to indicate that they are values from computer code,
rather than regular English words—are surrounded by backticks in the plain
text file.</p>
<p>The process of writing your document in this way can feel a little like the
old-fashioned style of dictating to a secretary. You can imagine yourself
dictating to the computer through the plain text. You have to say not just the
words, but also specify in the plain text the formatting that you want at each
spot. For example, when you get to a word that you want in italics, you must
tell the computer, “Start italics,” then say the word, then “End italics.” You
do this using the special formatting characters that are specified by the markup
language that you use. Later in this module we will point to resources for
discovering all of these formatting specifications for Markdown, the markup
language used by RMarkdown.</p>
<p>The final characteristics of knitted documents is that, to create the final
document, you will render the plain text document. That is the process that will
create an attractive final document. To visualize this, <strong>rendering</strong> is the
process that takes the document from the plain text format, as shown in the left
of Figure <a href="experimental-data-preprocessing.html#fig:xcmsbothversions">3.4</a>, to the final format, shown in the right
of that figure.</p>
<p>When you render the document, it will be run through two software programs. The
first will look only for sections with executable code, based on the character
combination that is used to mark these executable code sections. This first
software will execute that code and take any output—including data results,
figures, and tables—and insert those at the relevant spot in the document’s
file. Next, these output file from this software run will be run through another
software program. This second program will look for all the formatting
instructions and render the final document in an attractive format. This final
output can be in a number of file formats, depending what you specify in the
preamble, including a PDF document, an HTML file, or a Word document.</p>
<p>You should consider the final document, regardless of the output format, as
read-only. This means that you should never make edits or changes to the final
version of the document. Instead you should make any changes to your initial
plain text file. This is because the rendering process will overwrite any
previous versions of the final document. Therefore any changes that you have
made to your final document will be overwritten anytime you re-render from the
original plain text document.</p>
</div>
<div id="quotes" class="section level3" number="3.9.5">
<h3><span class="header-section-number">3.9.5</span> Quotes</h3>
<blockquote>
<p>“It’s very important to keep a project notebook containing detailed information
about the chronology of your computational work, steps you’ve taken, information
about why you’ve made decisions, and of course all pertinent information to
reproduce your work. Some scientists do this in a handwritten notebook, others in
Microsoft Word documents. As with README files, bioinformaticians usually like keeping
project notebooks in simple plain-text because these can be read, searched, and
edited from the command line and across network connections to servers. Plain text
is also a future-proof format: plain-text files written in the 1960s are still
readable today, whereas files from word processors only 10 years old can be
difficult or impossible to open and edit. Additionally, plain text project notebooks can
also be put under version control … While plain-text is easy to write in your
text editor, it can be inconvenient for collaborators unfamiliar with the command
line to read. A lightweight markup language called <em>Markdown</em> is a plain-text format
that is easy to read and painlessly incorporated into typed notes, and can also be
rendered to HTML or PDF.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Markdown originates from the simple formatting conventions used in plain-text
emails. Long before HTML crept into email, emails were embellished with simple
markup for emphasis, lists, and blocks of text. Over time, this became a defacto
plain-text email formatting scheme. This scheme is very intuitive: underscores or
asterisks that flank text indicate emphasis, and lists are simply lines of text
beginning with dashes.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Markdown is just plain-text, which means that it’s portable and programs to edit
and read it will exist. Anyone who’s written notes or papers in old versions of
word processors is likely familiar with the hassle of trying to share or update
out-of-date proprietary formats. For these reasons, Markdown makes for a simple
and elegant notebook format.” <span class="citation">(Buffalo 2015)</span></p>
</blockquote>
<blockquote>
<p>“Information, whether data or computer code, should be organized in such a way that
there is only one copy of each important unit of information.” <span class="citation">(Murrell 2009)</span></p>
</blockquote>
<blockquote>
<p>“A typical encounter with Bioconductor (Box 1) starts with a specific scientific need, for example, differential analysis of gene expression
from an RNA-seq experiment. The user identifies the appropriate
documented workflow, and because the workflow contains functioning code, the user runs a simple command to install the required
packages and replicate the analysis locally. From there, she proceeds
to adapt the workflow to her particular problem. To this end, additional documentation is available in the form of package vignettes
and manual pages.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“<strong>Case study: high-throughput sequencing data analysis.</strong> Analysis of
large-scale RNA or DNA sequencing data often begins with aligning reads to a
reference genome, which is followed by interpretation of the alignment patterns.
Alignment is handled by a variety of tools, whose output typically is delivered
as a BAM file. The Bioconductor packages Rsamtools and GenomicAlignments provide
a flexible interface for importing and manipulating the data in a BAM file, for
instance for quality assessment, visualization, event detection and
summarization. The regions of interest in such analyses are genes, transcripts,
enhancers or many other types of sequence intervals that can be identified by
their genomic coordinates. Bioconductor supports representation and analysis of
genomic intervals with a ‘Ranges’ infrastructure that encompasses data
structures, algorithms and utilities including arithmetic functions, set
operations and summarization (Fig. 1). It consists of several packages
including IRanges, GenomicRanges, GenomicAlignments, GenomicFeatures,
VariantAnnotation and rtracklayer. The packages are frequently updated for
functionality, performance and usability. The Ranges infrastructure was designed
to provide tools that are convenient for end users analyzing data while
retaining flexibility to serve as a foundation for the development of more
complex and specialized software. We have formalized the data structures to the
point that they enable interoperability, but we have also made them adaptable to
specific use cases by allowing additional, less formalized userdefined data
components such as application-defined annotation. Workflows can differ vastly
depending on the specific goals of the investigation, but a common pattern is
reduction of the data to a defined set of ranges in terms of quantitative and
qualitative summaries of the alignments at each of the sites. Examples include
detecting coverage peaks or concentrations in chromatin
immunoprecipitation–sequencing, counting the number of cDNA fragments that match
each transcript or exon (RNA-seq) and calling DNA sequence variants (DNA-seq).
Such summaries can be stored in an instance of the class GenomicRanges.”
<span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Visualization is essential to genomic data analysis. We distinguish among
three main scenarios, each having different requirements. The first is rapid
interactive data exploration in ‘discovery mode.’ The second is the recording,
reporting and discussion of initial results among research collaborators, often
done via web pages with interlinked plots and tool-tips providing interactive
functionality. Scripts are often provided alongside to document what was done.
The third is graphics for scientific publications and presentations that show
essential messages in intuitive and attractive forms. The R environment offers
powerful support for all these flavors of visualization—using either the various
R graphics devices or HTML5-based visualization interfaces that offer more
interactivity—and Bioconductor fully exploits these facilities. Visualization
in practice often requires that users perform computations on the data, for
instance, data transformation and filtering, summarization and dimension
reduction, or fitting of a statistical model. The needed expressivity is not
always easy to achieve in a point-and-click interface but is readily realized in
a high-level programming language. Moreover, many visualizations, such as heat
maps or principal component analysis plots, are linked to mathematical and
statistical models—for which access to a scientific computing library is
needed.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>" It can be surprisingly difficult to retrace the computational steps
performed in a genomics research project. One of the goals of Bioconductor is to
help scientists report their analyses in a way that allows exact recreation by a
third party of all computations that transform the input data into the results,
including figures, tables and numbers. The project’s contributions comprise an
emphasis on literate programming vignettes, the BiocStyle and ReportingTools
packages, the assembly of experiment data and annotation packages, and the
archiving and availability of all previously released packages. … Full remote
reproducibility remains a challenging problem, in particular for computations
that require large computing resources or access data through infrastructure
that is potentially transient or has restricted access (e.g., the cloud).
Nevertheless, many examples of fully reproducible research reports have been
produced with Bioconductor." <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“Using Bioconductor requires a willingness to modify and eventually compose
scripts in a high-level computer language, to make informed choices between
different algorithms and software packages, and to learn enough R to do the
unavoidable data wrangling and troubleshooting. Alternative and complementary
tools exist; in particular, users may be ready to trade some loss of
flexibility, automation or functionality for simpler interaction with the
software, such as by running single-purpose tools or using a point-and-click
interface. Workflow and data management systems such as Galaxy and Illumina
BaseSpace provide a way to assemble and deploy easy-touse analysis pipelines
from components from different languages and frameworks. The IPython notebook
provides an attractive interactive workbook environment. Although its origins
are with the Python programming language, it now supports many languages,
including R. In practice, many users will find a combination of platforms most
productive for them.” <span class="citation">(Wolfgang Huber et al. 2015)</span></p>
</blockquote>
<blockquote>
<p>“A lab manual is perhaps the best way to inform new lab members of the ins
and outs of the lab and to keep all members updated on protocols and
regulations. What could be included in a lab manual? Anything you do not
want to explain over and over, anything that will make the lab more
functional and that can make life easier for yourself and lab members.” <span class="citation">(LEIPS 2010)</span></p>
</blockquote>
<blockquote>
<p>“Most PIs wish the labs were more organized, but it is not a huge priority,
that is, until the first student leaves and no one can find a particular
cell line in the freezer boxes. Resolutions are made, the crisis passes,
and all goes on as before until the next person leaves. Although it is
probably inevitable that there will be some confusion when a long-time
lab member moves on, an organized lab will not be as affected as an unorganized
one.” <span class="citation">(LEIPS 2010)</span></p>
</blockquote>
</div>
<div id="discussion-questions-6" class="section level3" number="3.9.6">
<h3><span class="header-section-number">3.9.6</span> Discussion questions</h3>

</div>
</div>
<div id="rmarkdown-for-creating-reproducible-data-pre-processing-protocols" class="section level2" number="3.10">
<h2><span class="header-section-number">3.10</span> RMarkdown for creating reproducible data pre-processing protocols</h2>
<p>The R extension package RMarkdown can be used to create documents that combine
code and text in a ‘knitted’ document, and it has become a popular tool for
improving the computational reproducibility and efficiency of the data analysis
stage of research. This tool can also be used earlier in the research process,
however, to improve reproducibility of pre-processing steps. In this module, we
will provide detailed instructions on how to use RMarkdown in RStudio to create
documents that combine code and text. We will show how an RMarkdown document
describing a data pre-processing protocol can be used to efficiently apply the
same data pre-processing steps to different sets of raw data.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Define RMarkdown and the documents it can create</li>
<li>Explain how RMarkdown can be used to improve the reproducibility of research
projects at the data pre-processing phase</li>
<li>Create a document in RStudio using RMarkdown</li>
<li>Apply it to several different datasets with the same format</li>
</ul>
<div id="creating-knitted-documents-in-r" class="section level3" number="3.10.1">
<h3><span class="header-section-number">3.10.1</span> Creating knitted documents in R</h3>
<p>In the last module, we described what knitted documents are, as well as the
advantages of using knitted documents to create data pre-processing protocols
for common pre-processing tasks in your research group. In this module, we will
go into more detail about how you can create these documents using R and
RStudio, and in the next module we will walk through an example data
pre-processing protocol created using this method.</p>
<p>R has a special format for creating knitted documents called <strong>Rmarkdown</strong>.
Later in this module, we will provide details on creating these documents, but
first, we will provide some details of some of the specific conventions used in
this type of knitted documents. As with other knitted documents, Rmarkdown
documents are originally written in plain text. In the previous module, we
discussed how knitted documents will use a markup language within the plain text
for formatting; for RMarkdown files, the markup language used is Markdown. Since
the laguage is Markdown, the preamble for each document uses YAML. We described
in the previous module how knitted documents include executable code, along with
the formatted text. By default, executable code for RMarkdown files will be in
R, but there are also options to include executable code in the document in a
number of other programming languages.</p>
</div>
<div id="creating-and-exploring-an-rmarkdown-file" class="section level3" number="3.10.2">
<h3><span class="header-section-number">3.10.2</span> Creating and exploring an Rmarkdown file</h3>
<p><strong>Creating a new Rmarkdown file and exploring the template.</strong></p>
<p>First, we will explain how you can create a new Rmarkdown document using
RStudio. Like other plain text documents, an Rmarkdown file should be edited
using a text editor, rather than a word processor like Word or Google Docs. It
is easiest to use the Rstudio IDE as the text editor when creating and editing
an R markdown document, as this IDE has incorporated some helpful functionality
for working with plain text documents for Rmarkdown. The RStudio IDE can be
downloaded and installed as a free software, as long as you use the personal
version (RStudio creates higher-powered versions for corporate use). Since
RStudio is free and has many helpful tools for working with Rmarkdown, we will
focus on using this interface in our advice in this module.</p>
<p>In RStudio, you can create a number of types of new files through the
“File” menu. To create a new R markdown file, choose “New File” and then choose
“Rmarkdown” from the choices in that menu. Figure <a href="experimental-data-preprocessing.html#fig:rmarkdownnewfile">3.5</a>
shows an example of what this menu option looks like.</p>
<div class="figure"><span id="fig:rmarkdownnewfile"></span>
<img src="figures/rmarkdown_newfile.png" alt="RStudio pull-down menus to help you navigate to open a new Rmarkdown file." width="\textwidth" />
<p class="caption">
Figure 3.5: RStudio pull-down menus to help you navigate to open a new Rmarkdown file.
</p>
</div>
<p>This will open a window with some options you can specify some of the overall
information about the document (Figure <a href="experimental-data-preprocessing.html#fig:rmarkdownchoices">3.6</a>), including
the title and the author, and you can specify the output format that you would
like. Possible output formats include HTML, Word, and PDF. You should be able to
use the HTML and Word output formats without any additional software. If you
would like to use the PDF output, you will need to install one other piece of
software: Miktex for Windows, MacTex for Mac, or TeX Live for Linux. These are
all pieces of software with an underlying TeX engine and all are open-source and
free.</p>
<div class="figure"><span id="fig:rmarkdownchoices"></span>
<img src="figures/rmarkdown_choices.png" alt="Options available when you create a new Rmarkdown file in RStudio. You can specify information that will go into the document's preamble, including the title and authors and the format that the document will be output to (HTML, Word, or PDF)." width="\textwidth" />
<p class="caption">
Figure 3.6: Options available when you create a new Rmarkdown file in RStudio. You can specify information that will go into the document’s preamble, including the title and authors and the format that the document will be output to (HTML, Word, or PDF).
</p>
</div>
<p>Once you have selected the options in this menu you can choose the “Okay” button
(Figure <a href="experimental-data-preprocessing.html#fig:rmarkdownchoices">3.6</a>). This will open a new document. This
document, however, won’t be blank. Instead it will include an example document
written in Rmarkdown (Figure <a href="experimental-data-preprocessing.html#fig:rmarkdowntemplate">3.7</a>). This example
document helps you navigate how the Rmarkdown process works, by letting you test
out a sample document. It also gives you a starting point—once you understand
how the example document works, you can edit it and change it to convert it
into the document you would like to create.</p>
<div class="figure"><span id="fig:rmarkdowntemplate"></span>
<img src="figures/rmarkdown_template.png" alt="Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document." width="\textwidth" />
<p class="caption">
Figure 3.7: Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document.
</p>
</div>
<p>If you have not used Rmarkdown before, it is very helpful to try knitting this
example document before making changes, to explore how pieces in the document
align with elements in the rendered output document. Once you are
familiar with the line-up between elements in this file in the output document,
you delete parts of the example file and insert your own text and code.</p>
<p>We will walk you through exploring this example document, as well as aligning it
with the formatted output document (Figure <a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>).
First, to render this or any Rmarkdown document, if you are in RStudio you can
use the “Knit” button at the top of the file, as shown in Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdowntemplate2">3.9</a>. When you click on this button, it will render the
entire document to the output format you’ve selected (HTML, PDF, or Word). This
rendering process will both run the executable code and apply all formatting.
The final output (Figure <a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>, right) will pop up
in a new window. As you start with Rmarkdown, it is useful to look at this
output to see how it compares with the plain text Rmarkdown file (Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>, left).</p>
<div class="figure"><span id="fig:rmarkdownoriginalfinal"></span>
<img src="figures/rmarkdownoriginalfinal.png" alt="Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document." width="\textwidth" />
<p class="caption">
Figure 3.8: Example of the template Rmarkdown document that you will see when you create a new Rmarkdown file in RStudio. You can explore this template and try rendering (knitting) it. Once you are familiar with how this example works, you can edit the text and code to adapt it for your own document.
</p>
</div>
<div class="figure"><span id="fig:rmarkdowntemplate2"></span>
<img src="figures/rmarkdown_template2.png" alt="Example of the template Rmarkdown document, highlighting buttons in RStudio that you can use to facilitate working with the document. The 'knit' button, highlighted at the top of the figure, will render the entire document. The green arrow, highlighted lower in the figure within a code chunk, can be used to run the code in that specific code chunk." width="\textwidth" />
<p class="caption">
Figure 3.9: Example of the template Rmarkdown document, highlighting buttons in RStudio that you can use to facilitate working with the document. The ‘knit’ button, highlighted at the top of the figure, will render the entire document. The green arrow, highlighted lower in the figure within a code chunk, can be used to run the code in that specific code chunk.
</p>
</div>
<p>You will also notice, after you first render the document, that your working
directory has a new file with this output document. For example, if you are
working to create an HTML document using an Rmarkdown file called
“my_report.Rmd,” once you knit your Rmarkdown file, you will notice a new file
in your working directory called “my_report.html.” This new file is your output
file, the one that you would share with colleagues as a report. You should
consider this output document to be read only—in other words, you can read and
share this document, but you should not make any changes directly to this
document, since they will be overwritten anytime you re-render the original
Rmarkdown document.</p>
<p>Next, let’s compare the example Rmarkdown document (the one that is given when
you first open an Rmarkdown file in RStudio) with the output file that is
created when you render this example document (Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>). If you look at the output document (Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>, right), you can notice how different elements
align with pieces in the original Rmarkdown file (Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdownoriginalfinal">3.8</a>). For example, the output document includes a
header with the text “R Markdown.” This second-level header is created by the
Markdown notation in the original file of:</p>
<pre><code>## R Markdown</code></pre>
<p>This header is formatted in a larger font than other text, and on a separate
line—the exact formatting is specified within the style file for the Rmarkdown
document, and will be applied to all second-level headers in the document. You
can also see formatting specified through things like bold font for the word
“Knit,” through the Markdown syntax <code>**Knit**</code>, and a clickable link specified
through the syntax <code>&lt;http://rmarkdown.rstudio.com&gt;</code>. At the beginning of the
original document, you can see how elements like the title, author, date, and
output format are specified in the YAML. Finally, you can see that special
character combinations demarcate sections of executable code.</p>
<p><strong>Formatting text with Markdown.</strong></p>
<p>For the main text in an Rmarkdown document, all formatting is done using
Markdown as the markup language. Markdown is a popular markup language, in part
because it is a good bit simpler than other markup languages like HTML or LaTeX.
This simplicity means that it is not quite as expressive as other markup
languages. However, Markdown provides adequate formatting for 90% of the
formatting you will typically want to do for a research report or
pre-processeing protocol, and by staying simpler, it is much easier to learn the
Markdown syntax quickly compared to other markup languages.</p>
<p>As with other markup languages, Markdown uses special characters or combinations of characters to indicate formatting within the plain text of the original document. When the document is rendered, these markings are used by the software to create the formatting that you have specified in the final output document. Some example formatting symbols and conventions for Markdown include:</p>
<ul>
<li>to format a word or phrase in bold, surround it with two asterisks (<code>**</code>)</li>
<li>to format a word or phrase in italics, surround it with one asterisk (<code>*</code>)</li>
<li>to create a first-level header, put the header text on its own line, starting the line with <code>#</code></li>
<li>to create a second-level header, put the header text on its own line, starting the line with <code>##</code></li>
<li>separate paragraphs with empty lines</li>
<li>use hyphens to create bulleted lists</li>
</ul>
<p>One thing to keep in mind when using Markdown, in terms of formatting, is that
white space can be very important in specifying the formatting. For example when
you specify a new paragraph, you must leave a blank line from your previous
text. Similarly when you use a hash (<code>#</code>) to indicate a header, you must leave a
blank space after the hash before the word or phrase that you want to be used in
that header. To create a section header, you would write:</p>
<pre><code># Initial Data Inspection</code></pre>
<p>Meanwhile this:</p>
<pre><code>#Initial Data Inspection</code></pre>
<p>Would render to:</p>
<p>#Initial Data Inspection</p>
<p>Similarly, white space is needed to separate paragraphs. For example, this would create two paragraphs:</p>
<pre><code>This is a first paragraph. 

This is a second.</code></pre>
<p>Meanwhile this would create one:</p>
<pre><code>This is a first paragraph.
This is still part of the first paragraph.</code></pre>
<p>The syntax of Markdown is fairly simple and can be learned quickly. For more
details on this syntax, you can refer to the Rmarkdown reference guide at
<a href="https://rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf" class="uri">https://rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf</a>. The
basic formatting rules for Markdown are also covered any number of more
extensive resources for Rmarkdown that we will point you to later in this
module.</p>
<p><strong>Creating a preamble with YAML.</strong></p>
<p>In the previous module, we explained how knitted documents include a
preamble to specify some metadata about the document, including elements
like the title, authors, and output format. In R, this preamble is
created using YAML. In this subsection, we provide some more details
on using this YAML section in Rmarkdown documents.</p>
<p>When you initially create an Rmarkdown document, you select the output format
for it (e.g., pdf, Word, HTML). You can change this output selection once you’ve
created the document, however, by making a change in a special section at the
top of the RMarkdown document called the “YAML.” The YAML is a special section
at the top of an RMarkdown document (the original, plain text file, not the
rendered version). It is set off from the rest of the document using a special
combination of characters, using a process very similar to how executable code
is set off from other text with a special set of characters so it can be easily
identified by the software program that renders the document. For the YAML, this
combination of characters is three hyphens (<code>---</code>) on a line by themselves to
start the YAML section and then another three on a line by themselves to end it.
Here is an example of what the YAML might look like at the top of an RMarkdown
document:</p>
<pre><code>---
title: &quot;Laboratory report for example project&quot;
author: &quot;Brooke Anderson&quot;
date: &quot;1/12/2020&quot;
output: word_document
---</code></pre>
<p>Within the YAML itself, you can specify different options for your document.
You can change simple things like the title, author, and date, but you can
also change more complex things, including how the output document is rendered.
For each thing that you want to specify, you specify it with a special
keyword for that option and then a valid choice for that keyword. The idea
is very similar to setting parameter values in a function call in R. For
example, the <code>title:</code> keyword is a valid one in RMarkdown YAML. It allows you
to set the words that will be printed in the title space, using title formatting,
in your output document. It can take any string of characters, so you can put in
any text for the title that you’d like, as long as you surround it with quotation
marks. The <code>author:</code> and <code>date:</code> keywords work in similar ways. The <code>output:</code>
keyword allows you to specify the output that the document should be rendered to.
In this case, the keyword can only take one of a few set values, including
<code>word_document</code> to output a Word document, <code>pdf_document</code> to output a pdf
document (see later in this section for some more set-up required to make that
work), and <code>html_document</code> to output an HTML document.</p>
<p>As you start using RMarkdown, you will be able to do a lot without messing with
the YAML much. In fact, you can get a long way without ever changing the values
in the YAML from the default values they are given when you first create an
RMarkdown document. As you become more familiar with R, you may want to learn
more about how the YAML works and how you can use it to customize your
document—it turns out that quite a lot can be set in the YAML to do very
interesting customizations in your final rendered document. The book <em>R
Markdown: The Definitive Guide</em> [ref], which is available free online, has
sections discussing YAML choices for both HTML and pdf output, at
<a href="https://bookdown.org/yihui/rmarkdown/html-document.html" class="uri">https://bookdown.org/yihui/rmarkdown/html-document.html</a> and
<a href="https://bookdown.org/yihui/rmarkdown/pdf-document.html" class="uri">https://bookdown.org/yihui/rmarkdown/pdf-document.html</a>, respectively. There is
also a talk that Yihui Xie, the creator of RMarkdown, gave on this topic at a
past RStudio conference, available at
<a href="https://rstudio.com/resources/rstudioconf-2017/customizing-extending-r-markdown/" class="uri">https://rstudio.com/resources/rstudioconf-2017/customizing-extending-r-markdown/</a>.</p>
<p><strong>Executable code in Rmarkdown files.</strong></p>
<p>In the previous module, we described how knitted documents use special markers
to indicate where sections of executable code start and stop. In RMarkdown,
the markers you will use to indicate executable code look like this:</p>
<pre><code>```r{}
my_object &lt;- c(1, 2, 3)
```</code></pre>
<p>The first part of this marker (<code>```r{}</code>) indicates that a section of
code is starting. Then end part (<code>```</code>) indicates that the file is
moving back to regular Markdown. In the first marker, you can use the simplest
form (<code>```r{}</code>) to indicate the start of the code. However, you can also
include options to customize actions when the code in that section is executed.
There are many options that you can set for each code chunk. These options will
specify how the code in that section of executable code will be run and how the
output from running the code will be presented. These specifications are called
<strong>chunk options</strong>, and you specify them in the special character combination
where you mark the start of executable code. For example, you can specify that
the code should be printed in the document, but not executed, by setting the
<code>eval</code> parameter to <code>FALSE</code> with <code>```{r eval = FALSE}</code> as the marker to
start the code section.</p>
<p>The chunk options also include <code>echo</code>, which can be used to specify whether to
print the code in that code chunk when the document is rendered. For some
documents, it is useful to print out the code that is executed, where for other
documents you may not want that printed. For example, for a pre-processing
protocol, you are aiming to show yourself and others how the pre-processing was
done. In this case, it is very helpful to print out all of the code, so that
future researchers who read that protocol can clearly see each step. By
contrast, if you are using Rmarkdown to create a report or an article that is
focused on the results of your analysis, it may make more sense to instead hide
the code in the final document.</p>
<p>As part of the code options, you can also specify whether messages and warnings
created when running the code should be included in the document output, and
there are number of code chunk options that specify how tables and figures
rendered by the code should be shown. For more details on the possible options
that can be specified for how code is evaluated within an executable chunk of
code, you refer to the Rmarkdown cheat sheet available at
<a href="https://rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf" class="uri">https://rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf</a></p>
<p>RStudio has some functionality that is useful when you are working with code in
Rmarkdown documents. Within each code chuck are some buttons that can be used to
test out the code in that chunk of executable code. One is the green right arrow
key to the right at the top of the code chunk, highlighted in Figure
<a href="experimental-data-preprocessing.html#fig:rmarkdowntemplate2">3.9</a>. This button will run all of the code in that
chunk and show you the output in an output field that will open directly below
the code chunk. This functionality allows you to explore the code in your
document as you build it, rather than waiting until you are ready to render the
entire document. The button directly to the left of that button, which looks
like an upward-pointing arrow over a rectangle, will execute all code that comes
before this chunk in the document. This can be very helpful in making sure that
you have set up your environment to run this particular chunk of code.</p>
</div>
<div id="more-advanced-rmarkdown-functionality" class="section level3" number="3.10.3">
<h3><span class="header-section-number">3.10.3</span> More advanced Rmarkdown functionality</h3>
<p>The details and resources that we have covered so far focus on the basics of
Rmarkdown. You can get a lot done just with these basics. However, the Rmarkdown
system is very rich and allows complex functionality beyond these basics. In
this subsection, we will highlight just a few of the ways Rmarkdown can be used
in a more advanced way. Since this topic is so broad, we will focus on elements
that we have found to be particularly useful for biomedical researchers as they
become more advanced Rmarkdown users. For the most part, we will not go into
extensive detail about how to use these more advanced features in this module,
but instead point to resources where you can learn more as you are ready. If you
are just learning Rmarkdown, at this point it will be helpful to just know that
some of these advanced features are available, so you can come back and explore
them when you become familiar with the basics. However, we will provide more
details for two elements that we find particularly useful in creating data
pre-processing protocols: including bibliographical references and including
mathematical equations. In the next module, we walk through an example of using
RMarkdown, and in this example we show how you can use these two more advanced
features when creating a pre-processing protocol.</p>
<p><strong>Including bibliographical references.</strong></p>
<p>You can use Rmarkdown, in combination with something called BibTeX, as a
referencing system. This allows you to include bibliographical references in the
documents that you create, and Rmarkdown will handle the creation of the
references section and the numbering of the documents within your text.</p>
<p>To include references in RMarkdown documents, you can use something called
<strong>BibTeX</strong>. This has three components:</p>
<ol style="list-style-type: decimal">
<li>Create a plain text file with listings for each of
your references (<strong>BibTeX file</strong>). Save it with the
extension <code>.bib</code>.</li>
<li>In your RMarkdown document, include the filepath
to this BibTeX file.</li>
<li>In the text of the RMarkdown file, include a key and special character
combination anytime you want to reference a paper.</li>
</ol>
<p>Once you have a BibTeX file, you will include its
path in the YAML:</p>
<pre><code>  ---
  title: &quot;Reproducible Research with R&quot;
  author: &quot;Brooke Anderson&quot;
  date: &quot;1/25/2021&quot;
  output: beamer_presentation
  bibliography: mybibliography.bib
  ---</code></pre>
<p>It is easiest if you save the BibTeX file in the same directory
as the RMarkdown document.</p>
<p>You can add references in the text using a key for each paper
within special characters (<code>[@paper1, @paper2]</code>).</p>
<p>For example, you would write:</p>
<pre><code>This technique follows earlier work [@fox2020, 
@anderson2019].</code></pre>
<p>To create:</p>
<blockquote>
<p>``This technique follows earlier work (Fox et al. 2020, Anderson et al. 2019)."</p>
</blockquote>
<p>With full paper details included at the end of the document.</p>
<p>Put all the bibliographical details in the BibTeX file:</p>
<p>You will have a file named something like <code>mybibliography.bib</code>
with entries for each paper like this:</p>
<pre><code>@article{fox2020,
  title={Cyto-feature engineering: A pipeline for flow cytometry 
    analysis to uncover immune populations and associations with 
    disease},
  author={Fox, Amy and Dutt, Taru S and Karger, Burton and Rojas, 
    Mauricio and Obreg{\&#39;o}n-Henao, Andr{\&#39;e}s and 
    Anderson, G Brooke and Henao-Tamayo, Marcela},
  journal={Scientific Reports},
  volume={10},
  number={1},
  pages={1--12},
  year={2020}
}</code></pre>
<p>You can get these details from Google Scholar:</p>
<div class="figure"><span id="fig:unnamed-chunk-23"></span>
<img src="figures/google_scholar.png" alt="Example of using Google Scholar to get bibliographical information for a BibTeX file." width="\textwidth" />
<p class="caption">
Figure 3.10: Example of using Google Scholar to get bibliographical information for a BibTeX file.
</p>
</div>
<p><strong>Including mathematical equations.</strong></p>
<p>Next, you can include mathematical equations within an
Rmarkdown document. This allows you to use nice formatting whenever you need
to include an equation within your document.</p>
<p><strong>Including executable code in other languages.</strong></p>
<p>In your RMarkdown documents, you include executable code in special sections
(“chunks”) that are separated from the regular text using a special combination
of characters, as described earlier in this module and in the previous module.
By default, in Rmarkdown files the code in these chunks are executed using the R
programming language. However, you can also include executable code in a number
of other programming languages. For example, you could set some code chunks to
run Python, others to run Julia, and still others to run a shell script [?].</p>
<p>This can be very helpful if you have steps in you Python that use code in
different languages. For example, there may be a package [?] in Python that
works well for an early step in your data preprocessing, and then later
steps that are easier with general R functions. This presents no problem in
creating an RMarkdown data pre-processing protocol, as you can include
different steps using different languages.</p>
<p>The program that is used to run the code in a specific chunk is called the
“engine” for that chunk [?]. You can change the engine [where in the code
chunk markup]. [More on this.] When your RMarkdown document is rendered,
your computer will use the specified software to run each code chunk. Of course,
to run that piece of code, your computer must have that type of software installed
and available. For example, if you include a chunk of code that you’d like to
run with a Python engine, you must have Python on your computer.</p>
<p>While you can use many different software programs as the engine for each code
chunk, there are a few limitations with some programs. For many open-source
software programs, the results from running a chunk of code with that engine
will be available for later code chunks to use as an input. This is not the case,
however, for some proprietary software [?]. For example, if you use the SAS software
program as the engine for one of your code chunks, the output from running that code
will not be available to input to later code in the document [?].</p>
<p><strong>Caching code results.</strong></p>
<p>Some code can take a while to run, particularly if it is processing very large
datasets. By default, RMarkdown will re-run all code in the document every time
you render it. This is usually the best set-up, since it allows you to confirm
that the code is all executing as desired each time the code is rendered.
However, if you have steps that take a long time, this can make it so the
RMarkdown document takes a long time to render each time you render it.</p>
<p>To help with this problem, RMarkdown has a system that allows you to <em>cache</em>
results from some or all code chunks in the document. This is a really nice
system—it will check the inputs to that part of the code each time the
document is run. If those inputs have changed, it will take the time to re-run
that piece of code, to use the updated inputs. However, if the inputs have not
changed since the last time the document was rendered, then the last results for
that chunk of code will be pulled from memory and used, without re-running the
code in that chunk. This saves time most of the times that you render the
document, while taking the time to re-run the code when necessary, because the
inputs have changed and so the outputs may be different.</p>
<p>[More on caching from RMarkdown, including things to look out for, like the size
of the document as it grows to include cached values.]</p>
<p><strong>Outputting to other formats.</strong></p>
<p>You can use RMarkdown to create documents other than traditional reports.
Scientists might find the outputs of presentations and posters particularly
useful.</p>
<p>RMarkdown has allowed a pdf slide output for a long time. This output leverages
the “beamer” format from LaTeX. You can create a series of presentation slides
in RMarkdown, using Markdown to specify formatting, and then the document will
be rendered to pdf slides. These slides can be shown using pdf viewer software,
like Adobe Acrobat, set either to full screen or to the presentation option.
More recently, capability has been added to RMarkdown that allows you to create
PowerPoint slides. Again, you will start from an RMarkdown document, using
Markdown syntax to do things like divide content into separate slides.
Regardless of the output format you choose (pdf slides or PowerPoint), the code
to generate figures and tables in the presentation can be included directly in
the RMarkdown file, so it is re-run with the latest data each time you render
the presentation.</p>
<p>It is also possible to use RMarkdown to create scientific posters, although this
is a bit less common and there are fewer tutorial resources with instructions on
doing this. To find out more about creating scientific posters with Rmarkdown,
you can start by looking at the documentation for some R packages that have been
created for this process. Two include the <code>posterdown</code> package [ref], with
documentation available at
<a href="https://reposhub.com/python/miscellaneous/brentthorne-posterdown.html" class="uri">https://reposhub.com/python/miscellaneous/brentthorne-posterdown.html</a>, and the
<code>pagedown</code> package [ref], with documentation available at
<a href="https://github.com/rstudio/pagedown" class="uri">https://github.com/rstudio/pagedown</a>. There are also some blog posts available
where researchers describe how they created a poster with Rmarkdown; one
thorough one is “How to make a poster in R” by Wei Yang Tham, available at
<a href="https://wytham.rbind.io/post/making-a-poster-in-r/" class="uri">https://wytham.rbind.io/post/making-a-poster-in-r/</a>.</p>
<p>[Using journal article templates]</p>
<p>You can use RMarkdown to create much larger outputs, compared to simpler reports
and protocols. RMarkdown can now be used to create very large and dynamic
documents, including online books (which can also be rendered to pdf versions
suitable for printing), dashboard-style websites, and blogs. Once members of
your research group are familiar with the basics of RMarkdown, you may want to
explore using it to create these more complex outputs. The book format divides
content into chapters and special sections like appendices and references. It
includes a table of contents based on weblinks, so readers can easily navigate
the content. It uses a book format as its base that allows readers to do things
like change the font size and search the book text for keywords. The book
containing these modules is one example of using bookdown. If you would like to
explore using bookdown to create online books based on Rmarkdown files, there
are a number of resources available. There is an online book available with
extensive instructions on using this package, available at
<a href="https://bookdown.org/yihui/bookdown/" class="uri">https://bookdown.org/yihui/bookdown/</a>. There is also a helpful website with more
details on this package, <a href="https://bookdown.org/" class="uri">https://bookdown.org/</a>. The website include a gallery of
example books created with bookdown <a href="https://bookdown.org/home/archive/" class="uri">https://bookdown.org/home/archive/</a>, which
you can use to explore the types of books that can be created.</p>
<p>You can also use Rmarkdown documents to create webpages, with pages included for
blogs. This format allows you to create a very attractive website that includes
a blog section, where you can write and regularly post new blogs, keeping the
site dynamic. It is a nice entry point to developing and maintaining a website
for people who are learning to code in R but otherwise haven’t done much coding,
as you can do all the steps within RStudio. There are templates for these blogs
that are appropriate for creating personal or research group websites for
academics. These websites can be created to highlight the research and people in
your research lab. You can encourage students and postdocs to create personal
sites, to raise the profile of their research. In the past, we have even used
one as a central, unifying spot for a group study, with students contributing
blog posts as their graded assignment
(<a href="https://kind-neumann-789611.netlify.app/" class="uri">https://kind-neumann-789611.netlify.app/</a>). To learn how to create websites with
blogs, you can check the book <em>blogdown: Creating Websites with R Markdown</em>
[ref], which is available both in print and free online at
<a href="https://bookdown.org/yihui/blogdown/" class="uri">https://bookdown.org/yihui/blogdown/</a>. This process takes a bit of work to
initially get the website set up, but then allows for easy and straightforward
maintenance.</p>
<p>Finally, a simpler way to make basic web content with RMarkdown is through their
<em>flexdashboard</em> format. This format creates a smaller website that is focused on
sharing data results—you can see a gallery of examples at
<a href="https://rmarkdown.rstudio.com/flexdashboard/examples.html" class="uri">https://rmarkdown.rstudio.com/flexdashboard/examples.html</a>. This format is
excellent for creating a webpage that allows users to view complex, and
potentially interactive, results from data you’ve collected. It can be
particularly helpful for groups that need to quickly communicate regularly
updated data to viewers. During 2020, for example, many public health
departments maintained dashboard-style websites to share evolving data on
COVID-19 in the community. Using RMarkdown in this case has the key advantage of
allowing you to easily update the dashboard webpage as you get new or updated
data, since it is easy to re-run any data processing, analysis, and
visualization code in the document. To learn how to use RMarkdown to create
dashboard websites, you can check out RStudio’s flexdashboard site at
<a href="https://rmarkdown.rstudio.com/flexdashboard/index.html" class="uri">https://rmarkdown.rstudio.com/flexdashboard/index.html</a>. There is also guidance
available in one of the chapters of <em>R Markdown: The Definitive Guide</em> [ref]:
<a href="https://bookdown.org/yihui/rmarkdown/dashboards.html" class="uri">https://bookdown.org/yihui/rmarkdown/dashboards.html</a>.</p>
<p><strong>More complex formatting.</strong></p>
<p>As mentioned earlier, Markdown is a fairly simple markup language. Occasionally,
this simplicity means that you might not be able to create fancier formatting
that you might desire. There is a method that allows you to work around this
constraint in RMarkdown.</p>
<p>In Rmarkdown documents, when you need more complex formatting, you can shift
into a more complex markup language for part of the document. Markup languages
like LaTeX and HTML are much more expressive than Markdown, with many more
formatting choices possible. For example, there is functionality within
LaTeX and HTML to create much more complex tables than in Markdown. However,
there is a downside—when you include formatting specified in these
more complex markup languages, you will limit the output formats that you can
render the document to. For example, if you include LaTeX formatting within
an RMarkdown document, you must output the document to PDF, while if you
include HTML, you must output to an HTML file. Conversely, if you stick with
the simpler formatting available through the Markdown syntax, you can easily
switch the output format for your document among several choices.</p>
<p>The <em>R Markdown Cookbook</em> [ref] includes chapters on how to customize Rmarkdown
output through LaTeX
(<a href="https://bookdown.org/yihui/rmarkdown-cookbook/latex-output.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/latex-output.html</a>) and HTML
(<a href="https://bookdown.org/yihui/rmarkdown-cookbook/html-output.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/html-output.html</a>). These
customizations can include creating custom formats for the entire document (for
example, you can customize the appearance of a whole HTML document by
customizing the CSS style file for the document). They can also include
smaller-level customizations, like changing the citation style that is used in
conjunction with a BibTeX file by adding to the preamble for LaTeX output.</p>
<p>One area of customization that is particularly useful and simple to implement is
with customized tables. The Markdown syntax can create very simple tables, but
does not allow the creation of more complex tables. There is an R package called
<code>kableExtra</code> [ref] that allows you to create very attractive and complex tables
in RMarkdown documents.</p>
<p>This package leverages more of the power of underlying markup languages, rather
than the simpler Markdown language. If you remember, Markdown is pretty easy to
learn because it has a somewhat limited set of special characters and special
markings that you can use to specify formatting in your output document. This
basic set of functionality is often all you need, but for complex table
formatting, you will need more. There is much more available in the deeper
markup languages that you can use specifically to render pdf documents (software
derived from TeX) and the one that you can use specifically to render HTML (the
HTML markup language). As a result, you will need to create RMarkdown files that
are customized to a single output format (pdf or HTML) to take advantage of this
package.</p>
<p>You can install this package the same as any other R package from CRAN, using
<code>install.packages</code>. You will need to use then need to use <code>library("kableExtra)</code>
within your RMarkdown document before you use functions from the package. The
<code>kableExtra</code> package is extensively documented through two vignettes that come
with package, one if the output will be in pdf
(<a href="https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_pdf.pdf" class="uri">https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_pdf.pdf</a>)
and one if it will be in HTML
(<a href="https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html" class="uri">https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html</a>).
There is also information on using <code>kableExtra</code> available through <em>R Markdown
Cookbook</em> [ref]: <a href="https://bookdown.org/yihui/rmarkdown-cookbook/kableextra.html" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/kableextra.html</a>.</p>
</div>
<div id="learning-more-about-rmarkdown." class="section level3" number="3.10.4">
<h3><span class="header-section-number">3.10.4</span> Learning more about Rmarkdown.</h3>
<p>To learn more about RMarkdown, you can explore a number of excellent resources.
The most comprehensive are shared by RStudio, where RMarkdown’s developer
and maintainer, Yihui Xie, works. These resources are all freely available
online, and some are also available to buy as print books, if you prefer that
format.</p>
<p>First, you should check out the online tutorials that are provided by RStudio on
RMarkdown. These are available at RStudio’s RMarkdown page:
<a href="https://rmarkdown.rstudio.com/" class="uri">https://rmarkdown.rstudio.com/</a>. The page’s “Getting Started” section
(<a href="https://rmarkdown.rstudio.com/lesson-1.html" class="uri">https://rmarkdown.rstudio.com/lesson-1.html</a>) provides a nice introduction you
can work through to try out RMarkdown and practice the overview provided in the
last subsection of this module. The “Articles” section
(<a href="https://rmarkdown.rstudio.com/articles.html" class="uri">https://rmarkdown.rstudio.com/articles.html</a>) provides a number of other
documents to help you learn RMarkdown. RStudio’s RMarkdown page also includes a
“Gallery” (<a href="https://rmarkdown.rstudio.com/gallery.html" class="uri">https://rmarkdown.rstudio.com/gallery.html</a>). This resource allows you
to browse through example documents, so you can get a visual idea of what you
might want to create and then access the example code for a similar document.
This is a great resource for exploring the variety of documents that you can
create using RMarkdown.</p>
<p>To go more deeply into RMarkdown, there are two online books from some of the
same team that are available online. The first is <em>R Markdown: The Definitive
Guide</em> by Yihui Xie, J. J. Allaire, and Garrett Grolemund [ref]. This book is
available free online at <a href="https://bookdown.org/yihui/rmarkdown/" class="uri">https://bookdown.org/yihui/rmarkdown/</a>. It moves from
basics through very advanced functionality that you can implement with
RMarkdown, including several of the topics we highlight later in this
subsection.</p>
<p>The second online book to explore from this team is <em>R Markdown Cookbook</em>, by
Yihui Xie, Christophe Dervieux, and Emily Riederer [ref]. This book is available
free online at <a href="https://bookdown.org/yihui/rmarkdown-cookbook/" class="uri">https://bookdown.org/yihui/rmarkdown-cookbook/</a>. This book is a
helpful resource for dipping in to a specific section when you want to learn how
to achieve a specific task. Just like a regular cookbook has recipes that you
can explore and use one at a time, this book does not require a comprehensive
end-to-end read, but instead provides “recipes” with advice and instructions for
doing specific things. For example, if you want to figure out how to align a
figure that you create in the center of the page, rather than the left, you can
find a “recipe” in this book to do that.</p>
<!-- ### Potential quotes / refs  -->
<!-- > "WordPerfect was always the best word processor. Because it allowed for insight into -->
<!-- its very structure. You could hit a certain key combination and suddenly the screen -->
<!-- would split and you'd reveal the codes, the bolds and italics, and so forth,  -->
<!-- that would define your text when it was printed. It was beloved of legal secretaries -->
<!-- and journalists alike. Because when you work with words, at the practical, everyday -->
<!-- level, the ability to look under the hood is essential. Words are not simple. And -->
<!-- WordPerfect acknowledged that. Microsoft Word did not. Microsoft kept insisting that -->
<!-- what you saw on your screen was the way things *were*, and if your fonts just kept -->
<!-- sort of randonmly changing, well, you must have wanted it that way. Then along came  -->
<!-- HTML, and what I remember most was that sense of being back inside the file. Sure,  -->
<!-- HTML was a typographic nightmare, a bunch of unjustified Times New Roman in 12 pt on  -->
<!-- screens with chiclet-size pixels, but under the hood you could see all the pieces.  -->
<!-- Just like WordPerfect. That transparency was a wonderful thing, and it renewed -->
<!-- computing for me." [@ford2015on] -->
<!-- > "TeX was created by Donald E. Knuth, a professor at Stanford University who has -->
<!-- achieved international renown as a mathematician and computer scientist. -->
<!-- Knuth also has an aesthetic sense uncommon in his field, and his work output is -->
<!-- truly phenomenal. TeX is a happy byproduct of Knuth's mammoth enterprise,  -->
<!-- *The Art of Computer Programming*. This series of reference books, designed -->
<!-- to cover the whole gamut of programming concepts and techniques, is a  -->
<!-- *sine qua non* for all computer scientists." [@seroul2012beginner] -->
<!-- > "Roughly speaking, text processors fall into two categories: -->
<!-- (1) WYSIWYG systems: what you see is what you get. You see on the screen at all -->
<!-- times what the printed document will look like, and what you type has immediate -->
<!-- effect on the appearance of the document. (2)  markup systems, where you type your text  -->
<!-- interspersed with formatting instructions, but don't see their effect right away. You must run a program to examine the -->
<!-- resulting image, whether on paper or on the screen. In computer science jargon, -->
<!-- markup systems must compile the source file you type.  WYSIWYG systems have the obvious  -->
<!-- advantage of immediate feedback, but they -->
<!-- are not very precise: what is acceptable at a resolution of 300 dots per inch, for an -->
<!-- ephemeral publication such as a newsletter or flier, is no longer so for a book that -->
<!-- will be phototypeset at high resolution. The human eye is extraordinarily sensitive: -->
<!-- you can be bothered by the appearance of a text without being able to pinpoint why, -->
<!-- just as you can tell when someone plays the wrong note in an orchestra, without -->
<!-- being able to identify the CUlprit. One quickly leams in typesetting that the beauty, -->
<!-- legibility and comfortable reading of a text depend on minute details: each element -->
<!-- must be placed exactly right, within thousandths of an inch. For this type of work, -->
<!-- the advantage of immediate feedback vanishes: fine details of spacing, alignment, -->
<!-- and so on are much too small to be discernible at the screen's relatively low  -->
<!-- resolution, and even if it such were not the case, it would still be a monumental chore -->
<!-- to find the right place for everything by hand. For this reason it is not surprising that in the world of professional typesetting -->
<!-- markup systems are preferred. They automate the task of finding the right place -->
<!-- for each character with great precision. Naturally, this approach is less attractive for -->
<!-- beginners, since one can't see the results as one types, and must develop a feeling -->
<!-- for what the system will do. But nowadays, you can have the best of both worlds -->
<!-- by using a markup system with a WYSIWYG *front end*; we'll talk about such front -->
<!-- ends for TEX later on. TEX was developed in the late seventies and early eighties,  -->
<!-- before WYSIWYG systems were widespread. But were it to be redesigned now, it would  -->
<!-- still be a markup -->
<!-- language. To give you an idea of the precision with which TEX operates: the internal  -->
<!-- unit it uses for its calculations is about a hundred times smaller than the -->
<!-- wavelength of visible light! (That's right, a hundred times.) In other words, any -->
<!-- round-off error introduced in the calculations is invisible to the naked eye."  -->
<!-- [@seroul2012beginner] -->
<!-- > "You should be sure to understand the difference between a text editor and a text -->
<!-- processor. A text processor is a text editor together with formatting software that -->
<!-- allows you to switch fonts, do double columns, indent, and so on. A text editor -->
<!-- puts your text in a file on disk, and displays a portion of it on the screen. It doesn't -->
<!-- format your text at all. We insist on the difference because those accustomed to WYSIWYG systems are -->
<!-- often not aware of it: they only know text processors. Where can you find a text -->
<!-- editor? Just about everywhere. Every text processor includes a text editor which -->
<!-- you can use. But if you use your text processor as a text editor, be sure to save your -->
<!-- file using a 'save ASCII' or 'save text only' option, so that the text processor's own -->
<!-- formatting commands are stripped off. If you give TEX a file created without this -->
<!-- precaution, you'll get garbage, because TEX cannot digest your text processor's -->
<!-- commands." [@seroul2012beginner] -->
<!-- > "TeX enabled authors to encode their precise intent into their manuscripts: -->
<!-- This block of text is a computer program, while this word is a keyword in that -->
<!-- program. The language it used, called TeX markup, formalized the slow, -->
<!-- error-prone communication that is normally carried out with the printer over -->
<!-- repeated galley proofs." [@apte2019lingua] -->
<!-- > "The idea of writing markup inside text wasn’t especially novel; it has been -->
<!-- used from 1970’s runoff (the UNIX family of printer-preparation utilities) to -->
<!-- today’s HTML tags. TeX was new in that it captured key concepts necessary for -->
<!-- realistic typesetting and formalized them." [@apte2019lingua] -->
<!-- > "With these higher-level commands, the free TeX engine, and the LaTeX book, -->
<!-- the use of TeX exploded. The macro file has since evolved and changed names, but -->
<!-- authors still typically run the program called latex or its variants. Hence, -->
<!-- most people who write TeX manuscripts know the program as LaTeX and the commands -->
<!-- they use as LaTeX commands." [@apte2019lingua] -->
<!-- > "The effect of LaTeX on scientific and technical publishing has been profound. -->
<!-- Precise typesetting is critical, particularly for conveying concepts using -->
<!-- chemical and mathematical formulas, algorithms, and similar constructs. The -->
<!-- sheer volume of papers, journals, books, and other publications generated in the -->
<!-- modern world is far beyond the throughput possible via manual typesetting. And -->
<!-- TeX enables automation without losing precision. Thanks to LaTeX, book authors -->
<!-- can generate camera-ready copy on their own. Most academic and journal -->
<!-- publishers accept article manuscripts written in LaTeX, and there’s even an open -->
<!-- archive maintained by Cornell University where authors of papers in physics, -->
<!-- chemistry, and other disciplines can directly submit their LaTeX manuscripts for -->
<!-- open viewing. Over 10,000 manuscripts are submitted to this archive every month -->
<!-- from all over the world." [@apte2019lingua] -->
<!-- > "For many users, a practical difficulty with typesetting using TeX is -->
<!-- preparing the manuscripts. When TeX was first developed, technical authors were -->
<!-- accustomed to using plain-text editors like WordStar, vi, or Emacs with a -->
<!-- computer keyboard. The idea of marking up their text with commands and running -->
<!-- the manuscript through a typesetting engine felt natural to them. Today’s -->
<!-- typesetters, particularly desktop publishers, have a different mental model. -->
<!-- They expect to see the output in graphical form and then to visually make edits -->
<!-- with a mouse and keyboard, as they would in any WYSIWYG program. They might not -->
<!-- be too picky about the quality of the output, but they appreciate design -->
<!-- capabilities, such as the ability to flow text around curved outlines. Many -->
<!-- print products are now produced with tools like Microsoft Word for this very -->
<!-- reason. TeX authors cannot do the same work as easily." [@apte2019lingua] -->
<!-- > "Poor documentation can lead to irreproducibility and serious errors. There's -->
<!-- a vast amount of lurking complexity in bioinformatics work: complex workflows,  -->
<!-- multiple files, countless program parameters, and different software versions. -->
<!-- The best way to prevent this complexity from causing problems is to document -->
<!-- everything extensively. Documentation also makes your life easier when you need to  -->
<!-- go back and rerun an analysis, write detailed methods about your steps for a  -->
<!-- paper, or find the origin of some data in a directory." [@buffalo2015bioinformatics] -->
<!-- > "Scatterplots are useful for visualizing treatment-response comparisons ...,  -->
<!-- assocations between variables ..., or paired data (e.g., a disease biomarker in  -->
<!-- several patients before and after treatment)." [@holmes2018modern] -->
<!-- > "Sometimes we want to show the relationshiips between more than two variables. -->
<!-- Obvious choices for including additional dimensions are plot symbol shapes and -->
<!-- colors. ... Another way to show additional dimensions of the data is to show -->
<!-- multiple plots that result from repeatedly subsetting (or 'slicing') the data -->
<!-- based on one (or more) of the variables, so that we can visualize each part -->
<!-- separately. This is called faceting and it enables us to visualize data in up to -->
<!-- four or five dimensions. So we can, for instance, investigate whether the -->
<!-- observed patterns among the other variables are the same or different across the -->
<!-- range of the faceting variable." [@holmes2018modern] -->
<!-- > "You can add an enormous amount of information and expressivity by making your -->
<!-- plots interactive. ... The package `ggvis` is an attempt to extend the good -->
<!-- features of `ggplot2` into the realm of interactive graphics. In contrast to -->
<!-- `ggplot2`, which produces graphics into R's traditional graphics devices (PDF, -->
<!-- PNG, etc.), `ggvis` builds upon a JavaScript infrastructure called *Vega*, and -->
<!-- its plots are intended to be viewed in an HTML browser." [@holmes2018modern] -->
<!-- > "Heatmaps are a powerful way of visualizing large, matrix-like datasets and  -->
<!-- providing a quick overview of the patterns that might be in the data. There  -->
<!-- are a number of heatmap drawing functions in R; one that is convenient and  -->
<!-- produces good-looking output is the function `pheatmap` from the eponymous  -->
<!-- package." [@holmes2018modern] -->
<!-- > "Plots in which most points are huddled in one area, with much of the available -->
<!-- spaces sparesly populated, are difficult to read. If the histogram of the  -->
<!-- marginal distribution of a variable has a sharp peak and then long tails to  -->
<!-- one or both sides, transforming the data can be helpful. ... The plots in  -->
<!-- this chapter that involve microarray data use the logarithmic transformation -->
<!-- [footnote: 'We used it implicitly, since the data in the `ExpressionSet` object -->
<!-- `x` already came log-transformed']---not only in scatterplots... for the x- and -->
<!-- y-coordinates but also ... for the color scale that represents the  -->
<!-- expression fold change. The logarithm transformation is attractive because it has -->
<!-- a definite meaning---a move up or down by the same amount on a log-transformed -->
<!-- scale corresponds to the same multiplicative change on the original scale:  -->
<!-- log(ax) = log a + log x. Sometimes, however, the logarithm is not good enough,  -->
<!-- for instance when the data include zero or negative values, or when even on the  -->
<!-- logarithmic scale the data distribution is highly uneven." [@holmes2018modern] -->
<!-- > "To visualize genomic data, in addition to the general principles we have -->
<!-- discussed in this chapter, there are some specific considerations. The data are -->
<!-- usually associated with genomic coordinates. In fact, genomic coordinates offer -->
<!-- a great organizing principle for the integration of genomic data. ... The main  -->
<!-- challenge of genomic data visualization is the size of the genomes. We need -->
<!-- visualization at multiple scales, from whole genome down to the  -->
<!-- nucleotide level. It should be easy to zoom in and out, and we may need different -->
<!-- visualization strategies for the different size scales. It can be convenient to  -->
<!-- visualize biological molecules (genomes, genes, transcripts, proteins) in a linear -->
<!-- manner, although their embedding in the physical world can matter (a great deal)." -->
<!-- [@holmes2018modern] -->
<!-- > "Visualizing the data, either 'raw' or along the various steps of processing,  -->
<!-- summarization, and inference, is one of the most important activities in applied -->
<!-- statistics and, indeed, in science. It sometimes gets short shrift in textbooks -->
<!-- since there is not much deductive theory. However, there are many good (and bad) -->
<!-- practices, and once you pay attention to it, you will quickly see whether a  -->
<!-- certain graphic is effective in conveying its message, or what -->
<!-- choices you could make to create powerful and aesthetically attractive data -->
<!-- visualizations." [@holmes2018modern] -->
</div>
<div id="applied-exercise-7" class="section level3" number="3.10.5">
<h3><span class="header-section-number">3.10.5</span> Applied exercise</h3>

</div>
</div>
<div id="example-creating-a-reproducible-data-pre-processing-protocol" class="section level2" number="3.11">
<h2><span class="header-section-number">3.11</span> Example: Creating a reproducible data pre-processing protocol</h2>
<p>We will walk through an example of creating a reproducible protocol for the
automated gating of flow cytometry data for a project on the immunology of
tuberculosis lead by one of our Co-Is. This data pre-processing protocol was
created using RMarkdown and allows the efficient, transparent, and reproducible
gating of flow cytometry data for all experiments in the research group. We will
walk the trainees through how we developed the protocol initially, the final
pre-processing protocol, how we apply this protocol to new experimental data.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Explain how a reproducible data pre-processing protocol can be integrated into
a real research project</li>
<li>Understand how to design and implement a data pre-processing protocol to
replace manual or point-and-click data pre-processing tools</li>
</ul>
<div id="introduction-and-example-data" class="section level3" number="3.11.1">
<h3><span class="header-section-number">3.11.1</span> Introduction and example data</h3>
<p>In this module, we’ll provide advice and an example of how you can use the
tools for knitted documents to create a reproducible data preprocessing
protocol. This module builds on ideas and techniques that were introduced
in the last two modules, to help you put them into practical use for
data preprocessing that you do repeatedly for research data in your
laboratory.</p>
<p>In this module, we will use an example of a common pre-processing task in
immunological research: estimating the bacterial load in samples by plating at
different dilutions, identifying a good dilution for counting colony-forming
units (CFUs), and then back-calculating the estimated bacterial load in the
original sample based on the colonies counted at this dilution. This
experimental technique dates back to the late 1800s, with Koch, and continues to
be widely used in microbiology research and applications today
<span class="citation">(Ben-David and Davidson 2014)</span>. These data are originally from example data for an R
package called <code>bactcountr</code>, currently under development at
<a href="https://github.com/aef1004/bactcountr/tree/master/data" class="uri">https://github.com/aef1004/bactcountr/tree/master/data</a>.</p>
<p>These data represent data from a common laboratory task in immunological
research: estimating the bacterial load in samples by plating each sample at
different dilutions and counting colony-forming units at a “good” dilution for
each sample. For example, you may be testing out some drugs against an
infectious bacteria and want to know how successful different drugs are in
limiting bacterial load. You run an experiment and have samples from animals
treated with different drugs or under control. You want to know: How much viable
(i.e., replicating) bacteria are in each of your samples?</p>
<p>You can find out by plating the sample at different dilutions and
counting the colony-forming units (CFUs) that are cultured on each plate.
You put a sample on a plate with a medium they can grow on and then give them
time to grow. The idea is that individual bacteria from the original sample end
up randomly around the surface of the plate, and any that are viable (able to
reproduce) will form a new colony that, after a while, you’ll be able to see.</p>
<p>To count the number of colonies, you need a “just right” dilution (and you
typically won’t know which dilution this is for a sample until after plating) to
have a countable plate. If you have too high of a dilution (i.e., one with very
few viable bacteria), randomness will play a big role in the CFU count, and
you’ll estimate the original with more variability. If you have too low of a
dilution (i.e., one with lots of viable bacteria), it will be difficult to
identify separate colonies, and they may complete for resources. To translate
from diluted concentration to original concentration, you can then do a
back-calculation, incorporating both the number of colonies counted at that
dilution and how dilute the sample was. There is therefore some pre-processing
required (although it is fairly simple) to prepare the data collected by
counting the CFUs on each plate for use in statistical testings and to
combine with other experimental data.</p>
<p>We will use this example pre-processing protocol as an illustration in this
module. If you would like, you can access all the components and follow along
with the example, re-rendering it yourself on your own computer. The example
data are available as a csv file, downloadable
<a href="https://raw.githubusercontent.com/geanders/improve_repro/master/data/bactcountr_example_data/cfu_data.csv">here</a>.
You can open this file using spreadsheet software, or look at it directly in
RStudio.</p>
<p>The final pre-processing protocol for these data can be downloaded, including
both <a href="https://raw.githubusercontent.com/geanders/improve_repro/master/data/bactcountr_example_data/example_protocol.Rmd">the original RMarkdown
file</a>
and <a href="https://github.com/geanders/improve_repro/raw/master/data/bactcountr_example_data/example_protocol.pdf">the output PDF
document</a>.
Throughout this module, we will walk through elements of this document, to
provide an example as we explain the process of developing data pre-processing
modules for common tasks in your research group.</p>
<p>This example is intentionally simple, to allow a basic introduction to the
process using pre-processing tasks that are familiar to many laboratory-based
scientists and easy to explain to anyone who has not used plating experimental
work. However, the same general process can also be used to create
pre-processing protocols for data that are much larger or more complex or for
pre-processing pipelines that are much more involved.</p>
</div>
<div id="advice-on-designing-a-pre-processing-protocol" class="section level3" number="3.11.2">
<h3><span class="header-section-number">3.11.2</span> Advice on designing a pre-processing protocol</h3>
<p>Before you write your protocol in a knitted document, you should decide on the
content to include in the protocol. This section provides tips on this design
process.</p>
<p><strong>Defining input and output data for the protocol.</strong></p>
<p>The first step in designing the data pre-processing protocol is to decide on
what the starting point will be for the protocol (the data input) and what will
be the ending point (the data output). It may make sense to design a separate
protocol for each major type of data that you collect in your research
laboratory. Your input data for the protocol, under this design, might be the
data that is output from a specific type of equipment (e.g., flow cytometer) or
from a certain type of sample or measurement (e.g., metabolomics run on a mass
spectrometer), even if it is a fairly simple type of data (e.g., CFUs from
plating data, as used in the example protocol for this module). For example, say
you are working with data from a flow cytometer, metabolomics data measuremd
with a mass spectrometer, and bacterial load data measured by plating data and
counting colony forming units (CFUs). In this case, you may want to create three
pre-processing protocols: one for the flow data, one for the metabolomics data,
and one for the CFU data.</p>
<p>While pre-processing protocols for some types of data might be very complex,
others might be fairly simple. However, it is still worthwhile to develop a
protocol, as it allows you to pass along some of the details of pre-processing
the data that might have become “common sense” to longer-tenured members of your
research group. For example, The pre-processing protocol is fairly simple for
the CFU data used in the example protocol for this module. This protocol inputs
data collected in a plain-text delimited file (a csv file, in the example).
Within the protocol, there are be steps to convert initial measurements from
plating at different dilutions into estimates of the bacterial load in each
original sample. There are also sections in the protocol for exploratory data
analysis, to allow for quality assessment and control of the collected data as
part of the preprocessing. The output of the protocol is a simple data object (a
dataframe, in this example) with the bacterial load for each original sample.
These data are now ready to be used in tables and figures in the research report
or manuscript, as well as to explore associations with the experimental design
details (e.g., comparing bacterial load in treated versus untreated animals) or
merged with other types of experimental data (e.g., comparing immune cell
populations, as measured with flow cytometry data, with bacterial loads, as
measured from plating and counting CFUs).</p>
<p>Once you have identified the input data type to use for the protocol, you should
identify an example dataset from your laboratory that you can use to create the
protocol. This could be a dataset that you currently need to pre-process, in
which case the development of the protocol will serve a second purpose as
allowing you to complete this task at the same time. However, you may not have a
new set of data of this type that you need to pre-process, and in this case you
can build your protocol using a dataset from a previous experiment in your
laboratory. In this case, you may already have a record of the steps that you
used to pre-process the data previously, and these can be helpful as a starting
point as you draft the more thorough pre-processing protocol. You may want to
select an example dataset that you have already published or are getting ready
to publish, so you won’t feel awkard about making the data available for people
to practice with. If you don’t have an example dataset from your own
laboratory, you can explore example datasets that are already available, either
as data included with existing R packages or through open repositories,
including those hosted through national research institutions like the NIH. In
this case, be sure to cite the source of the data and include any available
information about the equipment that was used to collect it and the settings
used when the data were collected.</p>
<p>You can include the data and the RMarkdown file for the protocol in an RStudio
Project (see module [x]) and post this either publicly or privately on GitHub
(see modules [x]). This creates a “packet” of everything that a reader needs to
use to recreate what you did—they can download the whole GitHub repository and
will have a nice project directory on their computer with everything they need
to try out the protocol.</p>
<p>For the example protocol for this module, we want to pre-process data that were
collected “by hand” by counting CFUs on plates in the laboratory. These counts
were recorded in a plain text delimited file (a csv file) using spreadsheet
software. The spreadsheet was set up to ensure the data were recorded in a
“tidy” format, as described in module 2.3. The first few rows of the input data
look like:</p>
<pre><code>## # A tibble: 6 x 4
##   group replicate dilution  CFUs
##   &lt;dbl&gt; &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;
## 1     2 2-A              0    26
## 2     2 2-C              0     0
## 3     3 3-A              0     0
## 4     3 3-C              0     0
## 5     4 4-A              0     0
## 6     4 4-B              0     0</code></pre>
<p>Each row represents the number of bacterial colonies counted after plating a
certain sample at a certain dilution. Columns are included with values for the
experimental group of the sample (<code>group</code>), the specific ID of the sample within
that experimental group (<code>replicate</code>, e.g., <code>2-A</code> is mouse A in experimental
group 2), the dilution level for that plating (<code>dilution</code>), and the number of
bacterial colonies counted in that sample (<code>CFUs</code>).</p>
<p>When you have identified the input data type you will use for the protocol,
as well as selected an example dataset of this type to use to create the
protocol, you can include a subsection in the “Overview” section of the protocol
that describes these input data, what file format they are in, and how they
can be read into R for pre-processing (Figure <a href="experimental-data-preprocessing.html#fig:protocoldatainput">3.11</a>).</p>
<div class="figure"><span id="fig:protocoldatainput"></span>
<img src="figures/protocol_data_input.png" alt="Providing details on input data in the pre-processing protocol. Once you have an example data file for the type of data that will be input for the protocol, you can add a section that provides the code to read the data into R. You can also add code that will show the first few rows of the example dataset, as well as a description of the data. This figure shows examples of how these elements can be added to an RMarkdown file for a pre-processing protocol, and the associated elements in the final pdf of the protocol, using the example protocol for this module." width="\textwidth" />
<p class="caption">
Figure 3.11: Providing details on input data in the pre-processing protocol. Once you have an example data file for the type of data that will be input for the protocol, you can add a section that provides the code to read the data into R. You can also add code that will show the first few rows of the example dataset, as well as a description of the data. This figure shows examples of how these elements can be added to an RMarkdown file for a pre-processing protocol, and the associated elements in the final pdf of the protocol, using the example protocol for this module.
</p>
</div>
<p>For the data output, it often makes sense to plan for data in a format that is
appropriate for data analysis and for merging with other types of data collected
from the experiment.</p>
<p><strong>Outlining key tasks in pre-processing the input data.</strong></p>
<p>The next step is to outline the key tasks that are involved in moving from the
data input to the desired data output. For the plating data we are using for our
example, the key tasks to be included in the pre-processing protocol are:</p>
<ol style="list-style-type: decimal">
<li>Read the data into R</li>
<li>Explore the data and perform some quality checks</li>
<li>Identify a “good” dilution for each sample—one at which you have a
“countable” plate</li>
<li>Estimate the bacterial load in each original sample based on the CFUs counted
at the “good” dilution for the sample</li>
<li>Output data with the estimated bacterial load for each original sample</li>
</ol>
<p>Once you have this basic design, you can set up the pre-processing protocol to
include a separate section for each task, as well as an “Overview” section at
the beginning to describe the overall protocol, the data being pre-processed,
and the laboratory procedures used to collect those data. In RMarkdown, you can
create first-level section headers by putting the text for the header on its own
line and beginning that line with <code>#</code>, followed by a space. You should include a
blank line before and after the line with this header text. Figure
<a href="experimental-data-preprocessing.html#fig:protocolsections">3.12</a> shows how this is done in the example protocol for
this module, showing how text in the plain text RMarkdown file for the protocol
align with section headers in the final pdf output of the protocol.</p>
<div class="figure"><span id="fig:protocolsections"></span>
<img src="figures/protocol_sections.png" alt="Dividing an RMarkdown data pre-processing protocol into sections. This shows an example of creating section headers in a data pre-processing protocol created with RMarkdown, showing section headers in the example pre-procotcol for this module." width="\textwidth" />
<p class="caption">
Figure 3.12: Dividing an RMarkdown data pre-processing protocol into sections. This shows an example of creating section headers in a data pre-processing protocol created with RMarkdown, showing section headers in the example pre-procotcol for this module.
</p>
</div>
<p><strong>Adding code for pre-processing.</strong></p>
<p>For many of these steps, you likely have code—or can start drafting the
code—required for that step. If you were doing the pre-processing entirely
through a code script or interactively, you would exclusively write code for the
steps. A good next step, therefore, in designing your pre-protocol is to add in
the code for each step to conduct the pre-processing required in each section.</p>
<p>In RMarkdown, you can test this code as you write it. You insert each piece
of executable code within a special section, separated from the regular
text with special characters, as described in previous modules.</p>
<p>For any pre-processing steps that are straightforward (e.g., calculating the
dilution factor in the example module, which requires only simple mathematical
operations), you can directly write in the code required for the step.
For other pre-processing steps, however, the algorithm may be a bit more
complex. For example, complex algorithms have been developed for steps like
peak identification and alignment that are required when
pre-processing data from a mass spectrometer.</p>
<p>For these more complex tasks, you can start to explore available R packages for
performing the task. There are thousands of packages available that extend the
basic functionality of R, providing code implementations of algorithms in a
variety of scientific fields. Many of the R packages relevant for biological
data—especially high-throughput biological data—are available through a
repository called Bioconductor. These packages are all open-source (so you can
explore their code if you want to) and free. You can use vigettes and package
manuals for Bioconductor packages to identify the different functions you can
use for your pre-processing steps. Once you have identified a function for the
task, you can use the helpfile for the function to see how to use it. This help
documentation will allow you to determine all of the function’s parameters and
the choices you can select for each.</p>
<p>You can add each piece of code in the RMarkdown version of the protocol using
the standard method for RMarkdown (module [x]). Figure <a href="experimental-data-preprocessing.html#fig:protocolcode">3.13</a>
shows an example from the example protocol for this module. Here, we are using
code to help identify a “good” dilution for counting CFUs for each sample. The
code in included in an executable code chunk, and so it will be run each time
the protocol is rendered. Code comments are included in the code to provide
finer-level details about what the code is doing.</p>
<div class="figure"><span id="fig:protocolcode"></span>
<img src="figures/protocol_code.png" alt="Example of including code in a data pre-processing protocol created with RMarkdown. This figure shows how code can be included in the RMarkdown file for a pre-processing protocol (right), and the corresponding output in the final pdf of the protocol (left), for the code to identify a 'good' dilution for counting CFUs for each sample. Code comments are included to provide finer-level details on the code." width="\textwidth" />
<p class="caption">
Figure 3.13: Example of including code in a data pre-processing protocol created with RMarkdown. This figure shows how code can be included in the RMarkdown file for a pre-processing protocol (right), and the corresponding output in the final pdf of the protocol (left), for the code to identify a ‘good’ dilution for counting CFUs for each sample. Code comments are included to provide finer-level details on the code.
</p>
</div>
</div>
<div id="writing-data-pre-processing-protocols" class="section level3" number="3.11.3">
<h3><span class="header-section-number">3.11.3</span> Writing data pre-processing protocols</h3>
<p>Now that you have planned out the key components of the pre-processing protocol,
you can use RMarkdown’s functionality to flesh it out into a full pre-processing
protocol. This gives you the chance to move beyond a simple code script, and
instead include more thorough descriptions of what you’re doing at each step and
why you’re doing it. You can also include discussions of potential limitations
of the approach that you are taking in the pre-processing, as well as areas
where other research groups might use a different approach. These details can
help when it is time to write the Methods section for the paper describing your
results from an experiment using these data. They can also help your research
group identify pre-processing choices that might differ from other research
groups, which opens the opportunity to perform sensitivity analyses regarding
these pre-processing choices and ensure that your final conclusions are robust
across multiple reasonable pre-processing approaches.</p>
<p>Protocols are common for wet lab techniques, where they provide a “recipe” that
ensures consistency and reproducibility in those processes. Computational tasks,
including data pre-processing, can also be standardized through the creation and
use of protocol in your research group. While code scripts are becoming more
common as a means of recording data pre-processing steps, they are often not
as clear as a traditional protocol, in particular in terms of providing a
thorough description of what is being done at each step and why it is being
done that way. Data pre-processing protocols can provide these more thorough
descriptions, and by creating them with RMarkdown or with similar types
of “knitted” documents (module [x]), you can combine the executable code
used to pre-process the data with extensive documentation. As a further
advantage, the creation of these protocols will ensure that your research
group has thought carefully about each step of the process, rather than
relying on cobbling together bits and pieces of code they’ve found but don’t
fully understand. Just as the creation of a research protocol for a
clinical trial requires a careful consideration of each step of the ultimate
trial <span class="citation">(Al-JunDi and SAkkA 2016)</span>, the creation of data pre-processing protocols ensure
that each step in the process is carefully considered, and so helps to
ensure that each step of this process is conducted as carefully as the
steps taken in designing the experiment as a whole and each wet lab technique
conducted for the experiment.</p>
<blockquote>
<p>“Writing a research proposal is probably one of the most challenging and
difficult task as research is a new area for the majority of postgraduates and
new researchers. … Protocol writing allows the researcher to review and
critically evaluate the published literature on the interested topic, plan and
review the project steps and serves as a guide throughout the investigation.”
<span class="citation">(Al-JunDi and SAkkA 2016)</span></p>
</blockquote>
<p>A data-preprocessing protocol, in the sense we use it here, is essentially an
annotated recipe for each step in preparing your data from the initial, “raw”
state that is output from the laboratory equipment (or collected by hand) to a
state that is useful for answering important research questions. The exact
implementation of each step is given in code that can be re-used and adapted
with new data of a similar format. However, the code script is often not enough
to helpfully understand, share, and collaborate on the process. Instead, it’s
critical to also include descriptions written by humans and for humans. These
annotations can include descriptions of the code and how certain parameters are
standardized the algorithms in the code. They can also be used to justify
choices, and link them up both with characteristics of the data and equipment
for your experiment as well as with scientific principles that underlie the
choices. Protocols like this are critical to allow you to standardize the
process you use across many samples from one experiment, across different
experiments and projects in your research laboratory, and even across different
research laboratories.</p>
<p>As you begin adding text to your pre-processing protocol, you should keep in
mind these general aims. First, a good protocol provides adequate detail that
another researcher can fully reproduce the procedure <span class="citation">(Al-JunDi and SAkkA 2016)</span>. For a
protocol for a trial or wet lab technique, this means that the protocol should
allow another researcher to reproduce the process and get results that are
<em>comparable</em> to your results <span class="citation">(Al-JunDi and SAkkA 2016)</span>; for a data pre-processing
protocol, the protocol must include adequate details that another researcher,
provided they start with the same data, gets <em>identical</em> results (short of any
pre-processing steps that include some element of sampling or random-number
generation, e.g., Monte Carlo methods). This idea—being able to exactly
re-create the computational results from an earlier project—is refered to as
<strong>computational reproducbility</strong> [ref] and is considered a key component in
ensuring that research is fully reproducible [ref].</p>
<blockquote>
<p>“It should provide enough detail (methodology) that can allow another
investigator to do the study and arrive at comparable conclusions.”
<span class="citation">(Al-JunDi and SAkkA 2016)</span></p>
</blockquote>
<p>By creating the data pre-processing protocol as a knitted document (module [x])
using a tool like RMarkdown (module [x]), you can ensure that the protocol is
computationally reproducible. In an RMarkdown document, you include the code
examples as <em>executable</em> code—this means that the code is run every time you
render the document. You are therefore “checking” your code every time that you
run it. As the last step of your pre-processing protocol, you should output the
copy of the pre-processed data that you will use for any further analysis for
the project. You can use functions in R to output this to a plain text format,
for example a comma-separated delimited file (module [x]). Each time you render
the protocol, you will re-write this output file, and so this provides assurance
that the code in your protocol can be used to reproduce your output data (since
that’s how you yourself created that form of the data).</p>
<p>Figure <a href="experimental-data-preprocessing.html#fig:protocoloutput">3.14</a> provides an example from the example protocol
for this module. The RMarkdown file for the protocol includes code to write out
the final, pre-processed data to a comma-separated plain text file called
“processed_cfu_estimates.csv.” This code writes the output file into the same
directory where you’ve saved the RMarkdown file. Each time the RMarkdown file is
rendered to create the pdf version of the protocol, the input data will be
pre-processed from scratch, using the code throughout the protocol, and this
file will be overwritten with the data generated. This guarantees that the code
in the protocol can be used by anyone—you or other researchers—to reproduce
the final data from the protocol, and so guarantees that these data are
computationally reproducible.</p>
<div class="figure"><span id="fig:protocoloutput"></span>
<img src="figures/protocol_output_data.png" alt="Example of using code in pre-processing protocol to output the final, pre-processed data that will be used in further analysis for the research project. This example comes from the example protocol for this module, showing both the executable code included in the RMarkdown file for the protocol (right) and how this code is included in the final pdf of the protocol. Outputting the pre-processed data into a plain text file as the last step of the protocol helps ensure computational reproducibility for this step of working with experimental data." width="\textwidth" />
<p class="caption">
Figure 3.14: Example of using code in pre-processing protocol to output the final, pre-processed data that will be used in further analysis for the research project. This example comes from the example protocol for this module, showing both the executable code included in the RMarkdown file for the protocol (right) and how this code is included in the final pdf of the protocol. Outputting the pre-processed data into a plain text file as the last step of the protocol helps ensure computational reproducibility for this step of working with experimental data.
</p>
</div>
<p>Good protocols include not only <em>how</em> (for data pre-processing protocols, this
is the code), but also <em>why</em> each step is taken. This includes both higher-level
(i.e., what a larger question is being asked) and also at a fine level, for each
step in the process. [Analogy—cookbook that covers deeper principles for why
each step in process is done. <em>America’s Test Kitchen</em> recipes versus
back-of-the-package recipes.] A protocol should include some background, the
aims of the work, hypotheses to be tested, materials and methods, methods of
data collection and equipment to analyze samples <span class="citation">(Al-JunDi and SAkkA 2016)</span> (This
reference is discussing full protocols for a study, e.g., a clinicial trial, so
includes more steps that just pre-processing.)</p>
<p>When you write a protocol within RMarkdown, you can include references
very easily. Figure <a href="experimental-data-preprocessing.html#fig:protocolreferences">3.15</a> shows an example of the
elements you use to do this, showing each element in the example protocol for
this module.</p>
<div class="figure"><span id="fig:protocolreferences"></span>
<img src="figures/protocol_references.png" alt="Including references in a data pre-processing protocol created with RMarkdown. RMarkdown has a built-in referencing system that you can use, based on the BibTeX system for LaTeX. This figure shows examples from the example protocol for this module of the elements used for referencing. You create a BibTeX file with information about each reference, and then use the key for the reference within the text to cite that reference. All cited references will be printed at the end of the document; you can chose the header that you want for this reference section in the RMarkdown file ('References' in this example). In the YAML of the RMarkdown file, you specify the path to the BibTeX file (with the 'bibliography: ' key), so it can be linked in when the RMarkdown file is rendered." width="\textwidth" />
<p class="caption">
Figure 3.15: Including references in a data pre-processing protocol created with RMarkdown. RMarkdown has a built-in referencing system that you can use, based on the BibTeX system for LaTeX. This figure shows examples from the example protocol for this module of the elements used for referencing. You create a BibTeX file with information about each reference, and then use the key for the reference within the text to cite that reference. All cited references will be printed at the end of the document; you can chose the header that you want for this reference section in the RMarkdown file (‘References’ in this example). In the YAML of the RMarkdown file, you specify the path to the BibTeX file (with the ‘bibliography:’ key), so it can be linked in when the RMarkdown file is rendered.
</p>
</div>
<p>The process of writing a protocol forces you to think about each step in the
process, why you do it a certain way (include parameters you choose for
certain functions in a pipeline of code), and include justifications from
the literature for this reasoning. If done well, it should allow you to
quickly and thoroughly write the associated sections of Methods in research
reports and manuscripts and help you answer questions and challenges from
reviewers.</p>
<p>Writing this will also help you identify steps for which you are uncertain
how to proceed and what choices to make in customizing an analysis for your
research data. These are areas where you can search more deeply in the
literature to understand implications of certain choices and, if needed,
contact the researchers who developed and maintained associated software
packages to get advice.</p>
<blockquote>
<p>“Due to the slow growth rate and pathogenicity of mycobacteria, enumeration by
traditional reference methods like colony counting is notoriously
time-consuming, inconvenient and biohazardous.” <span class="citation">(Pathak et al. 2012)</span></p>
</blockquote>
<blockquote>
<p>“Traditionally, quantification of mycobacteria is done by seeding serial
dilutions of bacterial suspensions on suitable media such as Middlebrook 7H10
agar or Lowenstein Jensen followed by counting colony-forming units (CFU).
However, this method is hampered by the long generation time and the tendency of
mycobacteria to aggregate, resulting in multiple founders of a single colony and
an underestimation of the correct number of bacteria. Typically, the time
required for visible colonies to appear on 7H10 agar is 2–3 weeks for M.
tuberculosis and M. a. avium, while it takes about 4–8 weeks for M. a.
paratuberculosis. In addition, plating enough dilutions to make sure the results
can be reliably counted is a tedious task that gives piles of plates with
biohazardous bacteria. A further disadvantage of the colony counting method is
that it cannot be reliably conducted on frozen samples, which may be both more
practical and desirable in several research settings.” <span class="citation">(Pathak et al. 2012)</span></p>
</blockquote>
<blockquote>
<p>“Mycobacterium tuberculosis culture, a critical technique for routine diagnosis of tuberculosis, takes more than two weeks.” <span class="citation">(Ghodbane, Raoult, and Drancourt 2014)</span></p>
</blockquote>
<blockquote>
<p>“A major problem when dealing with tuberculosis has been a difficulty in
diagnosis due to slow growth of mycobacterial cultures, which subsequently
explains the slow process of evaluating the susceptibility of this microorganism
to antibiotics. Using current tools, a primary culture is obtained in two to
four weeks on average and antibiotic susceptibility is determined after an
additional two to four weeks. Therefore, four to eight weeks are needed to
obtain an isolate and determine its susceptibility to antibiotics.”
<span class="citation">(Ghodbane, Raoult, and Drancourt 2014)</span></p>
</blockquote>
<blockquote>
<p>“Quantification of viable bacteria is a crucial foundation for many types of
research. This seemingly simple task can be challenging, expensive, and
imprecise for Mycobacterium paratuberculosis, a slowly growing organism (&gt;24-h
generation time) with a strong tendency to form large clumps. Studies of
environmental survival, resistance to pasteurization or disinfectants, and
quantification of the pathogen in milk and feces from infected animals are just
a few examples that require precise and sensitive quantification of viable M.
paratuberculosis cells.” <span class="citation">(Shin et al. 2007)</span></p>
</blockquote>
<blockquote>
<p>“The main challenge in serial dilution experiments is the estimation of the
undiluted microorganisms counts <span class="math inline">\(n_0\)</span> from the measured <span class="math inline">\(\hat{n_j}\)</span>. There are
two competing processes (Tomasiewicz et al., 1980) that affect the accuracy of
the estimation: sampling errors and counting errors. Sampling errors are caused
by the statistical fluctuations of the population. For example, when sampling an
average of 100 colonies, the fluctuations in the number of the population are
expected to be <span class="math inline">\(\pm \sqrt{100}\)</span> when the sampling process is governed by a
Poisson probability (Poisson and Binomial distributions are often used in
statistical analysis to describe the dilution process (Hedges, 2002, Myers et
al., 1994)) where the standard deviation equals square-root of the mean; the
relative error (ratio of the standard deviation to the mean) is <span class="math inline">\(\sqrt{100} / 100 = 0.1\)</span>. Thus, the larger the sample size is, the smaller the relative
sampling error; hence, one would like to use a dilution plate with the largest
number (i.e., the least diluted sample, <span class="math inline">\(j \rightarrow 1\)</span>). However, as the
number of colonies increases, counting error is introduced due to the high
probability of two (or more) colonies to merge (due to overcrowding) and become
indistinguishable, and be erroneously counted as one colony. An optimum (a
‘sweet spot’) between these two processes (sampling and counting error) needs to
be found for using the optimal dilution (i.e., the optimal jth plate) with
which to estimate <span class="math inline">\(n_0\)</span>. Cells can grow into colonies in various ways. Wilson
(1922) states that when two cells are placed very close together only one cell
will develop, and when two cells are situated at a distance from each other both
cells may grow and then fuse into one colony. Either way, the end result is the
appearance of one colony which causes counting error.” <span class="citation">(Ben-David and Davidson 2014)</span></p>
</blockquote>
<p>For open-source software, resources to consult extend beyond the traditional
one. You can often find information on open-source software, including the
algorithms and principles that underlie the software, through peer-reviewed
publications in the scientific literature. However, you can also find
details—and often more thorough details—in <em>vignettes</em> that are published in
conjunction with the package. [More on vignettes and where to find them.]
Further, software is often presented at conferences and workshops, including the
yearly BioC [?] conference for the Bioconductor community. These talks and
workshops are sometimes available after the event as online recordings [some
examples of these]. You can consult books, as well, although these often
quickly become outdated for software that is rapidly evolving; an exception
is online books, which are becoming very popular to create through R’s
<code>bookdown</code> package and can be rapidly updates. Many of these books are
available through [bookdown’s gallery page].</p>
<p>An added advantage of data pre-processing protocols, created with knitted
documents, is that you can include steps and code for data quality
assessment. [More on exploratory data analysis] [Examples of how
exploratory data analysis could help when pre-processing biomedical
data—for example, identify outliers that might indicate equipment
was malfunctioning?]</p>
<p>Be sure that, in each step of your pre-processing protocol, you explain
<em>why</em> you are taking a certain step. For example, if you have a step
that aligns peaks across samples in metabolomics data [?], be sure to
explain that … [why it’s important to do this]. Include references to
the literature that justify this explanation; these references also
serve as further literature that someone else could read to understand
the step. This <em>why</em> is important even if the step is obvious to you—it
will allow you to pass along the task to others in your research group
without the new person needing to blindly trust each step.</p>
<p>It is particularly important to clearly explain what you are doing in
each step and how you are implementing your choices through code. Yes,
the code itself allows someone else to replicate what you did. However,
only those who are very, very familiar with the software program, including
any of the extension packages you include, can “read” the code directly
to understand what it’s doing. Further, even if you understand the code
very well when you create it, it is unlikely that you will stay at that
same level of comprehension in the future, as other tasks and challenges
take over that brain space. Explaining for humans, in text that augments
and accompanies the code, is also important because function names and
parameter names in code often are not easy to decipher. While excellent
programmers can sometimes create functions with clear and transparent
names, easy to translate to determine the task each is doing, this is
difficult in software development and is rare in practice. Human annotations,
written by and for humans, are critical to ensure that the steps will
be clear to you and others in the future when you revisit what was
done with this data and what you plan to do with future data.</p>
<p>You can begin to create this pre-processing protocol before you collect
any of your own research data. Example datasets exist online, and you
can often find an example that aligns with the type of data you will
be collecting and the format you will be collecting it in. [Places where
you could get this data—repositories, R package example datasets.]
[Characteristics that are important in your example dataset—same file
format that you will get from your equipment, key characteristics, like
number of experimental groups, like treatment categories and single vs
multiple time points]</p>
<p>If the format of the initial data is similar to the format you anticipate for
your data, you can create the code and explanations for key steps in your
pre-processing for that type of data. Often, you will be able to adapt the
RMarkdown document to change it from inputting the example data to inputting
your own experimental data with minimal complications, once your data
comes in. By thinking through and researching data pre-processing options
before the data is collected, you can save time in analyzing and presenting
your project results once you’ve completed the experimental data
collection for the project.</p>
<p>Further, with an example dataset, you can get a good approximation of the
format in which you will output data from the pre-processing steps.
This will allow you to begin planning the analysis and visualization
that you will use to combine the different types of data from your
experiment and use it to investigate important research hypotheses.
Again, if data follow standardized formats across steps in your process,
it will often be easy to adapt the code in the protocol to input the new
dataset that you created, without major changes to the code developed with
the example dataset.</p>
<p>For each step of the protocol, you can also include potential problems
that might come up in specific instances of the data you get from
future experiments. This can help you adapt the code in the protocol in
thoughtful ways as you apply it in the future to new data collected
for new studies and projects.</p>
<blockquote>
<p>“Given an unknown sample which contains <span class="math inline">\(n_0\)</span> colony forming units (CFUs), a
series of <span class="math inline">\(J\)</span> dilutions are made sequentially each with a dilution factor
<span class="math inline">\(\alpha\)</span>. From each of the J dilutions a fraction <span class="math inline">\(\alpha_p^{-1}\)</span> is taken and
spread (plated) on an agar plate (assay) where colonies are counted. Thus, in
general there are two dilution factors: <span class="math inline">\(\alpha\)</span> and <span class="math inline">\(\alpha_p\)</span>. For example,
<span class="math inline">\(\alpha = 10\)</span> indicates a 10-fold dilution, e.g., by diluting successively 0.1
ml of sample into 0.9 ml of media; and <span class="math inline">\(\alpha_p = 1\)</span> means that the entire
volume is spread (plated) on the agar plate. For an experiment with a larger
dilution factor <span class="math inline">\(\alpha_p\)</span>, multiple plates may be spread at the same dilution
stage. For example, <span class="math inline">\(\alpha_p = 20\)</span> represent a 5% plating of the dilution, and
thus up to 20 replicates could be created. At each dilution the true number of
colonies is <span class="math inline">\(n_j = n_0 \alpha^{-j} \alpha_p^{-1}\)</span> and the estimated number is
<span class="math inline">\(\hat{n_j}\)</span>. The estimated quantities are denoted with a ‘hat’ (estimated
quantities can be measured quantities, or quantities that are derived from
measured or sampled quantities); symbols without a ‘hat’ denote true quantities
(also known as population values in statistics) that do not contain any sampling
or measurement error. In this work both <span class="math inline">\(n_j\)</span> and <span class="math inline">\(n_0\)</span> are ‘counts,’ i.e.,
number of colonies. Knowing the aliquot volume, one can easily convert counts to
concentration (for example CFU/ml).” <span class="citation">(Ben-David and Davidson 2014)</span></p>
</blockquote>
<p>In your data pre-processing protocol, show the code that you use to implement
this choice and also explain clearly in the text why you made this choice and
what alternatives should be considered if data characteristics are different.
Write this as if you are explaining to a new research group member (or your
future self) how to think about this step in the pre-processing, why you’re
doing it the way your doing it, and what code is used to do it that way. You
should also include references that justify choices when they are
available—include these using BibTex. By doing this, you will make it much
easier on yourself when you write the Methods section of papers that report on
the data you have pre-processed, as you’ll already have draft information on
your pre-processing methods in your protocol.</p>
</div>
<div id="practice-quiz-4" class="section level3" number="3.11.4">
<h3><span class="header-section-number">3.11.4</span> Practice quiz</h3>
<!-- > "Quantitative estimation of the number of viable microorganisms in -->
<!-- bacteriological samples has been a mainstay of the microbiological laboratory -->
<!-- for more than one-hundred years, since Koch first described the technique (Koch, -->
<!-- 1883). Serial dilution techniques are routinely used in hospitals, public -->
<!-- health, virology, immunology, microbiology, pharmaceutical industry, and food -->
<!-- protection (American Public Health Association, 2005, Hollinger, 1993, Taswell, -->
<!-- 1984, Lin and Stephenson, 1998) for microorganisms that can grow on -->
<!-- bacteriological media and develop into colonies." [@ben2014estimation] -->
<!-- > "The objective of the serial dilution method is to estimate the concentration -->
<!-- (number of colonies, organisms, bacteria, or viruses) of an unknown sample by -->
<!-- counting the number of colonies cultured from serial dilutions of the sample, -->
<!-- and then back track the measured counts to the unknown concentration." -->
<!-- [@ben2014estimation] -->
<!-- > "The most widely used method for determining drug efficacy [for TB] in mice  -->
<!-- remains enumeration of the bacterial load in lungs and spleens, by counting -->
<!-- the colony forming units (CFU) of the organ homogenates on agar plates." -->
<!-- [@franzblau2012comprehensive] -->
<!-- > "Quantitative estimation of the number of viable microorganisms in -->
<!-- bacteriological samples has been a mainstay of the microbiological laboratory -->
<!-- for more than one-hundred years, since Koch first described the technique (Koch, -->
<!-- 1883). Serial dilution techniques are routinely used in hospitals, public -->
<!-- health, virology, immunology, microbiology, pharmaceutical industry, and food -->
<!-- protection (American Public Health Association, 2005, Hollinger, 1993, Taswell, -->
<!-- 1984, Lin and Stephenson, 1998) for microorganisms that can grow on -->
<!-- bacteriological media and develop into colonies." [@ben2014estimation] -->
<!-- > "The objective of the serial dilution method is to estimate the concentration -->
<!-- (number of colonies, organisms, bacteria, or viruses) of an unknown sample by -->
<!-- counting the number of colonies cultured from serial dilutions of the sample, -->
<!-- and then back track the measured counts to the unknown concentration." -->
<!-- [@ben2014estimation] -->
<!-- > "The indirect or viable count has, as a rule, been performed by a  -->
<!-- modification of Koch's original plating method. ... The modifications of  -->
<!-- Koch's method have been concerned with the medium used, the question of  -->
<!-- preliminary dilution, the methods of dilution and the exact technique of  -->
<!-- counting the plates. The majority of observers appear to have used agar,  -->
<!-- but [a few] seem to have preferred gelatin, though in some cases both  -->
<!-- media were employed. With regard to preliminary dilution, the earlier -->
<!-- workers generally preferred to plate out the original emulsion, while -->
<!-- of late the tendency has been in the opposite direction ... . The  -->
<!-- method of dilution has been subject to considerable variation; on the  -->
<!-- whole volumetric pipettes have been the most popular, but [some]  -->
<!-- used dropping pipettes, while [others] resorted ot the use of a standard -->
<!-- platinum loop. ... The important question of the counting of the plates -->
<!-- has naturally depended largely on whether or no a preliminary dilution  -->
<!-- of the emulsion was made. Where the number of colonies was very great,  -->
<!-- microscopic counting was adopted ... . Where on the contrary, dilution was -->
<!-- employed, the use of the microscope was no longer necessary, and counting  -->
<!-- was performed with the naked eye or with a magnifying glass ..."  -->
<!-- [@wilson1922proportion] -->
<!-- > "In perusing the results of previous workers, it was striking to  -->
<!-- observe the peculiar lack of attention which was paid to the estimation -->
<!-- of the experimental error involved in the methods employed. Probably this  -->
<!-- is to be attributed to the fact that in many cases in which the enumeration -->
<!-- of bacteria was undertaken, a relative, rather than an absolute accuracy -->
<!-- was essential. It was felt that the successful accomplishment of this -->
<!-- object could only be obtained by working out a technique in which the  -->
<!-- errors inherent in every step should be known with certainty."  -->
<!-- [@wilson1922proportion] -->
<!-- > "The plate count method for estimating bacterial populations is  -->
<!-- satisfactory for many comparative purposes if *relative* rather than  -->
<!-- *absolute* numbers of cells are wanted, although in some cases, because -->
<!-- of clumping, plate counts may not bear a constant relation to total  -->
<!-- counts even during the logarithmic growth phase (Jennison, 1937).  -->
<!-- This lack of agreement may be overcome, at least with some organisms,  -->
<!-- by proper shaking to break up clumps of cells (Ziegler and Halvorson,  -->
<!-- 1935)." [@jennison1940evaluation] -->
<!-- > "Generally speaking, the methods which have been employed may be classified -->
<!-- into (1) the direct and (2) the indirect. In the former the organisms are -->
<!-- counted directly under the microscope, in the latter the number of bacteria -->
<!-- present is calculated from an enumeration of the colonies which develop when  -->
<!-- an aliquot part of the emulsion in question is mixed with a nutrient medium -->
<!-- in a Petri dish, and incubated for a variable period of time. The former  -->
<!-- is designed to record the total number of organisms present, the latter only -->
<!-- the number which happens to be viable at the moment of sampling."  -->
<!-- [@wilson1922proportion] -->
<!-- > "Mtb is a member of the slow-growing pathogenic mycobacterial species, -->
<!-- characterized by a 12- to 24-hour division rate and prolonged culture period on -->
<!-- agar of up to 21 days. Why Mtb grows so slowly is not well understood. Proposed -->
<!-- mechanisms include limitation of nutrient uptake through the highly impermeable -->
<!-- cell wall and slow rates of RNA synthesis (96). During experimental infections, -->
<!-- its metabolism can shift from an aerobic, carbohydrate-metabolizing mode to one -->
<!-- that is microaerophilic and lipid metabolizing (25). Mycobacteria are -->
<!-- facultative intracellular bacteria that multiply within phagocytic cells, -->
<!-- particularly macrophages and monocytes. Although many mycobacterial species are -->
<!-- environmental, Mtb is strictly parasitic." [@sakamoto2012pathology] -->
<!-- > "For the testing of single compounds as well as short term mouse experiments,  -->
<!-- some investigators mentioned plating whole lungs while others plate the  -->
<!-- homogenate of a single lung lobe from each animal. After long treatment  -->
<!-- regimens when low bacterial numbers are expected, generally the whole lung is -->
<!-- homogenized and a sizeable fraction of up to one-half of the homogenate is plated.  -->
<!-- One group homogenizes whole lungs in a total of 2.5 mL PBS and then plates -->
<!-- the entire homogenate on five 7H11 plates (0.5 mL per plate). Another group -->
<!-- uses whole lungs and total spleen homogenized in 4 mL PBS supplemented with  -->
<!-- 0.05% Triton X-100." [@franzblau2012comprehensive] -->
<!-- > "Several issues regarding the endpoints and the mouse models themselves -->
<!-- were discussed. Regarding accurate endpoints in treatment trials, the  -->
<!-- method used for enumerating bacteria from organ homogenates should be  -->
<!-- carefully (re)considered given the possibility that some non-culturable  -->
<!-- bacteria in samples of animal tissues may not form colonies on solid agar.  -->
<!-- Liquid media, such as BACTEC or MGIT, was suggested as a method to enhance -->
<!-- sensitivity for finding low numbers of CFU and to gain some insight into  -->
<!-- the state of the bacillus at the time point examined. Liquid testing using  -->
<!-- the MGIT system is also now the method of choice for sputum evaluation to assess  -->
<!-- culture conversion for clinical trials. Several automated liquid culture -->
<!-- systems have shown greater sensitivity than the traditional solid-media -->
<!-- cultures the acknowledged increased in the mycobacterial recovery rate of  -->
<!-- liquid media, which is likely due to a more mycobacterial populations  -->
<!-- being able to recover in liquid culture than on solid media. This is -->
<!-- supported by in vivo and in vitro observations that subpopulations of  -->
<!-- M. tuberculosis with different states of metabolic activity co-exist in  -->
<!-- old liquid cultures, as well as in liquid cultures with bacterial growth  -->
<!-- from chronic infected mice that do not grow on solid media." -->
<!-- [@franzblau2012comprehensive] -->
<!-- > "Another topic of discussion in at least two laboratories was the issue -->
<!-- of drug carry-over at the time of enumeration of bacterial colonies  -->
<!-- in the organs. CLF and TMC-207 both have long half-lives, high tissue -->
<!-- distribution and tissue binding, and therefore drug might still be present -->
<!-- at the time of sacrificing the animals. The first indication that  -->
<!-- drug carry-over is an issue is observed when dilutions of organ  -->
<!-- homogenates do not have the expected reduction in bacterial number.  -->
<!-- Another indication might also be the lack of correlation between the bacterial -->
<!-- number and the gross pathology observation. Several methods have been described -->
<!-- to reduce the carry-over of drug in agar plates, including using LJ medium -->
<!-- or 7H11 with 5% bovine serum albumin (BSA) for TMC-207, and using 0.4% -->
<!-- activated charcoal for CLF. Relapse studies will then show the true -->
<!-- sterilizing potential of these drugs and drug combinations." [@franzblau2012comprehensive] -->
<!-- > "It was pointed out that colony forming units (CFU) measures by plating -->
<!-- organ homogenates on solid agar are often inaccurate since they do not -->
<!-- necessarity include just single bacilli but rather small clumps or -->
<!-- conglomerations. There was discussion around the hypothesis that these -->
<!-- clusters represent a type of biofilm. Hatfull et al. found these -->
<!-- M. tuberculosis biofilms in vitro to consist of bacteria surrounded by  -->
<!-- a layer of free mycolic acids. Although the biofilm hypothesis was not -->
<!-- accepted by everyone, it was agreed that the extracellular bacteria in these -->
<!-- micro-environments are unique and should be treated by drug treatment.  -->
<!-- Another topic of discussion was about the drug refractory nature of these -->
<!-- persisting bacilli being a drug penetration issue, by drugs not getting through -->
<!-- this fibrous rim and extracellular lipid matrix. The ability to determine -->
<!-- the penetration of TB drugs into a granuloma was seen as an important -->
<!-- gap in the current knowledge about TB drugs." [@franzblau2012comprehensive] -->
<!-- > "The activity of an investigational drug or regimen has been in recent years -->
<!-- mostly determined by the reduction of colony forming units (CFU) of M. -->
<!-- tuberculosis by dilution of organ homogenates on solid agar 7H11 plates. Plating -->
<!-- of the organ homogenates has always been the gold standard for quantifying drug -->
<!-- efficacy in vivo. However, drug discovery efforts are often times held up by -->
<!-- this time consuming step requiring an incubation period of the bacterial plates -->
<!-- of 3--4 weeks. For early drug discovery efforts, more efforts are being -->
<!-- investigated lately for indirect, but more rapid, methods to 'measure' the -->
<!-- bacterial load: such as the luciferase readout or fluorescence. These novel -->
<!-- detection methods will undoubtedly accelerate the TB drug discovery process by -->
<!-- delivering an immediate readout on the efficacy of an experimental compound at -->
<!-- either time of sacrifice (for luciferase readout) and even in live animals in -->
<!-- real time (fluorescence). Thorough validation, however, is not available as of -->
<!-- yet and would be required before these indirect methods can replace the -->
<!-- enumeration by CFU for all drug classes." [@franzblau2012comprehensive] -->
<!-- > "Another discussion took place at several occasions during this project, -->
<!-- pointing out that perhaps not all bacteria can be identified, cultured or -->
<!-- visualized by current methods. The discussion started with the recent -->
<!-- introduction of liquid culture media for the diagnostic evaluation of clinical -->
<!-- specimens for suspected tuberculosis. The question then became whether drugs or -->
<!-- regimens which preferentially kill certain bacillary populations would give rise -->
<!-- to a 'flawed' readout by only culturing certain subsets of bacteria on solid -->
<!-- agar. If liquid cultures of organ homogenates allow the growth of TB -->
<!-- subpopulations that will not grow on solid media, the parallel evaluation in -->
<!-- liquid and in solid cultures might offer an opportunity to study drug effects on -->
<!-- subpopulations in different metabolic states. ... This question might even be -->
<!-- more relevant in animal models with a greater variety in granulomatous lesion -->
<!-- types where bacilli are located in different environments (such as in necrotic, -->
<!-- closed or cavitary lesions) and hence might have different metabolic stages." -->
<!-- [@franzblau2012comprehensive] -->
<!-- > "The importance of serial dilution and colony counting is reflected by the -->
<!-- number of standard operating procedures and regulatory guidelines describing -->
<!-- this methodology. In all of these guidelines the optimal number ($\hat{n_j}$) of -->
<!-- colonies to be counted has been reported (Park and Williams, 1905, Wilson, 1922, -->
<!-- Jennison and Wadsworth, 1940, Tomasiewicz et al., 1980, FDA, 2001, Goldman and -->
<!-- Green, 2008) as 40--400, 200--400, 100--400, 25--250, 30--300. It is interesting -->
<!-- to note that these references do not specify the area in which the colonies -->
<!-- grow, nor the diameter of the particular organism assayed. The result is that -->
<!-- titration and counting colonies is done within a range that may be inadequate, -->
<!-- and may introduce considerable error." [@ben2014estimation] -->
<!-- > "The plate count method is based on viable cell counts. The plate count  -->
<!-- method is performed by diluting the original sample in serial dilution  -->
<!-- tubes, followed by the plating of aliquots of the prepared serial dilutions -->
<!-- into appropriate plate count agar plates by the pour plate or spread  -->
<!-- plate technique. The pour plate technique utilizes tempered molton plate -->
<!-- count agar poured into the respective plate and mixed with the diluted -->
<!-- aliquot sample in the plate, whereas the spread plate technique utilizes -->
<!-- the addition and spreading of the diluted aliquot sample on the surface -->
<!-- of the preformed solid plate count agar in the respective plant. ... -->
<!-- These prepared plate count agar plates are then optimally incubated, and the -->
<!-- colonies observed on these plate count agar plates are then counted as  -->
<!-- the number of CFUs. The counting of CFUs assumes that every colony is  -->
<!-- separate and founded by a single viable microbial cell. The total  -->
<!-- colony counts obtained in CFUs from the incubated agar plates and the  -->
<!-- respective dilution factor used can then be combined to calculate the  -->
<!-- original number of microorganisms in the sample in CFUs per mL. The typical  -->
<!-- counting ranges are 20--250 CFUs or 30--300 CFUs per standard plate count -->
<!-- agar plate. Additional considerations for counting CFUs are counting of  -->
<!-- plate spreaders, too numerous to count (TNTC) reporting and statistics,  -->
<!-- rounding and averaging of observed plate counts, limit of detection, and  -->
<!-- limit of quantification of plate countes (77--79). There are also  -->
<!-- optimal condition assumptions for the plate count method as changes to the  -->
<!-- plate count agar nutrient level or temperature can affect the surface growth of  -->
<!-- bacteria (80, 81). Primary equipment and materials used for this method  -->
<!-- are serial dilution tubes (bottles); Petri plates or dishes; pipettes;  -->
<!-- specific growth medium, diluents, and reagents; incubator and water bath -->
<!-- with appropriate optimal temperature setting; commercial colony counter -->
<!-- (manual or automate); and plate spreader or rod. Total bacteria and fungi -->
<!-- can be enumerated separately using the plate count method based on the  -->
<!-- type of culture medium utilized (82--86). Specific or selective culture -->
<!-- medium can also be used in place of the standard plate count agar media -->
<!-- for more specific microbial enumeration (87). Sources of error using this  -->
<!-- method are improper or inadequate preparation of the test samples,  -->
<!-- serial dilution error, suboptimal incubation conditions, undercounting -->
<!-- due to cell aggregation or clumping, and analyst error in the colony -->
<!-- counting or calculation of observed results." [@goldman2015practical] -->
<!-- > "Almost every textbook, laboratory manual and methods volume in  -->
<!-- microbiology contains the statement that plates for counting bacteria should -->
<!-- contain, when possible 30--300 colonies. The '30--300' concept has been  -->
<!-- so ingrained in our thinking that the limits are rarely questioned. ... -->
<!-- The '30--300' concept originated with two publications by Breed and  -->
<!-- Dotterrer (7,8); the text and tables presented in both publications are  -->
<!-- identical. The authors summarized results of a few early studies and  -->
<!-- then proceeded to more clearly define the problem and provide a solution." -->
<!-- [@tomasiewicz1980most] -->
<!-- > "Considering the plating method per se, the total error of the mean plate -->
<!-- count of a given dilution of cells is chiefly made up of two rather  -->
<!-- distinct sources of deviations: (a) the distribution or sampling error,  -->
<!-- sometimes inaccurately called the counting error, (i.e., variation in  -->
<!-- number of colonies, due to sampling, between replicate plates of the  -->
<!-- given dilution), and (b), the dilution error, (i.e., the errors of  -->
<!-- pipetting involved in reaching the given dilution). ... It is customary  -->
<!-- to measure the reliability of the plate count by calculating only the -->
<!-- distribution error, and assuming that the dilution error is small, constant,  -->
<!-- and unimportant. We shall show, however, that at best this dilution error  -->
<!-- is of about the same order of magnitude as the distribution error, and  -->
<!-- is, therefore, equally deserving of consideration in arriving at the total -->
<!-- error of plate counts. Furthermore, the dilution error increases with higher -->
<!-- dilutions, whereas the distribution error does not. Obviously, one must take into -->
<!-- account both sources of variation in evaluating the total error, as, for example, -->
<!-- in a problem involving significance of differences, in which the same -->
<!-- dilution might be employed." [@jennison1940evaluation] -->
<!-- > "Traditionally, quantification of mycobacteria is done by seeding serial -->
<!-- dilutions of bacterial suspensions on suitable media such as Middlebrook 7H10 -->
<!-- agar or Lowenstein Jensen followed by counting colony-forming units (CFU). -->
<!-- However, this method is hampered by the long generation time and the tendency of -->
<!-- mycobacteria to aggregate, resulting in multiple founders of a single colony and -->
<!-- an underestimation of the correct number of bacteria. Typically, the time -->
<!-- required for visible colonies to appear on 7H10 agar is 2--3 weeks for M. -->
<!-- tuberculosis and M. a. avium, while it takes about 4--8 weeks for M. a. -->
<!-- paratuberculosis. In addition, plating enough dilutions to make sure the results -->
<!-- can be reliably counted is a tedious task that gives piles of plates with -->
<!-- biohazardous bacteria. A further disadvantage of the colony counting method is -->
<!-- that it cannot be reliably conducted on frozen samples, which may be both more -->
<!-- practical and desirable in several research settings." [@pathak2012counting] -->
<!-- > "It is clear that if three tubes are put up from an emulsion containing a -->
<!-- comparatively small number of bacilli, the chances of obtaining a representative -->
<!-- sample must be smaller than if an emulsion be employed which contains a much -->
<!-- larger number of bacilli. Similarly with the tubes themselves. If only a few -->
<!-- bacilli are introduced, the chances of obtaining a correct idea of the exact -->
<!-- number are smaller than if a large number of bacilli are introduced. Thus the -->
<!-- greater the number of colonies per tube, the less is the error of sampling. That -->
<!-- this is not a mere theoretical consideration is shown from an examination of the -->
<!-- data accumulated during the progress of this work... the percentage deviation of -->
<!-- each individual tube from the arithmetic mean was considerably greater in the -->
<!-- case when a small number of colonies developed than in the case when a large -->
<!-- number of bacilli were inoculated. In other words the sampling error in the -->
<!-- former instance was large, in the latter comparatively small." -->
<!-- [@ben2014estimation] -->
<!-- > "We now come to consider the second factor determining the optimum number -->
<!-- of bacilli to be inoculated in putting up viable counts by the tube method,  -->
<!-- namely the error of overcrowding. More or less in proporation as the error  -->
<!-- of sampling decreases as the number of developing colonies increases, so the -->
<!-- error of overcrowding increases as the number of developing colonies increases. -->
<!-- The two vary in opposite directions; the greater the number of colonies the  -->
<!-- less the sampling error; the fewer the number of colonies, the less is the  -->
<!-- overcrowding error. A point must be chosen between the two which will permit -->
<!-- of the minimum combined error experienced. Before this could be done, however,  -->
<!-- it was necessary to ascertain the actual effect of overcrowding on the  -->
<!-- development of colonies in tube preparations. As mentioned above, this  -->
<!-- overcrowding error is one which seems to have been neglected by the  -->
<!-- majority of observers, or at any rate, not clearly recognized. It is obvious -->
<!-- that the greater the number of bacilli distributed in a given space, the less -->
<!-- is the interval between each of them, and the greater the chance of two  -->
<!-- being coincident. In every case in which two bacilli are coincident or are -->
<!-- placed very close together only one colony will develop. Further, when two  -->
<!-- bacilli are situated at such a distance from each other that each is able  -->
<!-- to develop, yet at such a distance that continued development of both  -->
<!-- will result in fusion, it is clear that a simple colony must arise. Whether -->
<!-- one continues to grow and the other desists or whether both develop, the  -->
<!-- result is the same---namely, the appearance of one colony in place of two  -->
<!-- bacilli. On pure a priori grounds one would expect this overcrowding factor  -->
<!-- to be of considerable importance in determining the number of colonies which -->
<!-- will develop in a given space. One would expect it to play but a small part -->
<!-- so long as comparatively few bacilli were inoculated, but as the number -->
<!-- of the latter increased so should the percentage which fails to develop into -->
<!-- colonies become greater." [@ben2014estimation] -->
<!-- > "So far as evaluating the distribution error is concerned, this is usually -->
<!-- done by calculating the standard deviation of the mean (standard error)  -->
<!-- of the replicate plates, assuming that the variation between such plates  -->
<!-- is that of random samples. Under good experimental conditions, this  -->
<!-- coefficient of variation will average $\pm$ 4--5 per cent (Jennison, 1937). -->
<!-- In order to test whether observed variations between replicate plates  -->
<!-- are due to chance or to technique, the $\chi^2$ ('chi square') test may  -->
<!-- be used (Wilson and Kullmann, 1931). The calculated value of $\chi^2$ will  -->
<!-- be distributed in a known manner if the replicate samples are from a  -->
<!-- Poisson series, that is, if their variation is that of random samples from  -->
<!-- the same population. Fisher, Thornton, and MacKenize (1922), and Fisher (1938), -->
<!-- have shown that a Poisson distribution is obtained in parallel plate counts -->
<!-- made under standardized experimental conditions. Both the $\chi^2$ test -->
<!-- and colculation of the standard error of replicate plates apply only to  -->
<!-- a given dilution; they do not account for errors involved in arriving at that -->
<!-- dilution." [@jennison1940evaluation] -->
<!-- > "Plates with over 500 colonies under-estimate the true count owing to the -->
<!-- overcrowding error. With careful workers the actual error of counting probably -->
<!-- does not become appreciable till there are about 300 colonies per plate, and for -->
<!-- some distance above this limit it will probably be counterbalanced by the -->
<!-- diminished sampling error. If many places, however, have to be counted, the -->
<!-- fatigue error, which seems to be mainly responsible for the failure of the -->
<!-- sampling error to decrease with increasing numbers of colonies in accordance -->
<!-- with theoretical expectations, becomes appreciable." -->
<!-- [@wilson1935bacteriological] -->
<!-- > "For automated equipment (10), the optiumum counting range may well vary with  -->
<!-- the instrument, particle (colony) size limits, range of colony sizes, etc. -->
<!-- Furthermore, even if automation is not used, appropriate numbers of colonies -->
<!-- that should be on a countable plate can vary widely, depending on many other -->
<!-- variables. With soil fungi, for example, maxima of from 25-100 colonies per  -->
<!-- plate have been suggested (17). Coliform analyses demand another range (24)." -->
<!-- [@tomasiewicz1980most] -->
<!-- > "Given an unknown sample which contains $n_0$ colony forming units (CFUs), a -->
<!-- series of $J$ dilutions are made sequentially each with a dilution factor -->
<!-- $\alpha$. From each of the J dilutions a fraction $\alpha_p^{-1}$ is taken and -->
<!-- spread (plated) on an agar plate (assay) where colonies are counted. Thus, in -->
<!-- general there are two dilution factors: $\alpha$ and $\alpha_p$. For example, -->
<!-- $\alpha = 10$ indicates a 10-fold dilution, e.g., by diluting successively 0.1 -->
<!-- ml of sample into 0.9 ml of media; and $\alpha_p = 1$ means that the entire -->
<!-- volume is spread (plated) on the agar plate. For an experiment with a larger -->
<!-- dilution factor $\alpha_p$, multiple plates may be spread at the same dilution -->
<!-- stage. For example, $\alpha_p = 20$ represent a 5% plating of the dilution, and -->
<!-- thus up to 20 replicates could be created. At each dilution the true number of -->
<!-- colonies is $n_j = n_0 \alpha^{-j} \alpha_p^{-1}$ and the estimated number is -->
<!-- $\hat{n_j}$. The estimated quantities are denoted with a 'hat' (estimated -->
<!-- quantities can be measured quantities, or quantities that are derived from -->
<!-- measured or sampled quantities); symbols without a 'hat' denote true quantities -->
<!-- (also known as population values in statistics) that do not contain any sampling -->
<!-- or measurement error. In this work both $n_j$ and $n_0$ are 'counts', i.e., -->
<!-- number of colonies. Knowing the aliquot volume, one can easily convert counts to -->
<!-- concentration (for example CFU/ml)." [@ben2014estimation] -->
<!-- > "All respondents [from labs working on tuberculosis] use statistical  -->
<!-- methods in conjunction with their animal trials and most do not consult with -->
<!-- a statistician. Treatment effects are mostly analyzed by a one-way ANOVA,  -->
<!-- followed by a T-test, Tukey or Dunnett's test. For relapse studies a  -->
<!-- Fisher Exact or Chi-square test is applied to compare relapse proportions -->
<!-- between groups." [@franzblau2012comprehensive] -->
<!-- > "Statistical methods are of key importance and different methods are used -->
<!-- depending on the question asked. Methods used were found to be very similar -->
<!-- across laboratories, and often assistance of a statistician was provided to  -->
<!-- answer new questions. Power analysis prior to the experiments is required -->
<!-- to determine the number of animals used in the experiment. The primary data -->
<!-- analysis for mouse models is usually a one-way analysis of variance (ANOVA) -->
<!-- of the log-10 CF bacterial loads, with t-distribution based contrasts  -->
<!-- comparing individual treatments. A Dunnett's test is an appropriate method -->
<!-- for comparing all new compounds to untreated controls, while controlling the -->
<!-- overall error rate in that set of ccomparisons is set at the usual 0.05 -->
<!-- level. Also Bonferroni and Tukey statistical tests are appropriate tests for -->
<!-- a pairwise comparison between treatment groups. Evidence of differential  -->
<!-- relapse based on detection (yes/no) of TB bacteria would be established using -->
<!-- a Fisher's exact test comparing rates of relapse between two experimental -->
<!-- groups. In experiments with n = 5--10 per group, power to identify  -->
<!-- significantly different relapse rates is very low. Therefore in order to  -->
<!-- increase the statistical power, more treatment groups are implemented -->
<!-- with various lengths of treatment, or a higher relapse rate is aimed -->
<!-- for in order to see significant differences between different treatment  -->
<!-- regimens. Relapse information obtained from mouse models should always be -->
<!-- interpreted with great care and seen as trends in relapse differences between -->
<!-- treatments." [@franzblau2012comprehensive] -->
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="experimental-data-recording.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/12-scripted_preprocessing.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["improve_repro.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
