<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Module 13 Selecting software options for pre-processing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>
<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, Gregory Robertson">
<meta name="generator" content="bookdown 0.40 with bs4_book()">
<meta property="og:title" content="Module 13 Selecting software options for pre-processing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing">
<meta property="og:type" content="book">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Module 13 Selecting software options for pre-processing | Improving the Reproducibility of Experimental Data Recording and Pre-Processing">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.7.0/transition.js"></script><script src="libs/bs3compat-0.7.0/tabs.js"></script><script src="libs/bs3compat-0.7.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
          margin-bottom: 0em;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<meta name="description" content="The previous module described some common themes and processes in pre-processing biomedical data. While we’ve covered some key processes of pre-processing, we haven’t talked yet about the tools...">
<meta property="og:description" content="The previous module described some common themes and processes in pre-processing biomedical data. While we’ve covered some key processes of pre-processing, we haven’t talked yet about the tools...">
<meta name="twitter:description" content="The previous module described some common themes and processes in pre-processing biomedical data. While we’ve covered some key processes of pre-processing, we haven’t talked yet about the tools...">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Improving the Reproducibility of Experimental Data Recording and Pre-Processing</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Overview of these modules</a></li>
<li class="book-part">Experimental Data Recording</li>
<li><a class="" href="module1.html"><span class="header-section-number">1</span> Separating data recording and analysis</a></li>
<li><a class="" href="module2.html"><span class="header-section-number">2</span> Principles and power of structured data formats</a></li>
<li><a class="" href="module3.html"><span class="header-section-number">3</span> The “tidy” data format</a></li>
<li><a class="" href="module4.html"><span class="header-section-number">4</span> Designing templates for “tidy” data collection</a></li>
<li><a class="" href="module5.html"><span class="header-section-number">5</span> Example: Creating a template for “tidy” data collection</a></li>
<li><a class="" href="module6.html"><span class="header-section-number">6</span> Organizing project files</a></li>
<li><a class="" href="module7.html"><span class="header-section-number">7</span> Creating project directory templates</a></li>
<li><a class="" href="module8.html"><span class="header-section-number">8</span> Example: Creating a project template</a></li>
<li><a class="" href="module9.html"><span class="header-section-number">9</span> Harnessing version control for transparent data recording</a></li>
<li><a class="" href="module10.html"><span class="header-section-number">10</span> Enhance the reproducibility of collaborative research with version control platforms</a></li>
<li><a class="" href="module11.html"><span class="header-section-number">11</span> Using Git and GitHub to implement version control</a></li>
<li class="book-part">Experimental Data Pre-processing</li>
<li><a class="" href="module12.html"><span class="header-section-number">12</span> Principles of pre-processing experimental data</a></li>
<li><a class="active" href="module12a.html"><span class="header-section-number">13</span> Selecting software options for pre-processing</a></li>
<li><a class="" href="module13.html"><span class="header-section-number">14</span> Introduction to scripted data pre-processing in R</a></li>
<li><a class="" href="module13a.html"><span class="header-section-number">15</span> Tips for improving reproducibility when writing R scripts</a></li>
<li><a class="" href="module14.html"><span class="header-section-number">16</span> Simplify scripted pre-processing through R’s “tidyverse” tools</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/geanders/improve_repro">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="module12a" class="section level1" number="13">
<h1>
<span class="header-section-number">Module 13</span> Selecting software options for pre-processing<a class="anchor" aria-label="anchor" href="#module12a"><i class="fas fa-link"></i></a>
</h1>
<p>The previous module described some common themes and processes in pre-processing
biomedical data. While we’ve covered some key processes of pre-processing, we
haven’t talked yet about the tools you can use to implement it. These are often
combined together into a <em>pipeline</em> (also called a <em>workflow</em>). These pipelines can
become fairly long and complex when you need to pre-process data that are
complex.</p>
<p>Most pre-processing pipelines will be run on the computer, with software tools. An
exception might be for very simple pre-processing tasks—one example is
generating the average cage weight for a group of mice based on the total cage
weight and the number of mice. However, even simple processes like this, which
can be done by hand, can also be done with a computer, and doing so can help
avoid errors and to provide a record of the calculation that was used for the
pre-processing.</p>
<p>You will have a choice about which type of software you use for pre-processing.
There are two key dimensions that separate these choices—first, whether the
software is point-and-click versus script-based, and, second, whether the
software is proprietary versus open-source. It is important to note
that, in some cases, it may make sense to develop a pipeline that chains
together a few different software programs to complete the required
pre-processing.</p>
<p>In this module, we’ll talk about the advantages and disadvantages of these
different types of software. For reproducibility and rigor, there are many
advantages to using software that is script-based and open-source for data
pre-processing, and so in later modules, we’ll provide more information on how
you can use this type of software for pre-processing biomedical data. We also
recognize, however, that there are some cases where such software may not be a
viable option for some or all of the data pre-processing for a project.</p>
<p><strong>Objectives.</strong> After this module, the trainee will be able to:</p>
<ul>
<li>Describe software approaches for pre-processing data</li>
<li>Compare the advantages and disadvantages of Graphical User Interface–based
versus scripted approaches and of open-source versus proprietary approaches
to pre-processing</li>
</ul>
<div id="gui-based-software-versus-script-based-software" class="section level2" number="13.1">
<h2>
<span class="header-section-number">13.1</span> GUI-based software versus script-based software<a class="anchor" aria-label="anchor" href="#gui-based-software-versus-script-based-software"><i class="fas fa-link"></i></a>
</h2>
<p>When you pick software for pre-processing, the first key dimension to consider
is whether the software is “point-and-click” or script-based.
Let’s start with a definition of each.</p>
<p>Point-and-click software is more formally known as <em>GUI-based software</em>, where
GUI stand for “graphical user interface”. These are programs where your hand is
on the mouse most of the time, and you use the mouse to select actions and
options from buttons and other widgets that are shown by the software on the
screen. This type of software is also sometimes called “widget-based”, as it is
built around widgets like drop-down menus and slider bars.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Jeffrey M Perkel, &lt;span&gt;“THE FUTURE OF SCIENTIFIC FIGURES,”&lt;/span&gt; &lt;em&gt;Nature&lt;/em&gt; 554, no. 7690 (2018): 133–34.&lt;/p&gt;"><sup>290</sup></a></span></p>
<p>A basic example of GUI-based software is your computer’s calendar application
(“application” is a common synonym for “software”). To navigate across dates on
your calendar, you use your mouse to click on arrows or dates.
The software includes some text entry—for example, if you add something to
your calendar, you can click on a textbox and enter a description of the
activity using your keyboard. However, the basic way that you navigate and use
the software is via your computer mouse.</p>
<p>Script-based software uses a script, rather than clickable buttons and graphics,
as its main interface. A script, in this case, is a line-by-line set of
instructions describing what actions you want the software to perform. With
script-based software, you typically keep your keys on the keyboard more often
than on the mouse. Many script-based software programs will also allow you to
also send the lines of instructions one at a time in an area referred to as a
<em>console</em>, which will then return the result from each line after you run it.
Script-based software is also sometimes called software that is “used
programatically.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Ibid.&lt;/p&gt;"><sup>291</sup></a></span> Several script-based software programs are
commonly used with biomedical data including R, Python, and Unix bash scripts,
as well as some less common but emerging software programs like Julia.</p>
<p>When comparing point-and-click software to script-based software for
pre-processing, there are a few advantages to point-and-click software, but
many more to script-based software. In terms of code rigor and reproducibility,
script-based software comes out well ahead, especially when used to its
full advantage.</p>
<p>Let’s start, though, by acknowledging some appealing features of point-and-click
software. These features likely contribute to its wide popularity and to the
fact that the vast majority of software that you use in your day-to-day life
outside of research is probably point-and-click.</p>
<p>First, GUI-based software is often easier to learn to use, at least in terms of
basic use. The visual icons help you navigate choices and actions in the
software. Most GUI-based software programs are designed to take underlying
processes and make them easier for a new user to access and use. They do this
through an interface that is visual, rather than language- and script-based.
Further, many people are most familiar with point-and-click software, since so
many everyday applications are of this type, and so its interface can feel more
familiar to users. They also are easier for a new user to pick up because they
typically provide a much smaller set of options than a full programming language
does.</p>
<p>By contrast, script-based software can take more investment of time and energy
to initially learn how to use. This is because they are coding languages, and
each one is just that—a language. It is built on a (often large) set of
vocabulary that you must learn to be proficient, as you must learn the names and
options for a large set of functions within the language. Further, it has rules
and logic you must learn in terms of options for how to structure and access
data and how the inputs and outputs of different functions can be chained
together to build pipelines for pre-processing and analysis.</p>
<p>Script-based software also requires you to be precise in this language. As Brian
Kernighan writes in his book <em>D is for Digital</em>:</p>
<blockquote>
<p>“A computer is the ultimate sorcerer’s apprentice, able to follow instructions
tirelessly and without error, but requiring painstaking accuracy in the
specification of what to do.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Brian W Kernighan, &lt;em&gt;D Is for Digital: What a Well-Informed Person Should Know about Computers and Communications&lt;/em&gt; (CreateSpace Independent Publishing Platform, 2011).&lt;/p&gt;"><sup>292</sup></a></span></p>
</blockquote>
<p>However, while there is a higher investment required to learn script-based
software versus point-and-click software, there is also a higher payoff from
that effort. Script-based software creates a full framework for you to
combine tools in interesting ways and to build new tools when you need them.
With point-and-click software, there’s always a layer between the user and
the computer logic, and you are constrained to only use tools that were
designed by the person who programmed the point-and-click software. By
contrast, with script-based software, you have more direct access to the
underlying computer logic, and with many popular script-based languages
(R, Python), you have extraordinary power and flexibility in what you can
ask the program to do.</p>
<p>As an analogy, think about traveling to a country where you don’t yet speak the
language. You have a few choices in how you could communicate. You could
memorize a few key phrases that you think you’ll need, or get a phrase book that
lists these key phrases. Another choice is to try to learn the language,
including learning the grammar of the language, and how thoughts are put
together into phrases. Learning the language, even at a basic level, will take
much more time. However, it will allow you much greater ability to express
yourself. If you only know set phrases, then you may know how to ask someone at
a bakery for loaf of bread, if the person who wrote the phrase book decided to
include that, but not how to ask at a hotel for an extra blanket, if that wasn’t
included. By contrast, if you’ve learned the language, you have learned how to
form a question, and so you can extrapolate to express a great variety of
things.</p>
<p>GUI-based software can be like using a phrase book for a foreign
language—if the person who developed the tool didn’t imagine something that
you need, you’re stuck. Scripted software is more like learning a language—you
have to learn the rules (grammar) and vocabulary (names of functions and
their parameters), but once you do, you can combine them to address a wide
variety of tasks, including things no one else has yet thought of.</p>
<p>In the late 1990s, a famous computer scientist named Richard Hamming wrote a
book called, “The Art and Science of Engineering”, in which he talks a lot about
the process of building things and the role that programming can play in this
process. He predicted at the time that by 2020, it will be the experts in a
particular field that do programming for that field, rather than experts in
computer programming trying to build tools for other fields.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Richard R Hamming, &lt;em&gt;The Art of Doing Science and Engineering: Learning to Learn&lt;/em&gt; (CRC Press, 1997).&lt;/p&gt;"><sup>293</sup></a></span>
He notes:</p>
<blockquote>
<p>“What is wanted in the long run, of course, is that the man with the problem
does the actual writing of the code with no human interface, as we all too often
have these days, between the person who knows the problem and the person who
knows the programming language. This date is unfortunately too far off to
do much good immediately, but I would think by the year 2020 it would be
fairly universal practice for the expert in the field of application to do
the actual program preparation rather than have experts in computers (and
ignorant in the field of application) do the program preparation.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Ibid.&lt;/p&gt;"><sup>294</sup></a></span></p>
</blockquote>
<p>The rise of open-source, scripted programs like Python and R is rapidly helping to
achieve this vision—scientists in a variety of fields now write their own
small software programs and tools, building on the framework of larger
open-source languages. Training programs in many scientific fields recommend
or require at least one course in programming in these languages, often
taught in conjunction with data analysis and data management.</p>
<p>Another element that has helped make script-based software more accessible is
the development of programming languages that are easier to learn and use. Very
early programming languages required the programmer to understand a lot about
how the computer was built and organized, including thinking about where and how
data were stored in the computer’s memory. As programming languages have
developed, such “low-level” languages have remained in use, as they often allow
for unmatched speed in processing. However, “higher-level” programming languages
have become more common, and while these might be somewhat slower in
computational processing power, they are much faster for humans to learn and to
create tools with, as they abstract away many of the details that make low-level
programming more difficult.</p>
<p>Because of the development of easier-to-learn high-level programming languages
like R and Python, it is possible for a scientist to become proficient in one of
these script-based programs in about a year. In our own experience, we have
found that often one semester of a dedicated course or serious self-study,
followed with several months of regularly applying the software to research
data, is enough for a scientist to become productive in using a script-based
software like R or Python for research. With another year or so of regular use,
scientists can often start making their own small software extensions to the
language. However, in a 2017 article on analyzing single-cell RNA-sequencing
data, the author noted that “relatively few biologists are comfortable working
in those environments”, referring to Unix and R,<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Perkel, &lt;span&gt;“Single-Cell Sequencing Made Simple.”&lt;/span&gt;&lt;/p&gt;"><sup>295</sup></a></span> and noted
that this was a barrier to using many of the available tools for working with
single-cell RNA-sequencing data at the time.</p>
<p>It is true that this is a substantially larger investment in training than a
short course or workshop, which might be adequate for learning the basics of
many GUI-based software programs. Time can be a critical barrier, especially for
scientists who are advanced in their career and may have minimal time for
further training. Further, it’s more of a barrier in analyzing some types of
biomedical data, due to the extreme size and complexity of the data.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Anton Nekrutenko and James Taylor, &lt;span&gt;“Next-Generation Sequencing Data Interpretation: Enhancing Reproducibility and Accessibility,”&lt;/span&gt; &lt;em&gt;Nature Reviews Genetics&lt;/em&gt; 13, no. 9 (2012): 667–72.&lt;/p&gt;"><sup>296</sup></a></span> However, it is much less of a time investment than it
takes to become an expert in a scientific field. It takes years of training to
become an expert in cellular biology or immunology, for example. Richard
Hamming’s vision was that the experts can ask the best and most creative
questions of the data, and that it is best to remove the barrier of a separate
person as the computer programmer, so that the expert can directly create the
program and leverage the full capabilities of the computer. Higher-level
programming languages now are accessible enough that this vision is playing out
across scientific fields.</p>
<p>Script-based approaches also encourage the user to learn how the underlying process
works. The approach encourages the user to think more like a car owner who gets
under the hood from time to time than like one who only drives the car. This
approach does take more time to learn and develop, but with the upside that the
user will often have a much deeper understanding of what is happenening in each
step, as well as how to fix or adjust different steps to fix a pipeline or
adapt one pipeline to meet another analysis need.</p>
<p>Another advantage of script-based software—and one that is related to the idea
of experts in a scientific field directly programming—is that often the most
cutting edge algorithms and pipelines will be available first in scripted
languages, and only later be added into point-and-click software programs. This
means that you may have earlier access to new algorithms and approaches if you
are comfortable coding in a script-based language.</p>
<p>For example, an article about single-cell RNA-sequencing from 2017 noted that,
at the time, there were “very few, if any, ‘plug-and-play’ packages” for working
with scRNA-seq data, and of those available, they were “user-friendly but have
the drawback that they are to some extent a ‘black box’, with little
transparency as to the precise algorithmic details and parameters employed.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Haque et al., &lt;span&gt;“A Practical Guide to Single-Cell RNA-Sequencing for Biomedical Research and Clinical Applications.”&lt;/span&gt;&lt;/p&gt;"><sup>297</sup></a></span> Similarly, another article in the same year noted that, at
the time, “most scRNA-seq tools exist as Unix programs or packages in the
programming language R”, although “some ready-to-use pipelines have been
developed.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Perkel, &lt;span&gt;“Single-Cell Sequencing Made Simple.”&lt;/span&gt;&lt;/p&gt;"><sup>298</sup></a></span></p>
<p>Another key advantage of script-based software is that, in writing the
script, you are thoroughly documenting the steps you took to pre-process the
data. When you create a code script, the script itself includes all
the steps and details of the process. In combination with information about the
version of all software used and the raw data input to the pipeline, it creates
a fully reproducible record of the data pre-processing and analysis.</p>
<p>This means both that you will be able to re-do all the steps yourself in the
future, if you need to, but that also that other researchers can explore and
replicate what you do. You may want to share your process with others in your
laboratory group, for example, so they can understand the choices you made and
steps you took in pre-processing the data. You may also want to share the
process with readers of the articles you publish, and this may in fact be
required by the journal. Well-documented code also makes it much easier to write
up the method section later in manuscripts that leveraged the data collected in
the experiment.</p>
<p>By contrast, while you could write down the steps that you took and the buttons
you pressed when using GUI-based software, it’s very easy to forget to
record a step. Some GUI-based programs are taking steps to try to
ameliorate this, allowing a user to save or download a full record that records
the steps taken in a given pipeline or allow the user to develop a full,
recorded workflow (one example is FlowJo Envoy’s workflow model for analyzing
data from flow cytometry). There are also some movements towards
“integrative frameworks”, which can help improve reproducibility for pipelines
that span different types of software (Galaxy, Gene Prof).<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Nekrutenko and Taylor, &lt;span&gt;“Next-Generation Sequencing Data Interpretation.”&lt;/span&gt;&lt;/p&gt;"><sup>299</sup></a></span></p>
<p>When you use a code script, it will not run if you forget a step or a detail of
that step. It is like writing a recipe that can be applied again and again. By
writing a script, you encode the process a single time, so you can take the time
to check and recheck to make sure that you’ve encoded the process correctly.
This helps in avoiding small errors when you do the pre-processing—if you are
punching numbers into a calculator over and over, it’s easy to mistype a number
or forget a step every now and then, while the code will ensure that the same
process is run every time and that it faithfully uses the numbers saved in the
data for each step, rather than relying on a person correctly entering each
number in the calculation.</p>
<p>Scripts can be used across projects, as well, and so they can ensure consistency
in the calculation across projects. If different people do the calculation in
the lab for different projects or experiments, and they are doing the
calculations by hand, they might each do the calculation slightly differently,
even if it’s only in small details like how they report rounded numbers. A
script will do the exact same thing every time it is applied. You can even share
your script with colleagues at other labs, if you want to ensure that your data
pre-processing is comparable for experiments conducted in different research
groups, and many scientific journals will allow supplemental material with
code used for data pre-processing and analysis, or links within the manuscript
to a repository of this code posted online.</p>
<p>There are also gains in efficiency when you use a script. This is often a gain
that fully pays back the investment in learning the software—it can make data
pre-processing and analysis much more efficient over the long term. For small
pre-processing steps, these might seem small for each experiment, and certainly
when you first write the script, it will likely take longer to write and test
the script than it would to just do the calculation by hand (even more if you’re
just starting to learn how to write code scripts). However, since the script can
be applied again and again, with very little extra work to apply it to new data,
you’ll save yourself time in the future, and over a lot of experiments and
projects, this can add up. This makes it particularly useful to write scripts
for pre-processing tasks that you find yourself doing again and again in the
lab.</p>
</div>
<div id="open-source-versus-proprietary-software" class="section level2" number="13.2">
<h2>
<span class="header-section-number">13.2</span> Open-source versus proprietary software<a class="anchor" aria-label="anchor" href="#open-source-versus-proprietary-software"><i class="fas fa-link"></i></a>
</h2>
<p>When selecting software for pre-processing, the other dimension to consider is
whether it is open-source or proprietary. <em>Open-source software</em> is software
where you can access, explore, and build on all the underlying code for the
software. It also most often is free. By contrast, the code that powers
<em>proprietary software</em> is typically kept private, so you can use the product but
cannot explore the way that it is built or extend it in the same way that you
can open-source software. In biomedical research, many script-based languages
are open-source, while many GUI-based programs are proprietary. However, this is
not a hard and fast rule, and there are examples of open-source GUI-based
software (for example, the Inkscape program for vector graphic design) as well
as proprietary script-based software (for example, Matlab and SAS). There are
advantages and disadvantages to both types of software, but in terms of rigor
and reproducibility, open-source software often has the advantage.</p>
<p>Transparency is a key element of reproducibility.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Ellen P Neff, &lt;span&gt;“On the Past, Present, and Future of in Vivo Science,”&lt;/span&gt; &lt;em&gt;Lab Animal&lt;/em&gt; 50, no. 10 (2021): 273–76.&lt;/p&gt;"><sup>300</sup></a></span> As Gordon
Lithgow and coauthors note in a commentary on reproducibility, “Improved
reproducibility comes from pinning down methods.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Gordon J Lithgow, Monica Driscoll, and Patrick Phillips, &lt;span&gt;“A Long Journey to Reproducible Results,”&lt;/span&gt; &lt;em&gt;Nature&lt;/em&gt; 548, no. 7668 (2017): 387–88.&lt;/p&gt;"><sup>301</sup></a></span> If the
algorithms of software can be investigated, then scientists who are using two
different programs (for example, one program in Python and one in R) can
determine if their choice of program is causing differences in their results. By
contrast, if two research groups use two different types of proprietary
software, the algorithms that underlie the processing are often kept secret and
so cannot be compared. In that case, if the two groups conduct the same
experiment and get different results, it’s impossible to rule out whether the
difference was caused by the choice of software.</p>
<p>Gordon Lithgow, Monica Driscoll, and Patrick Phillips wrote a commentary for
<em>Nature</em> describing their experiences in replicating research. They
describe the advice they give students who are trying to do an
experiment that should work:</p>
<blockquote>
<p>“If there is nothing wrong with the reagents and reproducibility is still an
issue, then as I like to tell students, there are two options: (1) the physical
constants of the universe and hence the laws of physics are in a state of flux
in their round-bottomed flask, or (2) the researcher is doing something wrong
and either doesn’t know it or doesn’t want to know it. Then I ask them which
explanation they think I’m leaning towards.”<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Bruce C Gibb, &lt;span&gt;“Reproducibility,”&lt;/span&gt; &lt;em&gt;Nature Chemistry&lt;/em&gt; 6, no. 8 (2014): 653–54.&lt;/p&gt;"><sup>302</sup></a></span></p>
</blockquote>
<p>If you get different results from another group, it is critical to have a
detailed description of the methods that each group used to figure out why
the groups are getting different results. Open-source software provides this
at the level of computational analysis, because the openness of the software
means anyone can explore the exact details of how each algorithm runs.</p>
<p>One key advantage of open-source software is that all code is open, so you can
dig down to figure out exactly how each step of the program works. Futher, in
many cases for open-source scientific software, the algorithms and their
principles have gone through peer review as part of the academic publication
process. With proprietary software, on the other hand, details of algorithms may
be considered protected intellectual property, and so it may be hard to find out
the details of how the underlying algorithms work.<span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Nekrutenko and Taylor, &lt;span&gt;“Next-Generation Sequencing Data Interpretation.”&lt;/span&gt;&lt;/p&gt;"><sup>303</sup></a></span> Also,
the algorithms may not have gone through peer-review, especially if they are
considered private intellectual property.</p>
<p>Another advantage of open-source software is that older versions of the software
are often well-archived and easily available to reinstall if you need to
reproduce an analysis that was done using an earlier version of the software.
Another advantage is that open-source software is often free. This makes it
economical to test out, and it means that trainees from a lab will have no
problem continuing to use the software as they move to new positions.
The cost with open-source software, then, comes not with the price to buy
the software, but with the investment that is required to learn it.</p>
<p>One facet where proprietary software has an advantage is that it will often have
more comprehensive company-based user support than open-source software. The
companies that make and sell proprietary software will usually have a user
support team to answer questions and help develop pipelines and may also offer
training programs or materials.</p>
<p>Some open-source software also has robust user support, although sometimes a bit
less organized under a common source. In some cases, this has developed as a
result of a large community of users who help each other. Message boards like
StackOverflow provides a forum for users to ask and respond to questions. Some
companies also exist that provide, as their business model, user support for
open-source software. While open-source software is usually free, these
companies make money by providing support for that software.</p>
<p>User support is sparser for some of smaller software packages that are developed
as extensions of open-source software. For example, many of the packages for
pre-processing types of biomedical data are built by small bioinformatics teams
or individuals at academic or research institutions. Often this software is
developed by a single person or very small team as one part of their job
profile, with limited resources for user support and for providing training.
These extensions build on larger, more supported open-source software (e.g., R
or Python), but the extension itself is built and maintained by a very small
team that may not have the capacity to respond quickly to user questions. Many
open-source software developers try to create helpful documentation in the form
of help files and package vignettes (tutorials on how to use the software they
created), but from a practical point of view it is difficult for small
open-source developers to provide the same level of user support that a large
proprietary software company can.</p>
<p>This is often the case with cutting-edge open-source software for biomedical
pre-processing. These just-developed software packages are less likely to be
comprehensively documented than longer-established software. Further, it can
take a while for the community of software users to develop once software is
available, and while this is a limitation of new software for both open-source
and proprietary languages, it can represent more of a problem for open-source
software, where there is typically not a company-based helpline and so the
community of users often represents one of the main sources for help and
troubleshooting.</p>
<!-- ### Discussion questions -->

</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="module12.html"><span class="header-section-number">12</span> Principles of pre-processing experimental data</a></div>
<div class="next"><a href="module13.html"><span class="header-section-number">14</span> Introduction to scripted data pre-processing in R</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#module12a"><span class="header-section-number">13</span> Selecting software options for pre-processing</a></li>
<li><a class="nav-link" href="#gui-based-software-versus-script-based-software"><span class="header-section-number">13.1</span> GUI-based software versus script-based software</a></li>
<li><a class="nav-link" href="#open-source-versus-proprietary-software"><span class="header-section-number">13.2</span> Open-source versus proprietary software</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/geanders/improve_repro/blob/master/13-preprocessing_tools.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/geanders/improve_repro/edit/master/13-preprocessing_tools.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Improving the Reproducibility of Experimental Data Recording and Pre-Processing</strong>" was written by Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, Gregory Robertson. </p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
