<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>
  <meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  
  <meta name="twitter:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="experimental-data-preprocessing.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Data Visualization in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.1</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html"><i class="fa fa-check"></i><b>2</b> Experimental Data Recording</a><ul>
<li class="chapter" data-level="2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#separating-data-recording-and-analysis"><i class="fa fa-check"></i><b>2.1</b> Separating data recording and analysis</a><ul>
<li class="chapter" data-level="2.1.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-versus-data-analysis"><i class="fa fa-check"></i><b>2.1.1</b> Data recording versus data analysis</a></li>
<li class="chapter" data-level="2.1.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#hazards-of-combining-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.2</b> Hazards of combining recording and analysis</a></li>
<li class="chapter" data-level="2.1.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#approaches-to-separate-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.3</b> Approaches to separate recording and analysis</a></li>
<li class="chapter" data-level="2.1.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions"><i class="fa fa-check"></i><b>2.1.4</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#principles-and-power-of-structured-data-formats"><i class="fa fa-check"></i><b>2.2</b> Principles and power of structured data formats</a><ul>
<li class="chapter" data-level="2.2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-standards"><i class="fa fa-check"></i><b>2.2.1</b> Data recording standards</a></li>
<li class="chapter" data-level="2.2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#defining-data-standards-for-a-research-group"><i class="fa fa-check"></i><b>2.2.2</b> Defining data standards for a research group</a></li>
<li class="chapter" data-level="2.2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#two-dimensional-structured-data-format"><i class="fa fa-check"></i><b>2.2.3</b> Two-dimensional structured data format</a></li>
<li class="chapter" data-level="2.2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#saving-two-dimensional-structured-data-in-plain-text-file-formats"><i class="fa fa-check"></i><b>2.2.4</b> Saving two-dimensional structured data in plain text file formats</a></li>
<li class="chapter" data-level="2.2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#occassions-for-more-complex-data-structures-and-file-formats"><i class="fa fa-check"></i><b>2.2.5</b> Occassions for more complex data structures and file formats</a></li>
<li class="chapter" data-level="2.2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#levels-of-standardizationresearch-group-to-research-community"><i class="fa fa-check"></i><b>2.2.6</b> Levels of standardization—research group to research community</a></li>
<li class="chapter" data-level="2.2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise"><i class="fa fa-check"></i><b>2.2.7</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#the-tidy-data-format"><i class="fa fa-check"></i><b>2.3</b> The ‘tidy’ data format</a><ul>
<li class="chapter" data-level="2.3.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-makes-data-tidy"><i class="fa fa-check"></i><b>2.3.1</b> What makes data “tidy”?</a></li>
<li class="chapter" data-level="2.3.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#why-make-your-data-tidy"><i class="fa fa-check"></i><b>2.3.2</b> Why make your data “tidy”?</a></li>
<li class="chapter" data-level="2.3.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-tidyverse-tools-with-data-in-the-tidy-data-format"><i class="fa fa-check"></i><b>2.3.3</b> Using tidyverse tools with data in the “tidy data” format</a></li>
<li class="chapter" data-level="2.3.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz"><i class="fa fa-check"></i><b>2.3.4</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#designing-templates-for-tidy-data-collection"><i class="fa fa-check"></i><b>2.4</b> Designing templates for “tidy” data collection</a><ul>
<li class="chapter" data-level="2.4.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1"><i class="fa fa-check"></i><b>2.4.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.4.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-1"><i class="fa fa-check"></i><b>2.4.2</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-creating-a-template-for-tidy-data-collection"><i class="fa fa-check"></i><b>2.5</b> Example: Creating a template for “tidy” data collection</a><ul>
<li class="chapter" data-level="2.5.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-1"><i class="fa fa-check"></i><b>2.5.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.5.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2"><i class="fa fa-check"></i><b>2.5.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.5.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-1"><i class="fa fa-check"></i><b>2.5.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#power-of-using-a-single-structured-project-directory-for-storing-and-tracking-research-project-files"><i class="fa fa-check"></i><b>2.6</b> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a><ul>
<li class="chapter" data-level="2.6.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-2"><i class="fa fa-check"></i><b>2.6.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.6.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-1"><i class="fa fa-check"></i><b>2.6.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.6.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz-1"><i class="fa fa-check"></i><b>2.6.3</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#creating-project-templates"><i class="fa fa-check"></i><b>2.7</b> Creating ‘Project’ templates</a><ul>
<li class="chapter" data-level="2.7.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-3"><i class="fa fa-check"></i><b>2.7.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.7.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-2"><i class="fa fa-check"></i><b>2.7.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.7.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-2"><i class="fa fa-check"></i><b>2.7.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-creating-a-project-template"><i class="fa fa-check"></i><b>2.8</b> Example: Creating a ‘Project’ template</a><ul>
<li class="chapter" data-level="2.8.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-4"><i class="fa fa-check"></i><b>2.8.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.8.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-3"><i class="fa fa-check"></i><b>2.8.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.8.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-2"><i class="fa fa-check"></i><b>2.8.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#harnessing-version-control-for-transparent-data-recording"><i class="fa fa-check"></i><b>2.9</b> Harnessing version control for transparent data recording</a><ul>
<li class="chapter" data-level="2.9.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-5"><i class="fa fa-check"></i><b>2.9.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.9.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-4"><i class="fa fa-check"></i><b>2.9.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.9.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-3"><i class="fa fa-check"></i><b>2.9.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#enhance-the-reproducibility-of-collaborative-research-with-version-control-platforms"><i class="fa fa-check"></i><b>2.10</b> Enhance the reproducibility of collaborative research with version control platforms</a><ul>
<li class="chapter" data-level="2.10.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-6"><i class="fa fa-check"></i><b>2.10.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.10.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-5"><i class="fa fa-check"></i><b>2.10.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.10.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-4"><i class="fa fa-check"></i><b>2.10.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.11" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-git-and-gitlab-to-implement-version-control"><i class="fa fa-check"></i><b>2.11</b> Using git and GitLab to implement version control</a><ul>
<li class="chapter" data-level="2.11.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-7"><i class="fa fa-check"></i><b>2.11.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.11.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-6"><i class="fa fa-check"></i><b>2.11.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.11.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-3"><i class="fa fa-check"></i><b>2.11.3</b> Applied exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html"><i class="fa fa-check"></i><b>3</b> Experimental Data Preprocessing</a><ul>
<li class="chapter" data-level="3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#principles-and-benefits-of-scripted-pre-processing-of-experimental-data"><i class="fa fa-check"></i><b>3.1</b> Principles and benefits of scripted pre-processing of experimental data</a><ul>
<li class="chapter" data-level="3.1.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-8"><i class="fa fa-check"></i><b>3.1.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.1.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-7"><i class="fa fa-check"></i><b>3.1.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.1.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#discussion-questions-5"><i class="fa fa-check"></i><b>3.1.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-to-scripted-data-pre-processing-in-r"><i class="fa fa-check"></i><b>3.2</b> Introduction to scripted data pre-processing in R</a><ul>
<li class="chapter" data-level="3.2.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-9"><i class="fa fa-check"></i><b>3.2.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.2.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-8"><i class="fa fa-check"></i><b>3.2.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.2.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-4"><i class="fa fa-check"></i><b>3.2.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#simplify-scripted-pre-processing-through-rs-tidyverse-tools"><i class="fa fa-check"></i><b>3.3</b> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a><ul>
<li class="chapter" data-level="3.3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-10"><i class="fa fa-check"></i><b>3.3.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-9"><i class="fa fa-check"></i><b>3.3.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-2"><i class="fa fa-check"></i><b>3.3.3</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-data-types-in-experimental-data-pre-processing"><i class="fa fa-check"></i><b>3.4</b> Complex data types in experimental data pre-processing</a><ul>
<li class="chapter" data-level="3.4.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-11"><i class="fa fa-check"></i><b>3.4.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.4.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-10"><i class="fa fa-check"></i><b>3.4.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.4.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-3"><i class="fa fa-check"></i><b>3.4.3</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-data-types-in-r-and-bioconductor"><i class="fa fa-check"></i><b>3.5</b> Complex data types in R and Bioconductor</a><ul>
<li class="chapter" data-level="3.5.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-12"><i class="fa fa-check"></i><b>3.5.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.5.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-11"><i class="fa fa-check"></i><b>3.5.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.5.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-5"><i class="fa fa-check"></i><b>3.5.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#example-converting-from-complex-to-tidy-data-formats"><i class="fa fa-check"></i><b>3.6</b> Example: Converting from complex to ‘tidy’ data formats</a><ul>
<li class="chapter" data-level="3.6.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-13"><i class="fa fa-check"></i><b>3.6.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.6.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-12"><i class="fa fa-check"></i><b>3.6.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.6.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-6"><i class="fa fa-check"></i><b>3.6.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-to-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.7</b> Introduction to reproducible data pre-processing protocols</a><ul>
<li class="chapter" data-level="3.7.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-14"><i class="fa fa-check"></i><b>3.7.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.7.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-13"><i class="fa fa-check"></i><b>3.7.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.7.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#discussion-questions-6"><i class="fa fa-check"></i><b>3.7.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#rmarkdown-for-creating-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.8</b> RMarkdown for creating reproducible data pre-processing protocols</a><ul>
<li class="chapter" data-level="3.8.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-15"><i class="fa fa-check"></i><b>3.8.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.8.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-14"><i class="fa fa-check"></i><b>3.8.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.8.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-7"><i class="fa fa-check"></i><b>3.8.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#example-creating-a-reproducible-data-pre-processing-protocol"><i class="fa fa-check"></i><b>3.9</b> Example: Creating a reproducible data pre-processing protocol</a><ul>
<li class="chapter" data-level="3.9.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-16"><i class="fa fa-check"></i><b>3.9.1</b> Subsection 1</a></li>
<li class="chapter" data-level="3.9.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-15"><i class="fa fa-check"></i><b>3.9.2</b> Subsection 2</a></li>
<li class="chapter" data-level="3.9.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-4"><i class="fa fa-check"></i><b>3.9.3</b> Practice quiz</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>4</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Improving the Reproducibility of Experimental Data Recording and Pre-Processing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> References</h1>

<div id="refs" class="references">
<div>
<p>AlTarawneh, Ghada, and Simon Thorne. 2017. “A Pilot Study Exploring Spreadsheet Risk in Scientific Research.” <em>arXiv Preprint arXiv:1703.09785</em>.</p>
</div>
<div>
<p>Altman, Douglas G., and J. Martin Bland. 1997. “Statistical Notes: Units of Analysis.” <em>BMJ</em> 314: 1874.</p>
</div>
<div>
<p>Anderson, Nicholas R, E Sally Lee, J Scott Brockenbrough, Mark E Minie, Sherrilynne Fuller, James Brinkley, and Peter Tarczy-Hornoch. 2007. “Issues in Biomedical Research Data Management and Analysis: Needs and Barriers.” <em>Journal of the American Medical Informatics Association</em> 14 (4): 478–88.</p>
</div>
<div>
<p>Apte, Poornima. n.d. “The Lingua Franca of Latex.” <em>Increment</em>. <a href="https://increment.com/open-source/the-lingua-franca-of-latex/">https://increment.com/open-source/the-lingua-franca-of-latex/</a>.</p>
</div>
<div>
<p>Arnold, Taylor. 2017. “A Tidy Data Model for Natural Language Processing using cleanNLP.” <em>The R Journal</em> 9 (2): 248–67.</p>
</div>
<div>
<p>Barga, Roger, Bill Howe, David Beck, Stuart Bowers, William Dobyns, Winston Haynes, Roger Higdon, et al. 2011. “Bioinformatics and Data-Intensive Scientific Discovery in the Beginning of the 21st Century.” <em>Omics: A Journal of Integrative Biology</em> 15 (4): 199–201.</p>
</div>
<div>
<p>Baumer, Ben. 2015. “A Data Science Course for Undergraduates: Thinking with Data.” <em>The American Statistician</em> 69 (4): 334–42.</p>
</div>
<div>
<p>Baumer, Benjamin S, Daniel T Kaplan, and Nicholas J Horton. 2017. <em>Modern Data Science with R</em>. Boca Raton: CRC Press.</p>
</div>
<div>
<p>Birch, David, David Lyford-Smith, and Yike Guo. 2018. “The Future of Spreadsheets in the Big Data Era.” <em>arXiv Preprint arXiv:1801.10231</em>.</p>
</div>
<div>
<p>Brazma, Alvis, Maria Krestyaninova, and Ugis Sarkans. 2006. “Standards for Systems Biology.” <em>Nature Reviews Genetics</em> 7 (8): 593.</p>
</div>
<div>
<p>Broman, Karl W, and Kara H Woo. 2018. “Data Organization in Spreadsheets.” <em>The American Statistician</em> 72 (1): 2–10.</p>
</div>
<div>
<p>Bryan, Jennifer. 2018. “Excuse Me, Do You Have a Moment to Talk About Version Control?” <em>The American Statistician</em> 72 (1): 20–27.</p>
</div>
<div>
<p>Bryan, Jennifer, and Hadley Wickham. 2017. “Data Science: A Three Ring Circus or a Big Tent?” <em>Journal of Computational and Graphical Statistics</em> 26 (4): 784–85.</p>
</div>
<div>
<p>Buffalo, Vince. 2015. <em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>. " O’Reilly Media, Inc.".</p>
</div>
<div>
<p>Campbell-Kelly, Martin. 2007. “Number Crunching Without Programming: The Evolution of Spreadsheet Usability.” <em>IEEE Annals of the History of Computing</em> 29 (3): 6–19.</p>
</div>
<div>
<p>Creeth, Richard. 1985. “Microcomputer Spreadsheets: Their Uses and Abuses.” <em>Journal of Accountancy (Pre-1986)</em> 159 (000006): 90.</p>
</div>
<div>
<p>Edwards, Paul N, Matthew S Mayernik, Archer L Batcheller, Geoffrey C Bowker, and Christine L Borgman. 2011. “Science Friction: Data, Metadata, and Collaboration.” <em>Social Studies of Science</em> 41 (5): 667–90.</p>
</div>
<div>
<p>Ellis, Shannon E, and Jeffrey T Leek. 2018. “How to Share Data for Collaboration.” <em>The American Statistician</em> 72 (1): 53–57.</p>
</div>
<div>
<p>Ford, Paul. 2014. “On File Formats, Very Briefly.” <em>The Manual</em>. <a href="https://themanual.org/read/issues/4/paul-ford/article">https://themanual.org/read/issues/4/paul-ford/article</a>.</p>
</div>
<div>
<p>———. 2015. “I Dreamed of a Perfect Database.” <em>New Republic</em>. <a href="https://newrepublic.com/article/124425/dreamed-perfect-database">https://newrepublic.com/article/124425/dreamed-perfect-database</a>.</p>
</div>
<div>
<p>Ghosh, Samik, Yukiko Matsuoka, Yoshiyuki Asai, Kun-Yi Hsin, and Hiroaki Kitano. 2011. “Software for Systems Biology: From Tools to Integrated Platforms.” <em>Nature Reviews Genetics</em> 12 (12): 821.</p>
</div>
<div>
<p>Goodman, Alyssa, Alberto Pepe, Alexander W Blocker, Christine L Borgman, Kyle Cranmer, Merce Crosas, Rosanne Di Stefano, et al. 2014. “Ten Simple Rules for the Care and Feeding of Scientific Data.” Public Library of Science.</p>
</div>
<div>
<p>Hermans, Felienne, Bas Jansen, Sohon Roy, Efthimia Aivaloglou, Alaaeddin Swidan, and David Hoepelman. 2016. “Spreadsheets Are Code: An Overview of Software Engineering Approaches Applied to Spreadsheets.” In <em>2016 Ieee 23rd International Conference on Software Analysis, Evolution, and Reengineering (Saner)</em>, 5:56–65. IEEE.</p>
</div>
<div>
<p>Hermans, Felienne, and Emerson Murphy-Hill. 2015. “Enron’s Spreadsheets and Related Emails: A Dataset and Analysis.” In <em>2015 Ieee/Acm 37th Ieee International Conference on Software Engineering</em>, 2:7–16. IEEE.</p>
</div>
<div>
<p>Hicks, Stephanie C, and Rafael A Irizarry. 2017. “A Guide to Teaching Data Science.” <em>The American Statistician</em> In Press. <a href="https://doi.org/10.1080/00031305.2017.1356747">https://doi.org/10.1080/00031305.2017.1356747</a>.</p>
</div>
<div>
<p>Hsieh, TC, KH Ma, and Anne Chao. 2016. “iNEXT: an R package for rarefaction and extrapolation of species diversity (Hill numbers).” <em>Methods in Ecology and Evolution</em> 7 (12): 1451–6.</p>
</div>
<div>
<p>Hunt, Andrew, David Thomas, and Ward Cunningham. 2000. <em>The Pragmatic Programmer: From Journeyman to Master</em>. Addison-Wesley Professional.</p>
</div>
<div>
<p>Irizarry, Rafael A., and Michael I. Love. 2016. <em>Data Analysis for the Life Sciences with R</em>. Chapman; Hall.</p>
</div>
<div>
<p>Janssens, Jeroen. 2014. <em>Data Science at the Command Line: Facing the Future with Time-Tested Tools</em>. " O’Reilly Media, Inc.".</p>
</div>
<div>
<p>Kaplan, Daniel. 2018. “Teaching Stats for Data Science.” <em>The American Statistician</em> 72 (1): 89–96.</p>
</div>
<div>
<p>Keller, Sallie, Gizem Korkmaz, Mark Orr, Aaron Schroeder, and Stephanie Shipp. 2017. “The Evolution of Data Quality: Understanding the Transdisciplinary Origins of Data Quality Concepts and Approaches.” <em>Annu. Rev. Stat. Appl</em> 4: 85–108.</p>
</div>
<div>
<p>Kernighan, Brian W, and Rob Pike. 1984. <em>The Unix Programming Environment</em>. Vol. 270. Prentice-Hall Englewood Cliffs, NJ.</p>
</div>
<div>
<p>Krishnan, Sanjay, Daniel Haas, Michael J Franklin, and Eugene Wu. 2016. “Towards Reliable Interactive Data Cleaning: A User Survey and Recommendations.” In <em>Proceedings of the Workshop on Human-in-the-Loop Data Analytics</em>, 9. ACM.</p>
</div>
<div>
<p>Leek, Jeff. 2012. “Toward Tidy Data Analysis.” <em>Simply Statistics</em>.</p>
</div>
<div>
<p>Levy, Steven. 1984. “A Spreadsheet Way of Knowledge.” <em>Harpers</em> 269: 58–64.</p>
</div>
<div>
<p>Lowndes, Julia S Stewart, Benjamin D Best, Courtney Scarborough, Jamie C Afflerbach, Melanie R Frazier, Casey C O’Hara, Ning Jiang, and Benjamin S Halpern. 2017. “Our Path to Better Science in Less Time Using Open Data Science Tools.” <em>Nature Ecology &amp; Evolution</em> 1 (6): 160.</p>
</div>
<div>
<p>Lynch, Clifford. 2008. “Big Data: How Do Your Data Grow?” <em>Nature</em> 455 (7209): 28.</p>
</div>
<div>
<p>Marwick, Ben, Carl Boettiger, and Lincoln Mullen. 2018. “Packaging Data Analytical Work Reproducibly Using R (and Friends).” <em>The American Statistician</em> 72 (1): 80–88.</p>
</div>
<div>
<p>McCullough, BD. 2001. “Does Microsoft Fix Errors in Excel?” In <em>Proceedings of the 2001 Joint Statistical Meetings</em>.</p>
</div>
<div>
<p>McCullough, B D, and Berry Wilson. 2002. “On the Accuracy of Statistical Procedures in Microsoft Excel 2000 and Excel Xp.” <em>Computational Statistics &amp; Data Analysis</em> 40 (4): 713–21.</p>
</div>
<div>
<p>McCullough, Bruce D. 1999. “Assessing the Reliability of Statistical Software: Part Ii.” <em>The American Statistician</em> 53 (2): 149–59.</p>
</div>
<div>
<p>McCullough, Bruce D, and David A Heiser. 2008. “On the Accuracy of Statistical Procedures in Microsoft Excel 2007.” <em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4570–8.</p>
</div>
<div>
<p>McCullough, Bruce D, and Berry Wilson. 1999. “On the Accuracy of Statistical Procedures in Microsoft Excel 97.” <em>Computational Statistics &amp; Data Analysis</em> 31 (1): 27–37.</p>
</div>
<div>
<p>———. 2005. “On the Accuracy of Statistical Procedures in Microsoft Excel 2003.” <em>Computational Statistics &amp; Data Analysis</em> 49 (4): 1244–52.</p>
</div>
<div>
<p>McMurdie, Paul J, and Susan Holmes. 2013. “Phyloseq: An R Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.” <em>PloS One</em> 8 (4): e61217.</p>
</div>
<div>
<p>McNamara, Amelia. 2016. “On the State of Computing in Statistics Education: Tools for Learning and for Doing.” <em>arXiv Preprint arXiv:1610.00984</em>.</p>
</div>
<div>
<p>Mélard, Guy. 2014. “On the Accuracy of Statistical Procedures in Microsoft Excel 2010.” <em>Computational Statistics</em> 29 (5): 1095–1128.</p>
</div>
<div>
<p>Michener, William K. 2015. “Ten Simple Rules for Creating a Good Data Management Plan.” <em>PLoS Computational Biology</em> 11 (10): e1004525.</p>
</div>
<div>
<p>Murrell, Paul. 2009. <em>Introduction to Data Technologies</em>. Chapman; Hall/CRC.</p>
</div>
<div>
<p>Myneni, Sahiti, and Vimla L Patel. 2010. “Organization of Biomedical Data for Collaborative Scientific Research: A Research Information Management System.” <em>International Journal of Information Management</em> 30 (3): 256–64.</p>
</div>
<div>
<p>Nardi, Bonnie A, and James R Miller. 1990. “The Spreadsheet Interface: A Basis for End User Programming.” In <em>Proceedings of the Ifip Tc13 Third Interational Conference on Human-Computer Interaction</em>, 977–83. North-Holland Publishing Co.</p>
</div>
<div>
<p>Nash, JC. 2006. “Spreadsheets in Statistical Practice—Another Look.” <em>The American Statistician</em> 60 (3): 287–89.</p>
</div>
<div>
<p>Pawlik, Aleksandra, Celia WG van Gelder, Aleksandra Nenadic, Patricia M Palagi, Eija Korpelainen, Philip Lijnzaad, Diana Marek, Susanna-Assunta Sansone, John Hancock, and Carole Goble. 2017. “Developing a Strategy for Computational Lab Skills Training Through Software and Data Carpentry: Experiences from the ELIXIR Pilot Action.” <em>F1000Research</em> 6: ELIXIR–1040.</p>
</div>
<div>
<p>Peng, Roger. 2018. “Teaching R to New Users—from Tapply to the Tidyverse.” <em>Simply Statistics</em>.</p>
</div>
<div>
<p>Powell, Stephen G, Kenneth R Baker, and Barry Lawson. 2009. “Errors in Operational Spreadsheets: A Review of the State of the Art.” In <em>2009 42nd Hawaii International Conference on System Sciences</em>, 1–8. IEEE.</p>
</div>
<div>
<p>Raymond, Eric S. 2003. <em>The Art of Unix Programming</em>. Addison-Wesley Professional.</p>
</div>
<div>
<p>Robinson, David. 2017. “Teach the Tidyverse to Beginners.” <em>Variance Explained</em>.</p>
</div>
<div>
<p>Ross, Zev, Hadley Wickham, and David Robinson. 2017. “Declutter Your R Workflow with Tidy Tools.” <em>PeerJ Preprints</em> 5: e3180v1.</p>
</div>
<div>
<p>Sansone, Susanna-Assunta, Philippe Rocca-Serra, Dawn Field, Eamonn Maguire, Chris Taylor, Oliver Hofmann, Hong Fang, et al. 2012. “Toward Interoperable Bioscience Data.” <em>Nature Genetics</em> 44 (2): 121.</p>
</div>
<div>
<p>Schadt, Eric E, Michael D Linderman, Jon Sorenson, Lawrence Lee, and Garry P Nolan. 2010. “Computational Solutions to Large-Scale Data Management and Analysis.” <em>Nature Reviews Genetics</em> 11 (9): 647.</p>
</div>
<div>
<p>Sedgwick, Philip. 2014. “Unit of Observation and Unit of Analysis.” <em>BMJ</em> 348: g3840.</p>
</div>
<div>
<p>Seroul, Raymond. 2012. <em>A Beginner’s Book of Tex</em>. Springer Science &amp; Business Media.</p>
</div>
<div>
<p>Silge, Julia, and David Robinson. 2016. “Tidytext: Text Mining and Analysis Using Tidy Data Principles in R.” <em>The Journal of Open Source Software</em> 1 (3).</p>
</div>
<div>
<p>———. 2017. <em>Text Mining with R: A Tidy Approach</em>. Sebastopol: O’Reilly Media.</p>
</div>
<div>
<p>Stander, Julian, and Luciana Dalla Valle. 2017. “On Enthusing Students About Big Data and Social Media Visualization and Analysis Using R, RStudio, and RMarkdown.” <em>Journal of Statistics Education</em> 25 (2): 60–67.</p>
</div>
<div>
<p>Teixeira, Ricardo, and Vasco Amaral. 2016. “On the Emergence of Patterns for Spreadsheets Data Arrangements.” In <em>Federation of International Conferences on Software Technologies: Applications and Foundations</em>, 333–45. Springer.</p>
</div>
<div>
<p>Topaloglou, Thodoros, Susan B Davidson, HV Jagadish, Victor M Markowitz, Evan W Steeg, and Mike Tyers. 2004. “Biological Data Management: Research, Practice and Opportunities.” In <em>Proceedings of the Thirtieth International Conference on Very Large Data Bases-Volume 30</em>, 1233–6. VLDB Endowment.</p>
</div>
<div>
<p>Tyner, Sam, François Briatte, and Heike Hofmann. 2017. “Network Visualization with ggplot2.” <em>The R Journal</em> 9 (1): 27–59.</p>
</div>
<div>
<p>U.S. Department of Health and Human Services, National Institutes of Health. 2016. “NIH-Wide Strategic Plan, Fiscal Years 2016-2020: Turning Discovery into Health.” <a href="https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf">https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf</a>.</p>
</div>
<div>
<p>———. 2018. “NIH Strategic Plan for Data Science.” <a href="https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf">https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf</a>.</p>
</div>
<div>
<p>Waltemath, Dagmar, and Olaf Wolkenhauer. 2016. “How Modeling Standards, Software, and Initiatives Support Reproducibility in Systems Biology and Systems Medicine.” <em>IEEE Transactions on Biomedical Engineering</em> 63 (10): 1999–2006.</p>
</div>
<div>
<p>Welsh, Eric A, Paul A Stewart, Brent M Kuenzi, and James A Eschrich. 2017. “Escape Excel: A Tool for Preventing Gene Symbol and Accession Conversion Errors.” <em>PloS One</em> 12 (9): e0185207.</p>
</div>
<div>
<p>Wickham, Hadley. 2014. “Tidy Data.” <em>Journal of Statistical Software</em> 59 (10): 1–23.</p>
</div>
<div>
<p>———. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. 2nd ed. New York: Springer.</p>
</div>
<div>
<p>———. 2017. “The Tidy Tools Manifesto.” <em>CRAN Vignette</em>.</p>
</div>
<div>
<p>Wickham, Hadley, and Garrett Grolemund. 2016. <em>R for Data Science: Import, Tidy, Transform, Visualize, and Model Data</em>. sebastopol: O’Reilly Media.</p>
</div>
<div>
<p>Wilkinson, Mark D, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. “The Fair Guiding Principles for Scientific Data Management and Stewardship.” <em>Scientific Data</em> 3.</p>
</div>
<div>
<p>Willekens, Frans. 2013. “Chronological Objects in Demographic Research.” <em>Demographic Research</em> 28: 649–80.</p>
</div>
<div>
<p>Wilson, G. 2014. “Software Carpentry: Lessons Learned.” <em>F1000Research</em> 3: 62–62.</p>
</div>
<div>
<p>Yin, Tengfei, Dianne Cook, and Michael Lawrence. 2012. “Ggbio: An R Package for Extending the Grammar of Graphics for Genomic Data.” <em>Genome Biology</em> 13 (8): R77.</p>
</div>
<div>
<p>Zeeberg, Barry R, Joseph Riss, David W Kane, Kimberly J Bussey, Edward Uchio, W Marston Linehan, J Carl Barrett, and John N Weinstein. 2004. “Mistaken Identifiers: Gene Name Errors Can Be Introduced Inadvertently When Using Excel in Bioinformatics.” <em>BMC Bioinformatics</em> 5 (1): 80.</p>
</div>
<div>
<p>Ziemann, Mark, Yotam Eren, and Assam El-Osta. 2016. “Gene Name Errors Are Widespread in the Scientific Literature.” <em>Genome Biology</em> 17 (1): 177.</p>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="experimental-data-preprocessing.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/20-example_protocol.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["improve_repro.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
