<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing</title>
  <meta name="description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 References | Improving the Reproducibility of Experimental Data Recording and Pre-Processing" />
  
  <meta name="twitter:description" content="Online book with modules for improving the reproducibility of experimental data recording and preprocessing." />
  

<meta name="author" content="Brooke Anderson, Michael Lyons, Mercedes Gonzalez-Juarrero, Marcela Henao-Tamayo, and Gregory Robertson" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="experimental-data-preprocessing.html"/>

<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Data Visualization in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.1</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html"><i class="fa fa-check"></i><b>2</b> Experimental Data Recording</a>
<ul>
<li class="chapter" data-level="2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module1"><i class="fa fa-check"></i><b>2.1</b> Separating data recording and analysis</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-versus-data-analysis"><i class="fa fa-check"></i><b>2.1.1</b> Data recording versus data analysis</a></li>
<li class="chapter" data-level="2.1.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#hazards-of-combining-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.2</b> Hazards of combining recording and analysis</a></li>
<li class="chapter" data-level="2.1.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#approaches-to-separate-recording-and-analysis"><i class="fa fa-check"></i><b>2.1.3</b> Approaches to separate recording and analysis</a></li>
<li class="chapter" data-level="2.1.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions"><i class="fa fa-check"></i><b>2.1.4</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module2"><i class="fa fa-check"></i><b>2.2</b> Principles and power of structured data formats</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#data-recording-standards"><i class="fa fa-check"></i><b>2.2.1</b> Data recording standards</a></li>
<li class="chapter" data-level="2.2.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#defining-data-standards-for-a-research-group"><i class="fa fa-check"></i><b>2.2.2</b> Defining data standards for a research group</a></li>
<li class="chapter" data-level="2.2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#two-dimensional-structured-data-format"><i class="fa fa-check"></i><b>2.2.3</b> Two-dimensional structured data format</a></li>
<li class="chapter" data-level="2.2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#saving-two-dimensional-structured-data-in-plain-text-file-formats"><i class="fa fa-check"></i><b>2.2.4</b> Saving two-dimensional structured data in plain text file formats</a></li>
<li class="chapter" data-level="2.2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#occassions-for-more-complex-data-structures-and-file-formats"><i class="fa fa-check"></i><b>2.2.5</b> Occassions for more complex data structures and file formats</a></li>
<li class="chapter" data-level="2.2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#levels-of-standardizationresearch-group-to-research-community"><i class="fa fa-check"></i><b>2.2.6</b> Levels of standardization—research group to research community</a></li>
<li class="chapter" data-level="2.2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise"><i class="fa fa-check"></i><b>2.2.7</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module3"><i class="fa fa-check"></i><b>2.3</b> The ‘tidy’ data format</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-makes-data-tidy"><i class="fa fa-check"></i><b>2.3.1</b> What makes data “tidy?”</a></li>
<li class="chapter" data-level="2.3.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#why-make-your-data-tidy"><i class="fa fa-check"></i><b>2.3.2</b> Why make your data “tidy?”</a></li>
<li class="chapter" data-level="2.3.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-tidyverse-tools-with-data-in-the-tidy-data-format"><i class="fa fa-check"></i><b>2.3.3</b> Using tidyverse tools with data in the “tidy data” format</a></li>
<li class="chapter" data-level="2.3.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz"><i class="fa fa-check"></i><b>2.3.4</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module4"><i class="fa fa-check"></i><b>2.4</b> Designing templates for “tidy” data collection</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#exampledata-on-rate-of-bacterial-growth"><i class="fa fa-check"></i><b>2.4.1</b> Example—Data on rate of bacterial growth</a></li>
<li class="chapter" data-level="2.4.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#features-that-make-data-collection-templates-untidy"><i class="fa fa-check"></i><b>2.4.2</b> Features that make data collection templates “untidy”</a></li>
<li class="chapter" data-level="2.4.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#converting-to-a-tidier-format-for-data-collection-templates"><i class="fa fa-check"></i><b>2.4.3</b> Converting to a “tidier” format for data collection templates</a></li>
<li class="chapter" data-level="2.4.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#learning-more-about-tidy-data-collection-in-the-laboratory"><i class="fa fa-check"></i><b>2.4.4</b> Learning more about tidy data collection in the laboratory</a></li>
<li class="chapter" data-level="2.4.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#creating-the-rules-for-collecting-data-in-the-same-time-each-time"><i class="fa fa-check"></i><b>2.4.5</b> Creating the rules for collecting data in the same time each time</a></li>
<li class="chapter" data-level="2.4.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1"><i class="fa fa-check"></i><b>2.4.6</b> Subsection 1</a></li>
<li class="chapter" data-level="2.4.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#dont-repeat-yourself"><i class="fa fa-check"></i><b>2.4.7</b> Don’t Repeat Yourself!</a></li>
<li class="chapter" data-level="2.4.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#dont-repeat-your-report-writing"><i class="fa fa-check"></i><b>2.4.8</b> Don’t repeat your report-writing!</a></li>
<li class="chapter" data-level="2.4.9" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#automating-reports"><i class="fa fa-check"></i><b>2.4.9</b> Automating reports</a></li>
<li class="chapter" data-level="2.4.10" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#scripts-and-automated-reports-as-simple-pipelines"><i class="fa fa-check"></i><b>2.4.10</b> Scripts and automated reports as simple pipelines</a></li>
<li class="chapter" data-level="2.4.11" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-1"><i class="fa fa-check"></i><b>2.4.11</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module5"><i class="fa fa-check"></i><b>2.5</b> Example: Creating a template for “tidy” data collection</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#example-datadata-on-rate-of-bacterial-growth"><i class="fa fa-check"></i><b>2.5.1</b> Example data—Data on rate of bacterial growth</a></li>
<li class="chapter" data-level="2.5.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#limiting-the-template-to-the-collection-of-data"><i class="fa fa-check"></i><b>2.5.2</b> Limiting the template to the collection of data</a></li>
<li class="chapter" data-level="2.5.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-sensible-choices-about-rows-and-columns"><i class="fa fa-check"></i><b>2.5.3</b> Making sensible choices about rows and columns</a></li>
<li class="chapter" data-level="2.5.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#avoiding-problematic-characters-or-formatting"><i class="fa fa-check"></i><b>2.5.4</b> Avoiding problematic characters or formatting</a></li>
<li class="chapter" data-level="2.5.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#moving-further-data-analysis-to-later-in-the-pipeline"><i class="fa fa-check"></i><b>2.5.5</b> Moving further data analysis to later in the pipeline</a></li>
<li class="chapter" data-level="2.5.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#exampledata-on-rate-of-bacterial-growth-1"><i class="fa fa-check"></i><b>2.5.6</b> Example—Data on rate of bacterial growth</a></li>
<li class="chapter" data-level="2.5.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#exampledata-on-bacteria-colony-forming-units"><i class="fa fa-check"></i><b>2.5.7</b> Example—Data on bacteria colony forming units</a></li>
<li class="chapter" data-level="2.5.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#exampledata-from-multiple-related-experiments"><i class="fa fa-check"></i><b>2.5.8</b> Example—Data from multiple related experiments</a></li>
<li class="chapter" data-level="2.5.9" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#issues-with-these-data-sets"><i class="fa fa-check"></i><b>2.5.9</b> Issues with these data sets</a></li>
<li class="chapter" data-level="2.5.10" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#final-tidy-examples"><i class="fa fa-check"></i><b>2.5.10</b> Final “tidy” examples</a></li>
<li class="chapter" data-level="2.5.11" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#options-for-recording-tidy-data"><i class="fa fa-check"></i><b>2.5.11</b> Options for recording tidy data</a></li>
<li class="chapter" data-level="2.5.12" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#examples-of-how-tidy-data-can-be-easily-analyzed-visualized"><i class="fa fa-check"></i><b>2.5.12</b> Examples of how “tidy” data can be easily analyzed / visualized</a></li>
<li class="chapter" data-level="2.5.13" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-1"><i class="fa fa-check"></i><b>2.5.13</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module6"><i class="fa fa-check"></i><b>2.6</b> Power of using a single structured ‘Project’ directory for storing and tracking research project files</a></li>
<li class="chapter" data-level="2.7" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#details-of-the-data-from-the-set-of-example-studies"><i class="fa fa-check"></i><b>2.7</b> Details of the data from the set of example studies</a></li>
<li class="chapter" data-level="2.8" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#creating-an-organized-directory-for-project-files"><i class="fa fa-check"></i><b>2.8</b> Creating an organized directory for project files</a></li>
<li class="chapter" data-level="2.9" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-the-project-directory-and-r-project"><i class="fa fa-check"></i><b>2.9</b> Making the project directory and R Project</a>
<ul>
<li class="chapter" data-level="2.9.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#organizing-project-files-through-the-file-system"><i class="fa fa-check"></i><b>2.9.1</b> Organizing project files through the file system</a></li>
<li class="chapter" data-level="2.9.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#organizing-files-within-a-project-directory"><i class="fa fa-check"></i><b>2.9.2</b> Organizing files within a project directory</a></li>
<li class="chapter" data-level="2.9.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#using-rstudio-projects-with-project-file-directories"><i class="fa fa-check"></i><b>2.9.3</b> Using RStudio Projects with project file directories</a></li>
<li class="chapter" data-level="2.9.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-1"><i class="fa fa-check"></i><b>2.9.4</b> Subsection 1</a></li>
<li class="chapter" data-level="2.9.5" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2"><i class="fa fa-check"></i><b>2.9.5</b> Subsection 2</a></li>
<li class="chapter" data-level="2.9.6" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#practice-quiz-1"><i class="fa fa-check"></i><b>2.9.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module7"><i class="fa fa-check"></i><b>2.10</b> Creating ‘Project’ templates</a>
<ul>
<li class="chapter" data-level="2.10.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-an-existing-file-directory-an-rstudio-project"><i class="fa fa-check"></i><b>2.10.1</b> Making an existing file directory an RStudio Project</a></li>
<li class="chapter" data-level="2.10.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#making-an-rstudio-project-template"><i class="fa fa-check"></i><b>2.10.2</b> Making an RStudio Project Template</a></li>
<li class="chapter" data-level="2.10.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-2"><i class="fa fa-check"></i><b>2.10.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.11" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module8"><i class="fa fa-check"></i><b>2.11</b> Example: Creating a ‘Project’ template</a>
<ul>
<li class="chapter" data-level="2.11.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-1-2"><i class="fa fa-check"></i><b>2.11.1</b> Subsection 1</a></li>
<li class="chapter" data-level="2.11.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#subsection-2-1"><i class="fa fa-check"></i><b>2.11.2</b> Subsection 2</a></li>
<li class="chapter" data-level="2.11.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-2"><i class="fa fa-check"></i><b>2.11.3</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="2.12" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module9"><i class="fa fa-check"></i><b>2.12</b> Harnessing version control for transparent data recording</a>
<ul>
<li class="chapter" data-level="2.12.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-is-version-control"><i class="fa fa-check"></i><b>2.12.1</b> What is version control?</a></li>
<li class="chapter" data-level="2.12.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#recording-data-in-the-laboratoryfrom-paper-to-computers"><i class="fa fa-check"></i><b>2.12.2</b> Recording data in the laboratory—from paper to computers</a></li>
<li class="chapter" data-level="2.12.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-3"><i class="fa fa-check"></i><b>2.12.3</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.13" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module10"><i class="fa fa-check"></i><b>2.13</b> Enhance the reproducibility of collaborative research with version control platforms</a>
<ul>
<li class="chapter" data-level="2.13.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#what-are-version-control-platforms"><i class="fa fa-check"></i><b>2.13.1</b> What are version control platforms?</a></li>
<li class="chapter" data-level="2.13.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#why-use-version-control-platforms"><i class="fa fa-check"></i><b>2.13.2</b> Why use version control platforms?</a></li>
<li class="chapter" data-level="2.13.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#how-to-use-github"><i class="fa fa-check"></i><b>2.13.3</b> How to use GitHub</a></li>
<li class="chapter" data-level="2.13.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#discussion-questions-4"><i class="fa fa-check"></i><b>2.13.4</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="2.14" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#module11"><i class="fa fa-check"></i><b>2.14</b> Using git and GitLab to implement version control</a>
<ul>
<li class="chapter" data-level="2.14.1" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#how-to-use-version-control"><i class="fa fa-check"></i><b>2.14.1</b> How to use version control</a></li>
<li class="chapter" data-level="2.14.2" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#leveraging-git-and-github-as-a-project-director"><i class="fa fa-check"></i><b>2.14.2</b> Leveraging git and GitHub as a project director</a></li>
<li class="chapter" data-level="2.14.3" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#leveraging-git-and-github-as-a-scientist-who-programs"><i class="fa fa-check"></i><b>2.14.3</b> Leveraging git and GitHub as a scientist who programs</a></li>
<li class="chapter" data-level="2.14.4" data-path="experimental-data-recording.html"><a href="experimental-data-recording.html#applied-exercise-3"><i class="fa fa-check"></i><b>2.14.4</b> Applied exercise</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html"><i class="fa fa-check"></i><b>3</b> Experimental Data Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module12"><i class="fa fa-check"></i><b>3.1</b> Principles and benefits of scripted pre-processing of experimental data</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#what-is-pre-processing"><i class="fa fa-check"></i><b>3.1.1</b> What is pre-processing?</a></li>
<li class="chapter" data-level="3.1.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#approaches-to-simple-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.2</b> Approaches to simple preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#approaches-to-more-complex-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.3</b> Approaches to more complex preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#scripting-preprocessing-tasks"><i class="fa fa-check"></i><b>3.1.4</b> Scripting preprocessing tasks</a></li>
<li class="chapter" data-level="3.1.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#potential-quotes"><i class="fa fa-check"></i><b>3.1.5</b> Potential quotes</a></li>
<li class="chapter" data-level="3.1.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#discussion-questions-5"><i class="fa fa-check"></i><b>3.1.6</b> Discussion questions</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module13"><i class="fa fa-check"></i><b>3.2</b> Introduction to scripted data pre-processing in R</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#compiled-versus-interpreted-programming-languages"><i class="fa fa-check"></i><b>3.2.1</b> Compiled versus interpreted programming languages</a></li>
<li class="chapter" data-level="3.2.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#code-scripts-versus-interactive-coding"><i class="fa fa-check"></i><b>3.2.2</b> Code scripts versus interactive coding</a></li>
<li class="chapter" data-level="3.2.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#process-of-building-a-code-script"><i class="fa fa-check"></i><b>3.2.3</b> Process of building a code script</a></li>
<li class="chapter" data-level="3.2.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-4"><i class="fa fa-check"></i><b>3.2.4</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module14"><i class="fa fa-check"></i><b>3.3</b> Simplify scripted pre-processing through R’s ‘tidyverse’ tools</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#limitations-of-object-oriented-programming"><i class="fa fa-check"></i><b>3.3.1</b> Limitations of object-oriented programming</a></li>
<li class="chapter" data-level="3.3.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-tidyverse-approach"><i class="fa fa-check"></i><b>3.3.2</b> The “tidyverse” approach</a></li>
<li class="chapter" data-level="3.3.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#how-to-tidyverse"><i class="fa fa-check"></i><b>3.3.3</b> How to “tidyverse”</a></li>
<li class="chapter" data-level="3.3.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-1-3"><i class="fa fa-check"></i><b>3.3.4</b> Subsection 1</a></li>
<li class="chapter" data-level="3.3.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-2"><i class="fa fa-check"></i><b>3.3.5</b> Subsection 2</a></li>
<li class="chapter" data-level="3.3.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-2"><i class="fa fa-check"></i><b>3.3.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module15"><i class="fa fa-check"></i><b>3.4</b> Complex data types in experimental data pre-processing</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction"><i class="fa fa-check"></i><b>3.4.1</b> Introduction</a></li>
<li class="chapter" data-level="3.4.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#data-structures-and-the-dataframe"><i class="fa fa-check"></i><b>3.4.2</b> Data structures and the dataframe</a></li>
<li class="chapter" data-level="3.4.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#more-complex-data-structures-in-r"><i class="fa fa-check"></i><b>3.4.3</b> More complex data structures in R</a></li>
<li class="chapter" data-level="3.4.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#limitations-to-complex-data-structures"><i class="fa fa-check"></i><b>3.4.4</b> Limitations to complex data structures</a></li>
<li class="chapter" data-level="3.4.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#complex-versus-simple-structures-for-storing-data"><i class="fa fa-check"></i><b>3.4.5</b> Complex versus simple structures for storing data</a></li>
<li class="chapter" data-level="3.4.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#practice-quiz-3"><i class="fa fa-check"></i><b>3.4.6</b> Practice quiz</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module16"><i class="fa fa-check"></i><b>3.5</b> Complex data types in R and Bioconductor</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#rs-list-data-structure-and-list-based-structures"><i class="fa fa-check"></i><b>3.5.1</b> R’s list data structure and list-based structures</a></li>
<li class="chapter" data-level="3.5.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#exploring-and-extracting-data-from-r-list-data-structures"><i class="fa fa-check"></i><b>3.5.2</b> Exploring and extracting data from R list data structures</a></li>
<li class="chapter" data-level="3.5.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#interfacing-between-object-based-and-tidyverse-workflows"><i class="fa fa-check"></i><b>3.5.3</b> Interfacing between object-based and tidyverse workflows</a></li>
<li class="chapter" data-level="3.5.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#extras"><i class="fa fa-check"></i><b>3.5.4</b> Extras</a></li>
<li class="chapter" data-level="3.5.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-3"><i class="fa fa-check"></i><b>3.5.5</b> Subsection 2</a></li>
<li class="chapter" data-level="3.5.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-5"><i class="fa fa-check"></i><b>3.5.6</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module17"><i class="fa fa-check"></i><b>3.6</b> Example: Converting from complex to ‘tidy’ data formats</a>
<ul>
<li class="chapter" data-level="3.6.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#combining-bioconductor-and-tidyverse-approaches-in-a-workflow"><i class="fa fa-check"></i><b>3.6.1</b> Combining Bioconductor and tidyverse approaches in a workflow</a></li>
<li class="chapter" data-level="3.6.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-biobroom-package"><i class="fa fa-check"></i><b>3.6.2</b> The <code>biobroom</code> package</a></li>
<li class="chapter" data-level="3.6.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#the-ggbio-package"><i class="fa fa-check"></i><b>3.6.3</b> The <code>ggbio</code> package</a></li>
<li class="chapter" data-level="3.6.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#subsection-2-4"><i class="fa fa-check"></i><b>3.6.4</b> Subsection 2</a></li>
<li class="chapter" data-level="3.6.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-6"><i class="fa fa-check"></i><b>3.6.5</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module18"><i class="fa fa-check"></i><b>3.7</b> Introduction to reproducible data pre-processing protocols</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introducing-reproducible-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.7.1</b> Introducing reproducible data pre-processing protocols</a></li>
<li class="chapter" data-level="3.7.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#using-knitted-documents-for-protocols"><i class="fa fa-check"></i><b>3.7.2</b> Using knitted documents for protocols</a></li>
<li class="chapter" data-level="3.7.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#how-knitted-documents-work"><i class="fa fa-check"></i><b>3.7.3</b> How knitted documents work</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module19"><i class="fa fa-check"></i><b>3.8</b> RMarkdown for creating reproducible data pre-processing protocols</a>
<ul>
<li class="chapter" data-level="3.8.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#creating-knitted-documents-in-r"><i class="fa fa-check"></i><b>3.8.1</b> Creating knitted documents in R</a></li>
<li class="chapter" data-level="3.8.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#creating-and-exploring-an-rmarkdown-file"><i class="fa fa-check"></i><b>3.8.2</b> Creating and exploring an Rmarkdown file</a></li>
<li class="chapter" data-level="3.8.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#more-advanced-rmarkdown-functionality"><i class="fa fa-check"></i><b>3.8.3</b> More advanced Rmarkdown functionality</a></li>
<li class="chapter" data-level="3.8.4" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#learning-more-about-rmarkdown."><i class="fa fa-check"></i><b>3.8.4</b> Learning more about Rmarkdown.</a></li>
<li class="chapter" data-level="3.8.5" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-7"><i class="fa fa-check"></i><b>3.8.5</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="3.9" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#module20"><i class="fa fa-check"></i><b>3.9</b> Example: Creating a reproducible data pre-processing protocol</a>
<ul>
<li class="chapter" data-level="3.9.1" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#introduction-and-example-data"><i class="fa fa-check"></i><b>3.9.1</b> Introduction and example data</a></li>
<li class="chapter" data-level="3.9.2" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#advice-on-designing-a-pre-processing-protocol"><i class="fa fa-check"></i><b>3.9.2</b> Advice on designing a pre-processing protocol</a></li>
<li class="chapter" data-level="3.9.3" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#writing-data-pre-processing-protocols"><i class="fa fa-check"></i><b>3.9.3</b> Writing data pre-processing protocols</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="experimental-data-preprocessing.html"><a href="experimental-data-preprocessing.html#applied-exercise-8"><i class="fa fa-check"></i><b>3.10</b> Applied exercise</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>4</b> References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Improving the Reproducibility of Experimental Data Recording and Pre-Processing</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> References</h1>

<div id="refs" class="references csl-bib-body hanging-indent">
<div class="csl-entry">
Al-JunDi, AzzAm, and SAlAh SAkkA. 2016. <span>“Protocol Writing in Clinical Research.”</span> <em>Journal of Clinical and Diagnostic Research: JCDR</em> 10 (11): ZE10.
</div>
<div class="csl-entry">
AlTarawneh, Ghada, and Simon Thorne. 2017. <span>“A Pilot Study Exploring Spreadsheet Risk in Scientific Research.”</span> <em>arXiv Preprint arXiv:1703.09785</em>.
</div>
<div class="csl-entry">
Altman, Douglas G., and J. Martin Bland. 1997. <span>“Statistical Notes: Units of Analysis.”</span> <em>BMJ</em> 314: 1874.
</div>
<div class="csl-entry">
Altschul, Stephen, Barry Demchak, Richard Durbin, Robert Gentleman, Martin Krzywinski, Heng Li, Anton Nekrutenko, et al. 2013. <span>“The Anatomy of Successful Computational Biology Software.”</span> <em>Nature Biotechnology</em> 31 (10): 894–97.
</div>
<div class="csl-entry">
Anderson, Nicholas R, E Sally Lee, J Scott Brockenbrough, Mark E Minie, Sherrilynne Fuller, James Brinkley, and Peter Tarczy-Hornoch. 2007. <span>“Issues in Biomedical Research Data Management and Analysis: Needs and Barriers.”</span> <em>Journal of the American Medical Informatics Association</em> 14 (4): 478–88.
</div>
<div class="csl-entry">
Arnold, Taylor. 2017. <span>“<span class="nocase">A Tidy Data Model for Natural Language Processing using cleanNLP</span>.”</span> <em><span>The R Journal</span></em> 9 (2): 248–67.
</div>
<div class="csl-entry">
Barga, Roger, Bill Howe, David Beck, Stuart Bowers, William Dobyns, Winston Haynes, Roger Higdon, et al. 2011. <span>“Bioinformatics and Data-Intensive Scientific Discovery in the Beginning of the 21st Century.”</span> <em>Omics: A Journal of Integrative Biology</em> 15 (4): 199–201.
</div>
<div class="csl-entry">
Bass, Andrew J., David G. Robinson, Steve Lianoglou, Emily Nelson, John D. Storey, and with contributions from Laurent Gatto. 2020. <em>Biobroom: Turn Bioconductor Objects into Tidy Data Frames</em>. <a href="https://github.com/StoreyLab/biobroom">https://github.com/StoreyLab/biobroom</a>.
</div>
<div class="csl-entry">
Baumer, Ben. 2015. <span>“A Data Science Course for Undergraduates: Thinking with Data.”</span> <em>The American Statistician</em> 69 (4): 334–42.
</div>
<div class="csl-entry">
Baumer, Benjamin S, Daniel T Kaplan, and Nicholas J Horton. 2017. <em>Modern Data Science with r</em>. Boca Raton: CRC Press.
</div>
<div class="csl-entry">
Ben-David, Avishai, and Charles E Davidson. 2014. <span>“Estimation Method for Serial Dilution Experiments.”</span> <em>Journal of Microbiological Methods</em> 107: 214–21.
</div>
<div class="csl-entry">
Birch, David, David Lyford-Smith, and Yike Guo. 2018. <span>“The Future of Spreadsheets in the Big Data Era.”</span> <em>arXiv Preprint arXiv:1801.10231</em>.
</div>
<div class="csl-entry">
Blischak, John D, Emily R Davenport, and Greg Wilson. 2016. <span>“A Quick Introduction to Version Control with Git and GitHub.”</span> <em>PLoS Computational Biology</em> 12 (1).
</div>
<div class="csl-entry">
Brazma, Alvis, Maria Krestyaninova, and Ugis Sarkans. 2006. <span>“Standards for Systems Biology.”</span> <em>Nature Reviews Genetics</em> 7 (8): 593.
</div>
<div class="csl-entry">
Broman, Karl W, and Kara H Woo. 2018. <span>“Data Organization in Spreadsheets.”</span> <em>The American Statistician</em> 72 (1): 2–10.
</div>
<div class="csl-entry">
Brown, Zack. 2018. <span>“A Git Origin Story.”</span> <em>Linux Journal</em>.
</div>
<div class="csl-entry">
Bryan, Jennifer. 2018. <span>“Excuse Me, Do You Have a Moment to Talk about Version Control?”</span> <em>The American Statistician</em> 72 (1): 20–27.
</div>
<div class="csl-entry">
Bryan, Jennifer, and Hadley Wickham. 2017. <span>“Data Science: A Three Ring Circus or a Big Tent?”</span> <em>Journal of Computational and Graphical Statistics</em> 26 (4): 784–85.
</div>
<div class="csl-entry">
Buffalo, Vince. 2015. <em>Bioinformatics Data Skills: Reproducible and Robust Research with Open Source Tools</em>. " O’Reilly Media, Inc.".
</div>
<div class="csl-entry">
Burns, Patrick. 2011. <em>The r Inferno</em>. Lulu. com.
</div>
<div class="csl-entry">
Butler, Declan. 2005. <span>“Electronic Notebooks: A New Leaf.”</span> <em>Nature</em> 436 (7047): 20–22.
</div>
<div class="csl-entry">
Campbell-Kelly, Martin. 2007. <span>“Number Crunching Without Programming: The Evolution of Spreadsheet Usability.”</span> <em>IEEE Annals of the History of Computing</em> 29 (3): 6–19.
</div>
<div class="csl-entry">
Chambers, JM. 2006. <span>“How S4 Methods Work.”</span> Technical report.
</div>
<div class="csl-entry">
Chen, Yunshun, Davis McCarthy, Mark Robinson, and Gordon K Smyth. 2014. <span>“edgeR: Differential Expression Analysis of Digital Gene Expression Data User’s Guide.”</span> <em>Bioconductor User’s Guide. Available Online: Http://Www. Bioconductor. Org/Packages/Release/Bioc/Vignettes/edgeR/Inst/Doc/edgeRUsersGuide.pdf (Accessed on 15 February 2021)</em>.
</div>
<div class="csl-entry">
Creeth, Richard. 1985. <span>“Microcomputer Spreadsheets: Their Uses and Abuses.”</span> <em>Journal of Accountancy (Pre-1986)</em> 159 (000006): 90.
</div>
<div class="csl-entry">
Edwards, Paul N, Matthew S Mayernik, Archer L Batcheller, Geoffrey C Bowker, and Christine L Borgman. 2011. <span>“Science Friction: Data, Metadata, and Collaboration.”</span> <em>Social Studies of Science</em> 41 (5): 667–90.
</div>
<div class="csl-entry">
Ellis, Shannon E, and Jeffrey T Leek. 2018. <span>“How to Share Data for Collaboration.”</span> <em>The American Statistician</em> 72 (1): 53–57.
</div>
<div class="csl-entry">
Ford, Paul. 2014. <span>“On File Formats, Very Briefly.”</span> <em>The Manual</em>. <a href="https://themanual.org/read/issues/4/paul-ford/article">https://themanual.org/read/issues/4/paul-ford/article</a>.
</div>
<div class="csl-entry">
———. 2015. <span>“I Dreamed of a Perfect Database.”</span> <em>New Republic</em>. <a href="https://newrepublic.com/article/124425/dreamed-perfect-database">https://newrepublic.com/article/124425/dreamed-perfect-database</a>.
</div>
<div class="csl-entry">
Fox, Amy, Taru S Dutt, Burton Karger, Mauricio Rojas, Andrés Obregón-Henao, G Brooke Anderson, and Marcela Henao-Tamayo. 2020. <span>“Cyto-Feature Engineering: A Pipeline for Flow Cytometry Analysis to Uncover Immune Populations and Associations with Disease.”</span> <em>Scientific Reports</em> 10 (1): 1–12.
</div>
<div class="csl-entry">
Gatto, Laurent. 2013. <span>“MSnbase Development.”</span>
</div>
<div class="csl-entry">
Gentleman, Robert. 2008. <em>R Programming for Bioinformatics</em>. CRC Press.
</div>
<div class="csl-entry">
Ghosh, Samik, Yukiko Matsuoka, Yoshiyuki Asai, Kun-Yi Hsin, and Hiroaki Kitano. 2011. <span>“Software for Systems Biology: From Tools to Integrated Platforms.”</span> <em>Nature Reviews Genetics</em> 12 (12): 821.
</div>
<div class="csl-entry">
Giles, Jim. 2012. <span>“The Digital Lab: Lab-Management Software and Electronic Notebooks Are Here–and This Time, It’s More Than Just Talk.”</span> <em>Nature</em> 481 (7382): 430–32.
</div>
<div class="csl-entry">
Goodman, Alyssa, Alberto Pepe, Alexander W Blocker, Christine L Borgman, Kyle Cranmer, Merce Crosas, Rosanne Di Stefano, et al. 2014. <span>“Ten Simple Rules for the Care and Feeding of Scientific Data.”</span> Public Library of Science.
</div>
<div class="csl-entry">
Grune, Dick. 1986. <span>“Concurrent Verions System, a Method for Independent Cooperation.”</span> <em>Technical Report, Vrije Universiteit</em>.
</div>
<div class="csl-entry">
Hammer, Paul, Michaela S Banck, Ronny Amberg, Cheng Wang, Gabriele Petznick, Shujun Luo, Irina Khrebtukova, Gary P Schroth, Peter Beyerlein, and Andreas S Beutler. 2010. <span>“mRNA-Seq with Agnostic Splice Site Discovery for Nervous System Transcriptomics Tested in Chronic Pain.”</span> <em>Genome Research</em> 20 (6): 847–60.
</div>
<div class="csl-entry">
Hermans, Felienne, Bas Jansen, Sohon Roy, Efthimia Aivaloglou, Alaaeddin Swidan, and David Hoepelman. 2016. <span>“Spreadsheets Are Code: An Overview of Software Engineering Approaches Applied to Spreadsheets.”</span> In <em>2016 IEEE 23rd International Conference on Software Analysis, Evolution, and Reengineering (SANER)</em>, 5:56–65. IEEE.
</div>
<div class="csl-entry">
Hermans, Felienne, and Emerson Murphy-Hill. 2015. <span>“Enron’s Spreadsheets and Related Emails: A Dataset and Analysis.”</span> In <em>2015 IEEE/ACM 37th IEEE International Conference on Software Engineering</em>, 2:7–16. IEEE.
</div>
<div class="csl-entry">
Hicks, Stephanie C, and Rafael A Irizarry. 2017. <span>“A Guide to Teaching Data Science.”</span> <em>The American Statistician</em> In Press. <a href="https://doi.org/10.1080/00031305.2017.1356747">https://doi.org/10.1080/00031305.2017.1356747</a>.
</div>
<div class="csl-entry">
Hicks, Stephanie C, Ruoxi Liu, Yuwei Ni, Elizabeth Purdom, and Davide Risso. 2021. <span>“Mbkmeans: Fast Clustering for Single Cell Data Using Mini-Batch k-Means.”</span> <em>PLOS Computational Biology</em> 17 (1): e1008625.
</div>
<div class="csl-entry">
Holmes, Susan, and Wolfgang Huber. 2018. <em>Modern Statistics for Modern Biology</em>. Cambridge University Press.
</div>
<div class="csl-entry">
Hsieh, TC, KH Ma, and Anne Chao. 2016. <span>“<span class="nocase">iNEXT: an R package for rarefaction and extrapolation of species diversity (Hill numbers)</span>.”</span> <em>Methods in Ecology and Evolution</em> 7 (12): 1451–56.
</div>
<div class="csl-entry">
Huber, W., V. J. Carey, R. Gentleman, S. Anders, M. Carlson, B. S. Carvalho, H. C. Bravo, et al. 2015. <span>“<span>O</span>rchestrating High-Throughput Genomic Analysis with <span>B</span>ioconductor.”</span> <em>Nature Methods</em> 12 (2): 115–21. <a href="http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html">http://www.nature.com/nmeth/journal/v12/n2/full/nmeth.3252.html</a>.
</div>
<div class="csl-entry">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. <span>“Orchestrating High-Throughput Genomic Analysis with Bioconductor.”</span> <em>Nature Methods</em> 12 (2): 115.
</div>
<div class="csl-entry">
Hunt, Andrew, David Thomas, and Ward Cunningham. 2000. <em>The Pragmatic Programmer: From Journeyman to Master</em>. Addison-Wesley Professional.
</div>
<div class="csl-entry">
Irizarry, Rafael A., and Michael I. Love. 2016. <em>Data Analysis for the Life Sciences with r</em>. Chapman; Hall.
</div>
<div class="csl-entry">
Irving, Francis. 2011. <span>“Astonishments, Ten, in the History of Version Control.”</span> <a href="https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/">https://www.flourish.org/2011/12/astonishments-ten-in-the-history-of-version-control/</a>.
</div>
<div class="csl-entry">
Janssens, Jeroen. 2014. <em>Data Science at the Command Line: Facing the Future with Time-Tested Tools</em>. " O’Reilly Media, Inc.".
</div>
<div class="csl-entry">
Kaplan, Daniel. 2018. <span>“Teaching Stats for Data Science.”</span> <em>The American Statistician</em> 72 (1): 89–96.
</div>
<div class="csl-entry">
Keller, Sallie, Gizem Korkmaz, Mark Orr, Aaron Schroeder, and Stephanie Shipp. 2017. <span>“The Evolution of Data Quality: Understanding the Transdisciplinary Origins of Data Quality Concepts and Approaches.”</span> <em>Annu. Rev. Stat. Appl</em> 4: 85–108.
</div>
<div class="csl-entry">
Kernighan, Brian W. 2011. <em>D Is for Digital: What a Well-Informed Person Should Know about Computers and Communications</em>. CreateSpace Independent Publishing Platform.
</div>
<div class="csl-entry">
Kernighan, Brian W, and Rob Pike. 1984. <em>The UNIX Programming Environment</em>. Vol. 270. Prentice-Hall Englewood Cliffs, NJ.
</div>
<div class="csl-entry">
Klemens, Ben. 2014. <em>21st Century c: C Tips from the New School</em>. " O’Reilly Media, Inc.".
</div>
<div class="csl-entry">
Krishnan, Sanjay, Daniel Haas, Michael J Franklin, and Eugene Wu. 2016. <span>“Towards Reliable Interactive Data Cleaning: A User Survey and Recommendations.”</span> In <em>Proceedings of the Workshop on Human-in-the-Loop Data Analytics</em>, 9. ACM.
</div>
<div class="csl-entry">
Kwok, Roberta. 2018. <span>“Lab Notebooks Go Digital.”</span> <em>Nature</em> 560 (7717): 269–70.
</div>
<div class="csl-entry">
Lawrence, Michael, and Martin Morgan. 2014. <span>“Scalable Genomics with r and Bioconductor.”</span> <em>Statistical Science: A Review Journal of the Institute of Mathematical Statistics</em> 29 (2): 214.
</div>
<div class="csl-entry">
Leek, Jeff. 2012. <span>“Toward Tidy Data Analysis.”</span> <em>Simply Statistics</em>.
</div>
<div class="csl-entry">
LEIPS, JEFF. 2010. <span>“At the Helm: Leading Your Laboratory.”</span> <em>Genetics Research</em> 92 (4): 325–26.
</div>
<div class="csl-entry">
Levy, Steven. 1984. <span>“A Spreadsheet Way of Knowledge.”</span> <em>Harpers</em> 269: 58–64.
</div>
<div class="csl-entry">
Lowndes, Julia S Stewart, Benjamin D Best, Courtney Scarborough, Jamie C Afflerbach, Melanie R Frazier, Casey C O’Hara, Ning Jiang, and Benjamin S Halpern. 2017. <span>“Our Path to Better Science in Less Time Using Open Data Science Tools.”</span> <em>Nature Ecology &amp; Evolution</em> 1 (6): 160.
</div>
<div class="csl-entry">
Lynch, Clifford. 2008. <span>“Big Data: How Do Your Data Grow?”</span> <em>Nature</em> 455 (7209): 28.
</div>
<div class="csl-entry">
Majumder, Erica L-W, Elizabeth M Billings, H Paul Benton, Richard L Martin, Amelia Palermo, Carlos Guijas, Markus M Rinschen, et al. 2021. <span>“Cognitive Analysis of Metabolomics Data for Systems Biology.”</span> <em>Nature Protocols</em> 16 (3): 1376–1418.
</div>
<div class="csl-entry">
Marwick, Ben, Carl Boettiger, and Lincoln Mullen. 2018. <span>“Packaging Data Analytical Work Reproducibly Using <span>R</span> (and Friends).”</span> <em>The American Statistician</em> 72 (1): 80–88.
</div>
<div class="csl-entry">
Mascarelli, Amanda. 2014. <span>“Research Tools: Jump Off the Page.”</span> <em>Nature</em> 507 (7493): 523–25.
</div>
<div class="csl-entry">
McCullough, B D, and Berry Wilson. 2002. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2000 and Excel XP.”</span> <em>Computational Statistics &amp; Data Analysis</em> 40 (4): 713–21.
</div>
<div class="csl-entry">
McCullough, BD. 2001. <span>“Does Microsoft Fix Errors in Excel?”</span> In <em>Proceedings of the 2001 Joint Statistical Meetings</em>.
</div>
<div class="csl-entry">
McCullough, Bruce D. 1999. <span>“Assessing the Reliability of Statistical Software: Part II.”</span> <em>The American Statistician</em> 53 (2): 149–59.
</div>
<div class="csl-entry">
McCullough, Bruce D, and David A Heiser. 2008. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2007.”</span> <em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4570–78.
</div>
<div class="csl-entry">
McCullough, Bruce D, and Berry Wilson. 1999. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 97.”</span> <em>Computational Statistics &amp; Data Analysis</em> 31 (1): 27–37.
</div>
<div class="csl-entry">
———. 2005. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2003.”</span> <em>Computational Statistics &amp; Data Analysis</em> 49 (4): 1244–52.
</div>
<div class="csl-entry">
McMurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An <span>R</span> Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> <em>PloS One</em> 8 (4): e61217.
</div>
<div class="csl-entry">
McNamara, Amelia. 2016. <span>“On the State of Computing in Statistics Education: Tools for Learning and for Doing.”</span> <em>arXiv Preprint arXiv:1610.00984</em>.
</div>
<div class="csl-entry">
Metz, Cade. 2015. <span>“How GitHub Conquered Google, Microsoft, and Everyone Else.”</span> <em>Wired Magazine</em>.
</div>
<div class="csl-entry">
Mélard, Guy. 2014. <span>“On the Accuracy of Statistical Procedures in Microsoft Excel 2010.”</span> <em>Computational Statistics</em> 29 (5): 1095–1128.
</div>
<div class="csl-entry">
Michener, William K. 2015. <span>“Ten Simple Rules for Creating a Good Data Management Plan.”</span> <em>PLoS Computational Biology</em> 11 (10): e1004525.
</div>
<div class="csl-entry">
Mittelbach, Frank, Michel Goossens, Johannes Braams, David Carlisle, and Chris Rowley. 2004. <em>The LATEX Companion</em>. Addison-Wesley Professional.
</div>
<div class="csl-entry">
Murrell, Paul. 2009. <em>Introduction to Data Technologies</em>. Chapman; Hall/CRC.
</div>
<div class="csl-entry">
Myneni, Sahiti, and Vimla L Patel. 2010. <span>“Organization of Biomedical Data for Collaborative Scientific Research: A Research Information Management System.”</span> <em>International Journal of Information Management</em> 30 (3): 256–64.
</div>
<div class="csl-entry">
Nardi, Bonnie A, and James R Miller. 1990. <span>“The Spreadsheet Interface: A Basis for End User Programming.”</span> In <em>Proceedings of the IFIP Tc13 Third Interational Conference on Human-Computer Interaction</em>, 977–83. North-Holland Publishing Co.
</div>
<div class="csl-entry">
Nash, JC. 2006. <span>“Spreadsheets in Statistical Practice—Another Look.”</span> <em>The American Statistician</em> 60 (3): 287–89.
</div>
<div class="csl-entry">
Pawlik, Aleksandra, Celia WG van Gelder, Aleksandra Nenadic, Patricia M Palagi, Eija Korpelainen, Philip Lijnzaad, Diana Marek, Susanna-Assunta Sansone, John Hancock, and Carole Goble. 2017. <span>“Developing a Strategy for Computational Lab Skills Training Through <span class="nocase">Software and Data Carpentry</span>: Experiences from the <span>ELIXIR Pilot</span> Action.”</span> <em>F1000Research</em> 6: ELIXIR–1040.
</div>
<div class="csl-entry">
Peng, Roger. 2018. <span>“Teaching r to New Users—from Tapply to the Tidyverse.”</span> <em>Simply Statistics</em>.
</div>
<div class="csl-entry">
Perez-Riverol, Yasset, Laurent Gatto, Rui Wang, Timo Sachsenberg, Julian Uszkoreit, Felipe da Veiga Leprevost, Christian Fufezan, et al. 2016. <span>“Ten Simple Rules for Taking Advantage of Git and GitHub.”</span> <em>PLoS Computational Biology</em> 12 (7).
</div>
<div class="csl-entry">
Perkel, Jeffrey. 2018. <span>“Git: The Reproducibility Tool Scientists Love to Hate.”</span> <em>Naturejobs Blog</em>.
</div>
<div class="csl-entry">
Perkel, Jeffrey M. 2011. <span>“Coding Your Way Out of a Problem.”</span> Nature Publishing Group.
</div>
<div class="csl-entry">
Powell, Kendall. 2012. <span>“A Lab App for That.”</span> <em>Nature</em> 484 (7395): 553–55.
</div>
<div class="csl-entry">
Powell, Stephen G, Kenneth R Baker, and Barry Lawson. 2009. <span>“Errors in Operational Spreadsheets: A Review of the State of the Art.”</span> In <em>2009 42nd Hawaii International Conference on System Sciences</em>, 1–8. IEEE.
</div>
<div class="csl-entry">
Quintelier, Katrien, Artuur Couckuyt, Annelies Emmaneel, Joachim Aerts, Yvan Saeys, and Sofie Van Gassen. 2021. <span>“Analyzing High-Dimensional Cytometry Data Using FlowSOM.”</span> <em>Nature Protocols</em> 16 (8): 3775–3801.
</div>
<div class="csl-entry">
Raymond, Eric. 2009. <span>“Understanding Version-Control Systems (DRAFT).”</span> <a href="http://www.catb.org/ esr/writings/version-control/version-control.html">http://www.catb.org/ esr/writings/version-control/version-control.html</a>.
</div>
<div class="csl-entry">
Raymond, Eric S. 2003. <em>The Art of Unix Programming</em>. Addison-Wesley Professional.
</div>
<div class="csl-entry">
Robinson, David. 2014. <span>“Broom: An r Package for Converting Statistical Analysis Objects into Tidy Data Frames.”</span> <em>arXiv Preprint arXiv:1412.3565</em>.
</div>
<div class="csl-entry">
———. 2017. <span>“Teach the Tidyverse to Beginners.”</span> <em>Variance Explained</em>.
</div>
<div class="csl-entry">
Robinson, Mark D, Davis J McCarthy, and Gordon K Smyth. 2010. <span>“edgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.”</span> <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.
</div>
<div class="csl-entry">
Ross, Zev, Hadley Wickham, and David Robinson. 2017. <span>“Declutter Your <span>R</span> Workflow with Tidy Tools.”</span> <em>PeerJ Preprints</em> 5: e3180v1.
</div>
<div class="csl-entry">
<span>“<span>RStudio Project Templates</span>.”</span> 2021. <a href="https://rstudio.github.io/rstudio-extensions/rstudio_project_templates.html">https://rstudio.github.io/rstudio-extensions/rstudio_project_templates.html</a>.
</div>
<div class="csl-entry">
Sansone, Susanna-Assunta, Philippe Rocca-Serra, Dawn Field, Eamonn Maguire, Chris Taylor, Oliver Hofmann, Hong Fang, et al. 2012. <span>“Toward Interoperable Bioscience Data.”</span> <em>Nature Genetics</em> 44 (2): 121.
</div>
<div class="csl-entry">
Schadt, Eric E, Michael D Linderman, Jon Sorenson, Lawrence Lee, and Garry P Nolan. 2010. <span>“Computational Solutions to Large-Scale Data Management and Analysis.”</span> <em>Nature Reviews Genetics</em> 11 (9): 647.
</div>
<div class="csl-entry">
Schrode, Nadine, Carina Seah, PJ Michael Deans, Gabriel Hoffman, and Kristen J Brennand. 2021. <span>“Analysis Framework and Experimental Design for Evaluating Synergy-Driving Gene Expression.”</span> <em>Nature Protocols</em> 16 (2): 812–40.
</div>
<div class="csl-entry">
Sedgwick, Philip. 2014. <span>“Unit of Observation and Unit of Analysis.”</span> <em>BMJ</em> 348: g3840.
</div>
<div class="csl-entry">
Silge, Julia, and David Robinson. 2016. <span>“Tidytext: Text Mining and Analysis Using Tidy Data Principles in <span>R</span>.”</span> <em>The Journal of Open Source Software</em> 1 (3).
</div>
<div class="csl-entry">
———. 2017. <em>Text Mining with r: A Tidy Approach</em>. Sebastopol: O’Reilly Media.
</div>
<div class="csl-entry">
Smith, Colin A. 2013. <span>“LC/MS Preprocessing and Analysis with Xcms.”</span> <em>Documentation of Bioconductor Xcms Package</em>.
</div>
<div class="csl-entry">
Spraul, V Anton. 2012. <em>Think Like a Programmer: An Introduction to Creative Problem Solving</em>. no starch press.
</div>
<div class="csl-entry">
Stander, Julian, and Luciana Dalla Valle. 2017. <span>“On Enthusing Students about Big Data and Social Media Visualization and Analysis Using <span class="nocase">R, RStudio, and RMarkdown</span>.”</span> <em>Journal of Statistics Education</em> 25 (2): 60–67.
</div>
<div class="csl-entry">
Target, Sinclair. 2018. <span>“Version Control Before Git with CVS.”</span> <a href="https://twobithistory.org/2018/07/07/cvs.html">https://twobithistory.org/2018/07/07/cvs.html</a>.
</div>
<div class="csl-entry">
Teixeira, Ricardo, and Vasco Amaral. 2016. <span>“On the Emergence of Patterns for Spreadsheets Data Arrangements.”</span> In <em>Federation of International Conferences on Software Technologies: Applications and Foundations</em>, 333–45. Springer.
</div>
<div class="csl-entry">
Tippmann, Sylvia. 2014. <span>“My Digital Toolbox: Nuclear Engineer Katy Huff on Version-Control Systems.”</span> <em>Nature News</em>.
</div>
<div class="csl-entry">
Topaloglou, Thodoros, Susan B Davidson, HV Jagadish, Victor M Markowitz, Evan W Steeg, and Mike Tyers. 2004. <span>“Biological Data Management: Research, Practice and Opportunities.”</span> In <em>Proceedings of the Thirtieth International Conference on Very Large Data Bases-Volume 30</em>, 1233–36. VLDB Endowment.
</div>
<div class="csl-entry">
Tyner, Sam, François Briatte, and Heike Hofmann. 2017. <span>“<span class="nocase">Network Visualization with ggplot2</span>.”</span> <em><span>The R Journal</span></em> 9 (1): 27–59.
</div>
<div class="csl-entry">
U.S. Department of Health and Human Services, National Institutes of Health. 2016. <span>“<span>NIH-Wide Strategic Plan, Fiscal Years 2016-2020: Turning Discovery Into Health</span>.”</span> <a href="https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf">https://www.nih.gov/sites/default/files/about-nih/strategic-plan-fy2016-2020-508.pdf</a>.
</div>
<div class="csl-entry">
———. 2018. <span>“<span class="nocase">NIH Strategic Plan for Data Science</span>.”</span> <a href="https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf">https://datascience.nih.gov/sites/default/files/NIH_Strategic_Plan_for_Data_Science_Final_508.pdf</a>.
</div>
<div class="csl-entry">
Van Dongen, Marc RC. 2012. <em>LATEX and Friends</em>. Springer Science &amp; Business Media.
</div>
<div class="csl-entry">
Waltemath, Dagmar, and Olaf Wolkenhauer. 2016. <span>“How Modeling Standards, Software, and Initiatives Support Reproducibility in Systems Biology and Systems Medicine.”</span> <em>IEEE Transactions on Biomedical Engineering</em> 63 (10): 1999–2006.
</div>
<div class="csl-entry">
Welsh, Eric A, Paul A Stewart, Brent M Kuenzi, and James A Eschrich. 2017. <span>“Escape Excel: A Tool for Preventing Gene Symbol and Accession Conversion Errors.”</span> <em>PloS One</em> 12 (9): e0185207.
</div>
<div class="csl-entry">
Wickham, Hadley. 2014. <span>“Tidy Data.”</span> <em>Journal of Statistical Software</em> 59 (10): 1–23.
</div>
<div class="csl-entry">
———. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. 2nd ed. New York: Springer.
</div>
<div class="csl-entry">
———. 2017. <span>“The Tidy Tools Manifesto.”</span> <em>CRAN Vignette</em>.
</div>
<div class="csl-entry">
———. 2019. <em>Advanced r, Second Edition</em>. CRC press.
</div>
<div class="csl-entry">
Wickham, Hadley, and Garrett Grolemund. 2016. <em>R for Data Science: Import, Tidy, Transform, Visualize, and Model Data</em>. sebastopol: O’Reilly Media.
</div>
<div class="csl-entry">
Wilkinson, Mark D, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, et al. 2016. <span>“The FAIR Guiding Principles for Scientific Data Management and Stewardship.”</span> <em>Scientific Data</em> 3.
</div>
<div class="csl-entry">
Willekens, Frans. 2013. <span>“Chronological Objects in Demographic Research.”</span> <em>Demographic Research</em> 28: 649–80.
</div>
<div class="csl-entry">
Wilson, G. 2014. <span>“<span>Software Carpentry</span>: Lessons Learned.”</span> <em>F1000Research</em> 3: 62–62.
</div>
<div class="csl-entry">
Xie, Yihui, Joseph J Allaire, and Garrett Grolemund. 2018. <em>R Markdown: The Definitive Guide</em>. CRC Press.
</div>
<div class="csl-entry">
Yin, Tengfei, Dianne Cook, and Michael Lawrence. 2012. <span>“Ggbio: An <span>R</span> Package for Extending the Grammar of Graphics for Genomic Data.”</span> <em>Genome Biology</em> 13 (8): R77.
</div>
<div class="csl-entry">
Zeeberg, Barry R, Joseph Riss, David W Kane, Kimberly J Bussey, Edward Uchio, W Marston Linehan, J Carl Barrett, and John N Weinstein. 2004. <span>“Mistaken Identifiers: Gene Name Errors Can Be Introduced Inadvertently When Using <span>Excel</span> in Bioinformatics.”</span> <em>BMC Bioinformatics</em> 5 (1): 80.
</div>
<div class="csl-entry">
Ziemann, Mark, Yotam Eren, and Assam El-Osta. 2016. <span>“Gene Name Errors Are Widespread in the Scientific Literature.”</span> <em>Genome Biology</em> 17 (1): 177.
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="experimental-data-preprocessing.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/20-example_protocol.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["improve_repro.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
